<!doctype html>
<html class="no-js" lang="en" data-content_root="../../../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../../genindex.html"><link rel="search" title="Search" href="../../../search.html">

    <!-- Generated with Sphinx 8.2.3 and Furo 2025.09.25 -->
        <title>decima.data.preprocess - decima 0.0.post1.dev1+g6e9347243 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=580074bf" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">decima 0.0.post1.dev1+g6e9347243 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  <span class="sidebar-brand-text">decima 0.0.post1.dev1+g6e9347243 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">Overview</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../api/decima.html">decima package</a><input aria-label="Toggle navigation of decima package" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.cli.html">decima.cli package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.core.html">decima.core package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.data.html">decima.data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.hub.html">decima.hub package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.interpret.html">decima.interpret package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.model.html">decima.model package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.plot.html">decima.plot package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.tools.html">decima.tools package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.train.html">decima.train package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.utils.html">decima.utils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.vep.html">decima.vep package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <h1>Source code for decima.data.preprocess</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">json</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">bioframe</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">bf</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">grelu.io.genome</span><span class="w"> </span><span class="kn">import</span> <span class="n">read_sizes</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">grelu.sequence.format</span><span class="w"> </span><span class="kn">import</span> <span class="n">intervals_to_strings</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">grelu.sequence.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_unique_length</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">mygene</span>


<span class="kn">from</span><span class="w"> </span><span class="nn">decima.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">DECIMA_CONTEXT_SIZE</span>


<div class="viewcode-block" id="merge_transcripts">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.merge_transcripts">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">merge_transcripts</span><span class="p">(</span><span class="n">gtf</span><span class="p">):</span>
    <span class="c1"># Get gene-level columns</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">gtf</span><span class="p">[[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_type&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Aggregate all features from the same gene</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">genes</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;gene_name&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>

    <span class="c1"># Get minimum start and maximum end</span>
    <span class="n">genes</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">min</span><span class="p">)</span>
    <span class="n">genes</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">max</span><span class="p">)</span>

    <span class="c1"># Merge all other columns</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_type&quot;</span><span class="p">]:</span>
        <span class="n">genes</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">genes</span></div>



<div class="viewcode-block" id="var_to_intervals">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.var_to_intervals">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">var_to_intervals</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">chr_end_pad</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="n">seq_len</span><span class="o">=</span><span class="n">DECIMA_CONTEXT_SIZE</span><span class="p">,</span> <span class="n">crop_coords</span><span class="o">=</span><span class="mi">163840</span><span class="p">):</span>
    <span class="n">sizes</span> <span class="o">=</span> <span class="n">read_sizes</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="c1"># Calculate interval size</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;The interval size is </span><span class="si">{</span><span class="n">seq_len</span><span class="si">}</span><span class="s2"> bases. Of these, </span><span class="si">{</span><span class="n">crop_coords</span><span class="si">}</span><span class="s2"> will be upstream of the gene start and </span><span class="si">{</span><span class="n">seq_len</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">crop_coords</span><span class="si">}</span><span class="s2"> will be downstream of the gene start.&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Create intervals around + strand genes</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_start&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">crop_coords</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">seq_len</span>

    <span class="c1"># Create interval around - strand genes</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_end&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">crop_coords</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">seq_len</span>

    <span class="c1"># shift sequences with start &lt; 0</span>
    <span class="n">crossing_start</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">start</span> <span class="o">&lt;</span> <span class="n">chr_end_pad</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">crossing_start</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">chr_end_pad</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">crossing_start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">crossing_start</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">seq_len</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">crossing_start</span><span class="p">)</span><span class="si">}</span><span class="s2"> intervals extended beyond the chromosome start and have been shifted&quot;</span><span class="p">)</span>

    <span class="c1"># shift sequences with end &gt; chromosome size</span>
    <span class="n">crossing_end</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">size</span> <span class="ow">in</span> <span class="n">sizes</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="n">max_end</span> <span class="o">=</span> <span class="n">size</span> <span class="o">-</span> <span class="n">chr_end_pad</span>
        <span class="n">drop</span> <span class="o">=</span> <span class="p">(</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">chrom</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">max_end</span><span class="p">)</span>
        <span class="n">crossing_end</span> <span class="o">+=</span> <span class="n">drop</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drop</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">max_end</span>
        <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drop</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drop</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">seq_len</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">crossing_end</span><span class="si">}</span><span class="s2"> intervals extended beyond the chromosome end and have been shifted&quot;</span><span class="p">)</span>

    <span class="c1"># gene start position on output sequence</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_mask_start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_start&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_mask_start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_end&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_mask_start</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_mask_start</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_length</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_length</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="c1"># Get gene end position on sequence</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;gene_mask_end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_mask_start</span> <span class="o">+</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_length</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">seq_len</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_mask_end</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_mask_end</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="c1"># Drop intervals with less than crop_coords upstream bases</span>
    <span class="n">drop</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_mask_start</span> <span class="o">&lt;</span> <span class="n">crop_coords</span>
    <span class="n">ad</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[:,</span> <span class="o">~</span><span class="n">drop</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">drop</span><span class="p">)</span><span class="si">}</span><span class="s2"> intervals did not extend far enough upstream of the TSS and have been dropped&quot;</span><span class="p">)</span>

    <span class="c1"># Check length</span>
    <span class="k">assert</span> <span class="n">get_unique_length</span><span class="p">(</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="p">)</span> <span class="o">==</span> <span class="n">seq_len</span>
    <span class="k">return</span> <span class="n">ad</span></div>



<div class="viewcode-block" id="assign_borzoi_folds">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.assign_borzoi_folds">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">assign_borzoi_folds</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">splits</span><span class="p">):</span>
    <span class="c1"># Extract gene intervals</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">})</span>

    <span class="c1"># Overlap with Borzoi splits</span>
    <span class="n">overlaps</span> <span class="o">=</span> <span class="n">bf</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">splits</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">overlaps</span> <span class="o">=</span> <span class="n">overlaps</span><span class="p">[[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;fold_&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">overlaps</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;fold&quot;</span><span class="p">]</span>

    <span class="c1"># List all overlapping folds for each gene</span>
    <span class="n">overlaps</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">overlaps</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;gene_id&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">fold</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">f</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))</span>
    <span class="p">)</span>
    <span class="n">overlaps</span> <span class="o">=</span> <span class="n">overlaps</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">overlaps</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">overlaps</span><span class="o">.</span><span class="n">fold</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">),</span> <span class="s2">&quot;fold&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span>

    <span class="c1"># Add back to AnnData</span>
    <span class="n">ind</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">overlaps</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">ind</span>
    <span class="k">return</span> <span class="n">ad</span></div>



<div class="viewcode-block" id="aggregate_anndata">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.aggregate_anndata">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">aggregate_anndata</span><span class="p">(</span>
    <span class="n">ad</span><span class="p">,</span>
    <span class="n">by_cols</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
        <span class="s2">&quot;tissue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;organ&quot;</span><span class="p">,</span>
        <span class="s2">&quot;disease&quot;</span><span class="p">,</span>
        <span class="s2">&quot;study&quot;</span><span class="p">,</span>
        <span class="s2">&quot;dataset&quot;</span><span class="p">,</span>
        <span class="s2">&quot;region&quot;</span><span class="p">,</span>
        <span class="s2">&quot;subregion&quot;</span><span class="p">,</span>
        <span class="s2">&quot;celltype_coarse&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">sum_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;n_cells&quot;</span><span class="p">],</span>
<span class="p">):</span>
    <span class="c1"># Get column names</span>
    <span class="n">obs_cols</span> <span class="o">=</span> <span class="n">by_cols</span> <span class="o">+</span> <span class="n">sum_cols</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="c1"># Format obs</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating new obs matrix&quot;</span><span class="p">)</span>
    <span class="n">obs</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_cols</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">by_cols</span><span class="p">:</span>
        <span class="n">obs</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">obs</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">sum_cols</span><span class="p">:</span>
        <span class="n">obs</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">obs</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="c1"># Create X</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">ad</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">columns</span><span class="o">=</span><span class="n">gene_names</span><span class="p">)</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">obs</span><span class="p">,</span>
            <span class="n">X</span><span class="p">,</span>
        <span class="p">],</span>
        <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Aggregate X</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Aggregating&quot;</span><span class="p">)</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">by_cols</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1"># Split off the obs again</span>
    <span class="n">obs</span> <span class="o">=</span> <span class="n">X</span><span class="p">[</span><span class="n">obs_cols</span><span class="p">]</span>
    <span class="n">obs</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;agg_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">obs</span><span class="p">))]</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">X</span><span class="p">[</span><span class="n">gene_names</span><span class="p">]</span>

    <span class="c1"># Create new anndata</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating new anndata&quot;</span><span class="p">)</span>
    <span class="n">new_ad</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">X</span><span class="p">),</span> <span class="n">obs</span><span class="o">=</span><span class="n">obs</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">new_ad</span></div>



<div class="viewcode-block" id="change_values">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.change_values">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">change_values</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">value_dict</span><span class="p">):</span>
    <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">value_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">==</span> <span class="n">k</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>
    <span class="k">return</span> <span class="n">df</span></div>



<div class="viewcode-block" id="get_frac_N">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.get_frac_N">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_frac_N</span><span class="p">(</span><span class="n">interval</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">):</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">intervals_to_strings</span><span class="p">(</span><span class="n">interval</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span></div>



<div class="viewcode-block" id="match_cellranger_2024">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.match_cellranger_2024">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">match_cellranger_2024</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">genes24</span><span class="p">):</span>
    <span class="n">matched</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">unmatched_genes</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unmatched_genes</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes unmatched.&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">unmatched_genes</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes24</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">genes24</span><span class="o">.</span><span class="n">gene_id</span><span class="p">[</span><span class="n">genes24</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">gene_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes24</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes24</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">gene</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">matched</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">matched</span><span class="si">}</span><span class="s2"> genes matched.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="match_ref_ad">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.match_ref_ad">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">match_ref_ad</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">ref_ad</span><span class="p">):</span>
    <span class="n">matched</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">unmatched_genes</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unmatched_genes</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes unmatched.&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">unmatched_genes</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">ref_ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">ref_ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_id</span><span class="p">[</span><span class="n">ref_ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">gene_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref_ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ref_ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">gene</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">matched</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">matched</span><span class="si">}</span><span class="s2"> genes matched.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="load_ncbi_string">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.load_ncbi_string">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">load_ncbi_string</span><span class="p">(</span><span class="n">string</span><span class="p">,</span> <span class="n">allow_dups</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">reports</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">string</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># Check the total count</span>
    <span class="k">if</span> <span class="n">reports</span> <span class="o">==</span> <span class="p">{</span><span class="s2">&quot;total_count&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reports</span><span class="p">[</span><span class="s2">&quot;reports&quot;</span><span class="p">]):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">curr_dict</span> <span class="o">=</span> <span class="p">{}</span>

                <span class="c1"># Get gene metadata</span>
                <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="s2">&quot;query&quot;</span> <span class="ow">in</span> <span class="n">r</span> <span class="k">else</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>
                <span class="n">r</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
                <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>
                <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span> <span class="o">+</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;chromosomes&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;gene_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span>

                <span class="c1"># Get ensembl ID</span>
                <span class="k">if</span> <span class="s2">&quot;ensembl_gene_ids&quot;</span> <span class="ow">in</span> <span class="n">r</span><span class="p">:</span>
                    <span class="n">eids</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;ensembl_gene_ids&quot;</span><span class="p">]</span>
                    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">eids</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
                    <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">eids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>

                <span class="c1"># Get genomic position</span>
                <span class="n">annot</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;annotations&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;assembly_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GRCh38.p14&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">loc</span> <span class="o">=</span> <span class="n">annot</span><span class="p">[</span><span class="s2">&quot;genomic_locations&quot;</span><span class="p">]</span>
                <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
                <span class="n">loc</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;genomic_range&quot;</span><span class="p">]</span>

                <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[</span><span class="s2">&quot;begin&quot;</span><span class="p">]</span>
                <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
                <span class="n">curr_dict</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="k">if</span> <span class="n">loc</span><span class="p">[</span><span class="s2">&quot;orientation&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;minus&quot;</span> <span class="k">else</span> <span class="s2">&quot;+&quot;</span>

                <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_dict</span><span class="p">)</span>

            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">pass</span>

        <span class="n">out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">allow_dups</span><span class="p">:</span>
            <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="p">[</span><span class="n">out</span><span class="o">.</span><span class="n">gene_name</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">gene_name</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()[</span><span class="n">out</span><span class="o">.</span><span class="n">gene_name</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span>
        <span class="n">out</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ncbi&quot;</span>
        <span class="k">return</span> <span class="n">out</span></div>



<div class="viewcode-block" id="match_ncbi">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.match_ncbi">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">match_ncbi</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">ncbi</span><span class="p">):</span>
    <span class="n">matched</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">unmatched_genes</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unmatched_genes</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes unmatched.&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">unmatched_genes</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">gene_name</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">ad</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ncbi</span><span class="o">.</span><span class="n">gene_name</span> <span class="o">==</span> <span class="n">gene</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">matched</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">matched</span><span class="si">}</span><span class="s2"> genes matched.&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="make_inputs">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.make_inputs">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">make_inputs</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">ad</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">DeprecationWarning</span><span class="p">(</span>
        <span class="s2">&quot;make_inputs() is deprecated and will be removed in a future version. &quot;</span>
        <span class="s2">&quot;Please use the DecimaResult.prepare_one_hot() directly instead.&quot;</span>
    <span class="p">)</span></div>



<div class="viewcode-block" id="return_ensembl">
<a class="viewcode-back" href="../../../api/decima.data.html#decima.data.preprocess.return_ensembl">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">return_ensembl</span><span class="p">(</span><span class="n">queries</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;gene_id&quot;</span><span class="p">):</span>
    <span class="n">mg</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">MyGeneInfo</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">fields</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;ensembl.gene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;symbol&quot;</span><span class="p">,</span>
        <span class="s2">&quot;type_of_gene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genomic_pos.chr&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genomic_pos.start&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genomic_pos.end&quot;</span><span class="p">,</span>
        <span class="s2">&quot;genomic_pos.strand&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">query</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">queries</span><span class="p">):</span>
        <span class="c1"># Make the query</span>
        <span class="k">if</span> <span class="n">on</span> <span class="o">==</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">:</span>
            <span class="n">hits</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ensembl.gene:</span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">fields</span><span class="o">=</span><span class="n">fields</span><span class="p">)[</span><span class="s2">&quot;hits&quot;</span><span class="p">]</span>
            <span class="n">hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">hit</span> <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span> <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;symbol&quot;</span> <span class="ow">in</span> <span class="n">hit</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;ensembl&quot;</span> <span class="ow">in</span> <span class="n">hit</span><span class="p">)]</span>

        <span class="k">elif</span> <span class="n">on</span> <span class="o">==</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">:</span>
            <span class="n">hits</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">fields</span><span class="o">=</span><span class="n">fields</span><span class="p">)[</span><span class="s2">&quot;hits&quot;</span><span class="p">]</span>
            <span class="n">hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">hit</span> <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span> <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;symbol&quot;</span> <span class="ow">in</span> <span class="n">hit</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;ensembl&quot;</span> <span class="ow">in</span> <span class="n">hit</span><span class="p">)]</span>
            <span class="n">hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">hit</span> <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">hit</span><span class="p">[</span><span class="s2">&quot;ensembl&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span>
            <span class="n">hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">hit</span> <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span> <span class="k">if</span> <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;ensembl&quot;</span><span class="p">][</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ENSG&quot;</span><span class="p">)]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">hits</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">hit</span> <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span> <span class="k">if</span> <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">query</span><span class="p">]</span>

        <span class="c1"># Format the query</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">hits</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">hit</span> <span class="o">=</span> <span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">hit</span><span class="p">[</span><span class="s2">&quot;genomic_pos&quot;</span><span class="p">])</span> <span class="ow">is</span> <span class="nb">dict</span><span class="p">:</span>
                <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;genomic_pos&quot;</span><span class="p">][</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]:</span>
                    <span class="n">hit</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;genomic_pos&quot;</span><span class="p">][</span><span class="n">col</span><span class="p">]</span>
                <span class="k">del</span> <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;genomic_pos&quot;</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">on</span> <span class="o">==</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">:</span>
                    <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">query</span>
                <span class="k">elif</span> <span class="n">on</span> <span class="o">==</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">:</span>
                    <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;ensembl&quot;</span><span class="p">][</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
                    <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">query</span>

                <span class="k">del</span> <span class="n">hit</span><span class="p">[</span><span class="s2">&quot;ensembl&quot;</span><span class="p">]</span>
                <span class="n">df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">hit</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">strand</span><span class="o">.</span><span class="n">map</span><span class="p">({</span><span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;-&quot;</span><span class="p">})</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;mygene&quot;</span>
    <span class="k">return</span> <span class="n">df</span></div>

</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Gokcen Eraslan
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../../_static/documentation_options.js?v=52fac5ed"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>