{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b0f8e93b",
   "metadata": {},
   "source": [
    "# Variant Effect Prediction with Motif Intpretation with Attributions "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91fc519b",
   "metadata": {},
   "source": [
    "## CLI API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7d51a63c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: decima vep-attribution [OPTIONS]\n",
      "\n",
      "  Predict variant effect and save to parquet\n",
      "\n",
      "  Examples:\n",
      "\n",
      "      >>> decima vep-attribution -v \"data/sample.vcf\" -o \"vep_results.h5\"\n",
      "\n",
      "Options:\n",
      "  -v, --variants PATH             Path to the variant file .vcf file. VCF file\n",
      "                                  need to be normalized. Try normalizing th\n",
      "                                  vcf file incase of an error. `bcftools norm\n",
      "                                  -f ref.fasta input.vcf.gz -o output.vcf.gz`\n",
      "  -o, --output_h5 PATH            Path to the output h5 file.\n",
      "  --tasks TEXT                    Tasks to predict. If not provided, all tasks\n",
      "                                  will be predicted.\n",
      "  --off-tasks TEXT                Tasks to contrast against. If not provided,\n",
      "                                  no contrast will be performed.\n",
      "  --model INTEGER                 Model to use for attribution analysis.\n",
      "                                  Available options: ['v1_rep0', 'v1_rep1',\n",
      "                                  'v1_rep2', 'v1_rep3', 'ensemble', 'rep0',\n",
      "                                  'rep1', 'rep2', 'rep3', 0, 1, 2, 3, '0',\n",
      "                                  '1', '2', '3'].\n",
      "  --metadata TEXT                 Path to the metadata anndata file or name of\n",
      "                                  the model. If not provided, the compabilite\n",
      "                                  metadata for the model will be used.\n",
      "                                  Default: ensemble.\n",
      "  --method TEXT                   Method to use for attribution analysis.\n",
      "                                  Available options: 'inputxgradient',\n",
      "                                  'saliency', 'integratedgradients'.\n",
      "  --transform [specificity|aggregate]\n",
      "                                  Transform to use for attribution analysis.\n",
      "                                  Available options: 'specificity',\n",
      "                                  'aggregate'.\n",
      "  --device TEXT                   Device to use. Default: None which\n",
      "                                  automatically selects the best device.\n",
      "  --num-workers INTEGER           Number of workers for the loader. Default: 4\n",
      "  --distance-type TEXT            Type of distance. Default: tss.\n",
      "  --min-distance FLOAT            Minimum distance from the end of the gene.\n",
      "                                  Default: 0.\n",
      "  --max-distance FLOAT            Maximum distance from the TSS. Default:\n",
      "                                  524288.\n",
      "  --gene-col TEXT                 Column name for gene names. Default: None.\n",
      "  --genome TEXT                   Genome build. Default: hg38.\n",
      "  --float-precision TEXT          Floating-point precision to be used in\n",
      "                                  calculations. Avaliable options include:\n",
      "                                  '16-true', '16-mixed', 'bf16-true',\n",
      "                                  'bf16-mixed', '32-true', '64-true', '32',\n",
      "                                  '16', and 'bf16'.\n",
      "  --help                          Show this message and exit.\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! decima vep-attribution --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "963b41a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "! decima vep -v \"data/sample.vcf\" -o \"vep_vcf_attributions.h5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a477e150",
   "metadata": {},
   "outputs": [],
   "source": [
    "! ls vep_vcf_attributions.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fc2c6db",
   "metadata": {},
   "source": [
    "\n",
    "## Python API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9309f1d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import torch\n",
    "from decima.core.attribution import VariantAttributionResult\n",
    "from decima.vep import variant_effect_attribution\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d9ad6a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_effect_attribution(\n",
    "    \"tests/data/test.vcf\",\n",
    "    \"vep_vcf_attributions.h5\",\n",
    "    model=0,\n",
    "    method=\"inputxgradient\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4126fb8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "with h5py.File(\"vep_vcf_attributions.h5\", \"r\") as f:\n",
    "    print(f.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3022232e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with VariantAttributionResult(\"vep_vcf_attributions.h5\", tss_distance=10_000, num_workers=1) as ar:\n",
    "    genes = ar.genes\n",
    "    variants = ar.variants\n",
    "    print(genes)\n",
    "    print(variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cb59750",
   "metadata": {},
   "outputs": [],
   "source": [
    "with VariantAttributionResult(\"vep_vcf_attributions.h5\", tss_distance=10_000, num_workers=1) as ar:\n",
    "    seqs_ref, attrs_ref, seqs_alt, attrs_alt = ar.load(variants, genes)\n",
    "    print(seqs_ref)\n",
    "    print(attrs_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b3b4ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "with VariantAttributionResult(\"vep_vcf_attributions.h5\", tss_distance=10_000, num_workers=1) as ar:\n",
    "    attribution_ref, attribution_alt = ar.load(variants, genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcc21b89",
   "metadata": {},
   "outputs": [],
   "source": [
    "with VariantAttributionResult(\"vep_vcf_attributions.h5\", tss_distance=10_000, num_workers=1) as ar:\n",
    "    df_peaks, df_motifs = ar.recursive_seqlet_calling(variants, genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e65e660",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eb265ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b983e884",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribution_ref.plot_seqlogo(relative_loc=291)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fae648d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribution_alt.plot_seqlogo(relative_loc=291)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "decima2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
