<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html"><link rel="search" title="Search" href="../search.html"><link rel="next" title="Variant Effect Prediction with Decima" href="2-variant-effect-prediction.html"><link rel="prev" title="decima.vep package" href="../api/decima.vep.html">

    <!-- Generated with Sphinx 8.2.3 and Furo 2025.09.25 -->
        <title>Attribution and Motifs Detection with Decima - decima unknown documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=580074bf" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">decima unknown documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <span class="sidebar-brand-text">decima unknown documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/decima.html">decima package</a><input aria-label="Toggle navigation of decima package" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.cli.html">decima.cli package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.core.html">decima.core package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.data.html">decima.data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.hub.html">decima.hub package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.interpret.html">decima.interpret package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.model.html">decima.model package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.plot.html">decima.plot package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.tools.html">decima.tools package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.train.html">decima.train package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.utils.html">decima.utils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.vep.html">decima.vep package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Tutorial: Attribution Motif Discovery</a></li>
<li class="toctree-l1"><a class="reference internal" href="2-variant-effect-prediction.html">Tutorial: Variant Effect Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="3-finetune.html">Tutorial: Finetuning</a></li>
<li class="toctree-l1"><a class="reference internal" href="4-modisco.html">Tutorial: Modisco</a></li>
<li class="toctree-l1"><a class="reference internal" href="5-gene-expression-prediction.html">Tutorial: Gene Expression Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/1-attribution-motif-discovery.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section class="tex2jax_ignore mathjax_ignore" id="attribution-and-motifs-detection-with-decima">
<h1>Attribution and Motifs Detection with Decima<a class="headerlink" href="#attribution-and-motifs-detection-with-decima" title="Link to this heading">¶</a></h1>
<p>This documentation demonstrates how to use Decima’s attribution analysis capabilities to identify important regulatory regions in genomic sequences and discover transcription factor binding motifs within those regions. Attribution analysis helps reveal which parts of the DNA sequence most strongly influence gene expression predictions, while <strong>motif scanning</strong> can identify specific transcription factor binding sites in these regions of interest.</p>
<section id="cli-api">
<h2>CLI API<a class="headerlink" href="#cli-api" title="Link to this heading">¶</a></h2>
<p>Let’s look at a simple example using Decima’s CLI API to analyze the SPI1 and BRD3 genes. SPI1 is a key transcription factor in myeloid cell development. We’ll examine its regulation across different monocyte and macrophage cell types where it is known to be important.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>attributions<span class="w"> </span>--help
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Traceback (most recent call last):
  File &quot;/home/runner/work/decima/decima/.tox/docs/bin/decima&quot;, line 3, in &lt;module&gt;
    from decima.cli import main
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/decima/__init__.py&quot;, line 3, in &lt;module&gt;
    from decima.core.result import DecimaResult
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/decima/core/result.py&quot;, line 2, in &lt;module&gt;
    import anndata
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/__init__.py&quot;, line 10, in &lt;module&gt;
    from ._core.anndata import AnnData
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/_core/anndata.py&quot;, line 27, in &lt;module&gt;
    from .. import utils
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/utils.py&quot;, line 15, in &lt;module&gt;
    from ._core.sparse_dataset import BaseCompressedSparseDataset
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/_core/sparse_dataset.py&quot;, line 32, in &lt;module&gt;
    from .._settings import settings
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/_settings.py&quot;, line 16, in &lt;module&gt;
    from .compat import is_zarr_v2, old_positionals
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/compat/__init__.py&quot;, line 18, in &lt;module&gt;
    from zarr import Array as ZarrArray  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/__init__.py&quot;, line 6, in &lt;module&gt;
    from zarr.api.synchronous import (
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/api/synchronous.py&quot;, line 7, in &lt;module&gt;
    import zarr.api.asynchronous as async_api
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/api/asynchronous.py&quot;, line 13, in &lt;module&gt;
    from zarr.core.array import (
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/core/array.py&quot;, line 28, in &lt;module&gt;
    from zarr.codecs._v2 import V2Codec
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/codecs/__init__.py&quot;, line 5, in &lt;module&gt;
    from zarr.codecs.crc32c_ import Crc32cCodec
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/codecs/crc32c_.py&quot;, line 8, in &lt;module&gt;
    from crc32c import crc32c
ModuleNotFoundError: No module named &#39;crc32c&#39;
</pre></div>
</div>
</div>
</div>
<p>This decima command analyzes gene attributions: <code class="docutils literal notranslate"><span class="pre">--genes</span> <span class="pre">&quot;SPI1,BRD3&quot;</span></code> specifies focusing on SPI1 and BRD3; <code class="docutils literal notranslate"><span class="pre">--tasks</span> <span class="pre">&quot;cell_type</span> <span class="pre">==</span> <span class="pre">'classical</span> <span class="pre">monocyte'&quot;</span></code> filters the analysis to classical monocytes only; and <code class="docutils literal notranslate"><span class="pre">--output_prefix</span></code> output_classical_monoctypes/ designates the output directory for the results. You can also pass <code class="docutils literal notranslate"><span class="pre">--off-tasks</span></code> that are cell types used as a contrast group when analyzing cell type specificity - they represent the cell types you want to compare against when determining. If you do not pass, <code class="docutils literal notranslate"><span class="pre">--tasks</span></code> argument all avaliable cells will be used for attribution calculation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>attributions<span class="w"> </span>--model<span class="w"> </span>v1_rep0<span class="w"> </span>--genes<span class="w"> </span><span class="s2">&quot;SPI1,BRD3&quot;</span><span class="w"> </span>--tasks<span class="w"> </span><span class="s2">&quot;cell_type == &#39;classical monocyte&#39;&quot;</span><span class="w"> </span>--output-prefix<span class="w"> </span>example/output_classical_monoctypes
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Traceback (most recent call last):
  File &quot;/home/runner/work/decima/decima/.tox/docs/bin/decima&quot;, line 3, in &lt;module&gt;
    from decima.cli import main
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/decima/__init__.py&quot;, line 3, in &lt;module&gt;
    from decima.core.result import DecimaResult
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/decima/core/result.py&quot;, line 2, in &lt;module&gt;
    import anndata
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/__init__.py&quot;, line 10, in &lt;module&gt;
    from ._core.anndata import AnnData
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/_core/anndata.py&quot;, line 27, in &lt;module&gt;
    from .. import utils
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/utils.py&quot;, line 15, in &lt;module&gt;
    from ._core.sparse_dataset import BaseCompressedSparseDataset
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/_core/sparse_dataset.py&quot;, line 32, in &lt;module&gt;
    from .._settings import settings
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/_settings.py&quot;, line 16, in &lt;module&gt;
    from .compat import is_zarr_v2, old_positionals
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/anndata/compat/__init__.py&quot;, line 18, in &lt;module&gt;
    from zarr import Array as ZarrArray  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/__init__.py&quot;, line 6, in &lt;module&gt;
    from zarr.api.synchronous import (
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/api/synchronous.py&quot;, line 7, in &lt;module&gt;
    import zarr.api.asynchronous as async_api
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/api/asynchronous.py&quot;, line 13, in &lt;module&gt;
    from zarr.core.array import (
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/core/array.py&quot;, line 28, in &lt;module&gt;
    from zarr.codecs._v2 import V2Codec
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/codecs/__init__.py&quot;, line 5, in &lt;module&gt;
    from zarr.codecs.crc32c_ import Crc32cCodec
  File &quot;/home/runner/work/decima/decima/.tox/docs/lib/python3.11/site-packages/zarr/codecs/crc32c_.py&quot;, line 8, in &lt;module&gt;
    from crc32c import crc32c
ModuleNotFoundError: No module named &#39;crc32c&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>example/output_classical_monoctypes*
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ls: cannot access &#39;example/output_classical_monoctypes*&#39;: No such file or directory
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">h5py</span>

<span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s2">&quot;example/output_classical_monoctypes.attributions.h5&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">][:])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">][:]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="s2">&quot;attribution&quot;</span><span class="p">][:]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FileNotFoundError</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="n">line</span> <span class="mi">3</span>
<span class="g g-Whitespace">      </span><span class="mi">1</span> <span class="kn">import</span><span class="w"> </span><span class="nn">h5py</span>
<span class="ne">----&gt; </span><span class="mi">3</span> <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="s2">&quot;example/output_classical_monoctypes.attributions.h5&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="g g-Whitespace">      </span><span class="mi">4</span>     <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">][:])</span>
<span class="g g-Whitespace">      </span><span class="mi">5</span>     <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">][:]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="nn">File ~/work/decima/decima/.tox/docs/lib/python3.11/site-packages/h5py/_hl/files.py:566,</span> in <span class="ni">File.__init__</span><span class="nt">(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, fs_page_size, page_buf_size, min_meta_keep, min_raw_keep, locking, alignment_threshold, alignment_interval, meta_block_size, track_times, **kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">557</span>     <span class="n">fapl</span> <span class="o">=</span> <span class="n">make_fapl</span><span class="p">(</span><span class="n">driver</span><span class="p">,</span> <span class="n">libver</span><span class="p">,</span> <span class="n">rdcc_nslots</span><span class="p">,</span> <span class="n">rdcc_nbytes</span><span class="p">,</span> <span class="n">rdcc_w0</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">558</span>                      <span class="n">locking</span><span class="p">,</span> <span class="n">page_buf_size</span><span class="p">,</span> <span class="n">min_meta_keep</span><span class="p">,</span> <span class="n">min_raw_keep</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">559</span>                      <span class="n">alignment_threshold</span><span class="o">=</span><span class="n">alignment_threshold</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">560</span>                      <span class="n">alignment_interval</span><span class="o">=</span><span class="n">alignment_interval</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">561</span>                      <span class="n">meta_block_size</span><span class="o">=</span><span class="n">meta_block_size</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">562</span>                      <span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">563</span>     <span class="n">fcpl</span> <span class="o">=</span> <span class="n">make_fcpl</span><span class="p">(</span><span class="n">track_order</span><span class="o">=</span><span class="n">track_order</span><span class="p">,</span> <span class="n">track_times</span><span class="o">=</span><span class="n">track_times</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">564</span>                      <span class="n">fs_strategy</span><span class="o">=</span><span class="n">fs_strategy</span><span class="p">,</span> <span class="n">fs_persist</span><span class="o">=</span><span class="n">fs_persist</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">565</span>                      <span class="n">fs_threshold</span><span class="o">=</span><span class="n">fs_threshold</span><span class="p">,</span> <span class="n">fs_page_size</span><span class="o">=</span><span class="n">fs_page_size</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">566</span>     <span class="n">fid</span> <span class="o">=</span> <span class="n">make_fid</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">userblock_size</span><span class="p">,</span> <span class="n">fapl</span><span class="p">,</span> <span class="n">fcpl</span><span class="p">,</span> <span class="n">swmr</span><span class="o">=</span><span class="n">swmr</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">568</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">libver</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">569</span>     <span class="bp">self</span><span class="o">.</span><span class="n">_libver</span> <span class="o">=</span> <span class="n">libver</span>

<span class="nn">File ~/work/decima/decima/.tox/docs/lib/python3.11/site-packages/h5py/_hl/files.py:241,</span> in <span class="ni">make_fid</span><span class="nt">(name, mode, userblock_size, fapl, fcpl, swmr)</span>
<span class="g g-Whitespace">    </span><span class="mi">239</span>     <span class="k">if</span> <span class="n">swmr</span> <span class="ow">and</span> <span class="n">swmr_support</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">240</span>         <span class="n">flags</span> <span class="o">|=</span> <span class="n">h5f</span><span class="o">.</span><span class="n">ACC_SWMR_READ</span>
<span class="ne">--&gt; </span><span class="mi">241</span>     <span class="n">fid</span> <span class="o">=</span> <span class="n">h5f</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">flags</span><span class="p">,</span> <span class="n">fapl</span><span class="o">=</span><span class="n">fapl</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">242</span> <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">&#39;r+&#39;</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">243</span>     <span class="n">fid</span> <span class="o">=</span> <span class="n">h5f</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">h5f</span><span class="o">.</span><span class="n">ACC_RDWR</span><span class="p">,</span> <span class="n">fapl</span><span class="o">=</span><span class="n">fapl</span><span class="p">)</span>

<span class="nn">File h5py/_objects.pyx:54,</span> in <span class="ni">h5py._objects.with_phil.wrapper</span><span class="nt">()</span>

<span class="nn">File h5py/_objects.pyx:55,</span> in <span class="ni">h5py._objects.with_phil.wrapper</span><span class="nt">()</span>

<span class="nn">File h5py/h5f.pyx:104,</span> in <span class="ni">h5py.h5f.open</span><span class="nt">()</span>

<span class="ne">FileNotFoundError</span>: [Errno 2] Unable to synchronously open file (unable to open file: name = &#39;example/output_classical_monoctypes.attributions.h5&#39;, errno = 2, error message = &#39;No such file or directory&#39;, flags = 0, o_flags = 0)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>head<span class="w">  </span>example/output_classical_monoctypes.seqlets.bed<span class="w"> </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-t<span class="w"> </span>-s<span class="w"> </span><span class="s1">$&#39;\t&#39;</span><span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chr11  47152015  47152022  neg.SPI1@-29953  3  .  -0.42814336717128754
chr11  47160163  47160167  neg.SPI1@-21805  3  .  -0.25199126452207565
chr11  47160309  47160319  neg.SPI1@-21659  3  .  -0.7467306312173605
chr11  47160323  47160333  neg.SPI1@-21645  3  .  -0.9774367886129767
chr11  47165053  47165060  neg.SPI1@-16915  3  .  -0.46786930691450834
chr11  47165593  47165606  pos.SPI1@-16375  3  .  1.7940459251403809
chr11  47165642  47165653  pos.SPI1@-16326  3  .  1.0624531591311097
chr11  47165653  47165664  neg.SPI1@-16315  3  .  -1.4263939633965492
chr11  47165664  47165670  neg.SPI1@-16304  3  .  -0.40090638771653175
chr11  47165690  47165703  pos.SPI1@-16278  3  .  1.1790284179151058
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>tail<span class="w"> </span>example/output_classical_monoctypes.motifs.tsv<span class="w"> </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-t<span class="w"> </span>-s<span class="w"> </span><span class="s1">$&#39;\t&#39;</span><span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>RARA.H13CORE.1.P.B        neg.BRD3@2651    166492  166511  -  7.690212368965149   0.000499333134939661   CTACCTTCCTCTACCCTCT      -0.00013111276999361028  -0.0005596191433799649  2652
RARA.H13CORE.1.P.B        neg.BRD3@32535   196372  196391  -  7.606964826583862   0.000499333134939661   TGAAGTTCCTCTCCCCTCC      -0.0005527154669260582   -0.0007040620484463215  32532
RARA.H13CORE.1.P.B        pos.BRD3@2657    166492  166511  -  7.690212368965149   0.000499333134939661   CTACCTTCCTCTACCCTCT      -0.00013111276999361028  -0.0005596191433799649  2652
SP1.H13CORE.0.P.B         pos.BRD3@46      163891  163905  -  4.692713022232056   0.000499375164508819   GCGCGCCGCCCCGC           -0.00021742627099488994  0.0003424161255357046   51
ANDR.H13CORE.2.P.B        neg.BRD3@2454    166291  166310  +  7.7568066120147705  0.0004996894640498803  TTTTTATTTTTGTTTTGTT      -0.0008813150420074204   -0.0014833488872296214  2451
ATOH1.H13CORE.0.P.B       pos.BRD3@6463    170307  170317  +  6.371594667434692   0.0004997253417968743  AGCAGTTGGC               0.00022125871619209646   -5.499589891702337e-05  6467
CGGBP1.H13CORE.0.PSGIB.A  neg.BRD3@614     164462  164473  -  9.182251572608948   0.000499725341796875   GGGGCGGCGGG              4.643183505818756e-05    0.0006390382217894364   622
ZN394.H13CORE.0.P.C       pos.BRD3@573     164409  164428  +  6.16656231880188    0.0004997442010790109  GCCGCCGGAGCCGCGAGGC      0.001674207092085118     0.003827750839971681    569
ZNF30.H13CORE.0.P.C       neg.BRD3@291     164117  164140  -  7.218540787696838   0.0004997881442108104  CGGGCGCCGAGCCCCGCCCCCGC  -0.0007112058181361215   -0.0011016258332200105  277
NR1H4.H13CORE.1.P.B       pos.BRD3@194003  357832  357850  -  7.282587766647339   0.0004999180964659896  CCTTGGAGGCAGTGACTC       0.0006504213249879993    0.0014391388014935464   193992
</pre></div>
</div>
</div>
</div>
<p>QC file (<code class="docutils literal notranslate"><span class="pre">qc.warnings.log</span></code>) is a quality control log file that contains warnings about prediction reliability for genes. Specifically, it warns when a gene has low correlation with the model’s predictions (Pearson correlation &lt; 0.7).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>head<span class="w"> </span>output_classical_monoctypes.warnings.qc.log
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>head: cannot open &#39;output_classical_monoctypes.warnings.qc.log&#39; for reading: No such file or directory
</pre></div>
</div>
</div>
</div>
<section id="cli-subcommands">
<h3>CLI Subcommands<a class="headerlink" href="#cli-subcommands" title="Link to this heading">¶</a></h3>
<p>The Decima CLI supports running the attribution analysis pipeline step by step using dedicated subcommands.
This modular approach allows you to execute each stage of the workflow independently, such as:</p>
<ol class="arabic simple">
<li><p>Generating model predictions for selected genes and cell types (<code class="docutils literal notranslate"><span class="pre">attributions-predict</span></code>).</p></li>
<li><p>Calling significant seqlets from the attributions (<code class="docutils literal notranslate"><span class="pre">attributions-recursive-seqlet-calling</span></code>).</p></li>
<li><p>Visualizing the results and motif logos (<code class="docutils literal notranslate"><span class="pre">attributions-plot</span></code>).
By chaining these subcommands, you can customize, debug, or parallelize each step of the analysis as needed.</p></li>
</ol>
<p>This cell demonstrates how to run the Decima CLI to generate attributions for selected genes and cell types. The following command runs the attributions-predict subcommand for model 0 and 1, focusing on the genes SPI1 and BRD3 in cells where the cell_type is ‘classical monocyte’. The results are saved with the specified output prefix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>attributions-predict<span class="w"> </span>--model<span class="w"> </span>v1_rep0<span class="w"> </span>--genes<span class="w"> </span><span class="s2">&quot;SPI1,BRD3&quot;</span><span class="w"> </span>--tasks<span class="w"> </span><span class="s2">&quot;cell_type == &#39;classical monocyte&#39;&quot;</span><span class="w"> </span>--output-prefix<span class="w"> </span>example/output_classical_monoctypes_0
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Using device: cuda
decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1899.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.8 (1769.0MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1315.5MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...:  50%|█████████         | 1/2 [00:02&lt;00:02,  2.14s/it]decima - WARNING - Gene BRD3 has low correlation with the model. Pearson: 0.3440624267844621. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...: 100%|██████████████████| 2/2 [00:03&lt;00:00,  1.84s/it]

</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>attributions-predict<span class="w"> </span>--model<span class="w"> </span>v1_rep1<span class="w"> </span>--genes<span class="w"> </span><span class="s2">&quot;SPI1,BRD3&quot;</span><span class="w"> </span>--tasks<span class="w"> </span><span class="s2">&quot;cell_type == &#39;classical monocyte&#39;&quot;</span><span class="w"> </span>--output-prefix<span class="w"> </span>example/output_classical_monoctypes_1
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Using device: cuda
decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1859.3MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1901.0MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep1:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1387.3MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...:  50%|█████████         | 1/2 [00:02&lt;00:02,  2.10s/it]decima - WARNING - Gene BRD3 has low correlation with the model. Pearson: 0.3440624267844621. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...: 100%|██████████████████| 2/2 [00:03&lt;00:00,  1.83s/it]

</pre></div>
</div>
</div>
</div>
<p>This cell runs the recursive seqlet calling step of the Decima attribution pipeline. It takes the attributions from two models (model 0 and model 1) for the genes SPI1 and BRD3in classical monocytes, and calls significant seqlets (regions with high attribution).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>attributions-recursive-seqlet-calling<span class="w"> </span>--attributions<span class="w"> </span><span class="s2">&quot;example/output_classical_monoctypes_0.attributions.h5,example/output_classical_monoctypes_1.attributions.h5&quot;</span><span class="w"> </span>--output-prefix<span class="w"> </span>example/output_classical_monoctypes
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1844.8MB/s)
decima - INFO - No genes provided, using all 2 genes in the attribution files.
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1874.6MB/s)
Computing recursive seqlet calling...: 100%|█████| 2/2 [00:00&lt;00:00, 846.31it/s]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.

</pre></div>
</div>
</div>
</div>
<p>The following cell runs the Decima CLI to generate plots for the attributions and discovered seqlets. It uses the output prefix from previous steps and generates visualizations for the specified genes (SPI1, BRD3), highlighting motif locations within 500bp of the transcription start site (TSS).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>attributions-plot<span class="w"> </span>--output-prefix<span class="w"> </span>example/output_classical_monoctypes<span class="w"> </span>-g<span class="w"> </span><span class="s2">&quot;SPI1,BRD3&quot;</span><span class="w"> </span>--tss-distance<span class="w"> </span><span class="m">500</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Plotting attributions...:   0%|                           | 0/2 [00:00&lt;?, ?it/s]<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1864.1MB/s)
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/plotnine/ggplot.py:615: PlotnineWarning: Saving 10 x 2 in image.
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/plotnine/ggplot.py:616: PlotnineWarning: Filename: example/output_classical_monoctypes_plots/SPI1.peaks.png
Plotting attributions...:  50%|█████████▌         | 1/2 [00:13&lt;00:13, 13.09s/it]<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1888.9MB/s)
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/plotnine/ggplot.py:615: PlotnineWarning: Saving 10 x 2 in image.
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/plotnine/ggplot.py:616: PlotnineWarning: Filename: example/output_classical_monoctypes_plots/BRD3.peaks.png
Plotting attributions...: 100%|███████████████████| 2/2 [00:24&lt;00:00, 12.45s/it]

</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">Image</span>

<span class="n">Image</span><span class="p">(</span><span class="s2">&quot;example/output_classical_monoctypes_plots/SPI1_seqlogos/SPI1@267.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/17eb7a775bf4fca5448afd3086cd42f91f041e2d5fea521462b11593760b58a1.png" src="../_images/17eb7a775bf4fca5448afd3086cd42f91f041e2d5fea521462b11593760b58a1.png" />
</div>
</div>
</section>
<section id="querying-cells">
<h3>Querying Cells<a class="headerlink" href="#querying-cells" title="Link to this heading">¶</a></h3>
<p>To obtain attributions, cells of interest must be selected using the query API. We support Pandas’ query API functionality on the cell metadata DataFrame. Here are examples of how to write queries:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>query-cell<span class="w"> </span>--help
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Usage: decima query-cell [OPTIONS] [QUERY]

  Query a cell using query string

  Examples:

      &gt;&gt;&gt; decima query-cell &#39;cell_type == &quot;classical monocyte&quot;&#39;     ...

      &gt;&gt;&gt; decima query-cell &#39;cell_type == &quot;classical monocyte&quot; and disease ==
      &quot;healthy&quot; and tissue == &quot;blood&quot;&#39;     ...

      &gt;&gt;&gt; decima query-cell &#39;cell_type.str.contains(&quot;monocyte&quot;) and disease ==
      &quot;healthy&quot;&#39;     ...

Options:
  --help  Show this message and exit.

</pre></div>
</div>
</div>
</div>
<p>Query cells of type “classical monocyte” using Pandas query syntax: <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>query-cell<span class="w"> </span><span class="s1">&#39;cell_type == &quot;classical monocyte&quot;&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-t<span class="w"> </span>-s<span class="w"> </span><span class="s1">$&#39;\t&#39;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1874.1MB/s)
          cell_type           tissue                            organ           disease                                         study                                 dataset      region  subregion  celltype_coarse  n_cells  total_counts        n_genes  size_factor         train_pearson       val_pearson         test_pearson
agg_4705  classical monocyte  alveolar system                   lung            COVID-19                                        GSE155249                             scimilarity  nan     nan                         7244     26544273.0          15325    34749.092791034054  0.946616874183219   0.8437000068912937  0.8506571540216992
agg_4706  classical monocyte  alveolar system                   lung            healthy                                         GSE155249                             scimilarity  nan     nan                         72       218105.0            9142     30484.31888978114   0.9102228263646758  0.8083487523192785  0.8047828694155461
agg_4707  classical monocyte  ampulla of uterine tube           fallopian tube  healthy                                         fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         78       550950.0            9639     30719.377971431015  0.9077670011915634  0.8045070167513724  0.7896845423359651
agg_4708  classical monocyte  aorta                             vasculature     Abdominal Aortic Aneurysm                       GSE166676                             scimilarity  nan     nan                         432      1091075.0           11192    32981.443348717905  0.9389265854768138  0.8357299205241656  0.830575965756882
agg_4709  classical monocyte  aorta                             vasculature     healthy                                         GSE166676                             scimilarity  nan     nan                         25       162858.0            8859     31216.275954364824  0.8819013257206973  0.7821403055329706  0.7646999711802146
agg_4710  classical monocyte  apex of heart                     heart           healthy                                         b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         397      1226515.0           12369    32022.563851814968  0.9469178617442242  0.8326145310572417  0.8365506153530168
agg_4711  classical monocyte  blood                             blood           COVID-19                                        03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         17462    78882609.0          15711    33080.17541357136   0.9536210517623883  0.8539379752673611  0.8485800714004562
agg_4712  classical monocyte  blood                             blood           COVID-19                                        7d7cabfd-1d1f-40af-96b7-26a0825a306d  scimilarity  nan     nan                         141914   659177004.0         15175    32282.29230923367   0.9575085257562032  0.8570056960758388  0.8507770895392532
agg_4713  classical monocyte  blood                             blood           COVID-19                                        GSE154567                             scimilarity  nan     nan                         8613     40239000.0          16023    34450.628692057784  0.9605287026438375  0.8551842525202262  0.8491670381176235
agg_4714  classical monocyte  blood                             blood           COVID-19                                        GSE158034                             scimilarity  nan     nan                         35       91390.0             7372     27446.425385592618  0.8476138307738649  0.7606096001026369  0.7256993661246048
agg_4715  classical monocyte  blood                             blood           COVID-19                                        GSE161918                             scimilarity  nan     nan                         163244   1023475761.0        15929    31151.84947891148   0.9361102289470363  0.8261601916328626  0.8168421752771801
agg_4716  classical monocyte  blood                             blood           COVID-19                                        GSE163668                             scimilarity  nan     nan                         8399     55036800.0          15792    33644.10626885235   0.9571638472088531  0.8529462339847145  0.8506441761860545
agg_4717  classical monocyte  blood                             blood           COVID-19                                        GSE166992                             scimilarity  nan     nan                         2238     12283507.0          14186    32596.636302802952  0.9567710210039843  0.8531485173368566  0.8416173697367906
agg_4718  classical monocyte  blood                             blood           COVID-19                                        ddfad306-714d-4cc0-9985-d9072820c530  scimilarity  nan     nan                         61002    230056884.0         16484    32520.14418628346   0.9487335053479237  0.8533686239486711  0.8454541123444707
agg_4719  classical monocyte  blood                             blood           COVID-19                                        eb735cc9-d0a7-48fa-b255-db726bf365af  scimilarity  nan     nan                         19777    105875381.0         15812    32330.088619084574  0.9558882745902155  0.8545238316898663  0.8468877639468763
agg_4720  classical monocyte  blood                             blood           HIV enteropathy                                 GSE157829                             scimilarity  nan     nan                         491      1449812.0           12290    33110.90004135926   0.9412108394642186  0.8352699509238034  0.8345507070277177
agg_4721  classical monocyte  blood                             blood           Myelofibrosis                                   GSE117824                             scimilarity  nan     nan                         357      1492491.0           11548    32726.985198452294  0.9446223529088382  0.8417521390049872  0.8328218073658378
agg_4722  classical monocyte  blood                             blood           NA                                              GSE132950                             scimilarity  nan     nan                         146      784054.0            10913    30417.15641845661   0.9276395863920666  0.8264978172767997  0.8176327551177259
agg_4723  classical monocyte  blood                             blood           NA                                              GSE135325                             scimilarity  nan     nan                         232      633533.0            11129    31159.105128910356  0.9369963391148282  0.8254811186623798  0.8207578599532835
agg_4724  classical monocyte  blood                             blood           NA                                              GSE150233                             scimilarity  nan     nan                         1141     2453545.0           12228    32204.245569759012  0.9354773292749718  0.8333534679658088  0.8202743631285762
agg_4725  classical monocyte  blood                             blood           NA                                              GSE151310                             scimilarity  nan     nan                         48       151358.0            8028     27001.118740317568  0.8873812787091045  0.7886356061906991  0.766461694552445
agg_4726  classical monocyte  blood                             blood           NA                                              GSE164378                             scimilarity  nan     nan                         54305    476237982.0         17463    34023.11682209347   0.9636663701487779  0.856267291847072   0.8496477594095655
agg_4727  classical monocyte  blood                             blood           NA                                              GSE164402                             scimilarity  nan     nan                         6577     33889420.0          14992    33855.14311643263   0.9502216042319906  0.846017695872854   0.8447747394204608
agg_4728  classical monocyte  blood                             blood           Sezary&#39;s disease                                GSE122703                             scimilarity  nan     nan                         35       148650.0            8487     29592.979037498706  0.8928094999389883  0.7911806688728295  0.7911936593448785
agg_4729  classical monocyte  blood                             blood           dengue disease                                  GSE145307                             scimilarity  nan     nan                         785      7639702.0           13722    33610.52078618725   0.9561427618691068  0.8544883780028308  0.8514781068765508
agg_4730  classical monocyte  blood                             blood           dengue disease                                  GSE154386                             scimilarity  nan     nan                         19173    143929741.0         16877    34242.50262506596   0.9586193824399223  0.8509705295166231  0.8546685528097621
agg_4731  classical monocyte  blood                             blood           drug hypersensitivity syndrome                  GSE132802                             scimilarity  nan     nan                         1269     7314697.0           13270    32574.34811388645   0.9570929839341253  0.8466339050741839  0.8442788242172967
agg_4732  classical monocyte  blood                             blood           fibrosis                                        GSE136103                             scimilarity  nan     nan                         1774     5003888.0           13389    31155.271000486402  0.9562933985421416  0.8435982250231042  0.8386367834560556
agg_4733  classical monocyte  blood                             blood           healthy                                         03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         32464    109280914.0         16158    33646.02843110038   0.9568728712803031  0.8545324533535094  0.8487445540580735
agg_4734  classical monocyte  blood                             blood           healthy                                         436154da-bcf1-4130-9c8b-120ff9a888f2  scimilarity  nan     nan                         76800    206490628.0         16683    30736.453324546856  0.955313650235467   0.8494127267867799  0.84210567312908
agg_4735  classical monocyte  blood                             blood           healthy                                         5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         1044     3638976.0           12306    30542.8977364245    0.9465384578921433  0.851991795617166   0.8308191715256182
agg_4736  classical monocyte  blood                             blood           healthy                                         DS000010023                           scimilarity  nan     nan                         243      362606.0            8414     25201.261024165953  0.8865446516098515  0.7621391376670988  0.7681818769796727
agg_4737  classical monocyte  blood                             blood           healthy                                         GSE122703                             scimilarity  nan     nan                         18       83417.0             7546     28194.725173315186  0.859225612396475   0.7640890699253299  0.7443681539461423
agg_4738  classical monocyte  blood                             blood           healthy                                         GSE130117                             scimilarity  nan     nan                         2017     7130588.0           13535    33078.53692191542   0.9553673450160365  0.851402109626239   0.8385936100758409
agg_4739  classical monocyte  blood                             blood           healthy                                         GSE132802                             scimilarity  nan     nan                         1601     9955248.0           13132    32063.630951743195  0.9478882791739611  0.8391025143866828  0.8303465877530952
agg_4740  classical monocyte  blood                             blood           healthy                                         GSE139324                             scimilarity  nan     nan                         2333     8331045.0           13985    31135.881287246768  0.9608208780142045  0.8473885992448625  0.8432790193723467
agg_4741  classical monocyte  blood                             blood           healthy                                         GSE145809                             scimilarity  nan     nan                         69       245221.0            8962     29135.67197629852   0.8825701041728526  0.7811799267734735  0.7818647625179129
agg_4742  classical monocyte  blood                             blood           healthy                                         GSE149313                             scimilarity  nan     nan                         2420     6974751.0           13143    29560.496854576566  0.9574598613513423  0.8505290963248237  0.8379199735887167
agg_4743  classical monocyte  blood                             blood           healthy                                         GSE153421                             scimilarity  nan     nan                         3691     15561725.0          14569    34377.465875728165  0.9636686704566925  0.8576434473562725  0.8511814190737197
agg_4744  classical monocyte  blood                             blood           healthy                                         GSE156989                             scimilarity  nan     nan                         13554    160011485.0         16915    34135.439844737564  0.9640667421350761  0.8577967800377495  0.8517975138366085
agg_4745  classical monocyte  blood                             blood           healthy                                         GSE157829                             scimilarity  nan     nan                         1619     6957811.0           13507    30199.39288988673   0.9484019976492215  0.8436979316400604  0.8347196616710685
agg_4746  classical monocyte  blood                             blood           healthy                                         GSE159113                             scimilarity  nan     nan                         1025     6298250.0           12083    27477.50809897617   0.9078020151513733  0.8121457150205226  0.7980372877810575
agg_4747  classical monocyte  blood                             blood           healthy                                         GSE161329                             scimilarity  nan     nan                         5654     25653579.0          14349    28848.0539929647    0.9549801428956252  0.8450430950674043  0.8406188789518544
agg_4748  classical monocyte  blood                             blood           healthy                                         GSE161738                             scimilarity  nan     nan                         2676     13801473.0          12825    33337.477050230416  0.9541962906717452  0.8512846409758499  0.8485408028961247
agg_4749  classical monocyte  blood                             blood           healthy                                         GSE163668                             scimilarity  nan     nan                         2644     10486314.0          14049    33786.96584264489   0.9597801578342394  0.8560775485935677  0.8512149509551471
agg_4750  classical monocyte  blood                             blood           healthy                                         GSE166992                             scimilarity  nan     nan                         7501     28033216.0          15079    33455.367364577316  0.9622273594219685  0.8558958139235102  0.8495571689751152
agg_4751  classical monocyte  blood                             blood           healthy                                         GSE167363                             scimilarity  nan     nan                         3135     14722635.0          14375    29977.24002819913   0.942417448875388   0.8368071803109702  0.8258536430202982
agg_4752  classical monocyte  blood                             blood           healthy                                         GSE168710                             scimilarity  nan     nan                         16484    104881872.0         16223    34107.336261357574  0.9398282119039322  0.8424821834537695  0.8372971004604842
agg_4753  classical monocyte  blood                             blood           healthy                                         GSE168732                             scimilarity  nan     nan                         770      2548822.0           12508    33411.30103713399   0.9552513581030765  0.8508279875038706  0.847461536110767
agg_4754  classical monocyte  blood                             blood           healthy                                         b0cf0afa-ec40-4d65-b570-ed4ceacc6813  scimilarity  nan     nan                         40975    300555227.0         15784    35938.85772500803   0.9622425892039956  0.853424173800979   0.8508714303589978
agg_4755  classical monocyte  blood                             blood           healthy                                         ddfad306-714d-4cc0-9985-d9072820c530  scimilarity  nan     nan                         8827     36073928.0          15131    33208.591584008376  0.9546118779961532  0.8543086616569785  0.8462739374830107
agg_4756  classical monocyte  blood                             blood           intracranial hypotension                        GSE138266                             scimilarity  nan     nan                         2503     9675804.0           14485    30160.767605621222  0.9452052724479383  0.8423537848756032  0.8326629487875993
agg_4757  classical monocyte  blood                             blood           mucocutaneous lymph node syndrome               GSE168732                             scimilarity  nan     nan                         5745     25930751.0          14822    33366.18751424575   0.9564515409367231  0.8556431530577528  0.8540185868162636
agg_4758  classical monocyte  blood                             blood           multiple sclerosis                              GSE138266                             scimilarity  nan     nan                         3988     13926825.0          14991    31442.03464388843   0.9522779953120408  0.847799219646348   0.8382058078578654
agg_4759  classical monocyte  blood                             blood           non-alcoholic fatty liver disease               GSE136103                             scimilarity  nan     nan                         8306     29424841.0          15410    32004.200489375227  0.9619190264492873  0.8478709124980346  0.8451242436776344
agg_4760  classical monocyte  blood                             blood           rheumatoid arthritis                            GSE159117                             scimilarity  nan     nan                         834      4637566.0           12079    31364.847230552205  0.9356058277176598  0.8232134520999813  0.8176333921128414
agg_4761  classical monocyte  blood                             blood           septic shock                                    GSE167363                             scimilarity  nan     nan                         3860     51041813.0          15830    31688.79561595612   0.948652055959824   0.8541736693569211  0.8427375237296424
agg_4762  classical monocyte  blood                             blood           systemic lupus erythematosus                    436154da-bcf1-4130-9c8b-120ff9a888f2  scimilarity  nan     nan                         200468   516575809.0         16896    30011.373010792136  0.9562030923644677  0.8480520393236465  0.844052374052952
agg_4763  classical monocyte  blood                             blood           systemic lupus erythematosus                    GSE142016                             scimilarity  nan     nan                         8268     22146620.0          14873    30889.72528098081   0.9588937962174496  0.8480448150326806  0.8395981302528143
agg_4764  classical monocyte  blood                             blood           systemic lupus erythematosus                    GSE153765                             scimilarity  nan     nan                         42       109982.0            7500     27335.812367044335  0.8566470004710053  0.7665719714665945  0.7374607536624445
agg_4765  classical monocyte  blood                             blood           systemic lupus erythematosus                    GSE156989                             scimilarity  nan     nan                         30367    310637290.0         17082    33485.308563356346  0.9623402060903008  0.8532487075078466  0.8473526649094757
agg_4766  classical monocyte  blood                             blood           thrombocytopenia                                GSE149313                             scimilarity  nan     nan                         2724     15059814.0          14328    30599.80301260898   0.9543473550386421  0.8520722945129096  0.8417995728182829
agg_4767  classical monocyte  bone                              bone            Langerhans Cell Histiocytosis                   GSE133704                             scimilarity  nan     nan                         439      1404680.0           11388    30817.807833507268  0.9358504157466769  0.830348562008033   0.826566269904344
agg_4769  classical monocyte  bone marrow                       bone marrow     NA                                              GSE162692                             scimilarity  nan     nan                         1234     4757721.0           13466    31707.380952189662  0.953852620063789   0.8503857428588029  0.8377674131460707
agg_4770  classical monocyte  bone marrow                       bone marrow     essential thrombocythemia                       GSE117824                             scimilarity  nan     nan                         1649     7825780.0           13487    32454.468003620656  0.9503614540027875  0.8479601234457582  0.838408163377457
agg_4772  classical monocyte  bone marrow                       bone marrow     healthy                                         GSE132509                             scimilarity  nan     nan                         610      2315570.0           12950    31768.06513427212   0.95159369508558    0.8517118261701931  0.836658919433696
agg_4773  classical monocyte  bone marrow                       bone marrow     healthy                                         GSE154109                             scimilarity  nan     nan                         531      1431388.0           11793    31377.450948003392  0.9490546933955852  0.8431566630120637  0.8370883160295727
agg_4774  classical monocyte  bone marrow                       bone marrow     healthy                                         GSE163278                             scimilarity  nan     nan                         1119     3970394.0           13361    32081.93302956569   0.9620394897163868  0.8531148861215617  0.8426785397396367
agg_4775  classical monocyte  bone marrow                       bone marrow     healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         151      8025584.0           12883    29444.78352863075   0.8440608178227607  0.740574910750325   0.7417328577454956
agg_4776  classical monocyte  bone marrow                       bone marrow     monoclonal gammopathy                           GSE163278                             scimilarity  nan     nan                         1010     3124102.0           12959    30344.72094719757   0.9581906874137958  0.8503948562041261  0.8391672246132192
agg_4777  classical monocyte  breast                            breast          healthy                                         GSE164898                             scimilarity  nan     nan                         136      641471.0            12971    34463.52724138501   0.9163324788498406  0.8116274576633968  0.7978555908123931
agg_4778  classical monocyte  breast                            breast          healthy                                         c9706a92-0e5f-46c1-96d8-20e42467f287  scimilarity  nan     nan                         98       1444245.0           13491    30678.263421880285  0.9165520953567395  0.8162053142576849  0.7994301225229256
agg_4779  classical monocyte  bronchus                          airway          COVID-19                                        03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         104      270108.0            8816     27933.501359354133  0.8884582873892427  0.7928474093439662  0.786626157931558
agg_4780  classical monocyte  bronchus                          airway          COVID-19                                        GSE168215                             scimilarity  nan     nan                         90       217928.0            8444     27417.845704249117  0.880738805029126   0.7855823736726739  0.7798557380738542
agg_4782  classical monocyte  bronchus                          airway          healthy                                         GSE158127                             scimilarity  nan     nan                         158      1158198.0           12643    34764.50196701077   0.9364512338084163  0.8259291909369686  0.8266638555276521
agg_4783  classical monocyte  cardiac muscle of left ventricle  heart           healthy                                         GSE156703                             scimilarity  nan     nan                         13       116181.0            9463     35695.66320276271   0.8542740960069863  0.7515621053395214  0.7561639038477878
agg_4784  classical monocyte  carotid artery segment            vasculature     atherosclerosis                                 GSE155512                             scimilarity  nan     nan                         58       515211.0            10839    32837.84505237503   0.9343565353773022  0.8190650931322585  0.8202426969221358
agg_4785  classical monocyte  caudate lobe of liver             liver           healthy                                         44531dd9-1388-4416-a117-af0a99de2294  scimilarity  nan     nan                         238      730016.0            11505    31342.386314731422  0.9217674983890346  0.8140551552218395  0.8040417787954989
agg_4786  classical monocyte  cortex of kidney                  kidney          healthy                                         120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         79       323010.0            10939    32378.76683324232   0.9028856137251035  0.7978822439778066  0.7839454035009307
agg_4787  classical monocyte  cortex of kidney                  kidney          healthy                                         a98b828a-622a-483a-80e0-15703678befd  scimilarity  nan     nan                         91       477355.0            10898    32358.068865763344  0.9328436291917394  0.8237810319569842  0.8195391931798526
agg_4789  classical monocyte  digestive tract                   gut             healthy                                         DS000011665                           scimilarity  nan     nan                         347      1679116.0           12155    33347.55517047197   0.9422556928648441  0.8417267634096297  0.84018452536733
agg_4790  classical monocyte  exocrine pancreas                 pancreas        healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         821      7824069.0           14847    36135.64587109593   0.9493709998172055  0.8440837716099078  0.8410246939313819
agg_4791  classical monocyte  fallopian tube                    fallopian tube  healthy                                         fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         131      734093.0            11240    33115.640434504094  0.9359103734457376  0.8339026306142181  0.8225901799509813
agg_4792  classical monocyte  fimbria of uterine tube           fallopian tube  healthy                                         fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         34       209362.0            7663     27635.733860508328  0.8560416684135254  0.7382997749370328  0.7459235366949488
agg_4794  classical monocyte  gingiva                           mouth           periodontitis                                   GSE152042                             scimilarity  nan     nan                         198      879477.0            11312    32107.813302914532  0.9416262876541264  0.8333723697695014  0.8279530215775117
agg_4795  classical monocyte  head of femur                     bone            healthy                                         GSE169396                             scimilarity  nan     nan                         450      3669304.0           13216    33082.323604222154  0.9529417082022753  0.8522346343107771  0.8359032081996703
agg_4797  classical monocyte  heart left ventricle              heart           NA                                              ENCODE                                scimilarity  nan     nan                         50       138407.30523254164  11428    41105.63651890687   0.8614790128015358  0.7843765107548858  0.7889308929671582
agg_4798  classical monocyte  heart left ventricle              heart           healthy                                         b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         192      585001.0            11159    31422.874036870588  0.9363985001217598  0.8226438123601741  0.8283173244446851
agg_4799  classical monocyte  heart right ventricle             heart           healthy                                         b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         316      936263.0            11904    32002.67227624691   0.9425900348990802  0.8306128730459813  0.8308116461021977
agg_4800  classical monocyte  ileum                             gut             Crohn&#39;s disease                                 17481d16-ee44-49e5-bcf0-28c0780d8c4a  scimilarity  nan     nan                         76       311515.0            9984     29687.611679190355  0.9103310804624617  0.8063080284939284  0.7916226068478351
agg_4801  classical monocyte  ileum                             gut             Crohn&#39;s disease                                 DS000011665                           scimilarity  nan     nan                         119      298206.0            8021     26013.286557459236  0.880272438867354   0.7572099232100128  0.7459713937143965
agg_4802  classical monocyte  inferior nasal concha             bone            chronic rhinosinusitis with nasal polyps        GSE156285                             scimilarity  nan     nan                         241      1048981.0           12463    35082.083353928334  0.9475981193848912  0.8330375982148592  0.833114773003936
agg_4803  classical monocyte  interventricular septum           heart           healthy                                         b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         442      1322725.0           12418    32235.5434681197    0.94751399102473    0.8340623939411483  0.8353858365852226
agg_4804  classical monocyte  isthmus of fallopian tube         fallopian tube  healthy                                         fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         62       318330.0            8642     29768.198512126502  0.8846027131590791  0.784220371073871   0.7739277341448829
agg_4805  classical monocyte  kidney                            kidney          NA                                              GSE145927                             scimilarity  nan     nan                         1789     6957949.0           15145    34853.400215205314  0.9586386429540185  0.8579622487738222  0.8524123833700197
agg_4806  classical monocyte  kidney                            kidney          acute kidney failure                            bcb61471-2a44-4d00-a0af-ff085512674c  scimilarity  nan     nan                         587      1589224.0           12335    32471.78147434854   0.9513927341618255  0.84278469020191    0.8402011848101985
agg_4807  classical monocyte  kidney                            kidney          chronic kidney disease                          bcb61471-2a44-4d00-a0af-ff085512674c  scimilarity  nan     nan                         134      440788.0            10831    32410.84600407974   0.9323603662663799  0.8153954443953603  0.8190416636069936
agg_4808  classical monocyte  kidney                            kidney          healthy                                         120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         762      4034828.0           14946    34015.00816823295   0.9520640694425091  0.848205266473767   0.836085027208869
agg_4809  classical monocyte  kidney                            kidney          healthy                                         DS000010415                           scimilarity  nan     nan                         55       127079.0            8055     27206.419037355434  0.8216238135756493  0.7570847030479543  0.72524174726152
agg_4810  classical monocyte  kidney                            kidney          healthy                                         GSE140989                             scimilarity  nan     nan                         174      563438.0            11016    29887.593299155575  0.914390252807459   0.8069762795735104  0.8086759072557722
agg_4811  classical monocyte  left cardiac atrium               heart           NA                                              ENCODE                                scimilarity  nan     nan                         59       225070.96128814947  12831    43727.214042795575  0.8938217272345973  0.8048974803645641  0.8168195119621123
agg_4812  classical monocyte  left cardiac atrium               heart           healthy                                         b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         450      1446734.0           12669    32433.20393285361   0.9467532324494885  0.8378571513042986  0.8360845068688169
agg_4813  classical monocyte  left lung                         lung            NA                                              ENCODE                                scimilarity  nan     nan                         16       40636.89964582212   7786     32627.503591080356  0.7956982592153874  0.7257891637293196  0.6945729831399207
agg_4814  classical monocyte  liver                             liver           Alagille syndrome                               GSE163650                             scimilarity  nan     nan                         92       188745.0            7678     24676.271768005652  0.8778490663838523  0.7727787177978541  0.7475297818459176
agg_4815  classical monocyte  liver                             liver           Biliary atresia                                 GSE163650                             scimilarity  nan     nan                         367      1615410.0           11423    31280.031237098276  0.9308940943983487  0.8275363405345841  0.8098394949074608
agg_4816  classical monocyte  liver                             liver           fibrosis                                        GSE136103                             scimilarity  nan     nan                         1053     5824229.0           14559    31950.801484792315  0.9632028296773183  0.8486875622591808  0.8450295232730788
agg_4817  classical monocyte  liver                             liver           healthy                                         GSE136103                             scimilarity  nan     nan                         2036     7668787.0           14818    32446.71686472893   0.9614468191909097  0.8455504517344443  0.8464065028454753
agg_4818  classical monocyte  liver                             liver           healthy                                         GSE159977                             scimilarity  nan     nan                         584      4703990.0           13644    34306.87421529999   0.9580119653448788  0.8463006395755477  0.8456137609888349
agg_4819  classical monocyte  liver                             liver           healthy                                         GSE163650                             scimilarity  nan     nan                         440      4840312.0           12272    31180.407161439263  0.9312379663603724  0.8198899526213368  0.8071334502842269
agg_4820  classical monocyte  liver                             liver           non-alcoholic fatty liver disease               GSE136103                             scimilarity  nan     nan                         675      3625081.0           13858    31875.772311414476  0.9607644852684971  0.8451153856892568  0.8434410362324897
agg_4821  classical monocyte  liver                             liver           non-alcoholic steatohepatitis                   GSE159977                             scimilarity  nan     nan                         818      5328417.0           13712    34244.79288663241   0.9625204736588413  0.8495046961098656  0.843851264103893
agg_4822  classical monocyte  lower lobe of left lung           lung            NA                                              ENCODE                                scimilarity  nan     nan                         119      332235.9213328175   13992    45607.18635453728   0.9075609223976155  0.8257092626936027  0.8254709531577699
agg_4823  classical monocyte  lower lobe of lung                lung            healthy                                         GSE169471                             scimilarity  nan     nan                         305      1224338.0           11342    28255.985922767635  0.9404343350984603  0.8261785132237449  0.8150341534919611
agg_4824  classical monocyte  lung                              lung            COVID-19                                        GSE145926                             scimilarity  nan     nan                         6755     29326462.0          15670    32143.238185602037  0.9347325303698273  0.8382075353537076  0.8350131792189084
agg_4825  classical monocyte  lung                              lung            COVID-19                                        GSE149878                             scimilarity  nan     nan                         1388     17477477.0          15453    32118.37391645824   0.9547396360778304  0.8488321261303191  0.8334226404565429
agg_4826  classical monocyte  lung                              lung            COVID-19                                        covid                                 scimilarity  nan     nan                         87       182436.0            8979     30922.571470240666  0.8944321462094855  0.7906545756360502  0.7941388516891991
agg_4827  classical monocyte  lung                              lung            Idiopathic pulmonary arterial hypertension      GSE169471                             scimilarity  nan     nan                         338      1099281.0           11441    29048.65706098467   0.9394268881459132  0.8205211245968264  0.8008915371926529
agg_4828  classical monocyte  lung                              lung            NA                                              GSE122960                             scimilarity  nan     nan                         2035     4747594.0           13592    31170.75054081323   0.947029489389696   0.8239671657460403  0.8229589179837169
agg_4829  classical monocyte  lung                              lung            NA                                              GSE150708                             scimilarity  nan     nan                         1711     18922764.0          15768    34651.58457127426   0.9197817700449655  0.8258478966731367  0.8332402246613281
agg_4830  classical monocyte  lung                              lung            NA                                              GSE159354                             scimilarity  nan     nan                         804      1319717.0           12267    30466.67987009986   0.9289888078126158  0.8179900347657229  0.8046761734394422
agg_4831  classical monocyte  lung                              lung            chronic obstructive pulmonary disease           DS000011735                           scimilarity  nan     nan                         1757     5736362.0           16385    37750.32029026267   0.8922650938609307  0.8174310943962452  0.7970204631208431
agg_4832  classical monocyte  lung                              lung            healthy                                         5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         1254     5397217.0           13298    31013.01576531177   0.9490555294982109  0.8457060773457411  0.8329330004620054
agg_4833  classical monocyte  lung                              lung            healthy                                         DS000011735                           scimilarity  nan     nan                         4653     16523051.0          17066    37593.98867222708   0.8985708908575675  0.8260014412964971  0.8051199142820423
agg_4834  classical monocyte  lung                              lung            healthy                                         GSE128033                             scimilarity  nan     nan                         1047     3646581.0           13185    29331.2035341724    0.9513045959002788  0.837770557527153   0.8238987539695043
agg_4835  classical monocyte  lung                              lung            healthy                                         GSE128169                             scimilarity  nan     nan                         1732     11798577.0          15051    32941.755873862814  0.9636814596401634  0.8539834251349044  0.8457626672394015
agg_4836  classical monocyte  lung                              lung            healthy                                         GSE132771                             scimilarity  nan     nan                         1601     4614408.0           13275    29761.03818692572   0.9531291409651284  0.8457644131717956  0.8310980214000753
agg_4837  classical monocyte  lung                              lung            healthy                                         GSE169471                             scimilarity  nan     nan                         498      1613976.0           11886    28956.213107123967  0.9433400854993297  0.8271067022019086  0.815595680192867
agg_4838  classical monocyte  lung                              lung            hypersensitivity pneumonitis                    GSE122960                             scimilarity  nan     nan                         374      1513589.0           11850    30594.494180377842  0.9436379625667726  0.8236248274374875  0.8201281668004226
agg_4839  classical monocyte  lung                              lung            idiopathic pulmonary fibrosis                   DS000011735                           scimilarity  nan     nan                         3273     11098539.0          16692    36983.80245044498   0.9002060376489591  0.825628591643822   0.8057309447193657
agg_4840  classical monocyte  lung                              lung            idiopathic pulmonary fibrosis                   GSE122960                             scimilarity  nan     nan                         795      2302741.0           12763    31758.949309599942  0.9481965424789537  0.8315974088368188  0.8291678346888611
agg_4841  classical monocyte  lung                              lung            idiopathic pulmonary fibrosis                   GSE128033                             scimilarity  nan     nan                         264      892053.0            10997    28549.410927787198  0.9388857876621541  0.8259212982973368  0.8088642639807162
agg_4842  classical monocyte  lung                              lung            idiopathic pulmonary fibrosis                   GSE132771                             scimilarity  nan     nan                         562      1301612.0           12354    30446.385456748263  0.9469963680992495  0.8353213795556867  0.820474733820443
agg_4844  classical monocyte  lung                              lung            idiopathic pulmonary fibrosis                   GSE143706                             scimilarity  nan     nan                         28       77859.0             5999     21933.076720558005  0.8162248508692449  0.6872290766118224  0.6675933221673609
agg_4845  classical monocyte  lung                              lung            idiopathic pulmonary fibrosis                   GSE146981                             scimilarity  nan     nan                         28       77859.0             5999     21933.076720558005  0.8151661654118738  0.6848887765520302  0.667713152235204
agg_4846  classical monocyte  lung                              lung            idiopathic pulmonary fibrosis                   GSE159354                             scimilarity  nan     nan                         963      1825354.0           12518    29731.366588446697  0.9431677835768482  0.8335606700375219  0.8147956336542009
agg_4847  classical monocyte  lung                              lung            interstitial lung disease                       GSE122960                             scimilarity  nan     nan                         255      622277.0            10480    29149.028142322823  0.9283467350849584  0.8054016349869099  0.8007862785744999
agg_4848  classical monocyte  lung                              lung            interstitial lung disease                       GSE128169                             scimilarity  nan     nan                         697      1972432.0           12243    29254.839846468705  0.9423878335093786  0.8300974626358707  0.8196994274197087
agg_4849  classical monocyte  lung                              lung            scleroderma                                     GSE128169                             scimilarity  nan     nan                         108      906362.0            11850    32908.696557354284  0.9438117150692724  0.8371885386238901  0.8225520820106291
agg_4850  classical monocyte  lung                              lung            scleroderma                                     GSE132771                             scimilarity  nan     nan                         98       335776.0            9515     28212.056049440183  0.9149834322044056  0.8124364794406282  0.7889831369493125
agg_4851  classical monocyte  lung                              lung            systemic scleroderma;interstitial lung disease  GSE159354                             scimilarity  nan     nan                         680      1244200.0           11364    27669.832360293723  0.9311681193836778  0.8218066897336025  0.801632654842017
agg_4852  classical monocyte  lung parenchyma                   lung            COVID-19                                        GSE158127                             scimilarity  nan     nan                         1028     2949423.0           13468    33486.58561312063   0.9573331305866764  0.8476148647425674  0.8402420746316328
agg_4853  classical monocyte  lung parenchyma                   lung            healthy                                         GSE158127                             scimilarity  nan     nan                         791      2735456.0           13260    33646.87553319058   0.9544351950657847  0.8399981618864122  0.8327646420552953
agg_4854  classical monocyte  lymph node                        lymph node      Langerhans Cell Histiocytosis                   GSE133704                             scimilarity  nan     nan                         41       112531.0            7250     25404.282603262254  0.8424914182716917  0.7490029862443883  0.7315420690462492
agg_4855  classical monocyte  mesenteric artery                 vasculature     healthy                                         GSE156341                             scimilarity  nan     nan                         49       408553.0            10083    30979.99239432764   0.9337200372314851  0.8169964046619824  0.8163109060481625
agg_4856  classical monocyte  mesenteric artery                 vasculature     type II diabetes mellitus                       GSE156341                             scimilarity  nan     nan                         107      869426.0            11343    33124.102127533     0.9473048055491107  0.8341783608252021  0.8308049817701406
agg_4857  classical monocyte  mesenteric lymph node             lymph node      healthy                                         7681c7d7-0168-4892-a547-6f02a6430ace  scimilarity  nan     nan                         23       211416.0            9219     31018.02041830142   0.9058403141644794  0.7914556298280883  0.7867496166249129
agg_4858  classical monocyte  muscle tissue                     muscle          healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         1800     50754027.0          16141    33323.26969602708   0.9316672745867468  0.8348295016474899  0.8219918807835366
agg_4859  classical monocyte  nasal cavity                      airway          COVID-19                                        03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         2907     12130592.0          15268    34541.15937505348   0.9398365208964624  0.8399826209015534  0.8406896100498221
agg_4860  classical monocyte  nasal cavity                      airway          healthy                                         03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         129      319231.0            10961    33948.695520725145  0.908031594986554   0.802885512146251   0.7998873890770253
agg_4861  classical monocyte  nasopharynx                       airway          nasopharyngeal neoplasm                         GSE150825                             scimilarity  nan     nan                         248      919710.0            11652    32725.826204558824  0.9483756316738621  0.8416477499125612  0.8434913191320811
agg_4862  classical monocyte  nose                              airway          chronic rhinosinusitis with nasal polyps        GSE156285                             scimilarity  nan     nan                         89       407982.0            10658    32874.53627471163   0.9356986809715955  0.8282644612568757  0.8193450383260489
agg_4863  classical monocyte  olfactory epithelium              airway          NA                                              GSE139522                             scimilarity  nan     nan                         152      645745.0            11496    32760.770927681508  0.9344670047519139  0.8326335387583638  0.8215291009295241
agg_4864  classical monocyte  omental fat pad                   peritoneum      obesity                                         GSE163830                             scimilarity  nan     nan                         248      603440.0            11376    33014.67391159742   0.9276969765366693  0.8135481953870003  0.8105166731324462
agg_4865  classical monocyte  omentum                           peritoneum      NA                                              GSE151889                             scimilarity  nan     nan                         106      233037.0            9833     30451.216970905818  0.9023265636700606  0.794410906944691   0.7787417127899898
agg_4868  classical monocyte  peritoneum                        peritoneum      NA                                              GSE130888                             scimilarity  nan     nan                         20547    75515682.0          16606    31611.601467579523  0.9640573553126023  0.8513803623213743  0.8467189186177884
agg_4869  classical monocyte  peritoneum                        peritoneum      healthy                                         GSE130888                             scimilarity  nan     nan                         297      509237.0            11456    32213.169493243313  0.9218188146119148  0.8055575625334659  0.8039145027939639
agg_4870  classical monocyte  prostate gland                    prostate        healthy                                         GSE145843                             scimilarity  nan     nan                         24       87555.0             5997     21432.99958526179   0.816337811891654   0.7246769991610698  0.6943045232082502
agg_4871  classical monocyte  prostate gland                    prostate        healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         220      2445216.0           12460    33911.259081133416  0.9301914834403244  0.8259560877683964  0.824440772955729
agg_4872  classical monocyte  renal medulla                     kidney          healthy                                         120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         21       101089.0            7354     26191.96315382653   0.8491245141516279  0.751292403843078   0.7310820608427115
agg_4873  classical monocyte  respiratory airway                airway          COVID-19                                        29f92179-ca10-4309-a32b-d383d80347c1  scimilarity  nan     nan                         24222    187246624.0         17810    38621.12130270373   0.911673853186318   0.8025805020768422  0.8054824859649656
agg_4874  classical monocyte  respiratory tract epithelium      airway          NA                                              GSE139522                             scimilarity  nan     nan                         69       371203.0            11152    33714.18718588416   0.9189481336748234  0.8044925205508522  0.8036843863420214
agg_4875  classical monocyte  right cardiac atrium              heart           healthy                                         b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         311      977934.0            11891    31570.89048372305   0.9357294628028995  0.8260631626226992  0.8224053236502149
agg_4876  classical monocyte  sigmoid colon                     gut             ulcerative colitis                              DS000010618                           scimilarity  nan     nan                         56       157830.0            7772     25795.237651990203  0.8795503731261843  0.7579451181338609  0.7495331251552053
agg_4877  classical monocyte  spleen                            spleen          HIV infection                                   GSE148796                             scimilarity  nan     nan                         48       118392.0            7120     25206.082214526155  0.8589535988735265  0.7723626781751426  0.7544543625246387
agg_4878  classical monocyte  spleen                            spleen          healthy                                         4d74781b-8186-4c9a-b659-ff4dc4601d91  scimilarity  nan     nan                         2166     7905128.0           13952    30832.98149016589   0.957626764427953   0.8498277489734691  0.8368540073560422
agg_4879  classical monocyte  spleen                            spleen          healthy                                         GSE148796                             scimilarity  nan     nan                         49       99684.0             6785     24492.260723886982  0.8477188947186437  0.7504685014947085  0.7295165069095441
agg_4880  classical monocyte  spleen                            spleen          healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         3483     86078540.0          17775    36266.36759486      0.9100465322095738  0.8097019968114763  0.8081945289287479
agg_4881  classical monocyte  subcutaneous adipose tissue       adipose         healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         2019     27914261.0          15739    35142.992419516115  0.9421666034710219  0.8357406603138717  0.8249091266671545
agg_4882  classical monocyte  synovial fluid                    synovial joint  juvenile idiopathic arthritis                   GSE160097                             scimilarity  nan     nan                         68       430194.0            9771     30748.449183442222  0.9095023395695137  0.8036624580038914  0.8039477211085259
agg_4883  classical monocyte  synovial fluid                    synovial joint  psoriatic arthritis                             GSE161500                             scimilarity  nan     nan                         675      4107697.0           12980    33508.9642503863    0.9516464026609945  0.8477381221507095  0.8460202124356817
agg_4884  classical monocyte  tertiary ovarian follicle         ovary           NA                                              GSE146512                             scimilarity  nan     nan                         100      296748.0            10411    33084.00694991347   0.9139378500397654  0.8064064097295065  0.8023402394768804
agg_4885  classical monocyte  testis                            testis          NA                                              GSE153819                             scimilarity  nan     nan                         17       97625.0             7958     29072.297180853668  0.855345538109863   0.764463114789657   0.7520891122089112
agg_4886  classical monocyte  thoracic lymph node               lymph node      healthy                                         62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         20194    147502356.0         17156    34278.43204322054   0.9584542376154844  0.8568164088302802  0.8468420651789338
agg_4887  classical monocyte  thymus                            thymus          healthy                                         62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         487      2692133.0           12969    33549.45213075861   0.9497232200872254  0.8549259958290932  0.843613914494647
agg_4888  classical monocyte  thymus                            thymus          healthy                                         83ed3be8-4cb9-43e6-9aaa-3fbbf5d1bd3a  scimilarity  nan     nan                         27       80042.0             6527     23800.833698344715  0.8448423801602987  0.7347769941739797  0.7181319535525903
agg_4889  classical monocyte  thymus                            thymus          healthy                                         de13e3e2-23b6-40ed-a413-e9e12d7d3910  scimilarity  nan     nan                         52       298983.0            9582     30002.684399867492  0.9175531872094376  0.8185046437128072  0.8176606503789746
agg_4890  classical monocyte  tonsil                            tonsil          healthy                                         GSE119506                             scimilarity  nan     nan                         321      1114339.0           11546    30516.473940957327  0.936749794981687   0.8307356996245999  0.8215841775914924
agg_4893  classical monocyte  trachea                           airway          healthy                                         03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         126      457580.0            11135    33734.327765812595  0.9299687996532033  0.8279876048002456  0.8230768581762299
agg_4894  classical monocyte  trachea                           airway          healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         130      8245622.0           13663    32584.555010387805  0.859406194907081   0.7522520630806427  0.7550758397192211
agg_4895  classical monocyte  transition zone of prostate       prostate        prostatic hypertrophy                           4b54248f-2165-477c-a027-dd55082e8818  scimilarity  nan     nan                         520      2949099.0           13618    29095.077520604846  0.9205242051339535  0.807648434373933   0.7862293702463532
agg_4896  classical monocyte  transverse colon                  gut             healthy                                         62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         503      2932999.0           12868    33265.22762476361   0.9476555210791636  0.8481166846094836  0.8392740144102075
agg_4897  classical monocyte  tympanic membrane                 ear             NA                                              GSE128892                             scimilarity  nan     nan                         33       153723.0            7971     26217.848855873086  0.8515891731868822  0.7526405790618411  0.7328434865867175
agg_4899  classical monocyte  upper lobe of lung                lung            healthy                                         GSE169471                             scimilarity  nan     nan                         180      594059.0            10222    27881.011904781462  0.9248153838159024  0.8106326147324702  0.8003438801026445
agg_4900  classical monocyte  urine                             urinary         healthy                                         GSE165396                             scimilarity  nan     nan                         20       109197.0            7299     25505.26771828214   0.8530560359297166  0.7505269381795711  0.7350625083451162
agg_4901  classical monocyte  uterus                            uterus          healthy                                         32f2fd23-ec74-486f-9544-e5b2f41725f5  scimilarity  nan     nan                         18       189472.0            9397     30891.07905591756   0.8870309576225303  0.7796319609750195  0.7827462282311286
agg_4902  classical monocyte  vasculature                       vasculature     healthy                                         e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         14537    224206261.0         17922    36651.90273209977   0.9438232475974557  0.8425086108952623  0.8343079266995497
agg_4903  classical monocyte  visceral fat                      adipose         obesity                                         GSE128518                             scimilarity  nan     nan                         74       196657.0            9100     28836.082756573305  0.8890453654508667  0.7810993233599736  0.7788838086708725
</pre></div>
</div>
</div>
</div>
<p>Query cells that:</p>
<ul class="simple">
<li><p>have “monocyte” in their cell type name (cell_type.str.contains(“monocyte”))</p></li>
<li><p>are from healthy donors (disease == “healthy”)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>query-cell<span class="w"> </span><span class="s1">&#39;cell_type.str.contains(&quot;monocyte&quot;) and disease == &quot;healthy&quot;&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-t<span class="w"> </span>-s<span class="w"> </span><span class="s1">$&#39;\t&#39;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1835.5MB/s)
          cell_type               tissue                            organ           disease  study                                 dataset      region  subregion  celltype_coarse  n_cells  total_counts  n_genes  size_factor         train_pearson       val_pearson         test_pearson
agg_4706  classical monocyte      alveolar system                   lung            healthy  GSE155249                             scimilarity  nan     nan                         72       218105.0      9142     30484.31888978114   0.9102228263646758  0.8083487523192785  0.8047828694155461
agg_4707  classical monocyte      ampulla of uterine tube           fallopian tube  healthy  fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         78       550950.0      9639     30719.377971431015  0.9077670011915634  0.8045070167513724  0.7896845423359651
agg_4709  classical monocyte      aorta                             vasculature     healthy  GSE166676                             scimilarity  nan     nan                         25       162858.0      8859     31216.275954364824  0.8819013257206973  0.7821403055329706  0.7646999711802146
agg_4710  classical monocyte      apex of heart                     heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         397      1226515.0     12369    32022.563851814968  0.9469178617442242  0.8326145310572417  0.8365506153530168
agg_4733  classical monocyte      blood                             blood           healthy  03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         32464    109280914.0   16158    33646.02843110038   0.9568728712803031  0.8545324533535094  0.8487445540580735
agg_4734  classical monocyte      blood                             blood           healthy  436154da-bcf1-4130-9c8b-120ff9a888f2  scimilarity  nan     nan                         76800    206490628.0   16683    30736.453324546856  0.955313650235467   0.8494127267867799  0.84210567312908
agg_4735  classical monocyte      blood                             blood           healthy  5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         1044     3638976.0     12306    30542.8977364245    0.9465384578921433  0.851991795617166   0.8308191715256182
agg_4736  classical monocyte      blood                             blood           healthy  DS000010023                           scimilarity  nan     nan                         243      362606.0      8414     25201.261024165953  0.8865446516098515  0.7621391376670988  0.7681818769796727
agg_4737  classical monocyte      blood                             blood           healthy  GSE122703                             scimilarity  nan     nan                         18       83417.0       7546     28194.725173315186  0.859225612396475   0.7640890699253299  0.7443681539461423
agg_4738  classical monocyte      blood                             blood           healthy  GSE130117                             scimilarity  nan     nan                         2017     7130588.0     13535    33078.53692191542   0.9553673450160365  0.851402109626239   0.8385936100758409
agg_4739  classical monocyte      blood                             blood           healthy  GSE132802                             scimilarity  nan     nan                         1601     9955248.0     13132    32063.630951743195  0.9478882791739611  0.8391025143866828  0.8303465877530952
agg_4740  classical monocyte      blood                             blood           healthy  GSE139324                             scimilarity  nan     nan                         2333     8331045.0     13985    31135.881287246768  0.9608208780142045  0.8473885992448625  0.8432790193723467
agg_4741  classical monocyte      blood                             blood           healthy  GSE145809                             scimilarity  nan     nan                         69       245221.0      8962     29135.67197629852   0.8825701041728526  0.7811799267734735  0.7818647625179129
agg_4742  classical monocyte      blood                             blood           healthy  GSE149313                             scimilarity  nan     nan                         2420     6974751.0     13143    29560.496854576566  0.9574598613513423  0.8505290963248237  0.8379199735887167
agg_4743  classical monocyte      blood                             blood           healthy  GSE153421                             scimilarity  nan     nan                         3691     15561725.0    14569    34377.465875728165  0.9636686704566925  0.8576434473562725  0.8511814190737197
agg_4744  classical monocyte      blood                             blood           healthy  GSE156989                             scimilarity  nan     nan                         13554    160011485.0   16915    34135.439844737564  0.9640667421350761  0.8577967800377495  0.8517975138366085
agg_4745  classical monocyte      blood                             blood           healthy  GSE157829                             scimilarity  nan     nan                         1619     6957811.0     13507    30199.39288988673   0.9484019976492215  0.8436979316400604  0.8347196616710685
agg_4746  classical monocyte      blood                             blood           healthy  GSE159113                             scimilarity  nan     nan                         1025     6298250.0     12083    27477.50809897617   0.9078020151513733  0.8121457150205226  0.7980372877810575
agg_4747  classical monocyte      blood                             blood           healthy  GSE161329                             scimilarity  nan     nan                         5654     25653579.0    14349    28848.0539929647    0.9549801428956252  0.8450430950674043  0.8406188789518544
agg_4748  classical monocyte      blood                             blood           healthy  GSE161738                             scimilarity  nan     nan                         2676     13801473.0    12825    33337.477050230416  0.9541962906717452  0.8512846409758499  0.8485408028961247
agg_4749  classical monocyte      blood                             blood           healthy  GSE163668                             scimilarity  nan     nan                         2644     10486314.0    14049    33786.96584264489   0.9597801578342394  0.8560775485935677  0.8512149509551471
agg_4750  classical monocyte      blood                             blood           healthy  GSE166992                             scimilarity  nan     nan                         7501     28033216.0    15079    33455.367364577316  0.9622273594219685  0.8558958139235102  0.8495571689751152
agg_4751  classical monocyte      blood                             blood           healthy  GSE167363                             scimilarity  nan     nan                         3135     14722635.0    14375    29977.24002819913   0.942417448875388   0.8368071803109702  0.8258536430202982
agg_4752  classical monocyte      blood                             blood           healthy  GSE168710                             scimilarity  nan     nan                         16484    104881872.0   16223    34107.336261357574  0.9398282119039322  0.8424821834537695  0.8372971004604842
agg_4753  classical monocyte      blood                             blood           healthy  GSE168732                             scimilarity  nan     nan                         770      2548822.0     12508    33411.30103713399   0.9552513581030765  0.8508279875038706  0.847461536110767
agg_4754  classical monocyte      blood                             blood           healthy  b0cf0afa-ec40-4d65-b570-ed4ceacc6813  scimilarity  nan     nan                         40975    300555227.0   15784    35938.85772500803   0.9622425892039956  0.853424173800979   0.8508714303589978
agg_4755  classical monocyte      blood                             blood           healthy  ddfad306-714d-4cc0-9985-d9072820c530  scimilarity  nan     nan                         8827     36073928.0    15131    33208.591584008376  0.9546118779961532  0.8543086616569785  0.8462739374830107
agg_4772  classical monocyte      bone marrow                       bone marrow     healthy  GSE132509                             scimilarity  nan     nan                         610      2315570.0     12950    31768.06513427212   0.95159369508558    0.8517118261701931  0.836658919433696
agg_4773  classical monocyte      bone marrow                       bone marrow     healthy  GSE154109                             scimilarity  nan     nan                         531      1431388.0     11793    31377.450948003392  0.9490546933955852  0.8431566630120637  0.8370883160295727
agg_4774  classical monocyte      bone marrow                       bone marrow     healthy  GSE163278                             scimilarity  nan     nan                         1119     3970394.0     13361    32081.93302956569   0.9620394897163868  0.8531148861215617  0.8426785397396367
agg_4775  classical monocyte      bone marrow                       bone marrow     healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         151      8025584.0     12883    29444.78352863075   0.8440608178227607  0.740574910750325   0.7417328577454956
agg_4777  classical monocyte      breast                            breast          healthy  GSE164898                             scimilarity  nan     nan                         136      641471.0      12971    34463.52724138501   0.9163324788498406  0.8116274576633968  0.7978555908123931
agg_4778  classical monocyte      breast                            breast          healthy  c9706a92-0e5f-46c1-96d8-20e42467f287  scimilarity  nan     nan                         98       1444245.0     13491    30678.263421880285  0.9165520953567395  0.8162053142576849  0.7994301225229256
agg_4782  classical monocyte      bronchus                          airway          healthy  GSE158127                             scimilarity  nan     nan                         158      1158198.0     12643    34764.50196701077   0.9364512338084163  0.8259291909369686  0.8266638555276521
agg_4783  classical monocyte      cardiac muscle of left ventricle  heart           healthy  GSE156703                             scimilarity  nan     nan                         13       116181.0      9463     35695.66320276271   0.8542740960069863  0.7515621053395214  0.7561639038477878
agg_4785  classical monocyte      caudate lobe of liver             liver           healthy  44531dd9-1388-4416-a117-af0a99de2294  scimilarity  nan     nan                         238      730016.0      11505    31342.386314731422  0.9217674983890346  0.8140551552218395  0.8040417787954989
agg_4786  classical monocyte      cortex of kidney                  kidney          healthy  120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         79       323010.0      10939    32378.76683324232   0.9028856137251035  0.7978822439778066  0.7839454035009307
agg_4787  classical monocyte      cortex of kidney                  kidney          healthy  a98b828a-622a-483a-80e0-15703678befd  scimilarity  nan     nan                         91       477355.0      10898    32358.068865763344  0.9328436291917394  0.8237810319569842  0.8195391931798526
agg_4789  classical monocyte      digestive tract                   gut             healthy  DS000011665                           scimilarity  nan     nan                         347      1679116.0     12155    33347.55517047197   0.9422556928648441  0.8417267634096297  0.84018452536733
agg_4790  classical monocyte      exocrine pancreas                 pancreas        healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         821      7824069.0     14847    36135.64587109593   0.9493709998172055  0.8440837716099078  0.8410246939313819
agg_4791  classical monocyte      fallopian tube                    fallopian tube  healthy  fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         131      734093.0      11240    33115.640434504094  0.9359103734457376  0.8339026306142181  0.8225901799509813
agg_4792  classical monocyte      fimbria of uterine tube           fallopian tube  healthy  fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         34       209362.0      7663     27635.733860508328  0.8560416684135254  0.7382997749370328  0.7459235366949488
agg_4795  classical monocyte      head of femur                     bone            healthy  GSE169396                             scimilarity  nan     nan                         450      3669304.0     13216    33082.323604222154  0.9529417082022753  0.8522346343107771  0.8359032081996703
agg_4798  classical monocyte      heart left ventricle              heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         192      585001.0      11159    31422.874036870588  0.9363985001217598  0.8226438123601741  0.8283173244446851
agg_4799  classical monocyte      heart right ventricle             heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         316      936263.0      11904    32002.67227624691   0.9425900348990802  0.8306128730459813  0.8308116461021977
agg_4803  classical monocyte      interventricular septum           heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         442      1322725.0     12418    32235.5434681197    0.94751399102473    0.8340623939411483  0.8353858365852226
agg_4804  classical monocyte      isthmus of fallopian tube         fallopian tube  healthy  fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         62       318330.0      8642     29768.198512126502  0.8846027131590791  0.784220371073871   0.7739277341448829
agg_4808  classical monocyte      kidney                            kidney          healthy  120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         762      4034828.0     14946    34015.00816823295   0.9520640694425091  0.848205266473767   0.836085027208869
agg_4809  classical monocyte      kidney                            kidney          healthy  DS000010415                           scimilarity  nan     nan                         55       127079.0      8055     27206.419037355434  0.8216238135756493  0.7570847030479543  0.72524174726152
agg_4810  classical monocyte      kidney                            kidney          healthy  GSE140989                             scimilarity  nan     nan                         174      563438.0      11016    29887.593299155575  0.914390252807459   0.8069762795735104  0.8086759072557722
agg_4812  classical monocyte      left cardiac atrium               heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         450      1446734.0     12669    32433.20393285361   0.9467532324494885  0.8378571513042986  0.8360845068688169
agg_4817  classical monocyte      liver                             liver           healthy  GSE136103                             scimilarity  nan     nan                         2036     7668787.0     14818    32446.71686472893   0.9614468191909097  0.8455504517344443  0.8464065028454753
agg_4818  classical monocyte      liver                             liver           healthy  GSE159977                             scimilarity  nan     nan                         584      4703990.0     13644    34306.87421529999   0.9580119653448788  0.8463006395755477  0.8456137609888349
agg_4819  classical monocyte      liver                             liver           healthy  GSE163650                             scimilarity  nan     nan                         440      4840312.0     12272    31180.407161439263  0.9312379663603724  0.8198899526213368  0.8071334502842269
agg_4823  classical monocyte      lower lobe of lung                lung            healthy  GSE169471                             scimilarity  nan     nan                         305      1224338.0     11342    28255.985922767635  0.9404343350984603  0.8261785132237449  0.8150341534919611
agg_4832  classical monocyte      lung                              lung            healthy  5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         1254     5397217.0     13298    31013.01576531177   0.9490555294982109  0.8457060773457411  0.8329330004620054
agg_4833  classical monocyte      lung                              lung            healthy  DS000011735                           scimilarity  nan     nan                         4653     16523051.0    17066    37593.98867222708   0.8985708908575675  0.8260014412964971  0.8051199142820423
agg_4834  classical monocyte      lung                              lung            healthy  GSE128033                             scimilarity  nan     nan                         1047     3646581.0     13185    29331.2035341724    0.9513045959002788  0.837770557527153   0.8238987539695043
agg_4835  classical monocyte      lung                              lung            healthy  GSE128169                             scimilarity  nan     nan                         1732     11798577.0    15051    32941.755873862814  0.9636814596401634  0.8539834251349044  0.8457626672394015
agg_4836  classical monocyte      lung                              lung            healthy  GSE132771                             scimilarity  nan     nan                         1601     4614408.0     13275    29761.03818692572   0.9531291409651284  0.8457644131717956  0.8310980214000753
agg_4837  classical monocyte      lung                              lung            healthy  GSE169471                             scimilarity  nan     nan                         498      1613976.0     11886    28956.213107123967  0.9433400854993297  0.8271067022019086  0.815595680192867
agg_4853  classical monocyte      lung parenchyma                   lung            healthy  GSE158127                             scimilarity  nan     nan                         791      2735456.0     13260    33646.87553319058   0.9544351950657847  0.8399981618864122  0.8327646420552953
agg_4855  classical monocyte      mesenteric artery                 vasculature     healthy  GSE156341                             scimilarity  nan     nan                         49       408553.0      10083    30979.99239432764   0.9337200372314851  0.8169964046619824  0.8163109060481625
agg_4857  classical monocyte      mesenteric lymph node             lymph node      healthy  7681c7d7-0168-4892-a547-6f02a6430ace  scimilarity  nan     nan                         23       211416.0      9219     31018.02041830142   0.9058403141644794  0.7914556298280883  0.7867496166249129
agg_4858  classical monocyte      muscle tissue                     muscle          healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         1800     50754027.0    16141    33323.26969602708   0.9316672745867468  0.8348295016474899  0.8219918807835366
agg_4860  classical monocyte      nasal cavity                      airway          healthy  03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         129      319231.0      10961    33948.695520725145  0.908031594986554   0.802885512146251   0.7998873890770253
agg_4869  classical monocyte      peritoneum                        peritoneum      healthy  GSE130888                             scimilarity  nan     nan                         297      509237.0      11456    32213.169493243313  0.9218188146119148  0.8055575625334659  0.8039145027939639
agg_4870  classical monocyte      prostate gland                    prostate        healthy  GSE145843                             scimilarity  nan     nan                         24       87555.0       5997     21432.99958526179   0.816337811891654   0.7246769991610698  0.6943045232082502
agg_4871  classical monocyte      prostate gland                    prostate        healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         220      2445216.0     12460    33911.259081133416  0.9301914834403244  0.8259560877683964  0.824440772955729
agg_4872  classical monocyte      renal medulla                     kidney          healthy  120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         21       101089.0      7354     26191.96315382653   0.8491245141516279  0.751292403843078   0.7310820608427115
agg_4875  classical monocyte      right cardiac atrium              heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         311      977934.0      11891    31570.89048372305   0.9357294628028995  0.8260631626226992  0.8224053236502149
agg_4878  classical monocyte      spleen                            spleen          healthy  4d74781b-8186-4c9a-b659-ff4dc4601d91  scimilarity  nan     nan                         2166     7905128.0     13952    30832.98149016589   0.957626764427953   0.8498277489734691  0.8368540073560422
agg_4879  classical monocyte      spleen                            spleen          healthy  GSE148796                             scimilarity  nan     nan                         49       99684.0       6785     24492.260723886982  0.8477188947186437  0.7504685014947085  0.7295165069095441
agg_4880  classical monocyte      spleen                            spleen          healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         3483     86078540.0    17775    36266.36759486      0.9100465322095738  0.8097019968114763  0.8081945289287479
agg_4881  classical monocyte      subcutaneous adipose tissue       adipose         healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         2019     27914261.0    15739    35142.992419516115  0.9421666034710219  0.8357406603138717  0.8249091266671545
agg_4886  classical monocyte      thoracic lymph node               lymph node      healthy  62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         20194    147502356.0   17156    34278.43204322054   0.9584542376154844  0.8568164088302802  0.8468420651789338
agg_4887  classical monocyte      thymus                            thymus          healthy  62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         487      2692133.0     12969    33549.45213075861   0.9497232200872254  0.8549259958290932  0.843613914494647
agg_4888  classical monocyte      thymus                            thymus          healthy  83ed3be8-4cb9-43e6-9aaa-3fbbf5d1bd3a  scimilarity  nan     nan                         27       80042.0       6527     23800.833698344715  0.8448423801602987  0.7347769941739797  0.7181319535525903
agg_4889  classical monocyte      thymus                            thymus          healthy  de13e3e2-23b6-40ed-a413-e9e12d7d3910  scimilarity  nan     nan                         52       298983.0      9582     30002.684399867492  0.9175531872094376  0.8185046437128072  0.8176606503789746
agg_4890  classical monocyte      tonsil                            tonsil          healthy  GSE119506                             scimilarity  nan     nan                         321      1114339.0     11546    30516.473940957327  0.936749794981687   0.8307356996245999  0.8215841775914924
agg_4893  classical monocyte      trachea                           airway          healthy  03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         126      457580.0      11135    33734.327765812595  0.9299687996532033  0.8279876048002456  0.8230768581762299
agg_4894  classical monocyte      trachea                           airway          healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         130      8245622.0     13663    32584.555010387805  0.859406194907081   0.7522520630806427  0.7550758397192211
agg_4896  classical monocyte      transverse colon                  gut             healthy  62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         503      2932999.0     12868    33265.22762476361   0.9476555210791636  0.8481166846094836  0.8392740144102075
agg_4899  classical monocyte      upper lobe of lung                lung            healthy  GSE169471                             scimilarity  nan     nan                         180      594059.0      10222    27881.011904781462  0.9248153838159024  0.8106326147324702  0.8003438801026445
agg_4900  classical monocyte      urine                             urinary         healthy  GSE165396                             scimilarity  nan     nan                         20       109197.0      7299     25505.26771828214   0.8530560359297166  0.7505269381795711  0.7350625083451162
agg_4901  classical monocyte      uterus                            uterus          healthy  32f2fd23-ec74-486f-9544-e5b2f41725f5  scimilarity  nan     nan                         18       189472.0      9397     30891.07905591756   0.8870309576225303  0.7796319609750195  0.7827462282311286
agg_4902  classical monocyte      vasculature                       vasculature     healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         14537    224206261.0   17922    36651.90273209977   0.9438232475974557  0.8425086108952623  0.8343079266995497
agg_6287  intermediate monocyte   head of femur                     bone            healthy  GSE169396                             scimilarity  nan     nan                         102      191075.0      7853     26179.035956297153  0.8330771518439726  0.7503209876663273  0.7113081302663875
agg_6289  intermediate monocyte   lung                              lung            healthy  5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         178      1172582.0     11314    30515.569680815937  0.9409051040470379  0.8435279441582394  0.8229749449946738
agg_6290  intermediate monocyte   spleen                            spleen          healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         60       252220.0      8236     26163.955486425446  0.836498746238646   0.7439980946911184  0.7168040239671246
agg_6291  intermediate monocyte   thymus                            thymus          healthy  GSE159745                             scimilarity  nan     nan                         29       82115.0       5987     22540.234815420707  0.7817665678679913  0.7043553158094606  0.6760717657015229
agg_6292  intermediate monocyte   vasculature                       vasculature     healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         162      1525515.0     11980    32186.1781582846    0.9269571292216415  0.8224395619954739  0.8055668264019443
agg_7919  non-classical monocyte  apex of heart                     heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         158      675731.0      11222    30876.899222310585  0.9301440448035162  0.8258961291954683  0.824159827752181
agg_7939  non-classical monocyte  blood                             blood           healthy  03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         7359     36418526.0    14851    33258.733090217436  0.9519276590597959  0.8501060197631037  0.850960653386464
agg_7940  non-classical monocyte  blood                             blood           healthy  436154da-bcf1-4130-9c8b-120ff9a888f2  scimilarity  nan     nan                         14619    54479703.0    15493    30191.499157063707  0.9490211472741342  0.8448211022633122  0.8418061408505518
agg_7941  non-classical monocyte  blood                             blood           healthy  5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         195      1239240.0     11002    30223.31815928311   0.9379359624101378  0.8397815574875395  0.8299059656859994
agg_7942  non-classical monocyte  blood                             blood           healthy  GSE130117                             scimilarity  nan     nan                         322      1200290.0     11182    31409.27333671809   0.9418345510352945  0.8405863822794143  0.8304108431330093
agg_7943  non-classical monocyte  blood                             blood           healthy  GSE132802                             scimilarity  nan     nan                         102      726593.0      10232    30435.311954214158  0.9262718659348109  0.8186907385315048  0.8147462064075649
agg_7944  non-classical monocyte  blood                             blood           healthy  GSE134004                             scimilarity  nan     nan                         21       123360.0      6990     24135.62187154626   0.8660944307658819  0.7608357820845066  0.7535994530765779
agg_7945  non-classical monocyte  blood                             blood           healthy  GSE139324                             scimilarity  nan     nan                         435      2395489.0     12428    30987.28210955269   0.9487843521961865  0.8347185640893492  0.8386196076774184
agg_7946  non-classical monocyte  blood                             blood           healthy  GSE149313                             scimilarity  nan     nan                         567      2445950.0     11737    29144.914765151065  0.9480753865144765  0.8398428245307203  0.8354975068540325
agg_7947  non-classical monocyte  blood                             blood           healthy  GSE153421                             scimilarity  nan     nan                         441      2114891.0     11966    32647.54233438482   0.9524227157990666  0.8453840087915955  0.8445318485082767
agg_7948  non-classical monocyte  blood                             blood           healthy  GSE156989                             scimilarity  nan     nan                         3151     40420221.0    15662    32233.769948686153  0.9558775883405131  0.8527879494870606  0.8461332941066178
agg_7949  non-classical monocyte  blood                             blood           healthy  GSE157829                             scimilarity  nan     nan                         144      890675.0      10657    29172.377435644317  0.9302786443476869  0.8283014100006509  0.8220722486967346
agg_7950  non-classical monocyte  blood                             blood           healthy  GSE161329                             scimilarity  nan     nan                         1118     7175719.0     12865    29244.857112487658  0.9476682128749678  0.8393859845018986  0.8392730294908749
agg_7951  non-classical monocyte  blood                             blood           healthy  GSE161738                             scimilarity  nan     nan                         1497     12143757.0    12362    32632.110821778042  0.9476207444575895  0.8463853459866758  0.8490136875251815
agg_7952  non-classical monocyte  blood                             blood           healthy  GSE163668                             scimilarity  nan     nan                         323      1716760.0     11769    32812.39505818172   0.9472316303194352  0.8438698542111769  0.8424970169398833
agg_7953  non-classical monocyte  blood                             blood           healthy  GSE166992                             scimilarity  nan     nan                         1613     7143035.0     13383    32605.389353410996  0.9532209129181096  0.8486599938147087  0.8451557766892072
agg_7954  non-classical monocyte  blood                             blood           healthy  GSE167363                             scimilarity  nan     nan                         458      3094035.0     12228    29678.03739631962   0.9435654504659038  0.8379948419319221  0.8241961683725367
agg_7955  non-classical monocyte  blood                             blood           healthy  GSE168710                             scimilarity  nan     nan                         75       701113.0      10776    32559.49721301115   0.9233871475920297  0.8250528447177075  0.8185333840139741
agg_7956  non-classical monocyte  blood                             blood           healthy  GSE168732                             scimilarity  nan     nan                         229      1242404.0     11269    31965.299177754878  0.9416742339845935  0.8377482237588458  0.8415500681607002
agg_7957  non-classical monocyte  blood                             blood           healthy  b0cf0afa-ec40-4d65-b570-ed4ceacc6813  scimilarity  nan     nan                         5897     43180935.0    14970    35595.18649267558   0.9543288699936718  0.8496535044071925  0.8536782531653531
agg_7970  non-classical monocyte  bone marrow                       bone marrow     healthy  GSE132509                             scimilarity  nan     nan                         28       86280.0       7085     25870.53624683356   0.8392355979135352  0.755387271124993   0.7303317003908475
agg_7971  non-classical monocyte  bone marrow                       bone marrow     healthy  GSE154109                             scimilarity  nan     nan                         50       289907.0      9147     28526.85661530531   0.9070952567396775  0.808361820498357   0.7964598438037417
agg_7972  non-classical monocyte  bone marrow                       bone marrow     healthy  GSE163278                             scimilarity  nan     nan                         127      682328.0      10897    31202.529821716787  0.9358619814006104  0.8294675777784289  0.8230468472234642
agg_7974  non-classical monocyte  breast                            breast          healthy  GSE164898                             scimilarity  nan     nan                         54       120275.0      7652     26423.010551181265  0.8534343500588628  0.755424451856796   0.7230125883881353
agg_7976  non-classical monocyte  cortex of kidney                  kidney          healthy  120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         63       401864.0      11464    32990.96311584531   0.9014527631813947  0.7931942866954178  0.7865930763896118
agg_7977  non-classical monocyte  cortex of kidney                  kidney          healthy  a98b828a-622a-483a-80e0-15703678befd  scimilarity  nan     nan                         161      772141.0      11062    31056.660350613587  0.9346806312815913  0.8317520409566359  0.8313984606487961
agg_7979  non-classical monocyte  fallopian tube                    fallopian tube  healthy  fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         16       222500.0      7755     27898.660349927846  0.8621134765305163  0.7601890611687369  0.744729424431635
agg_7980  non-classical monocyte  fimbria of uterine tube           fallopian tube  healthy  fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         28       244266.0      7735     27678.814270356754  0.8734809587140066  0.7648153554084077  0.76608428730196
agg_7982  non-classical monocyte  heart left ventricle              heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         57       213565.0      9322     29851.355460106315  0.8978296307325292  0.7939956162210045  0.776120799756924
agg_7983  non-classical monocyte  heart right ventricle             heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         124      613752.0      10880    30311.360727579897  0.9311389328985725  0.8253171421863371  0.8219320702682233
agg_7985  non-classical monocyte  interventricular septum           heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         144      524658.0      10853    31170.857886799495  0.9241302459076691  0.8138209035301205  0.8224136632163046
agg_7986  non-classical monocyte  isthmus of fallopian tube         fallopian tube  healthy  fc77d2ae-247d-44d7-aa24-3f4859254c2c  scimilarity  nan     nan                         12       86198.0       5668     23558.81110396413   0.793224614700874   0.7114672551137675  0.6825985710920001
agg_7990  non-classical monocyte  kidney                            kidney          healthy  120e86b4-1195-48c5-845b-b98054105eec  scimilarity  nan     nan                         214      1788808.0     13749    33794.71179753717   0.9324162382479619  0.8250902105825786  0.818301334304562
agg_7991  non-classical monocyte  kidney                            kidney          healthy  GSE140989                             scimilarity  nan     nan                         473      1769375.0     13008    31797.13462972748   0.9190676320157254  0.8182361375222441  0.8172110487458264
agg_7992  non-classical monocyte  left cardiac atrium               heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         82       357018.0      10073    29995.2881818953    0.916928150799629   0.8074046233049013  0.808753490409063
agg_7995  non-classical monocyte  liver                             liver           healthy  GSE136103                             scimilarity  nan     nan                         423      2383574.0     13122    31645.435149503457  0.9524231389227664  0.8376748793878739  0.8415142758287522
agg_7996  non-classical monocyte  liver                             liver           healthy  GSE159977                             scimilarity  nan     nan                         473      4877555.0     13370    33200.636271185205  0.9498326621251537  0.8392029926705806  0.8469259328198118
agg_7997  non-classical monocyte  liver                             liver           healthy  GSE163650                             scimilarity  nan     nan                         10       96148.0       6782     24958.605923030595  0.844996457443227   0.7380858347163884  0.7209649786125333
agg_8007  non-classical monocyte  lung                              lung            healthy  5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         576      3313836.0     12566    30969.255628748215  0.9438879246302309  0.8444989044016609  0.8358111974870654
agg_8008  non-classical monocyte  lung                              lung            healthy  DS000011735                           scimilarity  nan     nan                         169      779577.0      12885    36247.76234710614   0.8808886302153663  0.8102555656654346  0.7948721237235893
agg_8009  non-classical monocyte  lung                              lung            healthy  GSE128033                             scimilarity  nan     nan                         79       343546.0      9330     27406.22701580364   0.9035509606166174  0.7968173004217675  0.7863212826693502
agg_8010  non-classical monocyte  lung                              lung            healthy  GSE128169                             scimilarity  nan     nan                         276      2433151.0     12769    32027.902610765417  0.9547437308289124  0.8449880167209973  0.8438649522788478
agg_8011  non-classical monocyte  lung                              lung            healthy  GSE132771                             scimilarity  nan     nan                         37       151922.0      7860     26295.743163049112  0.8838770944570796  0.790431305730734   0.7607028661311058
agg_8012  non-classical monocyte  lung                              lung            healthy  GSE169471                             scimilarity  nan     nan                         27       103204.0      6854     23947.65597317191   0.8348392917103126  0.7488344663849291  0.7237935010019484
agg_8024  non-classical monocyte  lung parenchyma                   lung            healthy  GSE158127                             scimilarity  nan     nan                         309      1788828.0     12136    32078.56380429425   0.9418105467635777  0.8295808338281161  0.831837004521098
agg_8026  non-classical monocyte  muscle tissue                     muscle          healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         321      23778338.0    13169    27977.08919973388   0.8814221140322411  0.7869360982600341  0.7677337716492793
agg_8034  non-classical monocyte  right cardiac atrium              heart           healthy  b52eb423-5d0d-4645-b217-e1c6d38b2e72  scimilarity  nan     nan                         70       408965.0      10146    29746.121058681525  0.9252128588702657  0.8186384457806745  0.8103901461358153
agg_8036  non-classical monocyte  spleen                            spleen          healthy  4d74781b-8186-4c9a-b659-ff4dc4601d91  scimilarity  nan     nan                         336      1586973.0     11934    30580.798338873254  0.9471354298985436  0.8378586626071394  0.8322814736104384
agg_8039  non-classical monocyte  thoracic lymph node               lymph node      healthy  62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         1950     18888557.0    15221    33581.76389036113   0.9559331607078543  0.8537562341521224  0.8469325123405803
agg_8040  non-classical monocyte  thymus                            thymus          healthy  62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         68       441502.0      10301    31247.452676451114  0.917407487628822   0.8223254514108821  0.8197270854172879
agg_8041  non-classical monocyte  trachea                           airway          healthy  03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         12       97840.0       7863     29204.083428972655  0.8702163087134754  0.7685799000235117  0.7716054237714325
agg_8042  non-classical monocyte  trachea                           airway          healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         16       534304.0      8970     24807.157880956805  0.8232311272852844  0.7180336962121132  0.7003798955542161
agg_8044  non-classical monocyte  transverse colon                  gut             healthy  62ef75e4-cbea-454e-a0ce-998ec40223d3  scimilarity  nan     nan                         135      908452.0      11380    32804.296806012215  0.933818206129888   0.8353788654625934  0.8345447172426862
agg_8045  non-classical monocyte  upper lobe of lung                lung            healthy  GSE169471                             scimilarity  nan     nan                         32       143341.0      7269     24395.316476390286  0.8673226900308589  0.7523857428733156  0.7403920706432944
agg_8046  non-classical monocyte  vasculature                       vasculature     healthy  e5f58829-1a66-40b5-a624-9046778e74f5  scimilarity  nan     nan                         987      10267209.0    15019    35019.47682838501   0.9379668399645168  0.8304128458333673  0.8283221722893935
</pre></div>
</div>
</div>
</div>
<p>This query selects cells that are:</p>
<ul class="simple">
<li><p>classical monocytes (cell_type == “classical monocyte”)</p></li>
<li><p>from healthy donors (disease == “healthy”)</p></li>
<li><p>from blood tissue (tissue == “blood”)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>query-cell<span class="w"> </span><span class="s1">&#39;cell_type == &quot;classical monocyte&quot; and disease == &quot;healthy&quot; and tissue == &quot;blood&quot;&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-t<span class="w"> </span>-s<span class="w"> </span><span class="s1">$&#39;\t&#39;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1862.1MB/s)
          cell_type           tissue  organ  disease  study                                 dataset      region  subregion  celltype_coarse  n_cells  total_counts  n_genes  size_factor         train_pearson       val_pearson         test_pearson
agg_4733  classical monocyte  blood   blood  healthy  03f821b4-87be-4ff4-b65a-b5fc00061da7  scimilarity  nan     nan                         32464    109280914.0   16158    33646.02843110038   0.9568728712803031  0.8545324533535094  0.8487445540580735
agg_4734  classical monocyte  blood   blood  healthy  436154da-bcf1-4130-9c8b-120ff9a888f2  scimilarity  nan     nan                         76800    206490628.0   16683    30736.453324546856  0.955313650235467   0.8494127267867799  0.84210567312908
agg_4735  classical monocyte  blood   blood  healthy  5d445965-6f1a-4b68-ba3a-b8f765155d3a  scimilarity  nan     nan                         1044     3638976.0     12306    30542.8977364245    0.9465384578921433  0.851991795617166   0.8308191715256182
agg_4736  classical monocyte  blood   blood  healthy  DS000010023                           scimilarity  nan     nan                         243      362606.0      8414     25201.261024165953  0.8865446516098515  0.7621391376670988  0.7681818769796727
agg_4737  classical monocyte  blood   blood  healthy  GSE122703                             scimilarity  nan     nan                         18       83417.0       7546     28194.725173315186  0.859225612396475   0.7640890699253299  0.7443681539461423
agg_4738  classical monocyte  blood   blood  healthy  GSE130117                             scimilarity  nan     nan                         2017     7130588.0     13535    33078.53692191542   0.9553673450160365  0.851402109626239   0.8385936100758409
agg_4739  classical monocyte  blood   blood  healthy  GSE132802                             scimilarity  nan     nan                         1601     9955248.0     13132    32063.630951743195  0.9478882791739611  0.8391025143866828  0.8303465877530952
agg_4740  classical monocyte  blood   blood  healthy  GSE139324                             scimilarity  nan     nan                         2333     8331045.0     13985    31135.881287246768  0.9608208780142045  0.8473885992448625  0.8432790193723467
agg_4741  classical monocyte  blood   blood  healthy  GSE145809                             scimilarity  nan     nan                         69       245221.0      8962     29135.67197629852   0.8825701041728526  0.7811799267734735  0.7818647625179129
agg_4742  classical monocyte  blood   blood  healthy  GSE149313                             scimilarity  nan     nan                         2420     6974751.0     13143    29560.496854576566  0.9574598613513423  0.8505290963248237  0.8379199735887167
agg_4743  classical monocyte  blood   blood  healthy  GSE153421                             scimilarity  nan     nan                         3691     15561725.0    14569    34377.465875728165  0.9636686704566925  0.8576434473562725  0.8511814190737197
agg_4744  classical monocyte  blood   blood  healthy  GSE156989                             scimilarity  nan     nan                         13554    160011485.0   16915    34135.439844737564  0.9640667421350761  0.8577967800377495  0.8517975138366085
agg_4745  classical monocyte  blood   blood  healthy  GSE157829                             scimilarity  nan     nan                         1619     6957811.0     13507    30199.39288988673   0.9484019976492215  0.8436979316400604  0.8347196616710685
agg_4746  classical monocyte  blood   blood  healthy  GSE159113                             scimilarity  nan     nan                         1025     6298250.0     12083    27477.50809897617   0.9078020151513733  0.8121457150205226  0.7980372877810575
agg_4747  classical monocyte  blood   blood  healthy  GSE161329                             scimilarity  nan     nan                         5654     25653579.0    14349    28848.0539929647    0.9549801428956252  0.8450430950674043  0.8406188789518544
agg_4748  classical monocyte  blood   blood  healthy  GSE161738                             scimilarity  nan     nan                         2676     13801473.0    12825    33337.477050230416  0.9541962906717452  0.8512846409758499  0.8485408028961247
agg_4749  classical monocyte  blood   blood  healthy  GSE163668                             scimilarity  nan     nan                         2644     10486314.0    14049    33786.96584264489   0.9597801578342394  0.8560775485935677  0.8512149509551471
agg_4750  classical monocyte  blood   blood  healthy  GSE166992                             scimilarity  nan     nan                         7501     28033216.0    15079    33455.367364577316  0.9622273594219685  0.8558958139235102  0.8495571689751152
agg_4751  classical monocyte  blood   blood  healthy  GSE167363                             scimilarity  nan     nan                         3135     14722635.0    14375    29977.24002819913   0.942417448875388   0.8368071803109702  0.8258536430202982
agg_4752  classical monocyte  blood   blood  healthy  GSE168710                             scimilarity  nan     nan                         16484    104881872.0   16223    34107.336261357574  0.9398282119039322  0.8424821834537695  0.8372971004604842
agg_4753  classical monocyte  blood   blood  healthy  GSE168732                             scimilarity  nan     nan                         770      2548822.0     12508    33411.30103713399   0.9552513581030765  0.8508279875038706  0.847461536110767
agg_4754  classical monocyte  blood   blood  healthy  b0cf0afa-ec40-4d65-b570-ed4ceacc6813  scimilarity  nan     nan                         40975    300555227.0   15784    35938.85772500803   0.9622425892039956  0.853424173800979   0.8508714303589978
agg_4755  classical monocyte  blood   blood  healthy  ddfad306-714d-4cc0-9985-d9072820c530  scimilarity  nan     nan                         8827     36073928.0    15131    33208.591584008376  0.9546118779961532  0.8543086616569785  0.8462739374830107
</pre></div>
</div>
</div>
</div>
</section>
<section id="attribution-calling-with-custom-genes-and-sequences">
<h3>Attribution calling with custom genes and sequences<a class="headerlink" href="#attribution-calling-with-custom-genes-and-sequences" title="Link to this heading">¶</a></h3>
<p>In this section, we demonstrate how to call attributions using custom gene sequences.
You can provide your own FASTA file containing sequences of interest and run attribution analysis
for any set of genes or genomic regions, using the Decima command-line interface.
The following examples show how to inspect your FASTA file, run attributions, and explore the output files.
The FASTA header line for each sequence contains the gene name and the coordinates of the masked region used for attribution analysis.
For example, in the header:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>CD68|gene_mask_start=163840|gene_mask_end=166460
</pre></div>
</div>
<p>“CD68” is the gene name, “gene_mask_start” and “gene_mask_end” specify the start and end positions (relative to the input sequence) of the region that was masked and analyzed for attributions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>cat<span class="w"> </span>../tests/data/seqs.fasta<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span>-200
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&gt;CD68|gene_mask_start=163840|gene_mask_end=166460
CTCTGCAGAGAGCGAGGACGGTGTGTCTGCCAGCGCCTTTGACTTCACTGTCTCCAACTTTGTGGACAACCTGTATGGCTACCCGGAAGGCAAGGATGTGCTTCGGGAGACCATCAAGTTTATGTACACAGACTGGGCCGACCGGGACAATGGCGAAATGCGCCGCAAAACCCTGCTGGCGCTCTTTACTGACCACCAAT
&gt;SPI1|gene_mask_start=163840|gene_mask_end=187556
TGCCACTTTTAGATATGTTCATGGGTGCAGATACGGCTTTATTTATTTGAGACAGAGTTTCACTCTTGTTGCCCAGGGTGGAGTGCAGTGGTGCGATCTCAGCTCACTGCAGCCTTCGCCTCCCGGGTTGAAGCGATTCTTCTGCCTCAACCTCGAGTAGCTGGGATTATAGGCACCTGCCAGCATGCCTGGCTAATTTT
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>attributions<span class="w"> </span>--model<span class="w"> </span>v1_rep0<span class="w"> </span>--seqs<span class="w"> </span>../tests/data/seqs.fasta<span class="w"> </span>--tasks<span class="w"> </span><span class="s2">&quot;cell_type == &#39;classical monocyte&#39;&quot;</span><span class="w"> </span>--output-prefix<span class="w"> </span>example/output_custom_seqs
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Using device: cuda
decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1875.1MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1871.0MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1297.0MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████████████| 2/2 [00:04&lt;00:00,  2.09s/it]
decima - INFO - Saving sequences...
Saving sequences...: 2it [00:00, 4493.09it/s]
decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1856.1MB/s)
decima - INFO - No genes provided, using all 2 genes in the attribution files.
Computing recursive seqlet calling...: 100%|█████| 2/2 [00:00&lt;00:00, 543.20it/s]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.

</pre></div>
</div>
</div>
</div>
<p>The output file for custom sequence also contains additional file of <code class="docutils literal notranslate"><span class="pre">seqs.fasta</span></code> which contains custom seqs. To visualize prediction on IGV, first load this fasta file and its index (.fai) to IGV, then load bam and bed files.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>example/output_custom_seqs*
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>example/output_custom_seqs.attributions.bigwig
example/output_custom_seqs.attributions.h5
example/output_custom_seqs.motifs.tsv
example/output_custom_seqs.seqlets.bed
example/output_custom_seqs.seqs.fasta
example/output_custom_seqs.seqs.fasta.fai
example/output_custom_seqs.warnings.qc.log
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="python-user-api">
<h2>Python User API<a class="headerlink" href="#python-user-api" title="Link to this heading">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>example/output_classical_monoctypes.*
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>example/output_classical_monoctypes.attributions.bigwig
example/output_classical_monoctypes.attributions.h5
example/output_classical_monoctypes.motifs.tsv
example/output_classical_monoctypes.seqlets.bed
example/output_classical_monoctypes.warnings.qc.log
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.interpret.attributions</span><span class="w"> </span><span class="kn">import</span> <span class="n">AttributionResult</span>

<span class="k">with</span> <span class="n">AttributionResult</span><span class="p">(</span><span class="s2">&quot;example/output_classical_monoctypes.attributions.h5&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ar</span><span class="p">:</span>
    <span class="n">seqs</span><span class="p">,</span> <span class="n">attrs</span> <span class="o">=</span> <span class="n">ar</span><span class="o">.</span><span class="n">load</span><span class="p">([</span><span class="s2">&quot;SPI1&quot;</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;seqs:&quot;</span><span class="p">,</span> <span class="n">seqs</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;attrs:&quot;</span><span class="p">,</span> <span class="n">attrs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading attributions and sequences...: 100%|██████████| 1/1 [00:00&lt;00:00, 469.84it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>seqs: [[[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 1. ... 0. 0. 0.]
  [0. 1. 0. ... 1. 0. 1.]
  [1. 0. 0. ... 0. 1. 0.]]]
attrs: [[[-5.04877644e-05 -3.64095758e-05  7.14814905e-06 ... -1.40671709e-05
   -5.00998385e-06 -6.19171897e-06]
  [-5.04877644e-05 -3.64095758e-05 -2.14444472e-05 ... -1.40671709e-05
   -5.00998385e-06 -6.19171897e-06]
  [-5.04877644e-05  1.09228728e-04  7.14814905e-06 ...  4.22015128e-05
   -5.00998385e-06  1.85751569e-05]
  [ 1.51463293e-04 -3.64095758e-05  7.14814905e-06 ... -1.40671709e-05
    1.50299516e-05 -6.19171897e-06]]]
</pre></div>
</div>
</div>
</div>
<p>Let’s look at a simple example using Decima’s Python API to analyze the SPI1 gene, which is a key transcription factor in myeloid cell development. We’ll examine its regulation across different monocyte and macrophage cell types where it is known to be important.</p>
<p>First we choice the cells, we are interested in:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">AttributionResult</span><span class="p">(</span><span class="s2">&quot;example/output_classical_monoctypes.attributions.h5&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ar</span><span class="p">:</span>
    <span class="n">attribution</span> <span class="o">=</span> <span class="n">ar</span><span class="o">.</span><span class="n">load_attribution</span><span class="p">(</span><span class="s2">&quot;SPI1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1889.1MB/s)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="n">attribution</span><span class="o">.</span><span class="n">plot_seqlogo</span><span class="p">(</span><span class="n">relative_loc</span><span class="o">=</span><span class="mi">291</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a8075529a1a455f7b6595a2388c23173478683997206938a63428b961fedf9e3.png" src="../_images/a8075529a1a455f7b6595a2388c23173478683997206938a63428b961fedf9e3.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">attribution</span><span class="o">.</span><span class="n">plot_peaks</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/c9528e8d680a784d6469f3d3aec97b72e78408c09dd4e60669c7a3e34047886a.png"><img alt="../_images/c9528e8d680a784d6469f3d3aec97b72e78408c09dd4e60669c7a3e34047886a.png" src="../_images/c9528e8d680a784d6469f3d3aec97b72e78408c09dd4e60669c7a3e34047886a.png" style="width: 1000px; height: 200px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima</span><span class="w"> </span><span class="kn">import</span> <span class="n">predict_attributions_seqlet_calling</span>

<span class="n">device</span> <span class="o">=</span> <span class="s2">&quot;cuda&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">spi1_cell_types</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;classical monocyte&quot;</span><span class="p">,</span>
    <span class="s2">&quot;intermediate monocyte&quot;</span><span class="p">,</span>
    <span class="s2">&quot;non-classical monocyte&quot;</span><span class="p">,</span>
    <span class="s2">&quot;alveolar macrophage&quot;</span><span class="p">,</span>
    <span class="s2">&quot;macrophage&quot;</span><span class="p">,</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_attributions_seqlet_calling</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/attrs_SP1I_monoctypes&quot;</span><span class="p">,</span>
    <span class="n">genes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SPI1&quot;</span><span class="p">],</span>
    <span class="n">tasks</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;cell_type in </span><span class="si">{</span><span class="n">spi1_cell_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1882.9MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1896.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1362.8MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 1/1 [00:02&lt;00:00,  2.19s/it]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1879.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1872.4MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep1:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1184.0MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 1/1 [00:01&lt;00:00,  1.78s/it]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1869.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1906.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep2:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1223.8MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 1/1 [00:01&lt;00:00,  1.78s/it]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1839.5MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.8 (1767.9MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep3:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1362.2MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 1/1 [00:01&lt;00:00,  1.79s/it]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1798.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1875.8MB/s)
Computing recursive seqlet calling...: 100%|██████████| 1/1 [00:01&lt;00:00,  1.66s/it]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
</pre></div>
</div>
</div>
</div>
<p>Similar to command line you can use <code class="docutils literal notranslate"><span class="pre">predict_save_attributions</span></code> and <code class="docutils literal notranslate"><span class="pre">recursive_seqlet_calling</span></code> functions calls attirubitions and seqlets step by step.</p>
<section id="custom-sequences">
<h3>Custom Sequences<a class="headerlink" href="#custom-sequences" title="Link to this heading">¶</a></h3>
<p>Attributions for a custom sequence can be calculated by passing data frame with columns of <code class="docutils literal notranslate"><span class="pre">seq</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_mask_start</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_mask_end</span></code>. The index of the DataFrame will be used as gene names.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="n">df_seqs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../tests/data/seqs.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">df_seqs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seq</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
    </tr>
    <tr>
      <th>name</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CD68</th>
      <td>CTCTGCAGAGAGCGAGGACGGTGTGTCTGCCAGCGCCTTTGACTTC...</td>
      <td>163840</td>
      <td>166460</td>
    </tr>
    <tr>
      <th>SPI1</th>
      <td>TGCCACTTTTAGATATGTTCATGGGTGCAGATACGGCTTTATTTAT...</td>
      <td>163840</td>
      <td>187556</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_attributions_seqlet_calling</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/attrs_custom_seqs_monoctypes&quot;</span><span class="p">,</span>
    <span class="n">seqs</span><span class="o">=</span><span class="n">df_seqs</span><span class="p">,</span>  <span class="c1"># &lt;-- custom sequences</span>
    <span class="n">tasks</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;cell_type in </span><span class="si">{</span><span class="n">spi1_cell_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
<span class="p">)</span>
<span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>attrs_custom_seqs_monoctypes
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1898.7MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1844.3MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1285.9MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 2/2 [00:03&lt;00:00,  1.66s/it]
Saving sequences...: 2it [00:00, 3483.64it/s]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1928.9MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1873.0MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep1:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1232.0MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 2/2 [00:03&lt;00:00,  1.65s/it]
Saving sequences...: 2it [00:00, 2953.74it/s]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1889.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1901.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep2:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1255.7MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 2/2 [00:03&lt;00:00,  1.64s/it]
Saving sequences...: 2it [00:00, 3894.43it/s]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1874.7MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1874.1MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep3:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1374.8MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 2/2 [00:03&lt;00:00,  1.64s/it]
Saving sequences...: 2it [00:00, 2833.03it/s]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1888.2MB/s)
Computing recursive seqlet calling...: 100%|██████████| 2/2 [00:00&lt;00:00, 895.17it/s]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ls: cannot access &#39;attrs_custom_seqs_monoctypes&#39;: No such file or directory
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">grelu.sequence.format</span><span class="w"> </span><span class="kn">import</span> <span class="n">strings_to_one_hot</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">DECIMA_CONTEXT_SIZE</span>

<span class="n">DECIMA_CONTEXT_SIZE</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>524288
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seqs</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="n">strings_to_one_hot</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">DECIMA_CONTEXT_SIZE</span><span class="p">))]</span>
        <span class="p">),</span>  <span class="c1"># one-hot encoded sequence</span>
        <span class="n">torch</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">DECIMA_CONTEXT_SIZE</span><span class="p">),</span>  <span class="c1"># binary mask for the gene</span>
    <span class="p">],</span>
    <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">seqs</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>torch.Size([1, 5, 524288])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_attributions_seqlet_calling</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/attrs_custom_tensors_monoctypes&quot;</span><span class="p">,</span>
    <span class="n">seqs</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span>  <span class="c1"># &lt;-- custom sequences as torch.Tensor where (batch_size, 5, seq_len), second dimension is one-hot encoded sequence and binary mask for the gene</span>
    <span class="n">tasks</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;cell_type in </span><span class="si">{</span><span class="n">spi1_cell_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">threshold</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1885.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1888.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1335.1MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
Computing attributions...: 100%|██████████| 1/1 [00:01&lt;00:00,  1.79s/it]
Saving sequences...: 1it [00:00, 3806.08it/s]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1854.2MB/s)
Computing recursive seqlet calling...: 100%|██████████| 1/1 [00:00&lt;00:00, 1219.63it/s]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/memelite/fimo.py:406: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>example/attrs_custom_tensors_monoctypes*
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>example/attrs_custom_tensors_monoctypes.attributions.bigwig
example/attrs_custom_tensors_monoctypes.attributions.h5
example/attrs_custom_tensors_monoctypes.motifs.tsv
example/attrs_custom_tensors_monoctypes.seqlets.bed
example/attrs_custom_tensors_monoctypes.seqs.fasta
example/attrs_custom_tensors_monoctypes.seqs.fasta.fai
example/attrs_custom_tensors_monoctypes.warnings.qc.log
</pre></div>
</div>
</div>
</div>
</section>
<section id="advance-developer-api">
<h3>Advance Developer API<a class="headerlink" href="#advance-developer-api" title="Link to this heading">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">DecimaResult</span></code> provides a unified interface for working with Decima results in anndata format. It contains an AnnData structure storing cell x gene expression data and metadata. Through DecimaResult, users can load pre-trained models, compute attributions to  understand genomic regulation, and analyze results through visualizations or export to genomic file formats. The object provides convenient access to cell and gene annotations through its metadata properties.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima</span><span class="w"> </span><span class="kn">import</span> <span class="n">DecimaResult</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">DecimaResult</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1877.9MB/s)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">cell_metadata</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;cell_type.str.endswith(&#39;macrophage&#39;)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell_type</th>
      <th>tissue</th>
      <th>organ</th>
      <th>disease</th>
      <th>study</th>
      <th>dataset</th>
      <th>region</th>
      <th>subregion</th>
      <th>celltype_coarse</th>
      <th>n_cells</th>
      <th>total_counts</th>
      <th>n_genes</th>
      <th>size_factor</th>
      <th>train_pearson</th>
      <th>val_pearson</th>
      <th>test_pearson</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_4063</th>
      <td>alveolar macrophage</td>
      <td>alveolar system</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>GSE155249</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>1453</td>
      <td>8.001524e+06</td>
      <td>14711</td>
      <td>36293.472025</td>
      <td>0.943059</td>
      <td>0.837210</td>
      <td>0.849998</td>
    </tr>
    <tr>
      <th>agg_4064</th>
      <td>alveolar macrophage</td>
      <td>alveolar system</td>
      <td>lung</td>
      <td>healthy</td>
      <td>GSE155249</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>1279</td>
      <td>7.598244e+06</td>
      <td>13673</td>
      <td>34158.514496</td>
      <td>0.932819</td>
      <td>0.831024</td>
      <td>0.843684</td>
    </tr>
    <tr>
      <th>agg_4065</th>
      <td>alveolar macrophage</td>
      <td>left lung</td>
      <td>lung</td>
      <td>NA</td>
      <td>ENCODE</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>405</td>
      <td>3.000961e+06</td>
      <td>16595</td>
      <td>46501.375857</td>
      <td>0.936081</td>
      <td>0.847924</td>
      <td>0.845485</td>
    </tr>
    <tr>
      <th>agg_4066</th>
      <td>alveolar macrophage</td>
      <td>lingula of left lung</td>
      <td>lung</td>
      <td>healthy</td>
      <td>a3ffde6c-7ad2-498a-903c-d58e732f7470</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>854</td>
      <td>1.713753e+06</td>
      <td>15110</td>
      <td>42773.009735</td>
      <td>0.893927</td>
      <td>0.806000</td>
      <td>0.804835</td>
    </tr>
    <tr>
      <th>agg_4067</th>
      <td>alveolar macrophage</td>
      <td>lower lobe of left lung</td>
      <td>lung</td>
      <td>NA</td>
      <td>ENCODE</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>763</td>
      <td>1.344798e+07</td>
      <td>17973</td>
      <td>49020.804487</td>
      <td>0.940586</td>
      <td>0.854680</td>
      <td>0.863014</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_6644</th>
      <td>macrophage</td>
      <td>uterus</td>
      <td>uterus</td>
      <td>healthy</td>
      <td>32f2fd23-ec74-486f-9544-e5b2f41725f5</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>425</td>
      <td>4.340830e+06</td>
      <td>15233</td>
      <td>36624.136739</td>
      <td>0.954753</td>
      <td>0.850247</td>
      <td>0.843175</td>
    </tr>
    <tr>
      <th>agg_6645</th>
      <td>macrophage</td>
      <td>uterus</td>
      <td>uterus</td>
      <td>healthy</td>
      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>231</td>
      <td>3.007554e+07</td>
      <td>14787</td>
      <td>27615.762157</td>
      <td>0.839476</td>
      <td>0.730554</td>
      <td>0.719085</td>
    </tr>
    <tr>
      <th>agg_6646</th>
      <td>macrophage</td>
      <td>vasculature</td>
      <td>vasculature</td>
      <td>healthy</td>
      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>12497</td>
      <td>4.040685e+08</td>
      <td>18199</td>
      <td>36829.498964</td>
      <td>0.938862</td>
      <td>0.836819</td>
      <td>0.833474</td>
    </tr>
    <tr>
      <th>agg_6647</th>
      <td>macrophage</td>
      <td>visceral fat</td>
      <td>adipose</td>
      <td>obesity</td>
      <td>GSE128518</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>729</td>
      <td>2.078431e+06</td>
      <td>13760</td>
      <td>34188.716187</td>
      <td>0.941596</td>
      <td>0.827360</td>
      <td>0.823912</td>
    </tr>
    <tr>
      <th>agg_6648</th>
      <td>macrophage</td>
      <td>white adipose tissue</td>
      <td>adipose</td>
      <td>NA</td>
      <td>GSE128890</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>45</td>
      <td>1.381560e+05</td>
      <td>8257</td>
      <td>27604.748095</td>
      <td>0.859386</td>
      <td>0.745328</td>
      <td>0.745539</td>
    </tr>
  </tbody>
</table>
<p>325 rows × 16 columns</p>
</div></div></div>
</div>
<p>The results and metadata stored in anndata format which you can access directly if needed but most operation are supported by DecimaResult object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">anndata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 8856 × 18457
    obs: &#39;cell_type&#39;, &#39;tissue&#39;, &#39;organ&#39;, &#39;disease&#39;, &#39;study&#39;, &#39;dataset&#39;, &#39;region&#39;, &#39;subregion&#39;, &#39;celltype_coarse&#39;, &#39;n_cells&#39;, &#39;total_counts&#39;, &#39;n_genes&#39;, &#39;size_factor&#39;, &#39;train_pearson&#39;, &#39;val_pearson&#39;, &#39;test_pearson&#39;
    var: &#39;chrom&#39;, &#39;start&#39;, &#39;end&#39;, &#39;strand&#39;, &#39;gene_type&#39;, &#39;frac_nan&#39;, &#39;mean_counts&#39;, &#39;n_tracks&#39;, &#39;gene_start&#39;, &#39;gene_end&#39;, &#39;gene_length&#39;, &#39;gene_mask_start&#39;, &#39;gene_mask_end&#39;, &#39;frac_N&#39;, &#39;fold&#39;, &#39;dataset&#39;, &#39;gene_id&#39;, &#39;pearson&#39;, &#39;size_factor_pearson&#39;, &#39;ensembl_canonical_tss&#39;
    layers: &#39;preds&#39;, &#39;v1_rep0&#39;, &#39;v1_rep1&#39;, &#39;v1_rep2&#39;, &#39;v1_rep3&#39;
</pre></div>
</div>
</div>
</div>
<p>These are the cell metadata contained in the Decima object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">cell_metadata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell_type</th>
      <th>tissue</th>
      <th>organ</th>
      <th>disease</th>
      <th>study</th>
      <th>dataset</th>
      <th>region</th>
      <th>subregion</th>
      <th>celltype_coarse</th>
      <th>n_cells</th>
      <th>total_counts</th>
      <th>n_genes</th>
      <th>size_factor</th>
      <th>train_pearson</th>
      <th>val_pearson</th>
      <th>test_pearson</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>Amygdala excitatory</td>
      <td>Amygdala_Amygdala</td>
      <td>CNS</td>
      <td>healthy</td>
      <td>jhpce#tran2021</td>
      <td>brain_atlas</td>
      <td>Amygdala</td>
      <td>Amygdala</td>
      <td>NaN</td>
      <td>331</td>
      <td>1.592883e+07</td>
      <td>17000</td>
      <td>41431.465186</td>
      <td>0.942459</td>
      <td>0.841377</td>
      <td>0.865640</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>Amygdala excitatory</td>
      <td>Amygdala_Basolateral nuclear group (BLN) - lat...</td>
      <td>CNS</td>
      <td>healthy</td>
      <td>SCR_016152</td>
      <td>brain_atlas</td>
      <td>Amygdala</td>
      <td>Basolateral nuclear group (BLN) - lateral nucl...</td>
      <td>NaN</td>
      <td>11369</td>
      <td>2.952133e+08</td>
      <td>18080</td>
      <td>40765.341481</td>
      <td>0.943098</td>
      <td>0.838936</td>
      <td>0.861092</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>Amygdala excitatory</td>
      <td>Amygdala_Bed nucleus of stria terminalis and n...</td>
      <td>CNS</td>
      <td>healthy</td>
      <td>SCR_016152</td>
      <td>brain_atlas</td>
      <td>Amygdala</td>
      <td>Bed nucleus of stria terminalis and nearby - BNST</td>
      <td>NaN</td>
      <td>139</td>
      <td>2.593231e+06</td>
      <td>15418</td>
      <td>42556.387020</td>
      <td>0.952170</td>
      <td>0.854544</td>
      <td>0.866654</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>Amygdala excitatory</td>
      <td>Amygdala_Central nuclear group - CEN</td>
      <td>CNS</td>
      <td>healthy</td>
      <td>SCR_016152</td>
      <td>brain_atlas</td>
      <td>Amygdala</td>
      <td>Central nuclear group - CEN</td>
      <td>NaN</td>
      <td>3892</td>
      <td>9.946371e+07</td>
      <td>17959</td>
      <td>42884.641430</td>
      <td>0.959744</td>
      <td>0.863585</td>
      <td>0.881554</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>Amygdala excitatory</td>
      <td>Amygdala_Corticomedial nuclear group (CMN) - a...</td>
      <td>CNS</td>
      <td>healthy</td>
      <td>SCR_016152</td>
      <td>brain_atlas</td>
      <td>Amygdala</td>
      <td>Corticomedial nuclear group (CMN) - anterior c...</td>
      <td>NaN</td>
      <td>2945</td>
      <td>1.281619e+08</td>
      <td>17885</td>
      <td>41816.741933</td>
      <td>0.951365</td>
      <td>0.854304</td>
      <td>0.868902</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>vascular associated smooth muscle cell</td>
      <td>upper lobe of right lung</td>
      <td>lung</td>
      <td>NA</td>
      <td>ENCODE</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>21</td>
      <td>3.483375e+04</td>
      <td>8515</td>
      <td>35404.911768</td>
      <td>0.735213</td>
      <td>0.665647</td>
      <td>0.654491</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>vascular associated smooth muscle cell</td>
      <td>urinary bladder</td>
      <td>urinary</td>
      <td>healthy</td>
      <td>GSE129845</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>24</td>
      <td>8.498500e+04</td>
      <td>7337</td>
      <td>26189.415789</td>
      <td>0.809852</td>
      <td>0.690022</td>
      <td>0.656160</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>vascular associated smooth muscle cell</td>
      <td>uterus</td>
      <td>uterus</td>
      <td>NA</td>
      <td>ENCODE</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>272</td>
      <td>5.700762e+05</td>
      <td>14769</td>
      <td>44938.403867</td>
      <td>0.915329</td>
      <td>0.808941</td>
      <td>0.839993</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>vascular associated smooth muscle cell</td>
      <td>uterus</td>
      <td>uterus</td>
      <td>healthy</td>
      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>472</td>
      <td>1.089170e+07</td>
      <td>14514</td>
      <td>30145.422152</td>
      <td>0.852339</td>
      <td>0.717682</td>
      <td>0.727469</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>vascular associated smooth muscle cell</td>
      <td>vasculature</td>
      <td>vasculature</td>
      <td>healthy</td>
      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>
      <td>scimilarity</td>
      <td>nan</td>
      <td>nan</td>
      <td>NaN</td>
      <td>1853</td>
      <td>5.992697e+07</td>
      <td>16764</td>
      <td>36464.273371</td>
      <td>0.909855</td>
      <td>0.780413</td>
      <td>0.796351</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 16 columns</p>
</div></div></div>
</div>
<p>Similarly, these are the gene metadata contained in the Decima object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">gene_metadata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>gene_type</th>
      <th>frac_nan</th>
      <th>mean_counts</th>
      <th>n_tracks</th>
      <th>gene_start</th>
      <th>gene_end</th>
      <th>gene_length</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
      <th>frac_N</th>
      <th>fold</th>
      <th>dataset</th>
      <th>gene_id</th>
      <th>pearson</th>
      <th>size_factor_pearson</th>
      <th>ensembl_canonical_tss</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>STRADA</th>
      <td>chr17</td>
      <td>63381538</td>
      <td>63905826</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>0.000000</td>
      <td>2.208074</td>
      <td>7616</td>
      <td>63682336</td>
      <td>63741986</td>
      <td>59650</td>
      <td>163840</td>
      <td>223490</td>
      <td>0.000000</td>
      <td>['fold1']</td>
      <td>train</td>
      <td>ENSG00000266173</td>
      <td>0.469923</td>
      <td>0.476627</td>
      <td>63741799.0</td>
    </tr>
    <tr>
      <th>ETV4</th>
      <td>chr17</td>
      <td>43219172</td>
      <td>43743460</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>0.030873</td>
      <td>0.925863</td>
      <td>5004</td>
      <td>43527844</td>
      <td>43579620</td>
      <td>51776</td>
      <td>163840</td>
      <td>215616</td>
      <td>0.000000</td>
      <td>['fold1']</td>
      <td>train</td>
      <td>ENSG00000175832</td>
      <td>0.738092</td>
      <td>0.613281</td>
      <td>43546340.0</td>
    </tr>
    <tr>
      <th>USP25</th>
      <td>chr21</td>
      <td>15566185</td>
      <td>16090473</td>
      <td>+</td>
      <td>protein_coding</td>
      <td>0.000000</td>
      <td>3.650355</td>
      <td>8604</td>
      <td>15730025</td>
      <td>15880069</td>
      <td>150044</td>
      <td>163840</td>
      <td>313884</td>
      <td>0.000000</td>
      <td>['fold6']</td>
      <td>train</td>
      <td>ENSG00000155313</td>
      <td>0.905222</td>
      <td>0.784446</td>
      <td>15729982.0</td>
    </tr>
    <tr>
      <th>ZSWIM5</th>
      <td>chr1</td>
      <td>44945761</td>
      <td>45470049</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>0.000620</td>
      <td>2.190115</td>
      <td>6123</td>
      <td>45016399</td>
      <td>45306209</td>
      <td>289810</td>
      <td>163840</td>
      <td>453650</td>
      <td>0.000000</td>
      <td>['fold5']</td>
      <td>train</td>
      <td>ENSG00000162415</td>
      <td>0.961772</td>
      <td>0.795131</td>
      <td>45206605.0</td>
    </tr>
    <tr>
      <th>C21orf58</th>
      <td>chr21</td>
      <td>45963427</td>
      <td>46487715</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>0.000791</td>
      <td>1.650467</td>
      <td>7354</td>
      <td>46300181</td>
      <td>46323875</td>
      <td>23694</td>
      <td>163840</td>
      <td>187534</td>
      <td>0.000000</td>
      <td>['fold6']</td>
      <td>train</td>
      <td>ENSG00000160298</td>
      <td>0.645268</td>
      <td>0.412368</td>
      <td>46323870.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>NPDC1</th>
      <td>chr9</td>
      <td>136685731</td>
      <td>137210019</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>0.000000</td>
      <td>2.625285</td>
      <td>7852</td>
      <td>137039463</td>
      <td>137046179</td>
      <td>6716</td>
      <td>163840</td>
      <td>170556</td>
      <td>0.000000</td>
      <td>['fold3']</td>
      <td>test</td>
      <td>ENSG00000107281</td>
      <td>0.316322</td>
      <td>0.178204</td>
      <td>137046177.0</td>
    </tr>
    <tr>
      <th>ZNF425</th>
      <td>chr7</td>
      <td>148765876</td>
      <td>149290164</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>0.001048</td>
      <td>1.292957</td>
      <td>6511</td>
      <td>149102784</td>
      <td>149126324</td>
      <td>23540</td>
      <td>163840</td>
      <td>187380</td>
      <td>0.000000</td>
      <td>['fold7']</td>
      <td>train</td>
      <td>ENSG00000204947</td>
      <td>0.821292</td>
      <td>0.737081</td>
      <td>149126324.0</td>
    </tr>
    <tr>
      <th>COL5A1</th>
      <td>chr9</td>
      <td>134477934</td>
      <td>135002222</td>
      <td>+</td>
      <td>protein_coding</td>
      <td>0.002159</td>
      <td>1.492664</td>
      <td>6209</td>
      <td>134641774</td>
      <td>134844843</td>
      <td>203069</td>
      <td>163840</td>
      <td>366909</td>
      <td>0.000000</td>
      <td>['fold3']</td>
      <td>test</td>
      <td>ENSG00000130635</td>
      <td>0.766624</td>
      <td>0.456999</td>
      <td>134641803.0</td>
    </tr>
    <tr>
      <th>BRD3</th>
      <td>chr9</td>
      <td>133708087</td>
      <td>134232375</td>
      <td>-</td>
      <td>protein_coding</td>
      <td>0.000000</td>
      <td>3.190450</td>
      <td>8675</td>
      <td>134030305</td>
      <td>134068535</td>
      <td>38230</td>
      <td>163840</td>
      <td>202070</td>
      <td>0.004662</td>
      <td>['fold3']</td>
      <td>test</td>
      <td>ENSG00000169925</td>
      <td>0.344062</td>
      <td>0.280283</td>
      <td>134068026.0</td>
    </tr>
    <tr>
      <th>EVI5L</th>
      <td>chr19</td>
      <td>7666393</td>
      <td>8190681</td>
      <td>+</td>
      <td>protein_coding</td>
      <td>0.000000</td>
      <td>1.959605</td>
      <td>7570</td>
      <td>7830233</td>
      <td>7864976</td>
      <td>34743</td>
      <td>163840</td>
      <td>198583</td>
      <td>0.000000</td>
      <td>['fold3']</td>
      <td>test</td>
      <td>ENSG00000142459</td>
      <td>0.810152</td>
      <td>0.704828</td>
      <td>7830218.0</td>
    </tr>
  </tbody>
</table>
<p>18457 rows × 20 columns</p>
</div></div></div>
</div>
<p>You can also access the genes and cells:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">genes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;STRADA&#39;, &#39;ETV4&#39;, &#39;USP25&#39;, &#39;ZSWIM5&#39;, &#39;C21orf58&#39;, &#39;MIR497HG&#39;, &#39;CFAP74&#39;,
       &#39;GSE1&#39;, &#39;LPP&#39;, &#39;CLK1&#39;,
       ...
       &#39;STRIP2&#39;, &#39;TNFRSF1A&#39;, &#39;RBM14-RBM4&#39;, &#39;C1orf21&#39;, &#39;LINC00639&#39;, &#39;NPDC1&#39;,
       &#39;ZNF425&#39;, &#39;COL5A1&#39;, &#39;BRD3&#39;, &#39;EVI5L&#39;],
      dtype=&#39;object&#39;, length=18457)
</pre></div>
</div>
</div>
</div>
<p>Cell indexes can be also accessed:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">cells</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;agg_0&#39;, &#39;agg_1&#39;, &#39;agg_2&#39;, &#39;agg_3&#39;, &#39;agg_4&#39;, &#39;agg_5&#39;, &#39;agg_6&#39;, &#39;agg_7&#39;,
       &#39;agg_8&#39;, &#39;agg_9&#39;,
       ...
       &#39;agg_9528&#39;, &#39;agg_9529&#39;, &#39;agg_9530&#39;, &#39;agg_9531&#39;, &#39;agg_9532&#39;, &#39;agg_9533&#39;,
       &#39;agg_9535&#39;, &#39;agg_9536&#39;, &#39;agg_9537&#39;, &#39;agg_9538&#39;],
      dtype=&#39;object&#39;, length=8856)
</pre></div>
</div>
</div>
</div>
<p>Predicted gene expression for specific gene can be accessed:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">predicted_expression_matrix</span><span class="p">(</span><span class="n">genes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SPI1&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SPI1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>0.256442</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>0.221014</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>0.179371</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>0.219646</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>0.217516</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>0.493780</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>0.292091</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>0.370765</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>0.168036</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>0.239733</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 1 columns</p>
</div></div></div>
</div>
<p>Or for all the genes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">predicted_expression_matrix</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>STRADA</th>
      <th>ETV4</th>
      <th>USP25</th>
      <th>ZSWIM5</th>
      <th>C21orf58</th>
      <th>MIR497HG</th>
      <th>CFAP74</th>
      <th>GSE1</th>
      <th>LPP</th>
      <th>CLK1</th>
      <th>...</th>
      <th>STRIP2</th>
      <th>TNFRSF1A</th>
      <th>RBM14-RBM4</th>
      <th>C1orf21</th>
      <th>LINC00639</th>
      <th>NPDC1</th>
      <th>ZNF425</th>
      <th>COL5A1</th>
      <th>BRD3</th>
      <th>EVI5L</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>2.973438</td>
      <td>1.845565</td>
      <td>4.592531</td>
      <td>5.099802</td>
      <td>1.774879</td>
      <td>0.356812</td>
      <td>2.590836</td>
      <td>4.629774</td>
      <td>4.897171</td>
      <td>3.326940</td>
      <td>...</td>
      <td>2.836060</td>
      <td>0.297015</td>
      <td>1.883849</td>
      <td>4.293593</td>
      <td>1.463565</td>
      <td>3.183534</td>
      <td>2.340202</td>
      <td>2.374942</td>
      <td>2.911916</td>
      <td>3.230072</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>2.954213</td>
      <td>1.896726</td>
      <td>4.688557</td>
      <td>5.510440</td>
      <td>1.666929</td>
      <td>0.352725</td>
      <td>2.292625</td>
      <td>4.459535</td>
      <td>4.915286</td>
      <td>3.192858</td>
      <td>...</td>
      <td>3.125704</td>
      <td>0.242543</td>
      <td>1.908177</td>
      <td>4.439424</td>
      <td>1.236739</td>
      <td>3.494824</td>
      <td>2.425672</td>
      <td>2.054568</td>
      <td>2.713408</td>
      <td>3.491463</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>2.938851</td>
      <td>2.197247</td>
      <td>4.861410</td>
      <td>5.617520</td>
      <td>1.773381</td>
      <td>0.380867</td>
      <td>2.394917</td>
      <td>4.415038</td>
      <td>4.836399</td>
      <td>3.390717</td>
      <td>...</td>
      <td>3.082098</td>
      <td>0.263285</td>
      <td>2.006456</td>
      <td>4.383455</td>
      <td>1.208590</td>
      <td>4.013819</td>
      <td>2.408381</td>
      <td>2.297343</td>
      <td>2.892222</td>
      <td>3.695785</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>3.045972</td>
      <td>2.138573</td>
      <td>4.863791</td>
      <td>5.273604</td>
      <td>1.760097</td>
      <td>0.463555</td>
      <td>2.391702</td>
      <td>3.940975</td>
      <td>4.857763</td>
      <td>3.410926</td>
      <td>...</td>
      <td>2.882890</td>
      <td>0.290327</td>
      <td>1.922963</td>
      <td>4.550189</td>
      <td>1.430520</td>
      <td>3.693118</td>
      <td>2.297103</td>
      <td>2.121887</td>
      <td>2.626117</td>
      <td>3.223912</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>3.025518</td>
      <td>2.019096</td>
      <td>4.602948</td>
      <td>5.257001</td>
      <td>1.755338</td>
      <td>0.382190</td>
      <td>2.432810</td>
      <td>4.392480</td>
      <td>4.959488</td>
      <td>3.250500</td>
      <td>...</td>
      <td>3.082296</td>
      <td>0.258540</td>
      <td>2.038277</td>
      <td>4.464807</td>
      <td>1.249043</td>
      <td>3.665800</td>
      <td>2.400820</td>
      <td>2.255862</td>
      <td>2.925619</td>
      <td>3.471005</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>2.333562</td>
      <td>0.633322</td>
      <td>4.675825</td>
      <td>2.793023</td>
      <td>0.752030</td>
      <td>0.692083</td>
      <td>0.503531</td>
      <td>4.327948</td>
      <td>6.903193</td>
      <td>3.695593</td>
      <td>...</td>
      <td>0.549795</td>
      <td>2.270181</td>
      <td>1.563218</td>
      <td>4.395422</td>
      <td>0.550088</td>
      <td>1.330252</td>
      <td>1.044471</td>
      <td>3.759369</td>
      <td>2.491346</td>
      <td>1.872717</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>0.835037</td>
      <td>0.358773</td>
      <td>1.964896</td>
      <td>0.307449</td>
      <td>0.337240</td>
      <td>0.834196</td>
      <td>0.093885</td>
      <td>1.853794</td>
      <td>3.700790</td>
      <td>4.467302</td>
      <td>...</td>
      <td>0.176885</td>
      <td>1.370898</td>
      <td>1.022708</td>
      <td>3.400267</td>
      <td>0.052162</td>
      <td>1.908870</td>
      <td>0.253417</td>
      <td>1.448111</td>
      <td>1.622033</td>
      <td>1.064292</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>3.008039</td>
      <td>1.209324</td>
      <td>4.798392</td>
      <td>3.931870</td>
      <td>1.401328</td>
      <td>1.638555</td>
      <td>0.969720</td>
      <td>4.779201</td>
      <td>6.631931</td>
      <td>4.127797</td>
      <td>...</td>
      <td>1.174298</td>
      <td>1.870530</td>
      <td>2.506874</td>
      <td>5.151776</td>
      <td>0.967644</td>
      <td>1.809947</td>
      <td>2.205356</td>
      <td>4.244005</td>
      <td>2.974467</td>
      <td>2.659873</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>1.241936</td>
      <td>0.455059</td>
      <td>2.919995</td>
      <td>0.571672</td>
      <td>0.486448</td>
      <td>1.175586</td>
      <td>0.145397</td>
      <td>2.412148</td>
      <td>4.759118</td>
      <td>4.913945</td>
      <td>...</td>
      <td>0.371035</td>
      <td>1.361073</td>
      <td>1.668085</td>
      <td>4.005738</td>
      <td>0.078611</td>
      <td>1.571750</td>
      <td>0.508187</td>
      <td>2.067150</td>
      <td>2.323764</td>
      <td>1.429850</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>1.715507</td>
      <td>0.700955</td>
      <td>3.044732</td>
      <td>0.858696</td>
      <td>0.903406</td>
      <td>1.763168</td>
      <td>0.215304</td>
      <td>2.604478</td>
      <td>4.549708</td>
      <td>4.839124</td>
      <td>...</td>
      <td>0.594310</td>
      <td>1.801298</td>
      <td>2.075996</td>
      <td>3.933860</td>
      <td>0.165590</td>
      <td>1.970268</td>
      <td>0.993521</td>
      <td>2.232347</td>
      <td>2.473388</td>
      <td>1.902884</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 18457 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span><span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1332.8MB/s)
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DecimaResult(anndata=AnnData object with n_obs × n_vars = 8856 × 18457
    obs: &#39;cell_type&#39;, &#39;tissue&#39;, &#39;organ&#39;, &#39;disease&#39;, &#39;study&#39;, &#39;dataset&#39;, &#39;region&#39;, &#39;subregion&#39;, &#39;celltype_coarse&#39;, &#39;n_cells&#39;, &#39;total_counts&#39;, &#39;n_genes&#39;, &#39;size_factor&#39;, &#39;train_pearson&#39;, &#39;val_pearson&#39;, &#39;test_pearson&#39;
    var: &#39;chrom&#39;, &#39;start&#39;, &#39;end&#39;, &#39;strand&#39;, &#39;gene_type&#39;, &#39;frac_nan&#39;, &#39;mean_counts&#39;, &#39;n_tracks&#39;, &#39;gene_start&#39;, &#39;gene_end&#39;, &#39;gene_length&#39;, &#39;gene_mask_start&#39;, &#39;gene_mask_end&#39;, &#39;frac_N&#39;, &#39;fold&#39;, &#39;dataset&#39;, &#39;gene_id&#39;, &#39;pearson&#39;, &#39;size_factor_pearson&#39;, &#39;ensembl_canonical_tss&#39;
    layers: &#39;preds&#39;, &#39;v1_rep0&#39;, &#39;v1_rep1&#39;, &#39;v1_rep2&#39;, &#39;v1_rep3&#39;)
</pre></div>
</div>
</div>
</div>
<p>Prepare an input for th SPI1 genes.</p>
<p>Takes around ~10 seconds on GPU and ~5 minutes to call attributions on CPU.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">attrs</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">attributions</span><span class="p">(</span>
    <span class="n">gene</span><span class="o">=</span><span class="s2">&quot;SPI1&quot;</span><span class="p">,</span>
    <span class="n">tasks</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">query_cells</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cell_type in </span><span class="si">{</span><span class="n">spi1_cell_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">),</span>
    <span class="n">off_tasks</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">query_cells</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;organ == &quot;blood&quot; and cell_type not in </span><span class="si">{</span><span class="n">spi1_cell_types</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Attributions can be visualized and processed with attributions object:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">attrs</span><span class="o">.</span><span class="n">peaks</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peak</th>
      <th>start</th>
      <th>end</th>
      <th>attribution</th>
      <th>p-value</th>
      <th>from_tss</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>pos.SPI1@37</td>
      <td>163877</td>
      <td>163902</td>
      <td>12.738598</td>
      <td>2.189116e-11</td>
      <td>37</td>
    </tr>
    <tr>
      <th>1</th>
      <td>pos.SPI1@-121</td>
      <td>163719</td>
      <td>163744</td>
      <td>5.578619</td>
      <td>1.900737e-05</td>
      <td>-121</td>
    </tr>
    <tr>
      <th>2</th>
      <td>pos.SPI1@-57</td>
      <td>163783</td>
      <td>163803</td>
      <td>9.269885</td>
      <td>3.052090e-05</td>
      <td>-57</td>
    </tr>
    <tr>
      <th>3</th>
      <td>pos.SPI1@62</td>
      <td>163902</td>
      <td>163909</td>
      <td>1.285628</td>
      <td>3.080049e-05</td>
      <td>62</td>
    </tr>
    <tr>
      <th>4</th>
      <td>pos.SPI1@-79</td>
      <td>163761</td>
      <td>163765</td>
      <td>0.832053</td>
      <td>6.105891e-05</td>
      <td>-79</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>72</th>
      <td>neg.SPI1@-22</td>
      <td>163818</td>
      <td>163836</td>
      <td>-11.178307</td>
      <td>4.889269e-04</td>
      <td>-22</td>
    </tr>
    <tr>
      <th>73</th>
      <td>neg.SPI1@443</td>
      <td>164283</td>
      <td>164293</td>
      <td>-0.718594</td>
      <td>4.918033e-04</td>
      <td>443</td>
    </tr>
    <tr>
      <th>74</th>
      <td>neg.SPI1@32783</td>
      <td>196623</td>
      <td>196630</td>
      <td>-0.460660</td>
      <td>4.924140e-04</td>
      <td>32783</td>
    </tr>
    <tr>
      <th>75</th>
      <td>neg.SPI1@-21806</td>
      <td>142034</td>
      <td>142040</td>
      <td>-0.328668</td>
      <td>4.924140e-04</td>
      <td>-21806</td>
    </tr>
    <tr>
      <th>76</th>
      <td>neg.SPI1@31668</td>
      <td>195508</td>
      <td>195512</td>
      <td>-0.212833</td>
      <td>4.924140e-04</td>
      <td>31668</td>
    </tr>
  </tbody>
</table>
<p>136 rows × 6 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">attrs</span><span class="o">.</span><span class="n">peaks_to_bed</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>start</th>
      <th>end</th>
      <th>name</th>
      <th>score</th>
      <th>strand</th>
      <th>attribution</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>41</th>
      <td>chr11</td>
      <td>47216350</td>
      <td>47216357</td>
      <td>pos.SPI1@162219</td>
      <td>3</td>
      <td>.</td>
      <td>0.542525</td>
    </tr>
    <tr>
      <th>45</th>
      <td>chr11</td>
      <td>47257597</td>
      <td>47257605</td>
      <td>pos.SPI1@120971</td>
      <td>3</td>
      <td>.</td>
      <td>0.679172</td>
    </tr>
    <tr>
      <th>65</th>
      <td>chr11</td>
      <td>47257633</td>
      <td>47257637</td>
      <td>neg.SPI1@120939</td>
      <td>3</td>
      <td>.</td>
      <td>-0.220710</td>
    </tr>
    <tr>
      <th>62</th>
      <td>chr11</td>
      <td>47257734</td>
      <td>47257739</td>
      <td>neg.SPI1@120837</td>
      <td>3</td>
      <td>.</td>
      <td>-0.273576</td>
    </tr>
    <tr>
      <th>44</th>
      <td>chr11</td>
      <td>47345731</td>
      <td>47345736</td>
      <td>neg.SPI1@32840</td>
      <td>3</td>
      <td>.</td>
      <td>-0.296525</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>70</th>
      <td>chr11</td>
      <td>47395483</td>
      <td>47395492</td>
      <td>neg.SPI1@-16916</td>
      <td>3</td>
      <td>.</td>
      <td>-0.567591</td>
    </tr>
    <tr>
      <th>40</th>
      <td>chr11</td>
      <td>47400211</td>
      <td>47400221</td>
      <td>neg.SPI1@-21645</td>
      <td>3</td>
      <td>.</td>
      <td>-0.898548</td>
    </tr>
    <tr>
      <th>38</th>
      <td>chr11</td>
      <td>47400225</td>
      <td>47400235</td>
      <td>neg.SPI1@-21659</td>
      <td>3</td>
      <td>.</td>
      <td>-0.727634</td>
    </tr>
    <tr>
      <th>75</th>
      <td>chr11</td>
      <td>47400376</td>
      <td>47400382</td>
      <td>neg.SPI1@-21806</td>
      <td>3</td>
      <td>.</td>
      <td>-0.328668</td>
    </tr>
    <tr>
      <th>58</th>
      <td>chr11</td>
      <td>47400703</td>
      <td>47400709</td>
      <td>neg.SPI1@-22133</td>
      <td>3</td>
      <td>.</td>
      <td>-0.327103</td>
    </tr>
  </tbody>
</table>
<p>136 rows × 7 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">attrs</span><span class="o">.</span><span class="n">plot_peaks</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/bb32b1c7540af7b5de7e23e7c2a3a46e06932cf72ca407b0baecdc6cdd532928.png"><img alt="../_images/bb32b1c7540af7b5de7e23e7c2a3a46e06932cf72ca407b0baecdc6cdd532928.png" src="../_images/bb32b1c7540af7b5de7e23e7c2a3a46e06932cf72ca407b0baecdc6cdd532928.png" style="width: 1000px; height: 200px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="n">attrs</span><span class="o">.</span><span class="n">plot_seqlogo</span><span class="p">(</span><span class="n">relative_loc</span><span class="o">=-</span><span class="mi">45</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/0a0147bfe257595d918a0c2359b5972a0475f37ffdd99865bd5a74e29783108b.png" src="../_images/0a0147bfe257595d918a0c2359b5972a0475f37ffdd99865bd5a74e29783108b.png" />
</div>
</div>
<p>This comment takes around ~1 minutes and detects motifs in the attributions using FIMO. The motifs are ranked by their attribution scores:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_motifs</span> <span class="o">=</span> <span class="n">attrs</span><span class="o">.</span><span class="n">scan_motifs</span><span class="p">()</span>
<span class="n">df_motifs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>motif</th>
      <th>peak</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>score</th>
      <th>p-value</th>
      <th>matched_seq</th>
      <th>site_attr_score</th>
      <th>motif_attr_score</th>
      <th>from_tss</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2918</th>
      <td>KLF7.H13CORE.0.P.B</td>
      <td>neg.SPI1@1813</td>
      <td>165662</td>
      <td>165672</td>
      <td>+</td>
      <td>15.217368</td>
      <td>0.000000e+00</td>
      <td>GGGGGCGGGG</td>
      <td>0.009462</td>
      <td>0.027049</td>
      <td>1822</td>
    </tr>
    <tr>
      <th>7458</th>
      <td>KLF7.H13CORE.0.P.B</td>
      <td>pos.SPI1@1820</td>
      <td>165662</td>
      <td>165672</td>
      <td>+</td>
      <td>15.217368</td>
      <td>0.000000e+00</td>
      <td>GGGGGCGGGG</td>
      <td>0.009462</td>
      <td>0.027049</td>
      <td>1822</td>
    </tr>
    <tr>
      <th>2992</th>
      <td>KLF7.H13CORE.0.P.B</td>
      <td>neg.SPI1@1832</td>
      <td>165662</td>
      <td>165672</td>
      <td>+</td>
      <td>15.217368</td>
      <td>0.000000e+00</td>
      <td>GGGGGCGGGG</td>
      <td>0.009462</td>
      <td>0.027049</td>
      <td>1822</td>
    </tr>
    <tr>
      <th>2068</th>
      <td>RXRB.H13CORE.2.PS.A</td>
      <td>neg.SPI1@1182</td>
      <td>165019</td>
      <td>165030</td>
      <td>-</td>
      <td>12.213856</td>
      <td>0.000000e+00</td>
      <td>CCATGACCTCT</td>
      <td>-0.008296</td>
      <td>-0.024166</td>
      <td>1179</td>
    </tr>
    <tr>
      <th>3882</th>
      <td>ZNF746.H13CORE.0.PSG.A</td>
      <td>neg.SPI1@1917</td>
      <td>165744</td>
      <td>165770</td>
      <td>-</td>
      <td>26.355452</td>
      <td>7.435328e-10</td>
      <td>AGGGAGGAGGGAGGAAGGTGGGAGGA</td>
      <td>-0.010776</td>
      <td>-0.016254</td>
      <td>1904</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5618</th>
      <td>ZN394.H13CORE.0.P.C</td>
      <td>pos.SPI1@-144</td>
      <td>163691</td>
      <td>163710</td>
      <td>-</td>
      <td>6.136955</td>
      <td>4.997442e-04</td>
      <td>TCCCAGCTTCCTCTGGGCA</td>
      <td>0.028328</td>
      <td>0.048309</td>
      <td>-149</td>
    </tr>
    <tr>
      <th>5894</th>
      <td>ZN394.H13CORE.0.P.C</td>
      <td>pos.SPI1@-150</td>
      <td>163691</td>
      <td>163710</td>
      <td>-</td>
      <td>6.136955</td>
      <td>4.997442e-04</td>
      <td>TCCCAGCTTCCTCTGGGCA</td>
      <td>0.028328</td>
      <td>0.048309</td>
      <td>-149</td>
    </tr>
    <tr>
      <th>3069</th>
      <td>ZNF30.H13CORE.0.P.C</td>
      <td>neg.SPI1@1832</td>
      <td>165664</td>
      <td>165687</td>
      <td>+</td>
      <td>7.201793</td>
      <td>4.997881e-04</td>
      <td>GGGCGGGGAGGCCTGATGACTGT</td>
      <td>0.008337</td>
      <td>0.010769</td>
      <td>1824</td>
    </tr>
    <tr>
      <th>4702</th>
      <td>ZN543.H13CORE.0.P.C</td>
      <td>neg.SPI1@19730</td>
      <td>183562</td>
      <td>183583</td>
      <td>-</td>
      <td>6.760764</td>
      <td>4.998900e-04</td>
      <td>TGGGGAGACAGGTGGGGCACC</td>
      <td>-0.005565</td>
      <td>-0.008536</td>
      <td>19722</td>
    </tr>
    <tr>
      <th>1373</th>
      <td>CREB3.H13CORE.0.SM.B</td>
      <td>neg.SPI1@-22</td>
      <td>163819</td>
      <td>163833</td>
      <td>+</td>
      <td>1.754682</td>
      <td>4.999302e-04</td>
      <td>GCGGTGATGTCACC</td>
      <td>-0.206415</td>
      <td>-0.585105</td>
      <td>-21</td>
    </tr>
  </tbody>
</table>
<p>8595 rows × 11 columns</p>
</div></div></div>
</div>
<p>If you just want attribution tensor from input one_hot encoded sequence prepare your input and call attributions object:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">one_hot_seq</span><span class="p">,</span> <span class="n">gene_mask</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">prepare_one_hot</span><span class="p">(</span><span class="s2">&quot;SPI1&quot;</span><span class="p">)</span>
<span class="n">inputs</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">one_hot_seq</span><span class="p">,</span> <span class="n">gene_mask</span><span class="p">])</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">inputs</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># (batch_size, 5, seq_len)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>torch.Size([1, 5, 524288])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.interpret.attributer</span><span class="w"> </span><span class="kn">import</span> <span class="n">DecimaAttributer</span>

<span class="n">attributer</span> <span class="o">=</span> <span class="n">DecimaAttributer</span><span class="p">(</span>
    <span class="n">model</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">model</span><span class="p">,</span>
    <span class="n">tasks</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">query_cells</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cell_type in </span><span class="si">{</span><span class="n">spi1_cell_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">),</span>
    <span class="n">off_tasks</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">query_cells</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;organ == &quot;blood&quot; and cell_type not in </span><span class="si">{</span><span class="n">spi1_cell_types</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">),</span>
    <span class="n">transform</span><span class="o">=</span><span class="s2">&quot;specificity&quot;</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s2">&quot;inputxgradient&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">attrs</span> <span class="o">=</span> <span class="n">attributer</span><span class="o">.</span><span class="n">attribute</span><span class="p">(</span><span class="n">inputs</span><span class="o">=</span><span class="n">inputs</span><span class="p">)</span>

<span class="n">attrs</span>  <span class="c1"># (batch_size, 4, seq_len) gene mask is removed from final attributions</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>tensor([[[-0.0000e+00,  0.0000e+00, -0.0000e+00,  ..., -0.0000e+00,
           0.0000e+00,  0.0000e+00],
         [-0.0000e+00, -0.0000e+00, -2.7323e-05,  ...,  0.0000e+00,
           0.0000e+00,  0.0000e+00],
         [-0.0000e+00,  1.2544e-04,  0.0000e+00,  ...,  3.7231e-05,
          -0.0000e+00,  1.4978e-05],
         [ 1.7231e-04, -0.0000e+00,  0.0000e+00,  ..., -0.0000e+00,
           1.2230e-05, -0.0000e+00]]], device=&#39;cuda:0&#39;)
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="2-variant-effect-prediction.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Variant Effect Prediction with Decima</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../api/decima.vep.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">decima.vep package</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Gokcen Eraslan
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Attribution and Motifs Detection with Decima</a><ul>
<li><a class="reference internal" href="#cli-api">CLI API</a><ul>
<li><a class="reference internal" href="#cli-subcommands">CLI Subcommands</a></li>
<li><a class="reference internal" href="#querying-cells">Querying Cells</a></li>
<li><a class="reference internal" href="#attribution-calling-with-custom-genes-and-sequences">Attribution calling with custom genes and sequences</a></li>
</ul>
</li>
<li><a class="reference internal" href="#python-user-api">Python User API</a><ul>
<li><a class="reference internal" href="#custom-sequences">Custom Sequences</a></li>
<li><a class="reference internal" href="#advance-developer-api">Advance Developer API</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=3f1b9271"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>