<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html"><link rel="search" title="Search" href="../search.html"><link rel="next" title="Fine-tuning Borzoi to create a Decima model" href="3-finetune.html"><link rel="prev" title="Attribution and Motifs Detection with Decima" href="1-attribution-motif-discovery.html">

    <!-- Generated with Sphinx 8.2.3 and Furo 2025.12.19 -->
        <title>Variant Effect Prediction with Decima - decima 0.5.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=7bdb33bb" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">decima 0.5.2 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <span class="sidebar-brand-text">decima 0.5.2 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/decima.html">decima package</a><input aria-label="Toggle navigation of decima package" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.cli.html">decima.cli package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.core.html">decima.core package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.data.html">decima.data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.hub.html">decima.hub package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.interpret.html">decima.interpret package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.model.html">decima.model package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.plot.html">decima.plot package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.tools.html">decima.tools package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.train.html">decima.train package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.utils.html">decima.utils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.vep.html">decima.vep package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="1-attribution-motif-discovery.html">Tutorial: Attribution Motif Discovery</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Tutorial: Variant Effect Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="3-finetune.html">Tutorial: Finetuning</a></li>
<li class="toctree-l1"><a class="reference internal" href="4-modisco.html">Tutorial: Modisco</a></li>
<li class="toctree-l1"><a class="reference internal" href="5-gene-expression-prediction.html">Tutorial: Gene Expression Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/2-variant-effect-prediction.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section class="tex2jax_ignore mathjax_ignore" id="variant-effect-prediction-with-decima">
<h1>Variant Effect Prediction with Decima<a class="headerlink" href="#variant-effect-prediction-with-decima" title="Link to this heading">¶</a></h1>
<p>Decima’s Variant Effect Prediction (VEP) module allows you to predict the effects of genetic variants on gene expression. This tutorial demonstrates how to use the VEP functionality through both command-line interface (CLI) and Python API. The VEP module takes variant file as input (in TSV or VCF format) and predicts their effects on gene expression across different cell types and tissues if provided.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;CUDA_VISIBLE_DEVICES&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;0&quot;</span>
</pre></div>
</div>
</div>
</div>
<section id="cli-api">
<h2>CLI API<a class="headerlink" href="#cli-api" title="Link to this heading">¶</a></h2>
<p>CLI API for variant effect prediction on gene expression.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>vep<span class="w"> </span>--help
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Usage: decima vep [OPTIONS]

  Predict variant effect and save to parquet

  Examples:

      &gt;&gt;&gt; decima vep -v &quot;data/sample.vcf&quot; -o &quot;vep_results.parquet&quot;

      &gt;&gt;&gt; decima vep -v &quot;data/sample.vcf&quot; -o &quot;vep_results.parquet&quot; --tasks
      &quot;cell_type == &#39;classical monocyte&#39;&quot; # only predict for classical
      monocytes

      &gt;&gt;&gt; decima vep -v &quot;data/sample.vcf&quot; -o &quot;vep_results.parquet&quot; --device 0
      # use device gpu device 0

      &gt;&gt;&gt; decima vep -v &quot;data/sample.vcf&quot; -o &quot;vep_results.parquet&quot; --include-
      cols &quot;gene_name,gene_id&quot; # include gene_name and gene_id columns in the
      output

      &gt;&gt;&gt; decima vep -v &quot;data/sample.vcf&quot; -o &quot;vep_results.parquet&quot; --gene-col
      &quot;gene_name&quot; # use gene_name column as gene names if these option passed
      genes and variants mapped based on these column not based on the genomic
      locus based on the annotaiton.

Options:
  -v, --variants PATH       Path to the variant file .vcf file
  -o, --output_pq PATH      Path to the output parquet file.
  --tasks TEXT              Tasks to predict. If not provided, all tasks will
                            be predicted.
  --chunksize INTEGER       Number of variants to process in each chunk.
                            Loading variants in chunks is more memory
                            efficient.This chuck of variants will be process
                            and saved to output parquet file before contineus
                            to next chunk. Default: 10_000.
  --model INTEGER           Model to use for variant effect prediction either
                            replicate number or path to the model.
  --device TEXT             Device to use. Default: None which automatically
                            selects the best device.
  --batch-size INTEGER      Batch size for the model. Default: 1.
  --num-workers INTEGER     Number of workers for the loader. Default: 1.
  --max-distance FLOAT      Maximum distance from the TSS. Default: 524288.
  --max-distance-type TEXT  Type of maximum distance. Default: tss.
  --include-cols TEXT       Columns to include in the output in the original
                            tsv file to include in the output parquet file.
                            Default: None.
  --gene-col TEXT           Column name for gene names. Default: None.
  --genome TEXT             Genome build. Default: hg38.
  --help                    Show this message and exit.

</pre></div>
</div>
</div>
</div>
<p>The VEP module takes a VCF file as input, identifies variants near genes, and predicts their effects on gene expression in a cell type-specific manner. The results are saved as a parquet file containing the following columns:</p>
<ul class="simple">
<li><p>chrom: Chromosome where the variant is located</p></li>
<li><p>pos: Genomic position of the variant</p></li>
<li><p>ref: Reference allele</p></li>
<li><p>alt: Alternative allele</p></li>
<li><p>gene: Gene name</p></li>
<li><p>start: Gene start position</p></li>
<li><p>end: Gene end position</p></li>
<li><p>strand: Gene strand</p></li>
<li><p>gene_mask_start: Start position of gene mask</p></li>
<li><p>gene_mask_end: End position of gene mask</p></li>
<li><p>rel_pos: Relative position within gene</p></li>
<li><p>ref_tx: Reference transcript</p></li>
<li><p>alt_tx: Alternative transcript</p></li>
<li><p>tss_dist: Distance to transcription start site</p></li>
<li><p>cell_0, cell_1, etc.: Predicted gene expression changes for each cell type</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>vep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;data/sample.vcf&quot;</span><span class="w"> </span>-o<span class="w"> </span><span class="s2">&quot;vep_vcf_results.parquet&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>cat<span class="w"> </span>vep_vcf_results.parquet.warnings.log
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>unknown: 0 / 48 
allele_mismatch_with_reference_genome: 26 / 48 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;vep_vcf_results.parquet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
      <th>gene</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
      <th>...</th>
      <th>agg_9528</th>
      <th>agg_9529</th>
      <th>agg_9530</th>
      <th>agg_9531</th>
      <th>agg_9532</th>
      <th>agg_9533</th>
      <th>agg_9535</th>
      <th>agg_9536</th>
      <th>agg_9537</th>
      <th>agg_9538</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>...</td>
      <td>-0.000053</td>
      <td>-0.000153</td>
      <td>-0.000089</td>
      <td>-0.000016</td>
      <td>-0.000013</td>
      <td>-0.000040</td>
      <td>-0.000070</td>
      <td>7.629395e-05</td>
      <td>-0.000026</td>
      <td>-0.000069</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>...</td>
      <td>-0.000586</td>
      <td>-0.000891</td>
      <td>-0.000565</td>
      <td>-0.000311</td>
      <td>-0.000461</td>
      <td>-0.000252</td>
      <td>-0.000376</td>
      <td>-6.060898e-04</td>
      <td>-0.000454</td>
      <td>-0.000725</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>PERM1</td>
      <td>621645</td>
      <td>1145933</td>
      <td>-</td>
      <td>163840</td>
      <td>170729</td>
      <td>...</td>
      <td>-0.000565</td>
      <td>-0.000787</td>
      <td>-0.000515</td>
      <td>-0.000279</td>
      <td>-0.000354</td>
      <td>-0.000202</td>
      <td>-0.000342</td>
      <td>-5.399883e-04</td>
      <td>-0.000423</td>
      <td>-0.000597</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>HES4</td>
      <td>639724</td>
      <td>1164012</td>
      <td>-</td>
      <td>163840</td>
      <td>165050</td>
      <td>...</td>
      <td>-0.001453</td>
      <td>-0.001775</td>
      <td>-0.001403</td>
      <td>-0.000820</td>
      <td>-0.000896</td>
      <td>-0.000575</td>
      <td>-0.001113</td>
      <td>-1.119316e-03</td>
      <td>-0.001036</td>
      <td>-0.001202</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>FAM87B</td>
      <td>653531</td>
      <td>1177819</td>
      <td>+</td>
      <td>163840</td>
      <td>166306</td>
      <td>...</td>
      <td>0.000045</td>
      <td>0.000105</td>
      <td>0.000060</td>
      <td>-0.000030</td>
      <td>-0.000049</td>
      <td>0.000029</td>
      <td>0.000135</td>
      <td>1.490116e-07</td>
      <td>0.000057</td>
      <td>0.000001</td>
    </tr>
    <tr>
      <th>5</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>RNF223</td>
      <td>713858</td>
      <td>1238146</td>
      <td>-</td>
      <td>163840</td>
      <td>167179</td>
      <td>...</td>
      <td>-0.000561</td>
      <td>-0.000826</td>
      <td>-0.000536</td>
      <td>-0.000299</td>
      <td>-0.000324</td>
      <td>-0.000196</td>
      <td>-0.000374</td>
      <td>-3.589988e-04</td>
      <td>-0.000342</td>
      <td>-0.000426</td>
    </tr>
    <tr>
      <th>6</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>C1orf159</td>
      <td>755913</td>
      <td>1280201</td>
      <td>-</td>
      <td>163840</td>
      <td>198383</td>
      <td>...</td>
      <td>-0.000399</td>
      <td>-0.000585</td>
      <td>-0.000374</td>
      <td>-0.000208</td>
      <td>-0.000212</td>
      <td>-0.000126</td>
      <td>-0.000265</td>
      <td>-2.926290e-04</td>
      <td>-0.000266</td>
      <td>-0.000376</td>
    </tr>
    <tr>
      <th>7</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>SAMD11</td>
      <td>760088</td>
      <td>1284376</td>
      <td>+</td>
      <td>163840</td>
      <td>184493</td>
      <td>...</td>
      <td>0.001317</td>
      <td>0.001001</td>
      <td>0.000884</td>
      <td>0.001013</td>
      <td>0.000691</td>
      <td>0.001310</td>
      <td>0.001097</td>
      <td>5.598068e-04</td>
      <td>0.000501</td>
      <td>0.000827</td>
    </tr>
    <tr>
      <th>8</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>KLHL17</td>
      <td>796744</td>
      <td>1321032</td>
      <td>+</td>
      <td>163840</td>
      <td>168975</td>
      <td>...</td>
      <td>-0.000244</td>
      <td>-0.000214</td>
      <td>-0.000310</td>
      <td>-0.000153</td>
      <td>-0.000324</td>
      <td>-0.000168</td>
      <td>-0.000147</td>
      <td>-4.823208e-04</td>
      <td>-0.000211</td>
      <td>-0.000401</td>
    </tr>
    <tr>
      <th>9</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>PLEKHN1</td>
      <td>802642</td>
      <td>1326930</td>
      <td>+</td>
      <td>163840</td>
      <td>173223</td>
      <td>...</td>
      <td>0.000012</td>
      <td>-0.000113</td>
      <td>0.000002</td>
      <td>-0.000028</td>
      <td>-0.000248</td>
      <td>-0.000283</td>
      <td>-0.000314</td>
      <td>-1.890063e-04</td>
      <td>-0.000022</td>
      <td>-0.000029</td>
    </tr>
    <tr>
      <th>10</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>TTLL10-AS1</td>
      <td>819107</td>
      <td>1343395</td>
      <td>-</td>
      <td>163840</td>
      <td>170339</td>
      <td>...</td>
      <td>-0.000286</td>
      <td>-0.000486</td>
      <td>-0.000312</td>
      <td>-0.000150</td>
      <td>-0.000256</td>
      <td>-0.000124</td>
      <td>-0.000179</td>
      <td>-3.057718e-04</td>
      <td>-0.000183</td>
      <td>-0.000252</td>
    </tr>
    <tr>
      <th>11</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>ISG15</td>
      <td>837298</td>
      <td>1361586</td>
      <td>+</td>
      <td>163840</td>
      <td>177242</td>
      <td>...</td>
      <td>0.007558</td>
      <td>0.004934</td>
      <td>0.007966</td>
      <td>0.007595</td>
      <td>-0.001943</td>
      <td>0.001598</td>
      <td>0.007838</td>
      <td>2.528191e-03</td>
      <td>0.007137</td>
      <td>0.009534</td>
    </tr>
    <tr>
      <th>12</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>TNFRSF18</td>
      <td>846144</td>
      <td>1370432</td>
      <td>-</td>
      <td>163840</td>
      <td>166924</td>
      <td>...</td>
      <td>-0.000190</td>
      <td>-0.000248</td>
      <td>-0.000215</td>
      <td>-0.000113</td>
      <td>-0.000070</td>
      <td>-0.000076</td>
      <td>-0.000159</td>
      <td>-5.897880e-05</td>
      <td>-0.000117</td>
      <td>-0.000182</td>
    </tr>
    <tr>
      <th>13</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>TNFRSF4</td>
      <td>853705</td>
      <td>1377993</td>
      <td>-</td>
      <td>163840</td>
      <td>166653</td>
      <td>...</td>
      <td>-0.000217</td>
      <td>-0.000332</td>
      <td>-0.000225</td>
      <td>-0.000132</td>
      <td>-0.000127</td>
      <td>-0.000102</td>
      <td>-0.000146</td>
      <td>-2.127588e-04</td>
      <td>-0.000168</td>
      <td>-0.000276</td>
    </tr>
    <tr>
      <th>14</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>AGRN</td>
      <td>856280</td>
      <td>1380568</td>
      <td>+</td>
      <td>163840</td>
      <td>199838</td>
      <td>...</td>
      <td>-0.000643</td>
      <td>-0.000886</td>
      <td>-0.000513</td>
      <td>-0.001688</td>
      <td>-0.000346</td>
      <td>-0.000528</td>
      <td>-0.000894</td>
      <td>-4.293919e-04</td>
      <td>-0.001309</td>
      <td>-0.000301</td>
    </tr>
    <tr>
      <th>15</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>SDF4</td>
      <td>871619</td>
      <td>1395907</td>
      <td>-</td>
      <td>163840</td>
      <td>178999</td>
      <td>...</td>
      <td>-0.000031</td>
      <td>-0.000034</td>
      <td>-0.000048</td>
      <td>-0.000017</td>
      <td>0.000019</td>
      <td>-0.000002</td>
      <td>-0.000036</td>
      <td>-1.719594e-05</td>
      <td>-0.000026</td>
      <td>-0.000037</td>
    </tr>
    <tr>
      <th>16</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>C1QTNF12</td>
      <td>886274</td>
      <td>1410562</td>
      <td>-</td>
      <td>163840</td>
      <td>168116</td>
      <td>...</td>
      <td>-0.000396</td>
      <td>-0.000607</td>
      <td>-0.000421</td>
      <td>-0.000306</td>
      <td>-0.000294</td>
      <td>-0.000207</td>
      <td>-0.000237</td>
      <td>-3.867447e-04</td>
      <td>-0.000289</td>
      <td>-0.000498</td>
    </tr>
    <tr>
      <th>17</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>UBE2J2</td>
      <td>913437</td>
      <td>1437725</td>
      <td>-</td>
      <td>163840</td>
      <td>183816</td>
      <td>...</td>
      <td>-0.000809</td>
      <td>-0.001129</td>
      <td>-0.000884</td>
      <td>-0.000664</td>
      <td>-0.000741</td>
      <td>-0.000528</td>
      <td>-0.000512</td>
      <td>-7.908940e-04</td>
      <td>-0.000607</td>
      <td>-0.000990</td>
    </tr>
    <tr>
      <th>18</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>ACAP3</td>
      <td>949161</td>
      <td>1473449</td>
      <td>-</td>
      <td>163840</td>
      <td>181059</td>
      <td>...</td>
      <td>-0.000826</td>
      <td>-0.001292</td>
      <td>-0.000932</td>
      <td>-0.000710</td>
      <td>-0.000789</td>
      <td>-0.000506</td>
      <td>-0.000457</td>
      <td>-9.316206e-04</td>
      <td>-0.000666</td>
      <td>-0.001122</td>
    </tr>
    <tr>
      <th>19</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>INTS11</td>
      <td>964243</td>
      <td>1488531</td>
      <td>-</td>
      <td>163840</td>
      <td>176946</td>
      <td>...</td>
      <td>-0.000033</td>
      <td>-0.000067</td>
      <td>-0.000013</td>
      <td>-0.000010</td>
      <td>0.000030</td>
      <td>0.000017</td>
      <td>-0.000020</td>
      <td>1.147389e-04</td>
      <td>-0.000005</td>
      <td>-0.000017</td>
    </tr>
    <tr>
      <th>20</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>DVL1</td>
      <td>988970</td>
      <td>1513258</td>
      <td>-</td>
      <td>163840</td>
      <td>177982</td>
      <td>...</td>
      <td>-0.000297</td>
      <td>-0.000456</td>
      <td>-0.000343</td>
      <td>-0.000253</td>
      <td>-0.000310</td>
      <td>-0.000204</td>
      <td>-0.000172</td>
      <td>-3.470182e-04</td>
      <td>-0.000220</td>
      <td>-0.000371</td>
    </tr>
    <tr>
      <th>21</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>MXRA8</td>
      <td>1001329</td>
      <td>1525617</td>
      <td>-</td>
      <td>163840</td>
      <td>172928</td>
      <td>...</td>
      <td>-0.000118</td>
      <td>-0.000153</td>
      <td>-0.000130</td>
      <td>-0.000090</td>
      <td>-0.000128</td>
      <td>-0.000076</td>
      <td>-0.000053</td>
      <td>-2.347827e-04</td>
      <td>-0.000090</td>
      <td>-0.000214</td>
    </tr>
    <tr>
      <th>22</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GNAT2</td>
      <td>109259481</td>
      <td>109783769</td>
      <td>-</td>
      <td>163840</td>
      <td>180515</td>
      <td>...</td>
      <td>0.000099</td>
      <td>0.000077</td>
      <td>0.000057</td>
      <td>0.000040</td>
      <td>0.000018</td>
      <td>0.000047</td>
      <td>0.000075</td>
      <td>2.750456e-04</td>
      <td>0.000065</td>
      <td>0.000258</td>
    </tr>
    <tr>
      <th>23</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GNAT2</td>
      <td>109259481</td>
      <td>109783769</td>
      <td>-</td>
      <td>163840</td>
      <td>180515</td>
      <td>...</td>
      <td>0.003284</td>
      <td>0.005105</td>
      <td>0.002662</td>
      <td>0.001423</td>
      <td>0.001525</td>
      <td>0.001257</td>
      <td>0.001919</td>
      <td>5.420089e-03</td>
      <td>0.002166</td>
      <td>0.005263</td>
    </tr>
    <tr>
      <th>24</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>SYPL2</td>
      <td>109302706</td>
      <td>109826994</td>
      <td>+</td>
      <td>163840</td>
      <td>179428</td>
      <td>...</td>
      <td>0.002200</td>
      <td>0.001744</td>
      <td>0.001206</td>
      <td>0.001263</td>
      <td>0.001008</td>
      <td>-0.000117</td>
      <td>0.001218</td>
      <td>2.799034e-04</td>
      <td>0.001503</td>
      <td>0.001372</td>
    </tr>
    <tr>
      <th>25</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>SYPL2</td>
      <td>109302706</td>
      <td>109826994</td>
      <td>+</td>
      <td>163840</td>
      <td>179428</td>
      <td>...</td>
      <td>-0.004052</td>
      <td>-0.003465</td>
      <td>-0.003092</td>
      <td>-0.003744</td>
      <td>-0.003384</td>
      <td>-0.003265</td>
      <td>-0.003253</td>
      <td>1.600981e-03</td>
      <td>-0.001915</td>
      <td>-0.002728</td>
    </tr>
    <tr>
      <th>26</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>ATXN7L2</td>
      <td>109319639</td>
      <td>109843927</td>
      <td>+</td>
      <td>163840</td>
      <td>173165</td>
      <td>...</td>
      <td>0.000001</td>
      <td>0.000010</td>
      <td>0.000091</td>
      <td>-0.000079</td>
      <td>-0.000044</td>
      <td>-0.000061</td>
      <td>-0.000042</td>
      <td>-4.082918e-05</td>
      <td>-0.000022</td>
      <td>-0.000060</td>
    </tr>
    <tr>
      <th>27</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>ATXN7L2</td>
      <td>109319639</td>
      <td>109843927</td>
      <td>+</td>
      <td>163840</td>
      <td>173165</td>
      <td>...</td>
      <td>0.000536</td>
      <td>-0.000179</td>
      <td>0.001585</td>
      <td>0.001431</td>
      <td>0.003103</td>
      <td>0.001631</td>
      <td>0.000257</td>
      <td>2.081454e-03</td>
      <td>0.000913</td>
      <td>0.001677</td>
    </tr>
    <tr>
      <th>28</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>CYB561D1</td>
      <td>109330212</td>
      <td>109854500</td>
      <td>+</td>
      <td>163840</td>
      <td>172720</td>
      <td>...</td>
      <td>0.000584</td>
      <td>0.000577</td>
      <td>0.000674</td>
      <td>0.000536</td>
      <td>0.000167</td>
      <td>0.000397</td>
      <td>0.000507</td>
      <td>1.705885e-04</td>
      <td>0.000503</td>
      <td>0.000481</td>
    </tr>
    <tr>
      <th>29</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>CYB561D1</td>
      <td>109330212</td>
      <td>109854500</td>
      <td>+</td>
      <td>163840</td>
      <td>172720</td>
      <td>...</td>
      <td>0.003762</td>
      <td>0.002849</td>
      <td>0.001487</td>
      <td>0.002733</td>
      <td>-0.000099</td>
      <td>0.000631</td>
      <td>0.001469</td>
      <td>4.820824e-04</td>
      <td>0.002098</td>
      <td>-0.000244</td>
    </tr>
    <tr>
      <th>30</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GPR61</td>
      <td>109376032</td>
      <td>109900320</td>
      <td>+</td>
      <td>163840</td>
      <td>172374</td>
      <td>...</td>
      <td>0.000155</td>
      <td>0.000251</td>
      <td>0.000114</td>
      <td>0.000134</td>
      <td>-0.000137</td>
      <td>-0.000068</td>
      <td>0.000125</td>
      <td>4.252791e-05</td>
      <td>0.000117</td>
      <td>0.000137</td>
    </tr>
    <tr>
      <th>31</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GPR61</td>
      <td>109376032</td>
      <td>109900320</td>
      <td>+</td>
      <td>163840</td>
      <td>172374</td>
      <td>...</td>
      <td>0.000258</td>
      <td>0.000810</td>
      <td>0.000308</td>
      <td>0.000489</td>
      <td>0.000618</td>
      <td>0.000236</td>
      <td>0.000523</td>
      <td>6.991625e-04</td>
      <td>0.000155</td>
      <td>0.000536</td>
    </tr>
    <tr>
      <th>32</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM3</td>
      <td>109380590</td>
      <td>109904878</td>
      <td>-</td>
      <td>163840</td>
      <td>170946</td>
      <td>...</td>
      <td>-0.000651</td>
      <td>-0.000626</td>
      <td>-0.000446</td>
      <td>-0.000290</td>
      <td>-0.000174</td>
      <td>-0.000026</td>
      <td>-0.000154</td>
      <td>-3.945976e-04</td>
      <td>-0.000396</td>
      <td>-0.000499</td>
    </tr>
    <tr>
      <th>33</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM3</td>
      <td>109380590</td>
      <td>109904878</td>
      <td>-</td>
      <td>163840</td>
      <td>170946</td>
      <td>...</td>
      <td>0.014822</td>
      <td>0.029157</td>
      <td>0.015316</td>
      <td>0.007980</td>
      <td>0.015211</td>
      <td>0.007929</td>
      <td>0.010251</td>
      <td>2.181430e-02</td>
      <td>0.009214</td>
      <td>0.024373</td>
    </tr>
    <tr>
      <th>34</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GNAI3</td>
      <td>109384775</td>
      <td>109909063</td>
      <td>+</td>
      <td>163840</td>
      <td>233549</td>
      <td>...</td>
      <td>0.002052</td>
      <td>0.002637</td>
      <td>0.003581</td>
      <td>0.001084</td>
      <td>-0.001867</td>
      <td>-0.003804</td>
      <td>0.003472</td>
      <td>-1.005411e-03</td>
      <td>0.003400</td>
      <td>0.001397</td>
    </tr>
    <tr>
      <th>35</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GNAI3</td>
      <td>109384775</td>
      <td>109909063</td>
      <td>+</td>
      <td>163840</td>
      <td>233549</td>
      <td>...</td>
      <td>-0.002514</td>
      <td>-0.002176</td>
      <td>0.001273</td>
      <td>0.007376</td>
      <td>0.011271</td>
      <td>0.019305</td>
      <td>-0.002886</td>
      <td>2.035785e-02</td>
      <td>-0.001660</td>
      <td>0.006990</td>
    </tr>
    <tr>
      <th>36</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>AMPD2</td>
      <td>109452264</td>
      <td>109976552</td>
      <td>+</td>
      <td>163840</td>
      <td>179789</td>
      <td>...</td>
      <td>0.000870</td>
      <td>0.001771</td>
      <td>0.002003</td>
      <td>0.000641</td>
      <td>-0.002239</td>
      <td>-0.000491</td>
      <td>0.001610</td>
      <td>-6.132126e-04</td>
      <td>0.002603</td>
      <td>0.000714</td>
    </tr>
    <tr>
      <th>37</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>AMPD2</td>
      <td>109452264</td>
      <td>109976552</td>
      <td>+</td>
      <td>163840</td>
      <td>179789</td>
      <td>...</td>
      <td>0.015270</td>
      <td>0.004848</td>
      <td>0.011452</td>
      <td>0.028524</td>
      <td>0.009958</td>
      <td>0.013799</td>
      <td>0.007597</td>
      <td>2.641273e-02</td>
      <td>0.009691</td>
      <td>0.026180</td>
    </tr>
    <tr>
      <th>38</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM4</td>
      <td>109492259</td>
      <td>110016547</td>
      <td>+</td>
      <td>163840</td>
      <td>182577</td>
      <td>...</td>
      <td>-0.000336</td>
      <td>0.001399</td>
      <td>0.000511</td>
      <td>0.000450</td>
      <td>-0.000501</td>
      <td>-0.000243</td>
      <td>-0.001328</td>
      <td>4.479885e-04</td>
      <td>0.000544</td>
      <td>0.002848</td>
    </tr>
    <tr>
      <th>39</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM4</td>
      <td>109492259</td>
      <td>110016547</td>
      <td>+</td>
      <td>163840</td>
      <td>182577</td>
      <td>...</td>
      <td>-0.032543</td>
      <td>-0.033684</td>
      <td>-0.029430</td>
      <td>-0.050647</td>
      <td>-0.030112</td>
      <td>-0.021851</td>
      <td>-0.012187</td>
      <td>-2.968431e-02</td>
      <td>-0.037312</td>
      <td>-0.067244</td>
    </tr>
    <tr>
      <th>40</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM2</td>
      <td>109504182</td>
      <td>110028470</td>
      <td>+</td>
      <td>163840</td>
      <td>205369</td>
      <td>...</td>
      <td>-0.001145</td>
      <td>-0.001371</td>
      <td>0.000107</td>
      <td>-0.000453</td>
      <td>-0.001597</td>
      <td>-0.000798</td>
      <td>-0.001882</td>
      <td>1.063347e-04</td>
      <td>0.000106</td>
      <td>0.000086</td>
    </tr>
    <tr>
      <th>41</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM2</td>
      <td>109504182</td>
      <td>110028470</td>
      <td>+</td>
      <td>163840</td>
      <td>205369</td>
      <td>...</td>
      <td>0.006410</td>
      <td>0.010421</td>
      <td>0.008430</td>
      <td>0.010849</td>
      <td>-0.004994</td>
      <td>0.003090</td>
      <td>-0.004843</td>
      <td>-9.270430e-03</td>
      <td>0.008755</td>
      <td>0.011827</td>
    </tr>
    <tr>
      <th>42</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM1</td>
      <td>109523974</td>
      <td>110048262</td>
      <td>+</td>
      <td>163840</td>
      <td>185065</td>
      <td>...</td>
      <td>0.000849</td>
      <td>0.001993</td>
      <td>0.001033</td>
      <td>-0.000050</td>
      <td>-0.000534</td>
      <td>0.000095</td>
      <td>0.000409</td>
      <td>-3.244877e-04</td>
      <td>0.000842</td>
      <td>0.000388</td>
    </tr>
    <tr>
      <th>43</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM1</td>
      <td>109523974</td>
      <td>110048262</td>
      <td>+</td>
      <td>163840</td>
      <td>185065</td>
      <td>...</td>
      <td>0.009973</td>
      <td>0.012482</td>
      <td>0.003073</td>
      <td>0.019083</td>
      <td>0.010606</td>
      <td>0.000387</td>
      <td>0.003512</td>
      <td>7.757962e-03</td>
      <td>0.011860</td>
      <td>0.019928</td>
    </tr>
    <tr>
      <th>44</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>0.002434</td>
      <td>0.001946</td>
      <td>0.001824</td>
      <td>0.000374</td>
      <td>-0.000416</td>
      <td>-0.000871</td>
      <td>0.003142</td>
      <td>-1.690984e-03</td>
      <td>0.001557</td>
      <td>0.002596</td>
    </tr>
    <tr>
      <th>45</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>-0.005434</td>
      <td>0.003420</td>
      <td>-0.019319</td>
      <td>-0.000747</td>
      <td>0.041849</td>
      <td>0.022034</td>
      <td>-0.024061</td>
      <td>4.308212e-02</td>
      <td>0.005673</td>
      <td>-0.007195</td>
    </tr>
    <tr>
      <th>46</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>0.000210</td>
      <td>0.000521</td>
      <td>0.000256</td>
      <td>0.000118</td>
      <td>0.000335</td>
      <td>0.000174</td>
      <td>0.000136</td>
      <td>8.502305e-04</td>
      <td>0.000165</td>
      <td>0.000544</td>
    </tr>
    <tr>
      <th>47</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>-0.000382</td>
      <td>-0.000466</td>
      <td>-0.000408</td>
      <td>-0.000091</td>
      <td>-0.000512</td>
      <td>-0.000240</td>
      <td>-0.000283</td>
      <td>-1.361221e-04</td>
      <td>-0.000274</td>
      <td>-0.000745</td>
    </tr>
  </tbody>
</table>
<p>48 rows × 8870 columns</p>
</div></div></div>
</div>
<p>Alternatively, you can pass tsv file with following format where first 4 columns are <code class="docutils literal notranslate"><span class="pre">chrom</span></code>, <code class="docutils literal notranslate"><span class="pre">pos</span></code>, <code class="docutils literal notranslate"><span class="pre">ref</span></code>, <code class="docutils literal notranslate"><span class="pre">alt</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>cat<span class="w"> </span>data/variants.tsv<span class="w"> </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-t<span class="w"> </span>-s<span class="w"> </span><span class="s1">$&#39;\t&#39;</span><span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chrom  pos        ref  alt
chr1   1000018    G    A
chr1   1002308    T    C
chr1   109727471  A    C
chr1   109728286  TTT  G
chr1   109728807  T    GG
</pre></div>
</div>
</div>
</div>
<p>You can only run predictions for the variants closer to tss than 100kbp anyway these are the ones likely to be most impactful on the gene expression.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>vep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;data/variants.tsv&quot;</span><span class="w"> </span>-o<span class="w"> </span><span class="s2">&quot;vep_results.parquet&quot;</span><span class="w"> </span>--max-distance<span class="w"> </span>100_000<span class="w"> </span>--max-distance-type<span class="w"> </span><span class="s2">&quot;tss&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima.vep - INFO - Using device: cuda and genome: hg38
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">celikm5</span> (<span class=" -Color -Color-Yellow">celikm5-genentech</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1276.1MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.1 (1916.9MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1417.4MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:76: UserWarning: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: PossibleUserWarning: The &#39;predict_dataloader&#39; does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=95` in the `DataLoader` to improve performance.
Predicting DataLoader 0: 100%|██████████████████| 66/66 [00:13&lt;00:00,  4.88it/s]
decima.vep - INFO - Warnings:
decima.vep - INFO - allele_mismatch_with_reference_genome: 10 alleles out of 33 predictions mismatched with the genome file /home/celikm5/.local/share/genomes/hg38/hg38.fa.If this is not expected, please check if you are using the correct genome version.

</pre></div>
</div>
</div>
</div>
<p>If you have already have mapping genes and variant, you can use this mapping so predictions only will be conducted between this pairs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>vep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;data/variants_gene.tsv&quot;</span><span class="w"> </span>-o<span class="w"> </span><span class="s2">&quot;vep_gene_results.parquet&quot;</span><span class="w"> </span>--gene-col<span class="w"> </span><span class="s2">&quot;gene&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima.vep - INFO - Using device: cuda and genome: hg38
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">celikm5</span> (<span class=" -Color -Color-Yellow">celikm5-genentech</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1311.1MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.1 (1900.4MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1475.9MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:76: UserWarning: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: PossibleUserWarning: The &#39;predict_dataloader&#39; does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=95` in the `DataLoader` to improve performance.
Predicting DataLoader 0: 100%|████████████████████| 4/4 [00:01&lt;00:00,  3.79it/s]

</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;vep_gene_results.parquet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
      <th>gene</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
      <th>...</th>
      <th>agg_9528</th>
      <th>agg_9529</th>
      <th>agg_9530</th>
      <th>agg_9531</th>
      <th>agg_9532</th>
      <th>agg_9533</th>
      <th>agg_9535</th>
      <th>agg_9536</th>
      <th>agg_9537</th>
      <th>agg_9538</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>ISG15</td>
      <td>837298</td>
      <td>1361586</td>
      <td>+</td>
      <td>163840</td>
      <td>177242</td>
      <td>...</td>
      <td>-0.000746</td>
      <td>0.002301</td>
      <td>0.005067</td>
      <td>0.000135</td>
      <td>-0.000559</td>
      <td>-0.003155</td>
      <td>0.006503</td>
      <td>-0.001059</td>
      <td>-0.000566</td>
      <td>0.000964</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>ISG15</td>
      <td>837298</td>
      <td>1361586</td>
      <td>+</td>
      <td>163840</td>
      <td>177242</td>
      <td>...</td>
      <td>0.007558</td>
      <td>0.004934</td>
      <td>0.007966</td>
      <td>0.007595</td>
      <td>-0.001943</td>
      <td>0.001598</td>
      <td>0.007838</td>
      <td>0.002528</td>
      <td>0.007137</td>
      <td>0.009534</td>
    </tr>
  </tbody>
</table>
<p>2 rows × 8870 columns</p>
</div></div></div>
</div>
<p>The vep api reads n (default=10_000) number of variants from vcf file performs predictions on these variants, saves them to parquet file then performs predictios for next next chuck. You can change chucksize:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>vep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;data/sample.vcf&quot;</span><span class="w"> </span>-o<span class="w"> </span><span class="s2">&quot;vep_vcf_results.parquet&quot;</span><span class="w"> </span>--chunksize<span class="w"> </span><span class="m">1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima.vep - INFO - Using device: cuda and genome: hg38
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">celikm5</span> (<span class=" -Color -Color-Yellow">celikm5-genentech</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1365.3MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.2 (1851.4MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1323.7MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:76: UserWarning: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: PossibleUserWarning: The &#39;predict_dataloader&#39; does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=95` in the `DataLoader` to improve performance.
Predicting DataLoader 0: 100%|██████████████████| 44/44 [00:09&lt;00:00,  4.86it/s]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1340.8MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.1 (1944.0MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.4 (1576.5MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: PossibleUserWarning: The &#39;predict_dataloader&#39; does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=95` in the `DataLoader` to improve performance.
Predicting DataLoader 0: 100%|██████████████████| 26/26 [00:05&lt;00:00,  4.98it/s]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1359.5MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.1 (1939.7MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1425.5MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: PossibleUserWarning: The &#39;predict_dataloader&#39; does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=95` in the `DataLoader` to improve performance.
Predicting DataLoader 0: 100%|██████████████████| 26/26 [00:05&lt;00:00,  4.97it/s]
decima.vep - INFO - Warnings:
decima.vep - INFO - allele_mismatch_with_reference_genome: 26 alleles out of 48 predictions mismatched with the genome file /home/celikm5/.local/share/genomes/hg38/hg38.fa.If this is not expected, please check if you are using the correct genome version.

</pre></div>
</div>
</div>
</div>
<section id="python-api">
<h3>Python API<a class="headerlink" href="#python-api" title="Link to this heading">¶</a></h3>
<p>Similarly, variant effect prediction can be performed using the Python API as well.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.vep</span><span class="w"> </span><span class="kn">import</span> <span class="n">predict_variant_effect</span>

<span class="n">device</span> <span class="o">=</span> <span class="s2">&quot;cuda&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_variant</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s2">&quot;data/variants.tsv&quot;</span><span class="p">)</span>
<span class="n">df_variant</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>109728286</td>
      <td>TTT</td>
      <td>G</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>T</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Simply pass your dataframe to <code class="docutils literal notranslate"><span class="pre">predict_variat_effect</span></code> function which will return dataframe for the prediction. You can pass <code class="docutils literal notranslate"><span class="pre">tasks</span></code> query to subset predictions for specific cells. Moreover, by default decima model for replicate 0 is used to use other replicates pass model=<code class="docutils literal notranslate"><span class="pre">1</span></code> , <code class="docutils literal notranslate"><span class="pre">2</span></code> or <code class="docutils literal notranslate"><span class="pre">3</span></code> to use other replicates or pass your custom model. If you pass <code class="docutils literal notranslate"><span class="pre">include_cols</span></code> argument the columns in the input will maintained in the output. To further variants based on distance to tss use <code class="docutils literal notranslate"><span class="pre">max_dist_tss</span></code> argument.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_variant_effect</span><span class="p">(</span><span class="n">df_variant</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">celikm5</span> (<span class=" -Color -Color-Yellow">celikm5-genentech</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1339.8MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.2 (1791.0MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1495.3MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:76: UserWarning: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "a73a894cc53a4b2f98c395e3b8c6bf72", "version_major": 2, "version_minor": 0}</script><div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
      <th>gene</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
      <th>...</th>
      <th>agg_9528</th>
      <th>agg_9529</th>
      <th>agg_9530</th>
      <th>agg_9531</th>
      <th>agg_9532</th>
      <th>agg_9533</th>
      <th>agg_9535</th>
      <th>agg_9536</th>
      <th>agg_9537</th>
      <th>agg_9538</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>...</td>
      <td>-0.003487</td>
      <td>-0.006149</td>
      <td>-0.003175</td>
      <td>-0.002726</td>
      <td>-0.003147</td>
      <td>-0.001746</td>
      <td>-0.002291</td>
      <td>-0.005283</td>
      <td>-0.003393</td>
      <td>-0.006694</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>...</td>
      <td>-0.000096</td>
      <td>-0.000190</td>
      <td>-0.000135</td>
      <td>-0.000043</td>
      <td>-0.000076</td>
      <td>-0.000072</td>
      <td>-0.000106</td>
      <td>0.000029</td>
      <td>-0.000055</td>
      <td>-0.000119</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>...</td>
      <td>-0.002595</td>
      <td>-0.004256</td>
      <td>-0.002759</td>
      <td>-0.001601</td>
      <td>-0.002601</td>
      <td>-0.001238</td>
      <td>-0.001756</td>
      <td>-0.002918</td>
      <td>-0.001914</td>
      <td>-0.003244</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>...</td>
      <td>-0.000587</td>
      <td>-0.000894</td>
      <td>-0.000563</td>
      <td>-0.000324</td>
      <td>-0.000471</td>
      <td>-0.000258</td>
      <td>-0.000382</td>
      <td>-0.000683</td>
      <td>-0.000459</td>
      <td>-0.000757</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>PERM1</td>
      <td>621645</td>
      <td>1145933</td>
      <td>-</td>
      <td>163840</td>
      <td>170729</td>
      <td>...</td>
      <td>-0.002933</td>
      <td>-0.004738</td>
      <td>-0.003541</td>
      <td>-0.001943</td>
      <td>-0.002958</td>
      <td>-0.001370</td>
      <td>-0.002272</td>
      <td>-0.003326</td>
      <td>-0.002195</td>
      <td>-0.003648</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>77</th>
      <td>chr1</td>
      <td>109728286</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>-0.004391</td>
      <td>0.004506</td>
      <td>-0.017141</td>
      <td>0.000025</td>
      <td>0.042810</td>
      <td>0.024843</td>
      <td>-0.022073</td>
      <td>0.043973</td>
      <td>0.005934</td>
      <td>-0.006544</td>
    </tr>
    <tr>
      <th>78</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>T</td>
      <td>GG</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>-0.129386</td>
      <td>-0.144098</td>
      <td>-0.077484</td>
      <td>-0.091391</td>
      <td>-0.104009</td>
      <td>-0.034113</td>
      <td>-0.096020</td>
      <td>-0.063028</td>
      <td>-0.079460</td>
      <td>-0.102111</td>
    </tr>
    <tr>
      <th>79</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>0.000218</td>
      <td>0.000532</td>
      <td>0.000263</td>
      <td>0.000122</td>
      <td>0.000352</td>
      <td>0.000184</td>
      <td>0.000139</td>
      <td>0.000887</td>
      <td>0.000173</td>
      <td>0.000558</td>
    </tr>
    <tr>
      <th>80</th>
      <td>chr1</td>
      <td>109728286</td>
      <td>TTT</td>
      <td>G</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>-0.000218</td>
      <td>-0.000109</td>
      <td>-0.000204</td>
      <td>0.000009</td>
      <td>-0.000265</td>
      <td>-0.000104</td>
      <td>-0.000189</td>
      <td>0.000523</td>
      <td>-0.000149</td>
      <td>-0.000385</td>
    </tr>
    <tr>
      <th>81</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>T</td>
      <td>GG</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>-0.001127</td>
      <td>-0.002115</td>
      <td>-0.001278</td>
      <td>-0.000581</td>
      <td>-0.001344</td>
      <td>-0.000724</td>
      <td>-0.000569</td>
      <td>-0.003553</td>
      <td>-0.000857</td>
      <td>-0.002281</td>
    </tr>
  </tbody>
</table>
<p>82 rows × 8870 columns</p>
</div></div></div>
</div>
<p>You can predict and save predictions to file similar to CLI api based on dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_variant_effect</span><span class="p">(</span><span class="n">df_variant</span><span class="p">,</span> <span class="n">output_pq</span><span class="o">=</span><span class="s2">&quot;vep_results_py.parquet&quot;</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1289.8MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.2 (1774.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1430.2MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "f9c815b42b1b4ba6ad1edb5427a53db6", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;vep_results_py.parquet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
      <th>gene</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
      <th>...</th>
      <th>agg_9528</th>
      <th>agg_9529</th>
      <th>agg_9530</th>
      <th>agg_9531</th>
      <th>agg_9532</th>
      <th>agg_9533</th>
      <th>agg_9535</th>
      <th>agg_9536</th>
      <th>agg_9537</th>
      <th>agg_9538</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>...</td>
      <td>-0.003487</td>
      <td>-0.006149</td>
      <td>-0.003175</td>
      <td>-0.002726</td>
      <td>-0.003147</td>
      <td>-0.001746</td>
      <td>-0.002291</td>
      <td>-0.005283</td>
      <td>-0.003393</td>
      <td>-0.006694</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>...</td>
      <td>-0.000096</td>
      <td>-0.000190</td>
      <td>-0.000135</td>
      <td>-0.000043</td>
      <td>-0.000076</td>
      <td>-0.000072</td>
      <td>-0.000106</td>
      <td>0.000029</td>
      <td>-0.000055</td>
      <td>-0.000119</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>...</td>
      <td>-0.002595</td>
      <td>-0.004256</td>
      <td>-0.002759</td>
      <td>-0.001601</td>
      <td>-0.002601</td>
      <td>-0.001238</td>
      <td>-0.001756</td>
      <td>-0.002918</td>
      <td>-0.001914</td>
      <td>-0.003244</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>...</td>
      <td>-0.000587</td>
      <td>-0.000894</td>
      <td>-0.000563</td>
      <td>-0.000324</td>
      <td>-0.000471</td>
      <td>-0.000258</td>
      <td>-0.000382</td>
      <td>-0.000683</td>
      <td>-0.000459</td>
      <td>-0.000757</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>PERM1</td>
      <td>621645</td>
      <td>1145933</td>
      <td>-</td>
      <td>163840</td>
      <td>170729</td>
      <td>...</td>
      <td>-0.002933</td>
      <td>-0.004738</td>
      <td>-0.003541</td>
      <td>-0.001943</td>
      <td>-0.002958</td>
      <td>-0.001370</td>
      <td>-0.002272</td>
      <td>-0.003326</td>
      <td>-0.002195</td>
      <td>-0.003648</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>77</th>
      <td>chr1</td>
      <td>109728286</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>-0.004391</td>
      <td>0.004506</td>
      <td>-0.017141</td>
      <td>0.000025</td>
      <td>0.042810</td>
      <td>0.024843</td>
      <td>-0.022073</td>
      <td>0.043973</td>
      <td>0.005934</td>
      <td>-0.006544</td>
    </tr>
    <tr>
      <th>78</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>T</td>
      <td>GG</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>-0.129386</td>
      <td>-0.144098</td>
      <td>-0.077484</td>
      <td>-0.091391</td>
      <td>-0.104009</td>
      <td>-0.034113</td>
      <td>-0.096020</td>
      <td>-0.063028</td>
      <td>-0.079460</td>
      <td>-0.102111</td>
    </tr>
    <tr>
      <th>79</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>0.000218</td>
      <td>0.000532</td>
      <td>0.000263</td>
      <td>0.000122</td>
      <td>0.000352</td>
      <td>0.000184</td>
      <td>0.000139</td>
      <td>0.000887</td>
      <td>0.000173</td>
      <td>0.000558</td>
    </tr>
    <tr>
      <th>80</th>
      <td>chr1</td>
      <td>109728286</td>
      <td>TTT</td>
      <td>G</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>-0.000218</td>
      <td>-0.000109</td>
      <td>-0.000204</td>
      <td>0.000009</td>
      <td>-0.000265</td>
      <td>-0.000104</td>
      <td>-0.000189</td>
      <td>0.000523</td>
      <td>-0.000149</td>
      <td>-0.000385</td>
    </tr>
    <tr>
      <th>81</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>T</td>
      <td>GG</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>-0.001127</td>
      <td>-0.002115</td>
      <td>-0.001278</td>
      <td>-0.000581</td>
      <td>-0.001344</td>
      <td>-0.000724</td>
      <td>-0.000569</td>
      <td>-0.003553</td>
      <td>-0.000857</td>
      <td>-0.002281</td>
    </tr>
  </tbody>
</table>
<p>82 rows × 8870 columns</p>
</div></div></div>
</div>
<p>Or variant effect can be performed on vcf file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_variant_effect</span><span class="p">(</span><span class="s2">&quot;data/sample.vcf&quot;</span><span class="p">,</span> <span class="n">output_pq</span><span class="o">=</span><span class="s2">&quot;vep_results_vcf_py.parquet&quot;</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1272.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.2 (1862.4MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1514.0MB/s)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "6c19dcc02b5f43c89ea51643583f352d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;vep_results_vcf_py.parquet&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
      <th>gene</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
      <th>...</th>
      <th>agg_9528</th>
      <th>agg_9529</th>
      <th>agg_9530</th>
      <th>agg_9531</th>
      <th>agg_9532</th>
      <th>agg_9533</th>
      <th>agg_9535</th>
      <th>agg_9536</th>
      <th>agg_9537</th>
      <th>agg_9538</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>...</td>
      <td>-0.000096</td>
      <td>-0.000190</td>
      <td>-0.000135</td>
      <td>-0.000043</td>
      <td>-0.000076</td>
      <td>-0.000072</td>
      <td>-0.000106</td>
      <td>0.000029</td>
      <td>-0.000055</td>
      <td>-0.000119</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>...</td>
      <td>-0.000587</td>
      <td>-0.000894</td>
      <td>-0.000563</td>
      <td>-0.000324</td>
      <td>-0.000471</td>
      <td>-0.000258</td>
      <td>-0.000382</td>
      <td>-0.000683</td>
      <td>-0.000459</td>
      <td>-0.000757</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>PERM1</td>
      <td>621645</td>
      <td>1145933</td>
      <td>-</td>
      <td>163840</td>
      <td>170729</td>
      <td>...</td>
      <td>-0.000845</td>
      <td>-0.001181</td>
      <td>-0.000784</td>
      <td>-0.000442</td>
      <td>-0.000589</td>
      <td>-0.000313</td>
      <td>-0.000526</td>
      <td>-0.000856</td>
      <td>-0.000637</td>
      <td>-0.000881</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>HES4</td>
      <td>639724</td>
      <td>1164012</td>
      <td>-</td>
      <td>163840</td>
      <td>165050</td>
      <td>...</td>
      <td>-0.001767</td>
      <td>-0.002189</td>
      <td>-0.001705</td>
      <td>-0.001091</td>
      <td>-0.001184</td>
      <td>-0.000717</td>
      <td>-0.001306</td>
      <td>-0.001550</td>
      <td>-0.001349</td>
      <td>-0.001550</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>FAM87B</td>
      <td>653531</td>
      <td>1177819</td>
      <td>+</td>
      <td>163840</td>
      <td>166306</td>
      <td>...</td>
      <td>-0.000098</td>
      <td>0.000052</td>
      <td>-0.000040</td>
      <td>-0.000051</td>
      <td>0.000037</td>
      <td>0.000028</td>
      <td>-0.000024</td>
      <td>0.000093</td>
      <td>-0.000080</td>
      <td>-0.000203</td>
    </tr>
    <tr>
      <th>5</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>RNF223</td>
      <td>713858</td>
      <td>1238146</td>
      <td>-</td>
      <td>163840</td>
      <td>167179</td>
      <td>...</td>
      <td>-0.000603</td>
      <td>-0.000860</td>
      <td>-0.000584</td>
      <td>-0.000328</td>
      <td>-0.000361</td>
      <td>-0.000211</td>
      <td>-0.000395</td>
      <td>-0.000428</td>
      <td>-0.000371</td>
      <td>-0.000468</td>
    </tr>
    <tr>
      <th>6</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>C1orf159</td>
      <td>755913</td>
      <td>1280201</td>
      <td>-</td>
      <td>163840</td>
      <td>198383</td>
      <td>...</td>
      <td>-0.000358</td>
      <td>-0.000528</td>
      <td>-0.000338</td>
      <td>-0.000193</td>
      <td>-0.000205</td>
      <td>-0.000113</td>
      <td>-0.000245</td>
      <td>-0.000265</td>
      <td>-0.000241</td>
      <td>-0.000347</td>
    </tr>
    <tr>
      <th>7</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>SAMD11</td>
      <td>760088</td>
      <td>1284376</td>
      <td>+</td>
      <td>163840</td>
      <td>184493</td>
      <td>...</td>
      <td>0.001338</td>
      <td>0.000849</td>
      <td>0.000867</td>
      <td>0.000903</td>
      <td>0.001198</td>
      <td>0.001515</td>
      <td>0.001150</td>
      <td>0.001019</td>
      <td>0.000498</td>
      <td>0.001034</td>
    </tr>
    <tr>
      <th>8</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>KLHL17</td>
      <td>796744</td>
      <td>1321032</td>
      <td>+</td>
      <td>163840</td>
      <td>168975</td>
      <td>...</td>
      <td>-0.000518</td>
      <td>-0.000364</td>
      <td>-0.000522</td>
      <td>-0.000307</td>
      <td>-0.000214</td>
      <td>-0.000158</td>
      <td>-0.000330</td>
      <td>-0.000363</td>
      <td>-0.000348</td>
      <td>-0.000509</td>
    </tr>
    <tr>
      <th>9</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>PLEKHN1</td>
      <td>802642</td>
      <td>1326930</td>
      <td>+</td>
      <td>163840</td>
      <td>173223</td>
      <td>...</td>
      <td>0.000109</td>
      <td>0.000011</td>
      <td>0.000054</td>
      <td>0.000009</td>
      <td>-0.000056</td>
      <td>-0.000199</td>
      <td>-0.000207</td>
      <td>-0.000112</td>
      <td>0.000043</td>
      <td>0.000051</td>
    </tr>
    <tr>
      <th>10</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>TTLL10-AS1</td>
      <td>819107</td>
      <td>1343395</td>
      <td>-</td>
      <td>163840</td>
      <td>170339</td>
      <td>...</td>
      <td>-0.000278</td>
      <td>-0.000460</td>
      <td>-0.000305</td>
      <td>-0.000142</td>
      <td>-0.000220</td>
      <td>-0.000106</td>
      <td>-0.000170</td>
      <td>-0.000272</td>
      <td>-0.000170</td>
      <td>-0.000231</td>
    </tr>
    <tr>
      <th>11</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>ISG15</td>
      <td>837298</td>
      <td>1361586</td>
      <td>+</td>
      <td>163840</td>
      <td>177242</td>
      <td>...</td>
      <td>0.007892</td>
      <td>0.005266</td>
      <td>0.008324</td>
      <td>0.007764</td>
      <td>-0.001580</td>
      <td>0.002468</td>
      <td>0.008656</td>
      <td>0.002925</td>
      <td>0.007485</td>
      <td>0.010022</td>
    </tr>
    <tr>
      <th>12</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>TNFRSF18</td>
      <td>846144</td>
      <td>1370432</td>
      <td>-</td>
      <td>163840</td>
      <td>166924</td>
      <td>...</td>
      <td>-0.000308</td>
      <td>-0.000433</td>
      <td>-0.000336</td>
      <td>-0.000197</td>
      <td>-0.000195</td>
      <td>-0.000144</td>
      <td>-0.000220</td>
      <td>-0.000190</td>
      <td>-0.000199</td>
      <td>-0.000302</td>
    </tr>
    <tr>
      <th>13</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>TNFRSF4</td>
      <td>853705</td>
      <td>1377993</td>
      <td>-</td>
      <td>163840</td>
      <td>166653</td>
      <td>...</td>
      <td>-0.000255</td>
      <td>-0.000385</td>
      <td>-0.000261</td>
      <td>-0.000158</td>
      <td>-0.000164</td>
      <td>-0.000123</td>
      <td>-0.000158</td>
      <td>-0.000281</td>
      <td>-0.000191</td>
      <td>-0.000319</td>
    </tr>
    <tr>
      <th>14</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>AGRN</td>
      <td>856280</td>
      <td>1380568</td>
      <td>+</td>
      <td>163840</td>
      <td>199838</td>
      <td>...</td>
      <td>-0.000901</td>
      <td>-0.001054</td>
      <td>-0.000771</td>
      <td>-0.001977</td>
      <td>-0.000437</td>
      <td>-0.000555</td>
      <td>-0.001084</td>
      <td>-0.000583</td>
      <td>-0.001613</td>
      <td>-0.000536</td>
    </tr>
    <tr>
      <th>15</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>SDF4</td>
      <td>871619</td>
      <td>1395907</td>
      <td>-</td>
      <td>163840</td>
      <td>178999</td>
      <td>...</td>
      <td>-0.000146</td>
      <td>-0.000230</td>
      <td>-0.000180</td>
      <td>-0.000099</td>
      <td>-0.000132</td>
      <td>-0.000082</td>
      <td>-0.000092</td>
      <td>-0.000189</td>
      <td>-0.000112</td>
      <td>-0.000195</td>
    </tr>
    <tr>
      <th>16</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>C1QTNF12</td>
      <td>886274</td>
      <td>1410562</td>
      <td>-</td>
      <td>163840</td>
      <td>168116</td>
      <td>...</td>
      <td>-0.000458</td>
      <td>-0.000697</td>
      <td>-0.000491</td>
      <td>-0.000359</td>
      <td>-0.000367</td>
      <td>-0.000241</td>
      <td>-0.000268</td>
      <td>-0.000487</td>
      <td>-0.000344</td>
      <td>-0.000595</td>
    </tr>
    <tr>
      <th>17</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>UBE2J2</td>
      <td>913437</td>
      <td>1437725</td>
      <td>-</td>
      <td>163840</td>
      <td>183816</td>
      <td>...</td>
      <td>-0.000815</td>
      <td>-0.001141</td>
      <td>-0.000897</td>
      <td>-0.000674</td>
      <td>-0.000717</td>
      <td>-0.000515</td>
      <td>-0.000520</td>
      <td>-0.000780</td>
      <td>-0.000614</td>
      <td>-0.001010</td>
    </tr>
    <tr>
      <th>18</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>ACAP3</td>
      <td>949161</td>
      <td>1473449</td>
      <td>-</td>
      <td>163840</td>
      <td>181059</td>
      <td>...</td>
      <td>-0.000748</td>
      <td>-0.001163</td>
      <td>-0.000843</td>
      <td>-0.000615</td>
      <td>-0.000704</td>
      <td>-0.000456</td>
      <td>-0.000404</td>
      <td>-0.000761</td>
      <td>-0.000572</td>
      <td>-0.000986</td>
    </tr>
    <tr>
      <th>19</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>INTS11</td>
      <td>964243</td>
      <td>1488531</td>
      <td>-</td>
      <td>163840</td>
      <td>176946</td>
      <td>...</td>
      <td>0.000021</td>
      <td>0.000008</td>
      <td>0.000037</td>
      <td>0.000035</td>
      <td>0.000096</td>
      <td>0.000057</td>
      <td>0.000009</td>
      <td>0.000184</td>
      <td>0.000039</td>
      <td>0.000044</td>
    </tr>
    <tr>
      <th>20</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>DVL1</td>
      <td>988970</td>
      <td>1513258</td>
      <td>-</td>
      <td>163840</td>
      <td>177982</td>
      <td>...</td>
      <td>-0.000325</td>
      <td>-0.000491</td>
      <td>-0.000359</td>
      <td>-0.000260</td>
      <td>-0.000322</td>
      <td>-0.000207</td>
      <td>-0.000179</td>
      <td>-0.000336</td>
      <td>-0.000235</td>
      <td>-0.000382</td>
    </tr>
    <tr>
      <th>21</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>MXRA8</td>
      <td>1001329</td>
      <td>1525617</td>
      <td>-</td>
      <td>163840</td>
      <td>172928</td>
      <td>...</td>
      <td>-0.000077</td>
      <td>-0.000098</td>
      <td>-0.000090</td>
      <td>-0.000067</td>
      <td>-0.000092</td>
      <td>-0.000057</td>
      <td>-0.000034</td>
      <td>-0.000212</td>
      <td>-0.000063</td>
      <td>-0.000159</td>
    </tr>
    <tr>
      <th>22</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GNAT2</td>
      <td>109259481</td>
      <td>109783769</td>
      <td>-</td>
      <td>163840</td>
      <td>180515</td>
      <td>...</td>
      <td>0.000103</td>
      <td>0.000070</td>
      <td>0.000060</td>
      <td>0.000038</td>
      <td>0.000027</td>
      <td>0.000055</td>
      <td>0.000078</td>
      <td>0.000277</td>
      <td>0.000062</td>
      <td>0.000245</td>
    </tr>
    <tr>
      <th>23</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GNAT2</td>
      <td>109259481</td>
      <td>109783769</td>
      <td>-</td>
      <td>163840</td>
      <td>180515</td>
      <td>...</td>
      <td>0.003317</td>
      <td>0.005164</td>
      <td>0.002698</td>
      <td>0.001438</td>
      <td>0.001558</td>
      <td>0.001273</td>
      <td>0.001937</td>
      <td>0.005484</td>
      <td>0.002190</td>
      <td>0.005321</td>
    </tr>
    <tr>
      <th>24</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>SYPL2</td>
      <td>109302706</td>
      <td>109826994</td>
      <td>+</td>
      <td>163840</td>
      <td>179428</td>
      <td>...</td>
      <td>0.002225</td>
      <td>0.001688</td>
      <td>0.001366</td>
      <td>0.001251</td>
      <td>0.001342</td>
      <td>0.000177</td>
      <td>0.001402</td>
      <td>0.000480</td>
      <td>0.001663</td>
      <td>0.001764</td>
    </tr>
    <tr>
      <th>25</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>SYPL2</td>
      <td>109302706</td>
      <td>109826994</td>
      <td>+</td>
      <td>163840</td>
      <td>179428</td>
      <td>...</td>
      <td>-0.004184</td>
      <td>-0.003700</td>
      <td>-0.003180</td>
      <td>-0.004169</td>
      <td>-0.003774</td>
      <td>-0.003399</td>
      <td>-0.003382</td>
      <td>0.000949</td>
      <td>-0.002237</td>
      <td>-0.002981</td>
    </tr>
    <tr>
      <th>26</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>ATXN7L2</td>
      <td>109319639</td>
      <td>109843927</td>
      <td>+</td>
      <td>163840</td>
      <td>173165</td>
      <td>...</td>
      <td>-0.000055</td>
      <td>-0.000036</td>
      <td>0.000088</td>
      <td>-0.000085</td>
      <td>0.000037</td>
      <td>-0.000040</td>
      <td>-0.000056</td>
      <td>-0.000044</td>
      <td>-0.000046</td>
      <td>-0.000027</td>
    </tr>
    <tr>
      <th>27</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>ATXN7L2</td>
      <td>109319639</td>
      <td>109843927</td>
      <td>+</td>
      <td>163840</td>
      <td>173165</td>
      <td>...</td>
      <td>0.000644</td>
      <td>-0.000063</td>
      <td>0.001706</td>
      <td>0.001486</td>
      <td>0.003162</td>
      <td>0.001671</td>
      <td>0.000299</td>
      <td>0.002102</td>
      <td>0.001007</td>
      <td>0.001831</td>
    </tr>
    <tr>
      <th>28</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>CYB561D1</td>
      <td>109330212</td>
      <td>109854500</td>
      <td>+</td>
      <td>163840</td>
      <td>172720</td>
      <td>...</td>
      <td>0.000533</td>
      <td>0.000539</td>
      <td>0.000585</td>
      <td>0.000474</td>
      <td>0.000076</td>
      <td>0.000280</td>
      <td>0.000422</td>
      <td>-0.000033</td>
      <td>0.000436</td>
      <td>0.000369</td>
    </tr>
    <tr>
      <th>29</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>CYB561D1</td>
      <td>109330212</td>
      <td>109854500</td>
      <td>+</td>
      <td>163840</td>
      <td>172720</td>
      <td>...</td>
      <td>0.003902</td>
      <td>0.002963</td>
      <td>0.001609</td>
      <td>0.002841</td>
      <td>0.000024</td>
      <td>0.000726</td>
      <td>0.001582</td>
      <td>0.000497</td>
      <td>0.002237</td>
      <td>-0.000092</td>
    </tr>
    <tr>
      <th>30</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GPR61</td>
      <td>109376032</td>
      <td>109900320</td>
      <td>+</td>
      <td>163840</td>
      <td>172374</td>
      <td>...</td>
      <td>0.000144</td>
      <td>0.000244</td>
      <td>0.000101</td>
      <td>0.000130</td>
      <td>-0.000119</td>
      <td>-0.000066</td>
      <td>0.000112</td>
      <td>0.000059</td>
      <td>0.000113</td>
      <td>0.000135</td>
    </tr>
    <tr>
      <th>31</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GPR61</td>
      <td>109376032</td>
      <td>109900320</td>
      <td>+</td>
      <td>163840</td>
      <td>172374</td>
      <td>...</td>
      <td>0.000287</td>
      <td>0.000869</td>
      <td>0.000338</td>
      <td>0.000516</td>
      <td>0.000702</td>
      <td>0.000260</td>
      <td>0.000547</td>
      <td>0.000815</td>
      <td>0.000190</td>
      <td>0.000561</td>
    </tr>
    <tr>
      <th>32</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM3</td>
      <td>109380590</td>
      <td>109904878</td>
      <td>-</td>
      <td>163840</td>
      <td>170946</td>
      <td>...</td>
      <td>-0.000685</td>
      <td>-0.000671</td>
      <td>-0.000490</td>
      <td>-0.000305</td>
      <td>-0.000203</td>
      <td>-0.000045</td>
      <td>-0.000183</td>
      <td>-0.000423</td>
      <td>-0.000412</td>
      <td>-0.000544</td>
    </tr>
    <tr>
      <th>33</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM3</td>
      <td>109380590</td>
      <td>109904878</td>
      <td>-</td>
      <td>163840</td>
      <td>170946</td>
      <td>...</td>
      <td>0.014859</td>
      <td>0.029183</td>
      <td>0.015364</td>
      <td>0.007996</td>
      <td>0.015255</td>
      <td>0.007950</td>
      <td>0.010278</td>
      <td>0.021851</td>
      <td>0.009230</td>
      <td>0.024402</td>
    </tr>
    <tr>
      <th>34</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GNAI3</td>
      <td>109384775</td>
      <td>109909063</td>
      <td>+</td>
      <td>163840</td>
      <td>233549</td>
      <td>...</td>
      <td>0.002294</td>
      <td>0.002954</td>
      <td>0.003762</td>
      <td>0.001515</td>
      <td>-0.001879</td>
      <td>-0.003638</td>
      <td>0.004167</td>
      <td>-0.000977</td>
      <td>0.003568</td>
      <td>0.001709</td>
    </tr>
    <tr>
      <th>35</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GNAI3</td>
      <td>109384775</td>
      <td>109909063</td>
      <td>+</td>
      <td>163840</td>
      <td>233549</td>
      <td>...</td>
      <td>-0.002409</td>
      <td>-0.002181</td>
      <td>0.001515</td>
      <td>0.007504</td>
      <td>0.011276</td>
      <td>0.019385</td>
      <td>-0.002945</td>
      <td>0.020405</td>
      <td>-0.001569</td>
      <td>0.007003</td>
    </tr>
    <tr>
      <th>36</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>AMPD2</td>
      <td>109452264</td>
      <td>109976552</td>
      <td>+</td>
      <td>163840</td>
      <td>179789</td>
      <td>...</td>
      <td>0.000565</td>
      <td>0.001265</td>
      <td>0.001679</td>
      <td>-0.000034</td>
      <td>-0.002686</td>
      <td>-0.000878</td>
      <td>0.001329</td>
      <td>-0.001112</td>
      <td>0.002059</td>
      <td>0.000288</td>
    </tr>
    <tr>
      <th>37</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>AMPD2</td>
      <td>109452264</td>
      <td>109976552</td>
      <td>+</td>
      <td>163840</td>
      <td>179789</td>
      <td>...</td>
      <td>0.015521</td>
      <td>0.005221</td>
      <td>0.011660</td>
      <td>0.028807</td>
      <td>0.010368</td>
      <td>0.014050</td>
      <td>0.007810</td>
      <td>0.026553</td>
      <td>0.009914</td>
      <td>0.026500</td>
    </tr>
    <tr>
      <th>38</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM4</td>
      <td>109492259</td>
      <td>110016547</td>
      <td>+</td>
      <td>163840</td>
      <td>182577</td>
      <td>...</td>
      <td>-0.000140</td>
      <td>0.001370</td>
      <td>0.000755</td>
      <td>0.000528</td>
      <td>-0.000446</td>
      <td>-0.000184</td>
      <td>-0.001062</td>
      <td>0.000593</td>
      <td>0.000575</td>
      <td>0.002940</td>
    </tr>
    <tr>
      <th>39</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM4</td>
      <td>109492259</td>
      <td>110016547</td>
      <td>+</td>
      <td>163840</td>
      <td>182577</td>
      <td>...</td>
      <td>-0.032480</td>
      <td>-0.033616</td>
      <td>-0.029502</td>
      <td>-0.050562</td>
      <td>-0.030190</td>
      <td>-0.021981</td>
      <td>-0.012192</td>
      <td>-0.029633</td>
      <td>-0.037296</td>
      <td>-0.067220</td>
    </tr>
    <tr>
      <th>40</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM2</td>
      <td>109504182</td>
      <td>110028470</td>
      <td>+</td>
      <td>163840</td>
      <td>205369</td>
      <td>...</td>
      <td>-0.001281</td>
      <td>-0.001583</td>
      <td>-0.000117</td>
      <td>-0.000501</td>
      <td>-0.001837</td>
      <td>-0.001011</td>
      <td>-0.002222</td>
      <td>0.000031</td>
      <td>0.000068</td>
      <td>0.000062</td>
    </tr>
    <tr>
      <th>41</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM2</td>
      <td>109504182</td>
      <td>110028470</td>
      <td>+</td>
      <td>163840</td>
      <td>205369</td>
      <td>...</td>
      <td>0.006389</td>
      <td>0.010391</td>
      <td>0.008292</td>
      <td>0.010647</td>
      <td>-0.005511</td>
      <td>0.002638</td>
      <td>-0.004853</td>
      <td>-0.009829</td>
      <td>0.008542</td>
      <td>0.011590</td>
    </tr>
    <tr>
      <th>42</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM1</td>
      <td>109523974</td>
      <td>110048262</td>
      <td>+</td>
      <td>163840</td>
      <td>185065</td>
      <td>...</td>
      <td>0.001077</td>
      <td>0.002305</td>
      <td>0.001138</td>
      <td>-0.000056</td>
      <td>-0.000738</td>
      <td>0.000059</td>
      <td>0.000456</td>
      <td>-0.000396</td>
      <td>0.000904</td>
      <td>0.000512</td>
    </tr>
    <tr>
      <th>43</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM1</td>
      <td>109523974</td>
      <td>110048262</td>
      <td>+</td>
      <td>163840</td>
      <td>185065</td>
      <td>...</td>
      <td>0.009587</td>
      <td>0.012155</td>
      <td>0.002769</td>
      <td>0.018908</td>
      <td>0.010602</td>
      <td>0.000302</td>
      <td>0.003231</td>
      <td>0.007630</td>
      <td>0.011715</td>
      <td>0.019595</td>
    </tr>
    <tr>
      <th>44</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>0.001978</td>
      <td>0.001395</td>
      <td>0.001501</td>
      <td>0.000012</td>
      <td>-0.000538</td>
      <td>-0.000914</td>
      <td>0.002817</td>
      <td>-0.001998</td>
      <td>0.001149</td>
      <td>0.002069</td>
    </tr>
    <tr>
      <th>45</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>...</td>
      <td>-0.004803</td>
      <td>0.004060</td>
      <td>-0.018828</td>
      <td>-0.000383</td>
      <td>0.042374</td>
      <td>0.022364</td>
      <td>-0.023450</td>
      <td>0.043353</td>
      <td>0.006033</td>
      <td>-0.006614</td>
    </tr>
    <tr>
      <th>46</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>0.000218</td>
      <td>0.000532</td>
      <td>0.000263</td>
      <td>0.000122</td>
      <td>0.000352</td>
      <td>0.000184</td>
      <td>0.000139</td>
      <td>0.000887</td>
      <td>0.000173</td>
      <td>0.000558</td>
    </tr>
    <tr>
      <th>47</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>TTT</td>
      <td>G</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>...</td>
      <td>-0.000386</td>
      <td>-0.000480</td>
      <td>-0.000415</td>
      <td>-0.000092</td>
      <td>-0.000513</td>
      <td>-0.000241</td>
      <td>-0.000285</td>
      <td>-0.000141</td>
      <td>-0.000277</td>
      <td>-0.000758</td>
    </tr>
  </tbody>
</table>
<p>48 rows × 8870 columns</p>
</div></div></div>
</div>
</section>
</section>
<section id="developer-api">
<h2>Developer API<a class="headerlink" href="#developer-api" title="Link to this heading">¶</a></h2>
<p>To perform variant effect prediction, Decima creates dataset and dataloader from the given set of variants:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.data.dataset</span><span class="w"> </span><span class="kn">import</span> <span class="n">VariantDataset</span>

<span class="n">dataset</span> <span class="o">=</span> <span class="n">VariantDataset</span><span class="p">(</span><span class="n">df_variant</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_metadata:latest, 628.05MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1290.4MB/s)
</pre></div>
</div>
</div>
</div>
<p>Dataset prepares one_hot encoded sequence with gene mask which is ready to pass to the model:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>164
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;seq&#39;: tensor([[0., 1., 0.,  ..., 1., 0., 1.],
         [1., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 1.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 1., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.]]),
 &#39;warning&#39;: []}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;seq&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>torch.Size([5, 524288])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span><span class="o">.</span><span class="n">variants</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
      <th>gene</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>gene_mask_start</th>
      <th>gene_mask_end</th>
      <th>rel_pos</th>
      <th>ref_tx</th>
      <th>alt_tx</th>
      <th>tss_dist</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>40725</td>
      <td>C</td>
      <td>T</td>
      <td>-123115</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>FAM41C</td>
      <td>516455</td>
      <td>1040743</td>
      <td>-</td>
      <td>163840</td>
      <td>172672</td>
      <td>38435</td>
      <td>A</td>
      <td>G</td>
      <td>-125405</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>123131</td>
      <td>C</td>
      <td>T</td>
      <td>-40709</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>1002308</td>
      <td>T</td>
      <td>C</td>
      <td>NOC2L</td>
      <td>598861</td>
      <td>1123149</td>
      <td>-</td>
      <td>163840</td>
      <td>178946</td>
      <td>120841</td>
      <td>A</td>
      <td>G</td>
      <td>-42999</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>1000018</td>
      <td>G</td>
      <td>A</td>
      <td>PERM1</td>
      <td>621645</td>
      <td>1145933</td>
      <td>-</td>
      <td>163840</td>
      <td>170729</td>
      <td>145915</td>
      <td>C</td>
      <td>T</td>
      <td>-17925</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>77</th>
      <td>chr1</td>
      <td>109728286</td>
      <td>TTT</td>
      <td>G</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>180346</td>
      <td>TTT</td>
      <td>G</td>
      <td>16506</td>
    </tr>
    <tr>
      <th>78</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>T</td>
      <td>GG</td>
      <td>GSTM5</td>
      <td>109547940</td>
      <td>110072228</td>
      <td>+</td>
      <td>163840</td>
      <td>227488</td>
      <td>180867</td>
      <td>T</td>
      <td>GG</td>
      <td>17027</td>
    </tr>
    <tr>
      <th>79</th>
      <td>chr1</td>
      <td>109727471</td>
      <td>A</td>
      <td>C</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>507041</td>
      <td>T</td>
      <td>G</td>
      <td>343201</td>
    </tr>
    <tr>
      <th>80</th>
      <td>chr1</td>
      <td>109728286</td>
      <td>TTT</td>
      <td>G</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>506226</td>
      <td>AAA</td>
      <td>C</td>
      <td>342386</td>
    </tr>
    <tr>
      <th>81</th>
      <td>chr1</td>
      <td>109728807</td>
      <td>T</td>
      <td>GG</td>
      <td>ALX3</td>
      <td>109710224</td>
      <td>110234512</td>
      <td>-</td>
      <td>163840</td>
      <td>174642</td>
      <td>505705</td>
      <td>A</td>
      <td>CC</td>
      <td>341865</td>
    </tr>
  </tbody>
</table>
<p>82 rows × 14 columns</p>
</div></div></div>
</div>
<p>Let’s load model</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.hub</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_decima_model</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">load_decima_model</span><span class="p">(</span><span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact decima_rep0:latest, 2155.88MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.2 (1808.4MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact human_state_dict_fold0:latest, 709.30MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1487.8MB/s)
</pre></div>
</div>
</div>
</div>
<p>The model has <code class="docutils literal notranslate"><span class="pre">predict_on_dataset</span></code> method which performs prediction for the dataset object:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_on_dataset</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: PossibleUserWarning: The &#39;predict_dataloader&#39; does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=95` in the `DataLoader` to improve performance.
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "24aebc5e88744bd7b1361c1fde3bcd26", "version_major": 2, "version_minor": 0}</script></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">preds</span></code> are predicted expression for <code class="docutils literal notranslate"><span class="pre">alt</span> <span class="pre">-</span> <span class="pre">ref</span></code> alleles for each variant:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span><span class="p">[</span><span class="s2">&quot;expression&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(82, 8856)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span><span class="p">[</span><span class="s2">&quot;expression&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[-3.1028390e-03, -3.5525560e-03, -3.5931766e-03, ...,
        -5.2702725e-03, -3.3913553e-03, -6.6891909e-03],
       [ 1.1469424e-04,  1.7350912e-04,  8.1568956e-05, ...,
         1.3417006e-04,  1.1980534e-05, -2.7149916e-05],
       [ 6.0230494e-05, -8.2150102e-05, -2.6747584e-05, ...,
        -2.7078092e-03, -1.8020868e-03, -3.0442923e-03],
       ...,
       [ 5.8137625e-04,  6.4936280e-04,  4.9999356e-04, ...,
         8.4596872e-04,  1.6248971e-04,  5.3709745e-04],
       [ 2.6370510e-03,  2.7481169e-03,  1.8703863e-03, ...,
         5.3109229e-04, -1.4962628e-04, -3.7675351e-04],
       [-2.8469041e-03, -3.2258853e-03, -2.4763122e-03, ...,
        -3.5528541e-03, -8.5747242e-04, -2.2805035e-03]], dtype=float32)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span><span class="p">[</span><span class="s2">&quot;warnings&quot;</span><span class="p">]</span>  <span class="c1"># some of the variants does not match with the genome genome sequence.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Counter({&#39;allele_mismatch_with_reference_genome&#39;: 26, &#39;unknown&#39;: 0})
</pre></div>
</div>
</div>
</div>
<p>You can perform prediction for the individual alleles with directly using the api:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dl</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">DataLoader</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">batch</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">dl</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">batch</span><span class="p">[</span><span class="s2">&quot;seq&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># first allele and second allele</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>torch.Size([2, 5, 524288])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>

<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">():</span>
    <span class="n">preds</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">batch</span><span class="p">[</span><span class="s2">&quot;seq&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>The variant has little difference between reference and alternative alleles so it is likely neural based on the model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">preds</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">(),</span> <span class="n">preds</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;gene expression for ref allele&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;gene expression for alt allele&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0, 0.5, &#39;gene expression for alt allele&#39;)
</pre></div>
</div>
<img alt="../_images/0c83ceee171bd3f3fdb0290a607f5d1cf48287849e6507b5a6aaf434662f0f0d.png" src="../_images/0c83ceee171bd3f3fdb0290a607f5d1cf48287849e6507b5a6aaf434662f0f0d.png" />
</div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="3-finetune.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Fine-tuning Borzoi to create a Decima model</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="1-attribution-motif-discovery.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Attribution and Motifs Detection with Decima</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Gokcen Eraslan
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Variant Effect Prediction with Decima</a><ul>
<li><a class="reference internal" href="#cli-api">CLI API</a><ul>
<li><a class="reference internal" href="#python-api">Python API</a></li>
</ul>
</li>
<li><a class="reference internal" href="#developer-api">Developer API</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=8e576371"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>