<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html"><link rel="search" title="Search" href="../search.html"><link rel="next" title="Predicting Gene Expression with Decima" href="5-gene-expression-prediction.html"><link rel="prev" title="Fine-tuning Borzoi to create a Decima model" href="3-finetune.html">

    <!-- Generated with Sphinx 8.2.3 and Furo 2025.09.25 -->
        <title>Detecting and Clustering Seqlets with Modisco - decima 0.0.post1.dev1+ga9d6eec85 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=580074bf" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">decima 0.0.post1.dev1+ga9d6eec85 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <span class="sidebar-brand-text">decima 0.0.post1.dev1+ga9d6eec85 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/decima.html">decima package</a><input aria-label="Toggle navigation of decima package" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.cli.html">decima.cli package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.core.html">decima.core package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.data.html">decima.data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.hub.html">decima.hub package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.interpret.html">decima.interpret package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.model.html">decima.model package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.plot.html">decima.plot package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.tools.html">decima.tools package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.train.html">decima.train package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.utils.html">decima.utils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.vep.html">decima.vep package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="1-attribution-motif-discovery.html">Tutorial: Attribution Motif Discovery</a></li>
<li class="toctree-l1"><a class="reference internal" href="2-variant-effect-prediction.html">Tutorial: Variant Effect Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="3-finetune.html">Tutorial: Finetuning</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Tutorial: Modisco</a></li>
<li class="toctree-l1"><a class="reference internal" href="5-gene-expression-prediction.html">Tutorial: Gene Expression Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/4-modisco.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section class="tex2jax_ignore mathjax_ignore" id="detecting-and-clustering-seqlets-with-modisco">
<h1>Detecting and Clustering Seqlets with Modisco<a class="headerlink" href="#detecting-and-clustering-seqlets-with-modisco" title="Link to this heading">¶</a></h1>
<p>In this tutorial, we will explore how to use Modisco within the Decima framework to detect and cluster seqlets—short, high-scoring regions of DNA that are important for model predictions. Modisco (Motif Discovery from Importance Scores) is a powerful tool designed to analyze attribution scores from deep learning models. It identifies recurring patterns, or “motifs,” by first locating seqlets—short subsequences with high attribution—and then clustering these seqlets based on similarity. This process helps uncover biologically meaningful sequence motifs that drive the model’s predictions, providing insights into regulatory elements and sequence features learned by the model. We will walk through the steps of running Modisco on attribution data, from detecting seqlets to clustering them into motifs, and interpreting the results.</p>
<section id="cli-api">
<h2>CLI API<a class="headerlink" href="#cli-api" title="Link to this heading">¶</a></h2>
<p>In this tutorial, we’ll walk through a practical example of using Decima’s CLI API to analyze neuronal cells and uncover how they differ from non-neuronal cells based on the major regulators. Let’s first list avaliable cell types:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>query-cell<span class="w"> </span><span class="s2">&quot;cell_type.str.contains(&#39;neuron&#39;) and organ == &#39;CNS&#39; and disease == &#39;healthy&#39;&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1842.3MB/s)
	cell_type	tissue	organ	disease	study	dataset	region	subregion	celltype_coarse	n_cells	total_counts	n_genes	size_factor	train_pearson	val_pearson	test_pearson
agg_1112	CGE interneuron	Amygdala_Amygdala	CNS	healthy	jhpce#tran2021	brain_atlas	Amygdala	Amygdala		674	17421456.0	16954	42512.927350555474	0.9454484906394379	0.8549755331049285	0.8667663764982964
agg_1113	CGE interneuron	Amygdala_Basolateral nuclear group (BLN) - lateral nucleus - La	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Basolateral nuclear group (BLN) - lateral nucleus - La		3653	43170604.0	17556	43343.31426365281	0.9556543498602063	0.8580799006688026	0.8653460664237342
agg_1114	CGE interneuron	Amygdala_Bed nucleus of stria terminalis and nearby - BNST	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Bed nucleus of stria terminalis and nearby - BNST		618	6152593.0	16370	44713.64203304153	0.9520635955734341	0.852188720222159	0.858968806922992
agg_1115	CGE interneuron	Amygdala_Central nuclear group - CEN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Central nuclear group - CEN		4146	49400198.0	17671	44026.860058199665	0.9563052838304861	0.8593824369699917	0.8633745639889623
agg_1116	CGE interneuron	Amygdala_Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA		929	26964535.0	17136	42858.8950204464	0.9550498609794265	0.8605038580533132	0.8684496521271442
agg_1117	CGE interneuron	Amygdala_basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL		4827	53518064.0	17716	44281.06098179032	0.9559516728570443	0.8567887668317138	0.8635709259614972
agg_1118	CGE interneuron	Amygdala_basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM		3328	43659654.0	17730	45262.446914261665	0.9542446453364304	0.8568247176176679	0.8600818558559432
agg_1119	CGE interneuron	Amygdala_corticomedial nuclear group - CMN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	corticomedial nuclear group - CMN		4434	51772973.0	17758	44807.696973438295	0.9570021114498952	0.8556658443449039	0.8641266258892707
agg_1120	CGE interneuron	Basal ganglia_Body of the Caudate - CaB	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Body of the Caudate - CaB		187	3122777.0	15683	44838.93941730748	0.9494639797626426	0.850804855581977	0.8553423638918904
agg_1121	CGE interneuron	Basal ganglia_Globus pallidus (GP) - External segment of globus pallidus - GPe	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Globus pallidus (GP) - External segment of globus pallidus - GPe		107	1401962.0	14986	44440.308958941176	0.9432617394411168	0.8448997012052522	0.8476170302061452
agg_1122	CGE interneuron	Basal ganglia_Globus pallidus (GP) - Internal segment of globus pallidus - GPi	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Globus pallidus (GP) - Internal segment of globus pallidus - GPi		16	121804.0	10739	39259.57497605263	0.858884474221949	0.7726171135663182	0.7644669154369795
agg_1123	CGE interneuron	Basal ganglia_Nucleus Accumbens - NAC	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Nucleus Accumbens - NAC		92	1157081.0	14512	43915.45804097595	0.9406993659470231	0.8463012081482822	0.8484284417222941
agg_1124	CGE interneuron	Basal ganglia_Nucleus accumbens	CNS	healthy	jhpce#tran2021	brain_atlas	Basal ganglia	Nucleus accumbens		294	6261733.0	16363	44435.4512744186	0.939417438111905	0.8392186623819058	0.8527706997362122
agg_1125	CGE interneuron	Basal ganglia_Putamen - Pu	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Putamen - Pu		425	4717570.0	16183	45001.20451703623	0.9514341635683559	0.8539321245006076	0.855428781988558
agg_1126	CGE interneuron	Basal ganglia_septal nuclei - SEP	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	septal nuclei - SEP		1879	21882302.0	17497	46007.24312666675	0.9527114190907388	0.8532013539249315	0.8558978082659188
agg_1127	CGE interneuron	Basal ganglia_substantia innominata and nearby nuclei - SI	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	substantia innominata and nearby nuclei - SI		765	7857686.0	16676	45181.13280453827	0.9528570987403407	0.8528510578046272	0.857340287067751
agg_1128	CGE interneuron	Cerebral cortex_Anterior Olfactory Nucleus - AON	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior Olfactory Nucleus - AON		2403	39067991.0	17616	45931.36796447533	0.9546916552053556	0.8522756501950615	0.8582181016005722
agg_1130	CGE interneuron	Cerebral cortex_Anterior cingulate cortex	CNS	healthy	PRJNA434002	brain_atlas	Cerebral cortex	Anterior cingulate cortex		1644	7947983.0	15944	45552.80934538369	0.899805511496978	0.8093617097115822	0.8000548329236696
agg_1131	CGE interneuron	Cerebral cortex_Anterior cingulate cortex - ACC	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior cingulate cortex - ACC		3716	58054934.0	17749	45840.820597657774	0.9534435475418815	0.853738798083762	0.8559544023158703
agg_1132	CGE interneuron	Cerebral cortex_Anterior parahippocampal gyrus (AG) - Lateral entorhinal cortex - LEC	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior parahippocampal gyrus (AG) - Lateral entorhinal cortex - LEC		5912	81002476.0	17880	45675.62135987685	0.9541385127816103	0.8500039325320953	0.8568608826029541
agg_1133	CGE interneuron	Cerebral cortex_Anterior parahippocampal gyrus/posterior part (APH) - Medial entorhinal cortex - MEC	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior parahippocampal gyrus/posterior part (APH) - Medial entorhinal cortex - MEC		11071	153176261.0	18055	44934.06085029526	0.9564513772455941	0.8578279194724209	0.8613450453709655
agg_1134	CGE interneuron	Cerebral cortex_Caudal cingulate gyrus (CgGC) - A23	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Caudal cingulate gyrus (CgGC) - A23		4324	45019739.0	17656	45110.451368021866	0.9539821503006068	0.8559550389720095	0.857177067428263
agg_1135	CGE interneuron	Cerebral cortex_Cingulate gyrus/retrosplenial (CgGrs) - A29-A30	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Cingulate gyrus/retrosplenial (CgGrs) - A29-A30		6632	39702720.0	17615	46139.58724353013	0.9494599332961685	0.8495667411664938	0.8459029902797677
agg_1136	CGE interneuron	Cerebral cortex_Cuneus/caudal part - Peristriate Cortex - V2	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Cuneus/caudal part - Peristriate Cortex - V2		3560	45381187.0	17654	45224.39627648392	0.9537145582307842	0.8516769565904151	0.8558226304003664
agg_1137	CGE interneuron	Cerebral cortex_Cuneus/rostral part - Area Prostriata - Pro	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Cuneus/rostral part - Area Prostriata - Pro		2859	35796270.0	17617	46132.638278870865	0.953210222113289	0.8503846763672066	0.8517721300452203
agg_1138	CGE interneuron	Cerebral cortex_Dorsalateral prefrontal cortex	CNS	healthy	http://psychencode.org	brain_atlas	Cerebral cortex	Dorsalateral prefrontal cortex		1423	4616796.0	15660	48569.99403936421	0.9117588212407522	0.7869315086586617	0.8134432687774173
agg_1140	CGE interneuron	Cerebral cortex_Dorsolateral prefrontal cortex	CNS	healthy	GSE129308	brain_atlas	Cerebral cortex	Dorsolateral prefrontal cortex		3811	7049199.0	15518	45472.90946388328	0.9123433698437873	0.8126679354513144	0.8000875424359442
agg_1141	CGE interneuron	Cerebral cortex_Dorsolateral prefrontal cortex	CNS	healthy	SCR_002001	brain_atlas	Cerebral cortex	Dorsolateral prefrontal cortex		10725	68649325.0	14884	41705.65745832643	0.8781713598569268	0.8191804553888763	0.799212985469408
agg_1142	CGE interneuron	Cerebral cortex_Dorsolateral prefrontal cortex	CNS	healthy	jhpce#tran2021	brain_atlas	Cerebral cortex	Dorsolateral prefrontal cortex		567	12085629.0	16977	43519.60306913432	0.9478968648853212	0.8553751683830699	0.8666560185155503
agg_1145	CGE interneuron	Cerebral cortex_Entorhinal cortex	CNS	healthy	GSE160936	brain_atlas	Cerebral cortex	Entorhinal cortex		1047	4387962.0	15399	40719.46283580456	0.9286596800224921	0.8366923122084351	0.8507547006018672
agg_1147	CGE interneuron	Cerebral cortex_Frontal Cortex	CNS	healthy	GSE163122	brain_atlas	Cerebral cortex	Frontal Cortex		19	54018.0	8344	33326.7485689139	0.7992434996890754	0.7329329945323536	0.7380634826096151
agg_1148	CGE interneuron	Cerebral cortex_Frontal agranular insular cortex - FI	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Frontal agranular insular cortex - FI		4681	86805040.0	17857	44416.26306035089	0.9557815899225688	0.8595831773838496	0.8640557588613347
agg_1149	CGE interneuron	Cerebral cortex_Gyrus rectus (ReG) - Medial orbitofrontal cortex - A14	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Gyrus rectus (ReG) - Medial orbitofrontal cortex - A14		1279	31692146.0	17425	45084.21292275218	0.9550103469929688	0.8561271654196	0.8603893866482042
agg_1150	CGE interneuron	Cerebral cortex_Inferior frontal gyrus (IFG) - Ventrolateral prefrontal cortex - A44-A45	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Inferior frontal gyrus (IFG) - Ventrolateral prefrontal cortex - A44-A45		5033	69465714.0	17848	45994.239910992255	0.9554778167568956	0.8536153597606401	0.8555734610691637
agg_1151	CGE interneuron	Cerebral cortex_Inferior temporal gyrus - ITG	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Inferior temporal gyrus - ITG		5284	72800988.0	17847	45716.37075056562	0.9536311565746031	0.8544356903796938	0.8540222039431139
agg_1152	CGE interneuron	Cerebral cortex_Lingual gyrus (LiG) - Primary Visual Cortex - V1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Lingual gyrus (LiG) - Primary Visual Cortex - V1C		2183	22471869.0	17322	45577.36171821733	0.9524809146890161	0.8516653474121149	0.8531071638907382
agg_1153	CGE interneuron	Cerebral cortex_Long insular gyri (LIG) - Dysgranular insular cortex - Idg	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Long insular gyri (LIG) - Dysgranular insular cortex - Idg		5126	82685320.0	17916	46218.94033766634	0.9542202102627187	0.8510695199936927	0.8546741809365636
agg_1154	CGE interneuron	Cerebral cortex_Middle Temporal Gyrus - MTG	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Middle Temporal Gyrus - MTG		11311	151558742.0	18010	45732.36858827974	0.952730565143899	0.8519098774464178	0.8534353362507285
agg_1155	CGE interneuron	Cerebral cortex_Middle frontal gyrus (MFG) - A46	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Middle frontal gyrus (MFG) - A46		3951	46477496.0	17662	45688.617980936666	0.953364062745528	0.851636584925773	0.8545891176316116
agg_1157	CGE interneuron	Cerebral cortex_Occipital Cortex	CNS	healthy	GSE163122	brain_atlas	Cerebral cortex	Occipital Cortex		171	212716.0	12145	40982.11625428927	0.8832717415790446	0.7948885354834843	0.8225702992204573
agg_1158	CGE interneuron	Cerebral cortex_Occipitotemporal (fusiform) gyrus/temporal part (FuGt) - Temporal area TF	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Occipitotemporal (fusiform) gyrus/temporal part (FuGt) - Temporal area TF		3975	43199557.0	17636	45507.23819226017	0.9532879143459528	0.8533294099175049	0.8554793688418564
agg_1159	CGE interneuron	Cerebral cortex_Parietal lobe	CNS	healthy	PRJNA544731	brain_atlas	Cerebral cortex	Parietal lobe		325	1291000.0	13938	39998.241565767414	0.904521582278289	0.8193605423836186	0.8176628902062328
agg_1161	CGE interneuron	Cerebral cortex_Parietal operculum (PaO) - Gustatory cortex - A43	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Parietal operculum (PaO) - Gustatory cortex - A43		5024	78701516.0	17892	46142.555314647965	0.9539450591818139	0.8525012196667436	0.855167790430822
agg_1162	CGE interneuron	Cerebral cortex_Perirhinal gyrus (PRG) - A35-A36	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Perirhinal gyrus (PRG) - A35-A36		3012	51432821.0	17692	45131.31805705638	0.9560632331488161	0.8555943455175027	0.8614218812447207
agg_1163	CGE interneuron	Cerebral cortex_Perirhinal gyrus (PRG) -  A35-A36	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Perirhinal gyrus (PRG) -  A35-A36		144	3077583.0	15860	45585.48590990845	0.94770755111472	0.8524365900027768	0.8516278734318256
agg_1164	CGE interneuron	Cerebral cortex_Piriform cortex - Pir	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Piriform cortex - Pir		5304	73012905.0	17813	44873.944335107975	0.9556997945549233	0.8548111905904039	0.860258266899383
agg_1165	CGE interneuron	Cerebral cortex_Post-mortem dosolateral BA9	CNS	healthy	GSE144136	brain_atlas	Cerebral cortex	Post-mortem dosolateral BA9		2570	5965814.0	16155	47522.884224494985	0.9068086039182727	0.7989989458092506	0.7934264404955423
agg_1167	CGE interneuron	Cerebral cortex_Postcentral gyrus (PoCG) - Primary somatosensory cortex - S1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Postcentral gyrus (PoCG) - Primary somatosensory cortex - S1C		4884	62996504.0	17825	46019.31193181314	0.9532354131992904	0.8507754809917075	0.8524770203011584
agg_1168	CGE interneuron	Cerebral cortex_Posterior intermediate orbital gyrus (POrG) - Caudal division of OFCi - A13	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Posterior intermediate orbital gyrus (POrG) - Caudal division of OFCi - A13		4262	52913409.0	17753	45998.20253059246	0.953615905848413	0.8538107256636154	0.8558539119476164
agg_1169	CGE interneuron	Cerebral cortex_Posterior parahippocampal gyrus (PPH) - TH-TL	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Posterior parahippocampal gyrus (PPH) - TH-TL		4825	76436336.0	17892	46061.05240538032	0.9535840982569064	0.8528944265631686	0.8547035396100106
agg_1170	CGE interneuron	Cerebral cortex_Precentral gyrus (PrCG) - Primary motor cortex - M1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Precentral gyrus (PrCG) - Primary motor cortex - M1C		12727	146884792.0	17963	44470.47415111472	0.9547226046429305	0.8572875463413503	0.8600483658810508
agg_1176	CGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	GSE157827	brain_atlas	Cerebral cortex	Prefrontal cortex		5008	19848445.0	17304	45159.62227579555	0.9393600054559287	0.8396432923534439	0.8450459166741539
agg_1177	CGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	GSE167494	brain_atlas	Cerebral cortex	Prefrontal cortex		1794	13211024.0	17409	45724.65583715663	0.9469091235910225	0.8594468461006485	0.8649022466659783
agg_1178	CGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	PRJNA434002	brain_atlas	Cerebral cortex	Prefrontal cortex		1665	9481795.0	16016	45621.72012996503	0.8979547648941134	0.8073909907070784	0.7965111186378239
agg_1179	CGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	PRJNA544731	brain_atlas	Cerebral cortex	Prefrontal cortex		826	3367041.0	15274	43661.14958487554	0.9051999043253188	0.8133421639122885	0.8088287084809244
agg_1181	CGE interneuron	Cerebral cortex_Premotor cortex	CNS	healthy	PRJNA544731	brain_atlas	Cerebral cortex	Premotor cortex		350	1081458.0	13948	41367.59429187521	0.9021448372916091	0.8212426090282491	0.8128519813702438
agg_1183	CGE interneuron	Cerebral cortex_Rostral gyrus (RoG) - Dorsal division of MFC - A32	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Rostral gyrus (RoG) - Dorsal division of MFC - A32		5300	48080151.0	17740	45940.14992316566	0.9534549960902643	0.8524659349518056	0.853964641983153
agg_1184	CGE interneuron	Cerebral cortex_Short insular gyri - Granular insular cortex - Ig	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Short insular gyri - Granular insular cortex - Ig		3872	54863803.0	17796	46214.28719021868	0.9535166883859459	0.8496326796851502	0.85317863812959
agg_1186	CGE interneuron	Cerebral cortex_Somatosensory cortex	CNS	healthy	GSE160936	brain_atlas	Cerebral cortex	Somatosensory cortex		582	2485492.0	15425	42287.690403027314	0.9287968390307827	0.832253735985677	0.8485581697182518
agg_1187	CGE interneuron	Cerebral cortex_Subcallosal Gyrus (SCG) - Subgenual (subcallosal) division of MFC - A25	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Subcallosal Gyrus (SCG) - Subgenual (subcallosal) division of MFC - A25		4614	63037552.0	17855	45818.45271602795	0.9540811285074197	0.8550988434064752	0.8573155675073667
agg_1188	CGE interneuron	Cerebral cortex_Subgenual anterior cingulate cortex	CNS	healthy	jhpce#tran2021	brain_atlas	Cerebral cortex	Subgenual anterior cingulate cortex		2103	51030690.0	17508	42976.54360402118	0.9443716615080814	0.8547308776338655	0.8686303762745468
agg_1189	CGE interneuron	Cerebral cortex_Superior Temporal Gyrus - STG	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Superior Temporal Gyrus - STG		3973	39680274.0	17664	46796.560637020826	0.9516814599996662	0.846081904137383	0.8486666567787549
agg_1191	CGE interneuron	Cerebral cortex_Superior occipital gyrus (SOG) - Areas 19 and MT - A19	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Superior occipital gyrus (SOG) - Areas 19 and MT - A19		3558	49128087.0	17635	44582.24762740059	0.9546774331036844	0.8554153722545857	0.8592675654958597
agg_1192	CGE interneuron	Cerebral cortex_Supramarginal gyrus (SMG) - A40	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Supramarginal gyrus (SMG) - A40		5683	70588900.0	17849	45684.68704753316	0.9525630511152248	0.8524892474699114	0.856410536999382
agg_1193	CGE interneuron	Cerebral cortex_Supraparietal lobule (SPL) - Posterosuperior (dorsal) parietal cortex - A5-A7	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Supraparietal lobule (SPL) - Posterosuperior (dorsal) parietal cortex - A5-A7		4244	59929163.0	17802	45565.49384897568	0.9535026029597516	0.8517910261710867	0.855050582794054
agg_1195	CGE interneuron	Cerebral cortex_Temporal Cortex	CNS	healthy	GSE163122	brain_atlas	Cerebral cortex	Temporal Cortex		15	64527.0	8940	35023.97356147146	0.8093981501317319	0.7433994035544298	0.7365259401750139
agg_1197	CGE interneuron	Cerebral cortex_Temporal pole (TP) - Temporopolar area - A38	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Temporal pole (TP) - Temporopolar area - A38		5613	90000762.0	17933	45760.61062590786	0.955041605196879	0.8539103329628183	0.8586211004297455
agg_1198	CGE interneuron	Cerebral cortex_Transverse temporal gyrus (TTG) - Primary auditory cortex - A1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Transverse temporal gyrus (TTG) - Primary auditory cortex - A1C		3730	48356883.0	17721	45865.96252709382	0.952933906327049	0.850614355382243	0.8530717089075621
agg_1199	CGE interneuron	Cerebral cortex_Unclassified	CNS	healthy	GSE140231	brain_atlas	Cerebral cortex	Unclassified		1284	5120738.0	15935	46875.188661093554	0.8985067250057941	0.7975575635821707	0.787124459767903
agg_1201	CGE interneuron	Cerebral cortex_nan	CNS	healthy	GSE160936	brain_atlas	Cerebral cortex	nan		323	1134778.0	14542	42514.804335946086	0.9226715988771624	0.8225061115769186	0.8428913356673597
agg_1203	CGE interneuron	Cerebral cortex_occipital cortex	CNS	healthy	GSE148822	brain_atlas	Cerebral cortex	occipital cortex		1450	2365826.0	15947	44174.234808531546	0.9329289198690828	0.842346530000887	0.8588884951633767
agg_1205	CGE interneuron	Cerebral cortex_occipitotemporal cortex	CNS	healthy	GSE148822	brain_atlas	Cerebral cortex	occipitotemporal cortex		221	534585.0	13874	42991.07634366376	0.9140926425122817	0.8296886259606971	0.8383202288704407
agg_1206	CGE interneuron	Epithalamus_ETH	CNS	healthy	SCR_016152	brain_atlas	Epithalamus	ETH		57	462804.0	13029	42272.40776782958	0.9162630591666623	0.8259699566888652	0.8256469609995585
agg_1207	CGE interneuron	Grey matter_Cla	CNS	healthy	SCR_016152	brain_atlas	Grey matter	Cla		4536	60807216.0	17792	45427.60235087236	0.9543023070218183	0.8555569763795798	0.8561216537911441
agg_1210	CGE interneuron	Grey matter_Motor cortex	CNS	healthy	GSE174332	brain_atlas	Grey matter	Motor cortex		2248	20697132.0	17490	45333.17698450644	0.9493084819108504	0.8622889159965869	0.8654712716809753
agg_1216	CGE interneuron	Hippocampus_CA1	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA1		1056	24257946.0	17319	44618.662958680216	0.9555082946462914	0.8576587816177856	0.8652458087825374
agg_1217	CGE interneuron	Hippocampus_CA1-3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA1-3		1399	30321942.0	17654	45952.11000641589	0.9567527824344031	0.8585671841053573	0.8649575860777023
agg_1218	CGE interneuron	Hippocampus_CA2-3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA2-3		715	12822333.0	17047	45206.041748648764	0.9548881672748422	0.8555328863891355	0.8598074855657095
agg_1219	CGE interneuron	Hippocampus_Caudal Hippocampus - CA1-CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Caudal Hippocampus - CA1-CA3		4778	67753760.0	17845	45882.31531766433	0.955659926282955	0.8570173925421611	0.8592597569055781
agg_1220	CGE interneuron	Hippocampus_Caudal Hippocampus - CA4-DGC	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Caudal Hippocampus - CA4-DGC		3850	43983979.0	17665	44794.33524642079	0.9556424956632221	0.8578986973545148	0.862918582121493
agg_1221	CGE interneuron	Hippocampus_DG-CA4	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	DG-CA4		402	8642108.0	16547	43857.65063083102	0.9546136843144307	0.8563669085138821	0.8665502487018528
agg_1222	CGE interneuron	Hippocampus_Hippocampus	CNS	healthy	jhpce#tran2021	brain_atlas	Hippocampus	Hippocampus		198	5257259.0	16302	43694.06764200175	0.9439521677670791	0.8483678523458008	0.8615763326636526
agg_1228	CGE interneuron	Hippocampus_Rostral CA1-2	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral CA1-2		1446	42457430.0	17602	44976.25496962376	0.956097749881942	0.8592218643503801	0.8613344630958334
agg_1229	CGE interneuron	Hippocampus_Rostral CA1-CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral CA1-CA3		2477	34783394.0	17644	45816.66612130956	0.9537393499883875	0.8529414960671245	0.8563138700079675
agg_1230	CGE interneuron	Hippocampus_Rostral CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral CA3		1115	31640031.0	17496	43581.95994388098	0.9576084197006921	0.8638239733498098	0.8756803089236452
agg_1231	CGE interneuron	Hippocampus_Rostral DG-CA4	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral DG-CA4		1250	19699269.0	17238	44342.12441936283	0.9550084099177176	0.8575808754008807	0.8668830226576822
agg_1232	CGE interneuron	Hippocampus_Subicular cortex - Sub	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Subicular cortex - Sub		3585	27126836.0	17514	45615.54245224047	0.9544276922096665	0.8541096440988695	0.8547769440559505
agg_1233	CGE interneuron	Hippocampus_Uncal CA1-CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Uncal CA1-CA3		4243	24611820.0	17555	46345.80013735496	0.947789066922996	0.8475197737330911	0.8457748240749084
agg_1234	CGE interneuron	Hippocampus_Uncal DG-CA4	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Uncal DG-CA4		1003	7418678.0	16817	46113.96045898331	0.9466531252449502	0.8489206492374427	0.848386075106613
agg_1235	CGE interneuron	Hypothalamus_mammillary region of HTH (HTHma)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	mammillary region of HTH (HTHma)		19	169012.0	11320	40121.797006176486	0.8856662995090161	0.8037367990593358	0.802838778687612
agg_1236	CGE interneuron	Hypothalamus_mammillary region of HTH (HTHma) - mammillary nucleus - MN	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	mammillary region of HTH (HTHma) - mammillary nucleus - MN		25	779705.0	13829	40785.11147595983	0.9232268717942013	0.8409818874565197	0.8450233312927798
agg_1237	CGE interneuron	Hypothalamus_mammillary region of HTH (HTHma) - tuberal region of HTH - HTHtub	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	mammillary region of HTH (HTHma) - tuberal region of HTH - HTHtub		48	380963.0	13326	43897.297697504284	0.9182130253389306	0.82540737747863	0.8295891460044542
agg_1238	CGE interneuron	Hypothalamus_preoptic region of HTH - HTHpo	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo		161	2149160.0	14796	42700.563378464205	0.9453035875743634	0.8536411059732548	0.8555891746642673
agg_1239	CGE interneuron	Hypothalamus_preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)		174	1690731.0	14915	44460.50946676239	0.9435625577642032	0.840635467118368	0.8449369718665004
agg_1240	CGE interneuron	Hypothalamus_supraoptic region of HTH - HTHso	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	supraoptic region of HTH - HTHso		1423	31295463.0	17220	42601.00029777002	0.9509127659692103	0.8596009580473339	0.8682058813568818
agg_1241	CGE interneuron	Hypothalamus_supraoptic region of HTH - HTHso (anterior hypothalamic nucleus/AHN) - tuberal region of HTH - HTHtub (ventromedial and dorsomedial hypothalamic nucleic nuclei/VMH/DMH)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	supraoptic region of HTH - HTHso (anterior hypothalamic nucleus/AHN) - tuberal region of HTH - HTHtub (ventromedial and dorsomedial hypothalamic nucleic nuclei/VMH/DMH)		35	78228.0	10009	38635.09504756012	0.8334301185707277	0.7367611888109045	0.7452785744752756
agg_1242	CGE interneuron	Hypothalamus_supraoptic region of HTH - HTHso (anterior hypothalamic nucleus/AHN) - tuberal region of HTH - HTHtub (ventromedial hypothalamic nucleus/VMH)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	supraoptic region of HTH - HTHso (anterior hypothalamic nucleus/AHN) - tuberal region of HTH - HTHtub (ventromedial hypothalamic nucleus/VMH)		70	1101865.0	14717	45919.72980885698	0.9365659112653159	0.8300524413766043	0.8401280236667859
agg_1243	CGE interneuron	Hypothalamus_tuberal region of hypothalamus - HTHtub	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	tuberal region of hypothalamus - HTHtub		256	4585880.0	15703	42174.16231383817	0.9483005087846198	0.8557045973739881	0.8668683077339957
agg_1244	CGE interneuron	Midbrain_Inferior colliculus and nearby nuclei - IC	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Inferior colliculus and nearby nuclei - IC		13	164176.0	11372	40071.06497873283	0.8783321251348539	0.7942595405346409	0.804390317549318
agg_1245	CGE interneuron	Midbrain_Periaqueductal gray and Dorsal raphe nucleus - PAG-DR	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Periaqueductal gray and Dorsal raphe nucleus - PAG-DR		78	428240.0	13448	43645.67688472008	0.9206388104215625	0.8297756019894639	0.8304806953604367
agg_1246	CGE interneuron	Midbrain_Pretectal region - PTR	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Pretectal region - PTR		48	586996.0	13256	41778.929031273845	0.9246356299843965	0.8400121394611224	0.8391326953328622
agg_1247	CGE interneuron	Midbrain_Substantia Nigra - SN	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Substantia Nigra - SN		186	1644370.0	15226	44739.80581505666	0.9422287767005032	0.8413773929990623	0.850756673157757
agg_1248	CGE interneuron	Midbrain_Superior colliculus and nearby nuclei - SC	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Superior colliculus and nearby nuclei - SC		350	2935116.0	15645	44486.84828813827	0.948365632376795	0.8491916359469455	0.8522610942498586
agg_1249	CGE interneuron	Thalamus_Anterior nuclear complex - ANC	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Anterior nuclear complex - ANC		415	4628862.0	15887	43221.11709124231	0.9515883777834919	0.8559015219484319	0.8618820915379024
agg_1250	CGE interneuron	Thalamus_Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA		148	2591291.0	15227	43776.5705289588	0.9479673500619797	0.8515409693913362	0.8579637148668879
agg_1251	CGE interneuron	Thalamus_Lateral nuclear complex of thalamus (LNC) - ventral group of lateral nucleus - VLN	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Lateral nuclear complex of thalamus (LNC) - ventral group of lateral nucleus - VLN		20	465529.0	12608	40458.640982486155	0.9204919458250058	0.833472525331988	0.8363484000790021
agg_1252	CGE interneuron	Thalamus_Subthalamic nucleus and nearby - STH	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Subthalamic nucleus and nearby - STH		12	48930.0	8273	33601.57506628955	0.7916353056350358	0.717005097973611	0.7161454527217833
agg_1253	CGE interneuron	Thalamus_Ventral group of lateral nucleus (VLN) - ventral anterior nucleus of thalamus - VA - ventral lateral nucleus of thalamus - VL	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Ventral group of lateral nucleus (VLN) - ventral anterior nucleus of thalamus - VA - ventral lateral nucleus of thalamus - VL		80	405359.0	13163	42754.84355247164	0.9134407943724896	0.816077952619879	0.8215123320289275
agg_1254	CGE interneuron	Thalamus_intralaminar nuclear complex (ILN) - posterior group of intralaminar nuclei (PILN) - centromedian and parafasicular nuclei - CM and Pf	CNS	healthy	SCR_016152	brain_atlas	Thalamus	intralaminar nuclear complex (ILN) - posterior group of intralaminar nuclei (PILN) - centromedian and parafasicular nuclei - CM and Pf		33	533473.0	13189	40753.83942051242	0.9240632401164864	0.8358777478158795	0.8407518009319821
agg_1255	CGE interneuron	Thalamus_lateral nuclear complex of thalamus (LNC) - Pulvinar of thalamus - Pul	CNS	healthy	SCR_016152	brain_atlas	Thalamus	lateral nuclear complex of thalamus (LNC) - Pulvinar of thalamus - Pul		684	3973754.0	15983	43590.08238875969	0.9445663086463763	0.8503303286245829	0.8516656964747131
agg_1256	CGE interneuron	Thalamus_lateral nuclear complex of thalamus (LNC) - lateral posterior nucleus of thalamus - LP	CNS	healthy	SCR_016152	brain_atlas	Thalamus	lateral nuclear complex of thalamus (LNC) - lateral posterior nucleus of thalamus - LP		226	1410013.0	14621	42843.472174328854	0.9371470188185859	0.8481872653519125	0.8463956972968795
agg_1257	CGE interneuron	Thalamus_lateral nuclear complex of thalamus (LNC) - ventral posterior lateral nucleus - VPL	CNS	healthy	SCR_016152	brain_atlas	Thalamus	lateral nuclear complex of thalamus (LNC) - ventral posterior lateral nucleus - VPL		61	553114.0	13801	43686.118597136345	0.916540300502315	0.8204999882838292	0.8167449767586076
agg_1258	CGE interneuron	Thalamus_medial nuclear complex of thalamus (MNC) - mediodorsal nucleus of thalamus + reuniens nucleus (medioventral nucleus) of thalamus - MD + Re	CNS	healthy	SCR_016152	brain_atlas	Thalamus	medial nuclear complex of thalamus (MNC) - mediodorsal nucleus of thalamus + reuniens nucleus (medioventral nucleus) of thalamus - MD + Re		14	254289.0	11620	39589.30912901781	0.893805244087664	0.7964987890849986	0.7999827705781003
agg_1259	CGE interneuron	Thalamus_medial nuclear complex of thalamus (MNC) - mediodorsal nucleus of thalamus - MD	CNS	healthy	SCR_016152	brain_atlas	Thalamus	medial nuclear complex of thalamus (MNC) - mediodorsal nucleus of thalamus - MD		68	891523.0	14285	44531.72968036648	0.9341282492048123	0.8347986333134049	0.8409491999172779
agg_1260	CGE interneuron	Thalamus_posterior nuclear complex of thalamus (PoN) - lateral geniculate nucleus (LG)	CNS	healthy	SCR_016152	brain_atlas	Thalamus	posterior nuclear complex of thalamus (PoN) - lateral geniculate nucleus (LG)		1116	8879856.0	16606	43993.94471049856	0.9451924316131305	0.8514029501274426	0.8461715340569704
agg_1261	CGE interneuron	Thalamus_posterior nuclear complex of thalamus (PoN) - medial geniculate nuclei (MG)	CNS	healthy	SCR_016152	brain_atlas	Thalamus	posterior nuclear complex of thalamus (PoN) - medial geniculate nuclei (MG)		355	2234733.0	15185	43529.744792294245	0.9363821168798473	0.8477657453423078	0.8410011875784053
agg_1262	CGE interneuron	White matter_Unclassifed	CNS	healthy	GSE118257	brain_atlas	White matter	Unclassifed		48	102149.0	9997	38110.36556036987	0.8459352980352427	0.7713240018194049	0.7659342353072323
agg_1836	Eccentric medium spiny neuron	Amygdala_Amygdala	CNS	healthy	jhpce#tran2021	brain_atlas	Amygdala	Amygdala		253	3498891.0	15519	40448.34183203413	0.9349916890507771	0.8498156882148697	0.8717524868530316
agg_1837	Eccentric medium spiny neuron	Amygdala_Basolateral nuclear group (BLN) - lateral nucleus - La	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Basolateral nuclear group (BLN) - lateral nucleus - La		2614	37373934.0	17392	41638.12221921355	0.946974437450923	0.847991661796929	0.8728311397041211
agg_1838	Eccentric medium spiny neuron	Amygdala_Bed nucleus of stria terminalis and nearby - BNST	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Bed nucleus of stria terminalis and nearby - BNST		4070	61034153.0	17725	43366.55866013142	0.9449832628069204	0.8477951141899623	0.8647579750069853
agg_1839	Eccentric medium spiny neuron	Amygdala_Central nuclear group - CEN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Central nuclear group - CEN		3691	76250911.0	17745	42044.66933888139	0.9514448017876914	0.8543214296020579	0.876937120815823
agg_1840	Eccentric medium spiny neuron	Amygdala_Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA		312	9332001.0	16304	41613.063910534125	0.9461576605204305	0.8537505332441399	0.8718902967128986
agg_1841	Eccentric medium spiny neuron	Amygdala_basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL		991	14693462.0	17096	43349.253663206655	0.9466071511985032	0.8409559911723744	0.8654262663177494
agg_1842	Eccentric medium spiny neuron	Amygdala_basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM		2136	34214943.0	17486	43531.78569767113	0.9456787682949195	0.8401723315976578	0.8651359242111114
agg_1843	Eccentric medium spiny neuron	Amygdala_corticomedial nuclear group - CMN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	corticomedial nuclear group - CMN		3958	48755555.0	17694	43174.876652443105	0.9506933018097268	0.8516551225422615	0.8731869512180219
agg_1844	Eccentric medium spiny neuron	Basal ganglia_Body of the Caudate - CaB	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Body of the Caudate - CaB		1551	39202986.0	17424	41457.74016252917	0.9259798424126457	0.8260169421932617	0.8447155432709379
agg_1845	Eccentric medium spiny neuron	Basal ganglia_Globus pallidus (GP) - External segment of globus pallidus - GPe	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Globus pallidus (GP) - External segment of globus pallidus - GPe		1290	35374546.0	17372	41841.67685398771	0.9374765396080563	0.8387183698860925	0.8588994519905235
agg_1846	Eccentric medium spiny neuron	Basal ganglia_Nucleus Accumbens - NAC	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Nucleus Accumbens - NAC		2459	55330510.0	17609	42872.9089593033	0.9373280224975074	0.8361335100131294	0.855366592773061
agg_1847	Eccentric medium spiny neuron	Basal ganglia_Nucleus accumbens	CNS	healthy	jhpce#tran2021	brain_atlas	Basal ganglia	Nucleus accumbens		937	19344324.0	16952	41691.59203456692	0.9292516932598761	0.8324143458310376	0.8583836544325235
agg_1848	Eccentric medium spiny neuron	Basal ganglia_Putamen - Pu	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Putamen - Pu		1532	39824493.0	17432	41522.77853413851	0.9321374825366149	0.8343397129615645	0.8558908523690969
agg_1849	Eccentric medium spiny neuron	Basal ganglia_septal nuclei - SEP	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	septal nuclei - SEP		650	9239985.0	16787	43539.592830885056	0.9348948990808218	0.8383143024241715	0.8500947548924618
agg_1850	Eccentric medium spiny neuron	Basal ganglia_substantia innominata and nearby nuclei - SI	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	substantia innominata and nearby nuclei - SI		5395	95304701.0	17845	42671.29521993991	0.9418578412657839	0.8418476785493142	0.8629907394936818
agg_1851	Eccentric medium spiny neuron	Cerebral cortex_Anterior Olfactory Nucleus - AON	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior Olfactory Nucleus - AON		1904	20326709.0	17272	44304.954065111655	0.9364907667665184	0.8333917083634919	0.8547522246345102
agg_1852	Eccentric medium spiny neuron	Cerebral cortex_Dorsolateral prefrontal cortex	CNS	healthy	jhpce#tran2021	brain_atlas	Cerebral cortex	Dorsolateral prefrontal cortex		91	3594350.0	15539	41081.34436554239	0.935941817114317	0.8406435959927373	0.8686615915475352
agg_1853	Eccentric medium spiny neuron	Cerebral cortex_Frontal agranular insular cortex - FI	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Frontal agranular insular cortex - FI		63	2075917.0	14566	40645.474011381564	0.9323820955609043	0.833562300525369	0.8579783672917262
agg_1854	Eccentric medium spiny neuron	Cerebral cortex_Long insular gyri (LIG) - Dysgranular insular cortex - Idg	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Long insular gyri (LIG) - Dysgranular insular cortex - Idg		14	143115.0	10990	39309.121330738846	0.8589666904382315	0.7575177551708354	0.7780514721975762
agg_1855	Eccentric medium spiny neuron	Cerebral cortex_Perirhinal gyrus (PRG) - A35-A36	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Perirhinal gyrus (PRG) - A35-A36		1123	18373061.0	17218	43361.51675387646	0.9475325666266081	0.8421233213753856	0.8673991051714353
agg_1856	Eccentric medium spiny neuron	Cerebral cortex_Piriform cortex - Pir	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Piriform cortex - Pir		946	26226854.0	17096	41729.33133861903	0.9486876187321196	0.8511102034939563	0.8741164177181678
agg_1857	Eccentric medium spiny neuron	Cerebral cortex_Precentral gyrus (PrCG) - Primary motor cortex - M1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Precentral gyrus (PrCG) - Primary motor cortex - M1C		15	100437.0	10273	37957.500142213736	0.8619401611857915	0.7743646778169242	0.7940092190534478
agg_1858	Eccentric medium spiny neuron	Cerebral cortex_Subgenual anterior cingulate cortex	CNS	healthy	jhpce#tran2021	brain_atlas	Cerebral cortex	Subgenual anterior cingulate cortex		113	4106180.0	15885	41810.112484140525	0.9295512711427354	0.8391536584052428	0.8669023034905983
agg_1859	Eccentric medium spiny neuron	Grey matter_Cla	CNS	healthy	SCR_016152	brain_atlas	Grey matter	Cla		520	8411038.0	16478	43028.1498873797	0.9447777604776917	0.8431661644849828	0.8658352810792844
agg_1860	Eccentric medium spiny neuron	Hippocampus_CA1	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA1		84	1689249.0	14611	42308.78649896261	0.9401621388865626	0.835068666475771	0.8551525609823153
agg_1861	Eccentric medium spiny neuron	Hippocampus_CA1-3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA1-3		13	293094.0	12278	40954.80075031351	0.8992217596035066	0.7999397475207897	0.8231597817826668
agg_1862	Eccentric medium spiny neuron	Hippocampus_Caudal Hippocampus - CA4-DGC	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Caudal Hippocampus - CA4-DGC		47	608128.0	12920	39921.87665797351	0.8974663323548234	0.8012118782910058	0.8178549518262455
agg_1863	Eccentric medium spiny neuron	Hippocampus_Hippocampus	CNS	healthy	jhpce#tran2021	brain_atlas	Hippocampus	Hippocampus		31	1109961.0	14099	40557.14662520501	0.9273899109141328	0.8369796109567048	0.8600210861459844
agg_1864	Eccentric medium spiny neuron	Hypothalamus_preoptic region of HTH - HTHpo	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo		2330	51640744.0	17503	42679.76746648039	0.9409601827280252	0.840348392620232	0.8604579519561675
agg_1865	Eccentric medium spiny neuron	Hypothalamus_preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)		565	8224045.0	16394	43923.72334753606	0.9429273493241412	0.8372983150424523	0.8571441075789635
agg_1866	Eccentric medium spiny neuron	Hypothalamus_supraoptic region of HTH - HTHso	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	supraoptic region of HTH - HTHso		434	13111606.0	16456	40425.4238586538	0.9385925126325897	0.8460278149754488	0.8663811950707332
agg_1867	Eccentric medium spiny neuron	Midbrain_Unclassified	CNS	healthy	GSE157783	brain_atlas	Midbrain	Unclassified		14	114917.0	9818	36181.714010134245	0.8416212378179837	0.7532149744136349	0.7768524692302888
agg_1868	Eccentric medium spiny neuron	Thalamus_Anterior nuclear complex - ANC	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Anterior nuclear complex - ANC		746	7645113.0	16381	42815.79648292055	0.948484628173975	0.8495568731460875	0.8688807125927387
agg_1869	Eccentric medium spiny neuron	Thalamus_Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA		541	15930108.0	16787	41523.0893320909	0.9357186759722748	0.83560064088399	0.8572971207166471
agg_1870	Eccentric medium spiny neuron	Thalamus_Ventral group of lateral nucleus (VLN) - ventral anterior nucleus of thalamus - VA - ventral lateral nucleus of thalamus - VL	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Ventral group of lateral nucleus (VLN) - ventral anterior nucleus of thalamus - VA - ventral lateral nucleus of thalamus - VL		20	117737.0	10139	36547.396471567605	0.8471409795821925	0.7721611009725091	0.7850039546751
agg_2491	MGE interneuron	Amygdala_Amygdala	CNS	healthy	jhpce#tran2021	brain_atlas	Amygdala	Amygdala		409	13777101.0	16660	41359.14658121449	0.9444392422568889	0.8517551977241854	0.8676209228410131
agg_2492	MGE interneuron	Amygdala_Basolateral nuclear group (BLN) - lateral nucleus - La	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Basolateral nuclear group (BLN) - lateral nucleus - La		1061	21423455.0	17164	42654.248115680406	0.9542965361867625	0.8613064931579452	0.868815490419042
agg_2493	MGE interneuron	Amygdala_Bed nucleus of stria terminalis and nearby - BNST	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Bed nucleus of stria terminalis and nearby - BNST		166	1827695.0	15158	43981.651675853434	0.9448787686524807	0.8490275914218958	0.8533418570610546
agg_2494	MGE interneuron	Amygdala_Central nuclear group - CEN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Central nuclear group - CEN		1760	31053477.0	17481	43462.3208404647	0.9543411948300998	0.8600122768296773	0.8660896683249071
agg_2495	MGE interneuron	Amygdala_Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA		609	21863245.0	16897	41904.97759631928	0.9501737063270455	0.8554128895476135	0.8624338319882185
agg_2496	MGE interneuron	Amygdala_basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL		1057	19917276.0	17161	42604.6862273155	0.953798483025166	0.860651590728001	0.8698212056355575
agg_2497	MGE interneuron	Amygdala_basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM		1784	33119571.0	17549	43902.96306371541	0.9523415202656087	0.8535747260467098	0.8655401478053812
agg_2498	MGE interneuron	Amygdala_corticomedial nuclear group - CMN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	corticomedial nuclear group - CMN		1555	28602412.0	17506	43937.34054944343	0.9549364812212998	0.8576155015973854	0.8657168111994817
agg_2499	MGE interneuron	Basal ganglia_Globus pallidus (GP) - External segment of globus pallidus - GPe	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Globus pallidus (GP) - External segment of globus pallidus - GPe		14	363931.0	12637	41928.980123337795	0.9143027592379037	0.8152885964659543	0.8212241300910318
agg_2500	MGE interneuron	Basal ganglia_Nucleus Accumbens - NAC	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Nucleus Accumbens - NAC		20	385277.0	12598	41170.36162415581	0.9153758703420793	0.8200220156948643	0.8221207128459207
agg_2501	MGE interneuron	Basal ganglia_Nucleus accumbens	CNS	healthy	jhpce#tran2021	brain_atlas	Basal ganglia	Nucleus accumbens		266	5258847.0	16159	43224.232978201646	0.9409286116998999	0.8403845442432132	0.8558643354101686
agg_2502	MGE interneuron	Basal ganglia_Putamen - Pu	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Putamen - Pu		191	2779299.0	15551	43920.63540976047	0.9463633844002965	0.8496883400190438	0.8552019086045802
agg_2503	MGE interneuron	Basal ganglia_septal nuclei - SEP	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	septal nuclei - SEP		1429	24924668.0	17445	45173.37928793731	0.9503952330645382	0.8491060071505037	0.8559314584907406
agg_2504	MGE interneuron	Basal ganglia_substantia innominata and nearby nuclei - SI	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	substantia innominata and nearby nuclei - SI		667	7412740.0	16709	44750.547047095344	0.948134097855488	0.8506282530575365	0.8574254005579075
agg_2505	MGE interneuron	Cerebral cortex_Anterior Olfactory Nucleus - AON	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior Olfactory Nucleus - AON		3318	59973053.0	17718	44512.25053592313	0.9511866844152911	0.849120218442955	0.8591664689334718
agg_2507	MGE interneuron	Cerebral cortex_Anterior cingulate cortex	CNS	healthy	PRJNA434002	brain_atlas	Cerebral cortex	Anterior cingulate cortex		1995	11117873.0	16081	45179.21648388019	0.8965089340148632	0.804985295590617	0.7957027783688826
agg_2508	MGE interneuron	Cerebral cortex_Anterior cingulate cortex - ACC	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior cingulate cortex - ACC		5271	97368705.0	17869	44188.58976478878	0.9524429055696558	0.8495093943622023	0.8596610836215051
agg_2509	MGE interneuron	Cerebral cortex_Anterior parahippocampal gyrus (AG) - Lateral entorhinal cortex - LEC	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior parahippocampal gyrus (AG) - Lateral entorhinal cortex - LEC		3996	84774183.0	17847	44195.27877407958	0.9517972268850551	0.8516456686194328	0.85921931245538
agg_2510	MGE interneuron	Cerebral cortex_Anterior parahippocampal gyrus/posterior part (APH) - Medial entorhinal cortex - MEC	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior parahippocampal gyrus/posterior part (APH) - Medial entorhinal cortex - MEC		6404	140078426.0	17986	43152.00095627089	0.952323864729667	0.8540251497533076	0.8637559144241678
agg_2511	MGE interneuron	Cerebral cortex_Caudal cingulate gyrus (CgGC) - A23	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Caudal cingulate gyrus (CgGC) - A23		5856	63802214.0	17702	43234.31566597916	0.9507374146769353	0.8492333031420335	0.8580253936548816
agg_2512	MGE interneuron	Cerebral cortex_Cingulate gyrus/retrosplenial (CgGrs) - A29-A30	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Cingulate gyrus/retrosplenial (CgGrs) - A29-A30		7676	52516030.0	17650	44293.80571524706	0.9452372411663391	0.8405991797504577	0.8473866556985994
agg_2513	MGE interneuron	Cerebral cortex_Cuneus/caudal part - Peristriate Cortex - V2	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Cuneus/caudal part - Peristriate Cortex - V2		4257	65645451.0	17659	43356.56955307287	0.9483045548496131	0.843814478212617	0.8552982910193211
agg_2514	MGE interneuron	Cerebral cortex_Cuneus/rostral part - Area Prostriata - Pro	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Cuneus/rostral part - Area Prostriata - Pro		3639	53268136.0	17665	44440.51460839505	0.950149158075856	0.8446572675414704	0.8568222450605039
agg_2515	MGE interneuron	Cerebral cortex_Dorsalateral prefrontal cortex	CNS	healthy	http://psychencode.org	brain_atlas	Cerebral cortex	Dorsalateral prefrontal cortex		1922	7247656.0	15683	47343.60549703765	0.9155669978972271	0.7944935628936163	0.8164209537442745
agg_2517	MGE interneuron	Cerebral cortex_Dorsolateral prefrontal cortex	CNS	healthy	GSE129308	brain_atlas	Cerebral cortex	Dorsolateral prefrontal cortex		5070	11510390.0	15658	44053.416859390476	0.9083339555317534	0.807084771763412	0.7952863358833439
agg_2518	MGE interneuron	Cerebral cortex_Dorsolateral prefrontal cortex	CNS	healthy	SCR_002001	brain_atlas	Cerebral cortex	Dorsolateral prefrontal cortex		11974	89196550.0	14877	40375.335298792466	0.8811452953190944	0.8184790850728295	0.8016645574481662
agg_2519	MGE interneuron	Cerebral cortex_Dorsolateral prefrontal cortex	CNS	healthy	jhpce#tran2021	brain_atlas	Cerebral cortex	Dorsolateral prefrontal cortex		645	15038497.0	17036	42352.93122873921	0.9442194590303734	0.849789285525842	0.8665266730067818
agg_2522	MGE interneuron	Cerebral cortex_Entorhinal cortex	CNS	healthy	GSE160936	brain_atlas	Cerebral cortex	Entorhinal cortex		302	581183.0	12130	36486.8464868359	0.8895516477549692	0.8149478849985895	0.8337704432932246
agg_2524	MGE interneuron	Cerebral cortex_Frontal Cortex	CNS	healthy	GSE163122	brain_atlas	Cerebral cortex	Frontal Cortex		16	70172.0	8799	33974.63233778857	0.8118763535481881	0.7528303424561195	0.7481303580267099
agg_2525	MGE interneuron	Cerebral cortex_Frontal agranular insular cortex - FI	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Frontal agranular insular cortex - FI		5772	124639119.0	17927	43459.7015170545	0.9524032671776039	0.8534952362370954	0.8636876625391701
agg_2526	MGE interneuron	Cerebral cortex_Gyrus rectus (ReG) - Medial orbitofrontal cortex - A14	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Gyrus rectus (ReG) - Medial orbitofrontal cortex - A14		2060	50795925.0	17579	43698.71965590766	0.9513325519725236	0.8504598599771258	0.8580272600405687
agg_2527	MGE interneuron	Cerebral cortex_Inferior frontal gyrus (IFG) - Ventrolateral prefrontal cortex - A44-A45	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Inferior frontal gyrus (IFG) - Ventrolateral prefrontal cortex - A44-A45		6389	105812032.0	17893	44449.831210324366	0.9508588269489653	0.8461769145046513	0.8547104807046153
agg_2528	MGE interneuron	Cerebral cortex_Inferior temporal gyrus - ITG	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Inferior temporal gyrus - ITG		5185	75602611.0	17783	43909.19926818126	0.950810125857405	0.8496421991852124	0.8581300377260784
agg_2529	MGE interneuron	Cerebral cortex_Lingual gyrus (LiG) - Primary Visual Cortex - V1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Lingual gyrus (LiG) - Primary Visual Cortex - V1C		3677	55232602.0	17614	43708.35031940854	0.9456104506863592	0.8400672253493853	0.8503765007105082
agg_2530	MGE interneuron	Cerebral cortex_Long insular gyri (LIG) - Dysgranular insular cortex - Idg	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Long insular gyri (LIG) - Dysgranular insular cortex - Idg		5713	102196545.0	17903	44538.74990246346	0.9521999741486603	0.8494205945292841	0.8579708150521901
agg_2531	MGE interneuron	Cerebral cortex_Middle Temporal Gyrus - MTG	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Middle Temporal Gyrus - MTG		12656	191746732.0	18000	44044.23377219539	0.9502693570547087	0.8466137046651626	0.8555071247129131
agg_2532	MGE interneuron	Cerebral cortex_Middle frontal gyrus (MFG) - A46	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Middle frontal gyrus (MFG) - A46		5268	68233967.0	17711	43971.19142937117	0.9492405306461207	0.8451499686861522	0.8537464653796999
agg_2534	MGE interneuron	Cerebral cortex_Occipitotemporal (fusiform) gyrus/temporal part (FuGt) - Temporal area TF	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Occipitotemporal (fusiform) gyrus/temporal part (FuGt) - Temporal area TF		5450	64323983.0	17726	43720.55794484247	0.9515917735720116	0.8483050792693729	0.8589048783884909
agg_2535	MGE interneuron	Cerebral cortex_Parietal lobe	CNS	healthy	PRJNA544731	brain_atlas	Cerebral cortex	Parietal lobe		133	569401.0	12468	37595.49960109785	0.8907326466742803	0.8061940525939209	0.8092373729309823
agg_2537	MGE interneuron	Cerebral cortex_Parietal operculum (PaO) - Gustatory cortex - A43	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Parietal operculum (PaO) - Gustatory cortex - A43		7115	120047210.0	17919	44437.09301091113	0.9512477871447569	0.8451452826679898	0.8548666563138005
agg_2538	MGE interneuron	Cerebral cortex_Perirhinal gyrus (PRG) - A35-A36	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Perirhinal gyrus (PRG) - A35-A36		1932	37490198.0	17465	43259.31944751696	0.9541909686514075	0.8540303396121026	0.8654366180531673
agg_2539	MGE interneuron	Cerebral cortex_Perirhinal gyrus (PRG) -  A35-A36	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Perirhinal gyrus (PRG) -  A35-A36		117	3034731.0	15605	43526.22006701818	0.9473644788454095	0.8522954168059347	0.8593673572726558
agg_2540	MGE interneuron	Cerebral cortex_Piriform cortex - Pir	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Piriform cortex - Pir		3217	57727120.0	17686	43762.436474795206	0.9543255676211085	0.8545999342708804	0.863564845394952
agg_2541	MGE interneuron	Cerebral cortex_Post-mortem dosolateral BA9	CNS	healthy	GSE144136	brain_atlas	Cerebral cortex	Post-mortem dosolateral BA9		3575	10164175.0	16325	46806.33140707459	0.9069971617382123	0.7983175743353382	0.794815684585161
agg_2543	MGE interneuron	Cerebral cortex_Postcentral gyrus (PoCG) - Primary somatosensory cortex - S1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Postcentral gyrus (PoCG) - Primary somatosensory cortex - S1C		6470	100713734.0	17875	44372.39965330401	0.9497085444141861	0.8450959161936977	0.8535719902493208
agg_2544	MGE interneuron	Cerebral cortex_Posterior intermediate orbital gyrus (POrG) - Caudal division of OFCi - A13	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Posterior intermediate orbital gyrus (POrG) - Caudal division of OFCi - A13		5366	77974838.0	17815	44405.45901697979	0.9502945552662591	0.8463296941825013	0.8550375784224905
agg_2545	MGE interneuron	Cerebral cortex_Posterior parahippocampal gyrus (PPH) - TH-TL	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Posterior parahippocampal gyrus (PPH) - TH-TL		5263	87475500.0	17852	44327.56466231888	0.9519923289438857	0.8498492615752189	0.859266978972963
agg_2546	MGE interneuron	Cerebral cortex_Precentral gyrus (PrCG) - Primary motor cortex - M1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Precentral gyrus (PrCG) - Primary motor cortex - M1C		17544	263351574.0	18021	42751.448907934544	0.948569107423063	0.8463142396680147	0.8559714261188668
agg_2552	MGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	GSE157827	brain_atlas	Cerebral cortex	Prefrontal cortex		4610	21482625.0	17270	43731.54224890505	0.9363855712351603	0.8318806432247126	0.8412160935393553
agg_2553	MGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	GSE167494	brain_atlas	Cerebral cortex	Prefrontal cortex		1247	10376072.0	17204	44361.72553538063	0.9442885914414952	0.8517201236785358	0.8648145466877852
agg_2554	MGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	PRJNA434002	brain_atlas	Cerebral cortex	Prefrontal cortex		2511	14863691.0	16162	44673.20279790643	0.8989195129049853	0.8071982631305372	0.8006890898756347
agg_2555	MGE interneuron	Cerebral cortex_Prefrontal cortex	CNS	healthy	PRJNA544731	brain_atlas	Cerebral cortex	Prefrontal cortex		852	4039365.0	15379	43450.098832422605	0.9008645464806034	0.8076373172845014	0.8072548879496675
agg_2557	MGE interneuron	Cerebral cortex_Premotor cortex	CNS	healthy	PRJNA544731	brain_atlas	Cerebral cortex	Premotor cortex		311	1072054.0	13778	39874.7813160123	0.9016052282501837	0.8194735601775744	0.8136381571982871
agg_2559	MGE interneuron	Cerebral cortex_Rostral gyrus (RoG) - Dorsal division of MFC - A32	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Rostral gyrus (RoG) - Dorsal division of MFC - A32		4570	58857880.0	17731	44069.06350483119	0.9511720782778857	0.8463203838471274	0.8560199802608556
agg_2560	MGE interneuron	Cerebral cortex_Short insular gyri - Granular insular cortex - Ig	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Short insular gyri - Granular insular cortex - Ig		5331	82650726.0	17824	44394.120871907326	0.9509583744999757	0.8462445133654821	0.8571241887907932
agg_2562	MGE interneuron	Cerebral cortex_Somatosensory cortex	CNS	healthy	GSE160936	brain_atlas	Cerebral cortex	Somatosensory cortex		138	253647.0	11822	38266.141208792615	0.8767841346625923	0.7936019959715973	0.8134006209715183
agg_2563	MGE interneuron	Cerebral cortex_Subcallosal Gyrus (SCG) - Subgenual (subcallosal) division of MFC - A25	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Subcallosal Gyrus (SCG) - Subgenual (subcallosal) division of MFC - A25		6738	111747047.0	17933	44140.288432463174	0.9526334893013195	0.8515236536431087	0.8602446595188639
agg_2564	MGE interneuron	Cerebral cortex_Subgenual anterior cingulate cortex	CNS	healthy	jhpce#tran2021	brain_atlas	Cerebral cortex	Subgenual anterior cingulate cortex		1435	33906681.0	17248	41894.99018499041	0.9415120732425849	0.8510166771030189	0.8651258505145446
agg_2565	MGE interneuron	Cerebral cortex_Superior Temporal Gyrus - STG	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Superior Temporal Gyrus - STG		4693	53745497.0	17740	45035.09601291676	0.9482478863607943	0.8418675581822387	0.8511996261886916
agg_2567	MGE interneuron	Cerebral cortex_Superior occipital gyrus (SOG) - Areas 19 and MT - A19	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Superior occipital gyrus (SOG) - Areas 19 and MT - A19		4640	67271109.0	17631	42831.45523317643	0.9503792869302212	0.8486348220775979	0.859360564195124
agg_2568	MGE interneuron	Cerebral cortex_Supramarginal gyrus (SMG) - A40	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Supramarginal gyrus (SMG) - A40		7155	95138168.0	17848	43973.385161914906	0.9504336303557819	0.8461929748397125	0.8578464760178754
agg_2569	MGE interneuron	Cerebral cortex_Supraparietal lobule (SPL) - Posterosuperior (dorsal) parietal cortex - A5-A7	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Supraparietal lobule (SPL) - Posterosuperior (dorsal) parietal cortex - A5-A7		5630	80193333.0	17797	43742.70207026511	0.9512327178952547	0.8461941179433469	0.8579482598163595
agg_2572	MGE interneuron	Cerebral cortex_Temporal pole (TP) - Temporopolar area - A38	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Temporal pole (TP) - Temporopolar area - A38		5665	110350485.0	17940	44086.89577677079	0.953708858544873	0.8513824540838272	0.8602810204886354
agg_2573	MGE interneuron	Cerebral cortex_Transverse temporal gyrus (TTG) - Primary auditory cortex - A1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Transverse temporal gyrus (TTG) - Primary auditory cortex - A1C		4799	73523279.0	17770	43918.33248256951	0.9502853491865487	0.8458671001525015	0.858757045378173
agg_2574	MGE interneuron	Cerebral cortex_Unclassified	CNS	healthy	GSE140231	brain_atlas	Cerebral cortex	Unclassified		1219	5729012.0	15984	45916.35535647439	0.8957775348525301	0.797054655931438	0.7889163143857955
agg_2575	MGE interneuron	Cerebral cortex_nan	CNS	healthy	GSE160936	brain_atlas	Cerebral cortex	nan		39	82069.0	9206	34840.24746186593	0.8190571869697097	0.7306144898877015	0.7292907453780283
agg_2577	MGE interneuron	Cerebral cortex_occipital cortex	CNS	healthy	GSE148822	brain_atlas	Cerebral cortex	occipital cortex		117	246364.0	12197	40372.85168056861	0.88745766532023	0.7948783763579894	0.821386649779637
agg_2579	MGE interneuron	Cerebral cortex_occipitotemporal cortex	CNS	healthy	GSE148822	brain_atlas	Cerebral cortex	occipitotemporal cortex		70	189834.0	11628	39769.16515969363	0.8818384749014652	0.8031748269991748	0.814351561311026
agg_2580	MGE interneuron	Epithalamus_ETH	CNS	healthy	SCR_016152	brain_atlas	Epithalamus	ETH		12	71885.0	9527	36809.85853759243	0.824422869787946	0.7440895655139182	0.7553679675372315
agg_2581	MGE interneuron	Grey matter_Cla	CNS	healthy	SCR_016152	brain_atlas	Grey matter	Cla		3513	60764194.0	17729	44332.18321962197	0.9527358283259111	0.8525521251468033	0.859701251466556
agg_2584	MGE interneuron	Grey matter_Motor cortex	CNS	healthy	GSE174332	brain_atlas	Grey matter	Motor cortex		3176	40684160.0	17707	43957.55438189328	0.9446516045218407	0.8563721111404717	0.8618527207167933
agg_2590	MGE interneuron	Hippocampus_CA1	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA1		1542	50256065.0	17586	43026.10606620708	0.9497570037975281	0.8548983714813951	0.862305286905649
agg_2591	MGE interneuron	Hippocampus_CA1-3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA1-3		955	32592685.0	17658	44364.0592027126	0.952760200558601	0.858208550001937	0.8654308552711507
agg_2592	MGE interneuron	Hippocampus_CA2-3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA2-3		405	12427466.0	16877	43427.36652371909	0.9479364599220131	0.8507258501619306	0.8597478080960826
agg_2593	MGE interneuron	Hippocampus_Caudal Hippocampus - CA1-CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Caudal Hippocampus - CA1-CA3		2761	52247733.0	17723	44356.44255419876	0.9482748850948762	0.8491246279731572	0.8545886489562906
agg_2594	MGE interneuron	Hippocampus_Caudal Hippocampus - CA4-DGC	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Caudal Hippocampus - CA4-DGC		1176	20323114.0	17216	43747.53358399856	0.9473473754132059	0.8467759551726008	0.8565315424490701
agg_2595	MGE interneuron	Hippocampus_DG-CA4	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	DG-CA4		79	2492901.0	15246	41900.67343524716	0.9395119838587522	0.8493514298178396	0.8589159044795706
agg_2596	MGE interneuron	Hippocampus_Hippocampus	CNS	healthy	jhpce#tran2021	brain_atlas	Hippocampus	Hippocampus		59	2189102.0	15324	42472.04317459105	0.9370775068073622	0.8447930718708813	0.8564417124517391
agg_2602	MGE interneuron	Hippocampus_Rostral CA1-2	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral CA1-2		955	41524730.0	17477	43171.938875131724	0.9497339744218967	0.8531800823276372	0.8629519962286984
agg_2603	MGE interneuron	Hippocampus_Rostral CA1-CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral CA1-CA3		2237	48236091.0	17688	44226.34157174449	0.9488678749901845	0.8499180831182793	0.8572963944833775
agg_2604	MGE interneuron	Hippocampus_Rostral CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral CA3		447	16772147.0	17103	42318.88302627937	0.9465119706715757	0.8572332937695742	0.8677509453487087
agg_2605	MGE interneuron	Hippocampus_Rostral DG-CA4	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral DG-CA4		277	7865435.0	16546	42430.00131802814	0.9441027592674103	0.8494269573509831	0.8637403065288809
agg_2606	MGE interneuron	Hippocampus_Subicular cortex - Sub	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Subicular cortex - Sub		4153	46428577.0	17703	44106.187747494434	0.9523849166612751	0.8547240397425142	0.8596520862346101
agg_2607	MGE interneuron	Hippocampus_Uncal CA1-CA3	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Uncal CA1-CA3		2465	21980770.0	17489	44863.81070779017	0.9447083817652235	0.8462578379895598	0.8490324522059823
agg_2608	MGE interneuron	Hippocampus_Uncal DG-CA4	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Uncal DG-CA4		335	3643860.0	16148	44055.05574062089	0.9406193799345944	0.8432095721836758	0.8484634729799445
agg_2609	MGE interneuron	Hypothalamus_preoptic region of HTH - HTHpo	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo		22	453470.0	12617	40599.60295443444	0.9219576086495076	0.8291191541534322	0.8337400408636516
agg_2610	MGE interneuron	Hypothalamus_preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)		16	201362.0	11334	39197.85979581934	0.8912937813689751	0.8032225542810923	0.8073296298291398
agg_2611	MGE interneuron	Hypothalamus_supraoptic region of HTH - HTHso	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	supraoptic region of HTH - HTHso		1291	34587509.0	17244	41775.70339134899	0.947695036391916	0.8568901031945508	0.8623938798420291
agg_2612	MGE interneuron	Hypothalamus_supraoptic region of HTH - HTHso (anterior hypothalamic nucleus/AHN) - tuberal region of HTH - HTHtub (ventromedial hypothalamic nucleus/VMH)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	supraoptic region of HTH - HTHso (anterior hypothalamic nucleus/AHN) - tuberal region of HTH - HTHtub (ventromedial hypothalamic nucleus/VMH)		13	201539.0	11989	42042.031681340275	0.8926159239254491	0.7900437896326049	0.7976018773826827
agg_2613	MGE interneuron	Hypothalamus_tuberal region of hypothalamus - HTHtub	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	tuberal region of hypothalamus - HTHtub		42	1250352.0	13813	39859.463099401524	0.9325616164071415	0.8405974496687145	0.8524085965817827
agg_2614	MGE interneuron	Midbrain_Periaqueductal gray and Dorsal raphe nucleus - PAG-DR	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Periaqueductal gray and Dorsal raphe nucleus - PAG-DR		48	279415.0	12088	40422.48239700562	0.9015998774703433	0.8083821951494623	0.8145886349206347
agg_2615	MGE interneuron	Midbrain_Substantia Nigra - SN	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Substantia Nigra - SN		24	645604.0	13566	41518.8703956369	0.9212646512029814	0.8310832019975152	0.8374969841343043
agg_2616	MGE interneuron	Midbrain_Superior colliculus and nearby nuclei - SC	CNS	healthy	SCR_016152	brain_atlas	Midbrain	Superior colliculus and nearby nuclei - SC		79	869016.0	13900	41943.6713340587	0.9335775223214702	0.836230212994662	0.8410600633997231
agg_2617	MGE interneuron	Thalamus_Anterior nuclear complex - ANC	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Anterior nuclear complex - ANC		63	1689760.0	14392	40935.765146316844	0.9385291782898756	0.8460285020935135	0.854094876359309
agg_2618	MGE interneuron	Thalamus_Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA		23	483935.0	12795	41330.96925666164	0.9202418171464944	0.8261269502174232	0.8253395913468182
agg_2619	MGE interneuron	Thalamus_Lateral nuclear complex of thalamus (LNC) - ventral group of lateral nucleus - VLN	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Lateral nuclear complex of thalamus (LNC) - ventral group of lateral nucleus - VLN		15	644299.0	12682	38988.345127600296	0.9101440823448651	0.820951753505128	0.8290122191651943
agg_2620	MGE interneuron	Thalamus_lateral nuclear complex of thalamus (LNC) - Pulvinar of thalamus - Pul	CNS	healthy	SCR_016152	brain_atlas	Thalamus	lateral nuclear complex of thalamus (LNC) - Pulvinar of thalamus - Pul		59	907181.0	14156	42737.54158697253	0.9230276788817333	0.8260353938939361	0.8272422303517075
agg_2621	MGE interneuron	Thalamus_lateral nuclear complex of thalamus (LNC) - lateral posterior nucleus of thalamus - LP	CNS	healthy	SCR_016152	brain_atlas	Thalamus	lateral nuclear complex of thalamus (LNC) - lateral posterior nucleus of thalamus - LP		34	339865.0	12131	39372.36226065725	0.9034367683239668	0.8112611977762655	0.8142495801086926
agg_2622	MGE interneuron	Thalamus_medial nuclear complex of thalamus (MNC) - mediodorsal nucleus of thalamus - MD	CNS	healthy	SCR_016152	brain_atlas	Thalamus	medial nuclear complex of thalamus (MNC) - mediodorsal nucleus of thalamus - MD		20	414741.0	12853	41547.316148435246	0.9089247248871526	0.813595574745492	0.8152544778775622
agg_2623	MGE interneuron	Thalamus_posterior nuclear complex of thalamus (PoN) - lateral geniculate nucleus (LG)	CNS	healthy	SCR_016152	brain_atlas	Thalamus	posterior nuclear complex of thalamus (PoN) - lateral geniculate nucleus (LG)		57	768227.0	13434	40339.8513441331	0.9219632514083475	0.8263116083745812	0.8349479724650578
agg_2624	MGE interneuron	White matter_Unclassifed	CNS	healthy	GSE118257	brain_atlas	White matter	Unclassifed		94	289790.0	11627	41098.134330016575	0.8867654909586433	0.8059356433623136	0.7851502438087337
agg_2653	Medium spiny neuron	Amygdala_Amygdala	CNS	healthy	jhpce#tran2021	brain_atlas	Amygdala	Amygdala		855	16257020.0	16794	40209.40354842709	0.9307548981959526	0.8488615811730392	0.8689605369066653
agg_2654	Medium spiny neuron	Amygdala_Basolateral nuclear group (BLN) - lateral nucleus - La	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Basolateral nuclear group (BLN) - lateral nucleus - La		330	4992951.0	15782	40919.21981550283	0.9427614178501177	0.8466272519178762	0.8656673560635975
agg_2655	Medium spiny neuron	Amygdala_Bed nucleus of stria terminalis and nearby - BNST	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Bed nucleus of stria terminalis and nearby - BNST		14834	231645818.0	18115	44026.663188107996	0.9453419860163621	0.8461001317058413	0.8674271490373949
agg_2656	Medium spiny neuron	Amygdala_Central nuclear group - CEN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Central nuclear group - CEN		4702	152024109.0	17933	41825.65400418454	0.9506498903765198	0.8583036078560143	0.87445330234001
agg_2657	Medium spiny neuron	Amygdala_Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA	CNS	healthy	SCR_016152	brain_atlas	Amygdala	Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA		76	1988757.0	14473	41009.37497395409	0.9374639079699163	0.8459652724097986	0.8661217643696482
agg_2658	Medium spiny neuron	Amygdala_basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basolateral nucleus (basal nucleus) - BL		41	681582.0	13700	42390.62517590265	0.930301178896936	0.8335883138129021	0.8498381291011784
agg_2659	Medium spiny neuron	Amygdala_basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM	CNS	healthy	SCR_016152	brain_atlas	Amygdala	basolateral nuclear group (BLN) - basomedial nucleus (accessory basal nucleus) - BM		588	10857517.0	16752	43407.2880995957	0.9413954461547651	0.8491087638861763	0.860187656740274
agg_2660	Medium spiny neuron	Amygdala_corticomedial nuclear group - CMN	CNS	healthy	SCR_016152	brain_atlas	Amygdala	corticomedial nuclear group - CMN		1563	18589476.0	17293	43767.421985488276	0.9475050406764797	0.8500945691307341	0.868188681008702
agg_2661	Medium spiny neuron	Basal ganglia_Body of the Caudate - CaB	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Body of the Caudate - CaB		25597	734900951.0	18169	41031.74091339427	0.9130017546169741	0.8094631607687462	0.8330040905587449
agg_2662	Medium spiny neuron	Basal ganglia_Globus pallidus (GP) - External segment of globus pallidus - GPe	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Globus pallidus (GP) - External segment of globus pallidus - GPe		11913	421587642.0	18127	41409.19215617373	0.9245205942991978	0.820006356524453	0.8458336457726571
agg_2663	Medium spiny neuron	Basal ganglia_Globus pallidus (GP) - Internal segment of globus pallidus - GPi	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Globus pallidus (GP) - Internal segment of globus pallidus - GPi		91	3689378.0	15264	40334.966924483146	0.9258738064983694	0.8220183202800466	0.8471167198181737
agg_2664	Medium spiny neuron	Basal ganglia_Nucleus Accumbens - NAC	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Nucleus Accumbens - NAC		22095	556326311.0	18164	42779.16646063976	0.9241007801139076	0.8160594037096315	0.8419305863541408
agg_2665	Medium spiny neuron	Basal ganglia_Nucleus accumbens	CNS	healthy	jhpce#tran2021	brain_atlas	Basal ganglia	Nucleus accumbens		9837	271665829.0	17889	41582.21333344588	0.919599303144343	0.8154028670692957	0.8472977098332947
agg_2666	Medium spiny neuron	Basal ganglia_Putamen - Pu	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	Putamen - Pu		26169	694166058.0	18170	40849.22367869802	0.9127313224001196	0.8084587312170607	0.8346523973425407
agg_2667	Medium spiny neuron	Basal ganglia_septal nuclei - SEP	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	septal nuclei - SEP		3686	70583279.0	17863	42869.752774801855	0.9221592113278905	0.8179030901469408	0.8415064589677047
agg_2668	Medium spiny neuron	Basal ganglia_substantia innominata and nearby nuclei - SI	CNS	healthy	SCR_016152	brain_atlas	Basal ganglia	substantia innominata and nearby nuclei - SI		20978	406430174.0	18152	43171.70819204389	0.9356278573993547	0.8355914647959416	0.8580501444882963
agg_2669	Medium spiny neuron	Cerebral cortex_Anterior Olfactory Nucleus - AON	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Anterior Olfactory Nucleus - AON		7146	83678191.0	17914	43899.304433061516	0.9197892629833866	0.8118279846788765	0.838725791865812
agg_2670	Medium spiny neuron	Cerebral cortex_Frontal agranular insular cortex - FI	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Frontal agranular insular cortex - FI		47	1673650.0	14282	40604.468196450885	0.9310316824387254	0.8387474025258079	0.8630838302859096
agg_2671	Medium spiny neuron	Cerebral cortex_Perirhinal gyrus (PRG) - A35-A36	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Perirhinal gyrus (PRG) - A35-A36		60	1311204.0	14729	43400.63525982828	0.9426515623287082	0.8452776244969754	0.8631129728359953
agg_2672	Medium spiny neuron	Cerebral cortex_Piriform cortex - Pir	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Piriform cortex - Pir		528	13895788.0	16601	41718.898509735314	0.9449645834556754	0.8544810746353554	0.8773903973083115
agg_2673	Medium spiny neuron	Cerebral cortex_Precentral gyrus (PrCG) - Primary motor cortex - M1C	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Precentral gyrus (PrCG) - Primary motor cortex - M1C		13	214428.0	11454	38695.26349160653	0.8924953155417973	0.7998199730247992	0.8148967949524837
agg_2674	Medium spiny neuron	Cerebral cortex_Subcallosal Gyrus (SCG) - Subgenual (subcallosal) division of MFC - A25	CNS	healthy	SCR_016152	brain_atlas	Cerebral cortex	Subcallosal Gyrus (SCG) - Subgenual (subcallosal) division of MFC - A25		30	418073.0	13063	41585.729968004634	0.8871186460605669	0.7821038738334904	0.816336464253731
agg_2675	Medium spiny neuron	Cerebral cortex_Subgenual anterior cingulate cortex	CNS	healthy	jhpce#tran2021	brain_atlas	Cerebral cortex	Subgenual anterior cingulate cortex		456	24338636.0	17126	41522.34752505473	0.9333821901136405	0.8462980778537625	0.8674585783450169
agg_2677	Medium spiny neuron	Grey matter_Cla	CNS	healthy	SCR_016152	brain_atlas	Grey matter	Cla		441	9430332.0	16466	41163.602945269886	0.921979391685106	0.8201910372875032	0.842843688151027
agg_2678	Medium spiny neuron	Hippocampus_CA1	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	CA1		40	1049125.0	13959	42169.675836731396	0.9349494469318459	0.8391200271799821	0.8542758399265774
agg_2679	Medium spiny neuron	Hippocampus_Caudal Hippocampus - CA4-DGC	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Caudal Hippocampus - CA4-DGC		430	6612281.0	15981	41097.34699445523	0.9098959524061087	0.8090072110159198	0.8280726832719776
agg_2680	Medium spiny neuron	Hippocampus_Rostral DG-CA4	CNS	healthy	SCR_016152	brain_atlas	Hippocampus	Rostral DG-CA4		17	222139.0	11847	39862.89795022213	0.8691071021059862	0.7714996002895943	0.7962472128461958
agg_2681	Medium spiny neuron	Hypothalamus_preoptic region of HTH - HTHpo	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo		5346	128564800.0	17878	43046.84627750286	0.9441056048503704	0.8421353341456254	0.866689965845226
agg_2682	Medium spiny neuron	Hypothalamus_preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	preoptic region of HTH - HTHpo (medial preoptic nucleus/MPN) - supraoptic region of HTH - HTHso (paraventricular nucleus/PV)		1202	21586591.0	17227	44983.99150274918	0.9507961158008243	0.8480169975890431	0.8669422509999024
agg_2683	Medium spiny neuron	Hypothalamus_supraoptic region of HTH - HTHso	CNS	healthy	SCR_016152	brain_atlas	Hypothalamus	supraoptic region of HTH - HTHso		435	13475264.0	16512	40730.18309280418	0.9328893799640929	0.8411270417125518	0.8653686705020206
agg_2684	Medium spiny neuron	Midbrain_Unclassified	CNS	healthy	GSE157783	brain_atlas	Midbrain	Unclassified		13	138924.0	10081	36049.57615815546	0.8472115551978691	0.7544352004514613	0.7771766770288353
agg_2685	Medium spiny neuron	Thalamus_Anterior nuclear complex - ANC	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Anterior nuclear complex - ANC		926	12027800.0	16849	43612.739817779046	0.9535465002330261	0.861691575979908	0.8753609519126483
agg_2686	Medium spiny neuron	Thalamus_Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Lateral nuclear complex (LNC) - ventral anterior nucleus of thalamus - VA		2743	114224823.0	17803	40855.63847966631	0.9268757598914684	0.8230943566844251	0.848168195849154
agg_2687	Medium spiny neuron	Thalamus_Ventral group of lateral nucleus (VLN) - ventral anterior nucleus of thalamus - VA - ventral lateral nucleus of thalamus - VL	CNS	healthy	SCR_016152	brain_atlas	Thalamus	Ventral group of lateral nucleus (VLN) - ventral anterior nucleus of thalamus - VA - ventral lateral nucleus of thalamus - VL		14	108117.0	10209	37484.94947048898	0.8438314846351331	0.7444397677227123	0.766869932545645


</pre></div>
</div>
</div>
</div>
<p>The main arguments for the <code class="docutils literal notranslate"><span class="pre">decima</span> <span class="pre">modisco</span></code> CLI command allow you to customize motif discovery for your specific biological question. The <code class="docutils literal notranslate"><span class="pre">--top-n-markers</span></code> argument restricts the analysis to the top N marker genes, focusing on the most distinguishing genes for the selected cell types if not provided all genes are used. The <code class="docutils literal notranslate"><span class="pre">--max-seqlets</span></code> parameter sets an upper limit on the number of seqlets (short, high-scoring regions) extracted per metacluster, which can help manage memory usage and computation time. The <code class="docutils literal notranslate"><span class="pre">--tss-distance</span></code> option defines the window size (in base pairs) around the transcription start site to consider for motif discovery.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">--tasks</span></code> argument lets you specify cell types or conditions of interest using a query string, such as <code class="docutils literal notranslate"><span class="pre">&quot;cell_type.str.contains('neuron')</span> <span class="pre">and</span> <span class="pre">organ</span> <span class="pre">==</span> <span class="pre">'CNS'</span> <span class="pre">and</span> <span class="pre">disease</span> <span class="pre">==</span> <span class="pre">'healthy'&quot;</span></code> to select healthy CNS neurons. The <code class="docutils literal notranslate"><span class="pre">--transform</span></code> argument determines how attributions are aggregated or contrasted. For example, <code class="docutils literal notranslate"><span class="pre">&quot;specificity&quot;</span></code> compares the selected tasks (e.g., neurons) against the background (e.g., all other cells), highlighting features specific to the chosen group. Alternatively, <code class="docutils literal notranslate"><span class="pre">&quot;aggregate&quot;</span></code> simply sums attributions across the selected tasks.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code> and <code class="docutils literal notranslate"><span class="pre">--num-workers</span></code> options control parallelization and resource usage during computation. The <code class="docutils literal notranslate"><span class="pre">--model</span></code> argument selects which model replicate or checkpoint to use, and the <code class="docutils literal notranslate"><span class="pre">-o</span></code> or <code class="docutils literal notranslate"><span class="pre">--output-prefix</span></code> sets the path prefix for output files and reports.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>modisco<span class="w">  </span><span class="err">\</span>
    <span class="o">--</span><span class="n">top</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">markers</span> <span class="mi">50</span> \
    <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">seqlets</span> <span class="mi">5000</span> \
    <span class="o">--</span><span class="n">tss</span><span class="o">-</span><span class="n">distance</span> <span class="mi">5000</span>  \
    <span class="o">--</span><span class="n">tasks</span> <span class="s2">&quot;cell_type.str.contains(&#39;neuron&#39;) and organ == &#39;CNS&#39; and disease == &#39;healthy&#39;&quot;</span> \
    <span class="o">--</span><span class="n">transform</span> <span class="s2">&quot;specificity&quot;</span> \
    <span class="o">--</span><span class="n">batch</span><span class="o">-</span><span class="n">size</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">model</span> <span class="n">v1_rep0</span> \
    <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">workers</span> <span class="mi">8</span> \
    <span class="o">-</span><span class="n">o</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_neurons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Using device: cuda
decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1802.9MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1860.8MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (418.6MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
100%|████████████████████████████████████████████| 1/1 [00:00&lt;00:00, 295.08it/s]
Computing attributions...:   8%|█▎               | 4/50 [00:04&lt;00:46,  1.01s/it]decima - WARNING - Gene XRCC5 has low correlation with the model. Pearson: 0.4621304058925145. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  10%|█▋               | 5/50 [00:05&lt;00:43,  1.04it/s]decima - WARNING - Gene FNIP1 has low correlation with the model. Pearson: 0.044592478397006044. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  40%|██████▍         | 20/50 [00:18&lt;00:26,  1.14it/s]decima - WARNING - Gene LRRFIP1 has low correlation with the model. Pearson: 0.4621130932978845. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  70%|███████████▏    | 35/50 [00:31&lt;00:13,  1.14it/s]decima - WARNING - Gene CHD4 has low correlation with the model. Pearson: 0.27165706153002717. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...: 100%|████████████████| 50/50 [00:44&lt;00:00,  1.11it/s]
decima - INFO - Loading metadata
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1818.0MB/s)
100%|████████████████████████████████████████████| 1/1 [00:00&lt;00:00, 316.03it/s]
Loading attributions and sequences...: 100%|████| 50/50 [00:02&lt;00:00, 21.45it/s]
2025-10-06 10:26:59,580 - modisco-lite - INFO - Running TFMoDISco version 2.4.0
2025-10-06 10:26:59,583 - modisco-lite - INFO - Extracting seqlets for 50 tasks:
2025-10-06 10:26:59,583 - modisco-lite - INFO - - Smoothing and splitting tracks
2025-10-06 10:26:59,620 - modisco-lite - INFO - - Computing null values with Laplacian null model
2025-10-06 10:26:59,679 - modisco-lite - INFO - - Computing isotonic thresholds
2025-10-06 10:26:59,701 - modisco-lite - INFO - - Refining thresholds
2025-10-06 10:26:59,999 - modisco-lite - INFO - - Extracting seqlets
2025-10-06 10:27:00,048 - modisco-lite - INFO - - Extracting 1300 positive seqlets
2025-10-06 10:27:00,055 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 1300 seqlets
100%|██████████████████████████████████████| 1300/1300 [00:05&lt;00:00, 218.21it/s]
2025-10-06 10:27:13,633 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 1300 seqlets and 1300 neighbors
100%|██████████████████████████████████████| 1300/1300 [00:02&lt;00:00, 559.64it/s]
100%|██████████████████████████████████████| 1300/1300 [00:02&lt;00:00, 557.36it/s]
2025-10-06 10:27:18,390 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 10:27:19,094 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 10:27:21,693 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    9.0s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    9.5s remaining:  1.1min
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    9.5s remaining:   41.2s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    9.6s remaining:   28.7s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    9.6s remaining:   21.0s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    9.9s remaining:   16.5s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    9.9s remaining:   12.8s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:   10.0s remaining:   10.0s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:   11.7s remaining:    9.1s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:   12.0s remaining:    7.2s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:   12.2s remaining:    5.6s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:   12.3s remaining:    4.1s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:   13.0s remaining:    3.0s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:   13.3s remaining:    1.9s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:   14.3s finished
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.16468593856466973
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.16468593856466973
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.16377292460972134
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.16468828375575995
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.16468593856466973
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.1646968690389451
2025-10-06 10:27:36,075 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.16468828375575995
2025-10-06 10:27:36,076 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 3/3 [00:05&lt;00:00,  1.78s/it]
2025-10-06 10:27:41,414 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 1185 seqlets
100%|█████████████████████████████████████| 1185/1185 [00:00&lt;00:00, 1271.38it/s]
2025-10-06 10:27:42,698 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 1185 seqlets and 1185 neighbors
100%|██████████████████████████████████████| 1185/1185 [00:06&lt;00:00, 193.55it/s]
100%|██████████████████████████████████████| 1185/1185 [00:06&lt;00:00, 193.05it/s]
2025-10-06 10:27:55,064 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 10:27:57,534 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    2.1s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    2.5s remaining:   17.4s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    2.6s remaining:   11.2s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    2.7s remaining:    8.1s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    2.9s remaining:    6.3s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    3.0s remaining:    5.0s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    3.0s remaining:    3.9s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    3.1s remaining:    3.1s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    4.4s remaining:    3.4s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    5.1s remaining:    3.0s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    5.2s remaining:    2.4s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    5.4s remaining:    1.8s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    5.6s remaining:    1.3s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    5.6s remaining:    0.8s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    7.4s finished
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.1447236813116312
2025-10-06 10:28:05,005 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 3/3 [00:05&lt;00:00,  1.77s/it]
2025-10-06 10:28:10,304 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|████| 3/3 [00:16&lt;00:00,  5.66s/it]
2025-10-06 10:28:36,203 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|███████████████████| 26/26 [00:00&lt;00:00, 15400.64it/s]
Computing subpatterns:: 100%|███████████████████| 21/21 [00:00&lt;00:00, 23.69it/s]
2025-10-06 10:28:37,094 - modisco-lite - INFO - - Extracting 2179 negative seqlets
2025-10-06 10:28:37,107 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 2179 seqlets
100%|██████████████████████████████████████| 2179/2179 [00:02&lt;00:00, 759.06it/s]
2025-10-06 10:28:41,409 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 2179 seqlets and 2179 neighbors
100%|██████████████████████████████████████| 2179/2179 [00:03&lt;00:00, 549.49it/s]
100%|██████████████████████████████████████| 2179/2179 [00:03&lt;00:00, 555.47it/s]
2025-10-06 10:28:49,466 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 10:28:50,642 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 10:28:55,022 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:   12.1s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:   12.4s remaining:  1.4min
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:   12.9s remaining:   56.0s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:   13.0s remaining:   39.0s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:   13.0s remaining:   28.6s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:   13.2s remaining:   22.1s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:   15.0s remaining:   19.2s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:   16.9s remaining:   16.9s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:   17.9s remaining:   13.9s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:   19.0s remaining:   11.4s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:   19.1s remaining:    8.7s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:   19.5s remaining:    6.5s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:   19.6s remaining:    4.5s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:   20.6s remaining:    2.9s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:   21.9s finished
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.23181184941890648
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.23184146969421768
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.23184146969421768
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.23184095848063233
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.23181879129667568
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.23184095848063233
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.23066065746101844
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.23184095848063233
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.23179422289645368
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.23184054137243731
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.23184237532370044
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.23179422289645368
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.2318425961250881
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.2306421194753479
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.23184146969421768
2025-10-06 10:29:17,003 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.23180437648381458
2025-10-06 10:29:17,003 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 4/4 [00:04&lt;00:00,  1.14s/it]
2025-10-06 10:29:21,548 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 1866 seqlets
100%|██████████████████████████████████████| 1866/1866 [00:01&lt;00:00, 934.64it/s]
2025-10-06 10:29:24,082 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 1866 seqlets and 1866 neighbors
100%|██████████████████████████████████████| 1866/1866 [00:09&lt;00:00, 193.42it/s]
100%|██████████████████████████████████████| 1866/1866 [00:09&lt;00:00, 193.65it/s]
2025-10-06 10:29:43,529 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 10:29:47,620 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    3.6s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    4.2s remaining:   29.3s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    4.5s remaining:   19.6s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    4.7s remaining:   14.2s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    5.6s remaining:   12.4s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    6.4s remaining:   10.6s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    6.5s remaining:    8.4s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    7.1s remaining:    7.1s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    7.9s remaining:    6.1s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    9.0s remaining:    5.4s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    9.1s remaining:    4.1s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:   10.3s remaining:    3.4s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:   10.5s remaining:    2.4s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:   11.0s remaining:    1.6s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:   13.1s finished
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.19642208321503218
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.19645519080864812
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.1964546385546664
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.19640213963417552
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.19645519080864812
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.19639717078576602
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.19646935425451065
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.1964585874585927
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.19639717078576602
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.19639717078576602
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.19647132160678013
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.19646917667389605
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.19647132160678013
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.19647132160678013
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.19647132160678013
2025-10-06 10:30:00,793 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.19647234353742565
2025-10-06 10:30:00,793 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 4/4 [00:03&lt;00:00,  1.08it/s]
2025-10-06 10:30:04,505 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|████| 4/4 [00:19&lt;00:00,  4.82s/it]
2025-10-06 10:30:39,298 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|███████████████████| 26/26 [00:00&lt;00:00, 14885.60it/s]
Computing subpatterns:: 100%|███████████████████| 22/22 [00:02&lt;00:00,  9.44it/s]
Creating modisco logos for pos_patterns: 100%|██| 21/21 [00:31&lt;00:00,  1.52s/it]
Creating modisco logos for neg_patterns: 100%|██| 22/22 [00:48&lt;00:00,  2.23s/it]
Generating patterns dataframe: 100%|█████████████| 2/2 [00:00&lt;00:00, 197.28it/s]
Reading patterns for pos_patterns: 100%|██████| 21/21 [00:00&lt;00:00, 3236.82it/s]
Reading patterns for neg_patterns: 100%|██████| 22/22 [00:00&lt;00:00, 3260.13it/s]
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    7.9s
[Parallel(n_jobs=8)]: Done   2 tasks      | elapsed:    7.9s
[Parallel(n_jobs=8)]: Done   3 tasks      | elapsed:    8.1s
[Parallel(n_jobs=8)]: Done   4 tasks      | elapsed:    9.3s
[Parallel(n_jobs=8)]: Done   5 tasks      | elapsed:   10.9s
[Parallel(n_jobs=8)]: Done   6 tasks      | elapsed:   12.5s
[Parallel(n_jobs=8)]: Done   7 tasks      | elapsed:   13.4s
[Parallel(n_jobs=8)]: Done   8 tasks      | elapsed:   14.1s
[Parallel(n_jobs=8)]: Done   9 tasks      | elapsed:   25.2s
[Parallel(n_jobs=8)]: Done  10 tasks      | elapsed:   25.6s
[Parallel(n_jobs=8)]: Done  11 tasks      | elapsed:   27.4s
[Parallel(n_jobs=8)]: Done  12 tasks      | elapsed:   33.9s
[Parallel(n_jobs=8)]: Done  13 tasks      | elapsed:   34.6s
[Parallel(n_jobs=8)]: Done  14 tasks      | elapsed:   38.8s
[Parallel(n_jobs=8)]: Done  15 tasks      | elapsed:   39.7s
[Parallel(n_jobs=8)]: Done  16 tasks      | elapsed:   47.6s
[Parallel(n_jobs=8)]: Done  17 tasks      | elapsed:   48.5s
[Parallel(n_jobs=8)]: Done  18 tasks      | elapsed:   50.0s
[Parallel(n_jobs=8)]: Done  19 tasks      | elapsed:  1.0min
[Parallel(n_jobs=8)]: Done  20 tasks      | elapsed:  1.0min
[Parallel(n_jobs=8)]: Done  21 tasks      | elapsed:  1.3min
[Parallel(n_jobs=8)]: Done  22 tasks      | elapsed:  1.3min
[Parallel(n_jobs=8)]: Done  23 tasks      | elapsed:  1.4min
[Parallel(n_jobs=8)]: Done  24 tasks      | elapsed:  1.5min
[Parallel(n_jobs=8)]: Done  25 tasks      | elapsed:  1.6min
[Parallel(n_jobs=8)]: Done  26 tasks      | elapsed:  1.8min
[Parallel(n_jobs=8)]: Done  27 tasks      | elapsed:  1.9min
[Parallel(n_jobs=8)]: Done  28 tasks      | elapsed:  1.9min
[Parallel(n_jobs=8)]: Done  29 out of  43 | elapsed:  1.9min remaining:   54.4s
[Parallel(n_jobs=8)]: Done  30 out of  43 | elapsed:  1.9min remaining:   49.1s
[Parallel(n_jobs=8)]: Done  31 out of  43 | elapsed:  2.1min remaining:   48.2s
[Parallel(n_jobs=8)]: Done  32 out of  43 | elapsed:  2.1min remaining:   42.8s
[Parallel(n_jobs=8)]: Done  33 out of  43 | elapsed:  2.2min remaining:   40.8s
[Parallel(n_jobs=8)]: Done  34 out of  43 | elapsed:  2.3min remaining:   36.6s
[Parallel(n_jobs=8)]: Done  35 out of  43 | elapsed:  2.3min remaining:   31.8s
[Parallel(n_jobs=8)]: Done  36 out of  43 | elapsed:  2.3min remaining:   27.3s
[Parallel(n_jobs=8)]: Done  37 out of  43 | elapsed:  2.4min remaining:   23.2s
[Parallel(n_jobs=8)]: Done  38 out of  43 | elapsed:  2.5min remaining:   20.0s
[Parallel(n_jobs=8)]: Done  39 out of  43 | elapsed:  2.6min remaining:   15.9s
[Parallel(n_jobs=8)]: Done  40 out of  43 | elapsed:  2.7min remaining:   12.1s
[Parallel(n_jobs=8)]: Done  41 out of  43 | elapsed:  2.8min remaining:    8.1s
[Parallel(n_jobs=8)]: Done  43 out of  43 | elapsed:  3.3min finished
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1845.4MB/s)
Processing pos_patterns patterns...: 100%|█████| 21/21 [00:00&lt;00:00, 589.83it/s]
Processing neg_patterns patterns...: 100%|█████| 22/22 [00:00&lt;00:00, 572.37it/s]

</pre></div>
</div>
</div>
</div>
<p>The analysis creates the following files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">example/modisco_neurons.h5</span></code>: HDF5 file containing the discovered motifs and seqlets generated by modisco.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">example/modisco_neurons_patterns.h5</span></code>: HDF5 file with processed motif patterns generated by TF-Modisco see TF-modisco API.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">example/modisco_neurons_report/</span></code>: Directory containing HTML and image reports of discovered motifs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">example/modisco_neurons_seqlets.bed</span></code>: BED file with genomic locations of seqlets.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">example/modisco_neurons_*.attributions.h5</span></code>: HDF5 file containing the raw attribution scores for each input sequence and gene for each replicate. This file is used as input for motif discovery and can be inspected to understand which regions of the genome are most important for model predictions.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">example/modisco_neurons_*.attributions.bigwig</span></code>: Average attribution scores for each base across all input sequences, stored in BigWig format for visualization in genome browsers.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>example/modisco_neurons*
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>example/modisco_neurons_0.attributions.bigwig
example/modisco_neurons_0.attributions.h5
example/modisco_neurons_0.warnings.qc.log
example/modisco_neurons_1.attributions.bigwig
example/modisco_neurons_1.attributions.h5
example/modisco_neurons_1.warnings.qc.log
example/modisco_neurons.attributions.bigwig
example/modisco_neurons.attributions.h5
example/modisco_neurons.modisco.h5
example/modisco_neurons.seqlets.bed
example/modisco_neurons.warnings.qc.log

example/modisco_neurons_report:
ATF1.H13CORE.1.P.B.png	       REST.H13CORE.0.P.B.png
ATF2.H13CORE.0.PSM.A.png       RXRA.H13CORE.3.P.B.png
ATF2.H13CORE.1.P.B.png	       SATB1.H13CORE.0.P.B.png
ATF3.H13CORE.2.P.B.png	       SNAI2.H13CORE.0.PSM.A.png
ATF4.H13CORE.2.SM.B.png        SP2.H13CORE.1.P.B.png
ATMIN.H13CORE.0.P.B.png        SP3.H13CORE.0.P.B.png
CPEB1.H13CORE.1.S.B.png        SP4.H13CORE.2.P.B.png
CREB1.H13CORE.0.PSM.A.png      SPDEF.H13CORE.1.SM.B.png
CREB5.H13CORE.0.P.B.png        SPI1.H13CORE.0.P.B.png
CREB5.H13CORE.1.SM.B.png       SPIB.H13CORE.0.P.B.png
CTCFL.H13CORE.0.P.B.png        SPIB.H13CORE.2.SM.B.png
DMRTA.H13CORE.0.SM.B.png       STAT1.H13CORE.1.P.B.png
ELF1.H13CORE.0.PSM.A.png       STAT2.H13CORE.0.P.B.png
ELF4.H13CORE.0.PS.A.png        TAF1.H13CORE.0.P.B.png
ELK3.H13CORE.0.PSM.A.png       TCF21.H13CORE.0.PS.A.png
ELK4.H13CORE.0.PSM.A.png       TF7L1.H13CORE.2.S.C.png
ERG.H13CORE.0.P.B.png	       trimmed_logos
ERR1.H13CORE.0.PSM.A.png       VEZF1.H13CORE.1.P.B.png
ERR3.H13CORE.0.PSM.A.png       Z585A.H13CORE.0.P.C.png
ETV5.H13CORE.1.PM.A.png        ZBT14.H13CORE.0.P.C.png
ETV6.H13CORE.0.PS.A.png        ZBT14.H13CORE.1.S.B.png
FAM200B.H13CORE.0.PSGIB.A.png  ZEB1.H13CORE.0.P.B.png
FEV.H13CORE.0.S.B.png	       ZEB2.H13CORE.0.P.B.png
FOSB.H13CORE.0.P.B.png	       ZKSC5.H13CORE.0.P.C.png
FOS.H13CORE.2.S.C.png	       ZN100.H13CORE.1.P.B.png
FOSL1.H13CORE.0.P.B.png        ZN223.H13CORE.0.P.C.png
FOSL2.H13CORE.0.P.B.png        ZN250.H13CORE.0.P.C.png
FOXC1.H13CORE.1.S.C.png        ZN331.H13CORE.0.P.B.png
FOXC2.H13CORE.0.SM.B.png       ZN362.H13CORE.0.P.C.png
FOXI1.H13CORE.0.SM.B.png       ZN454.H13CORE.1.SM.B.png
FOXL1.H13CORE.0.S.B.png        ZN467.H13CORE.0.P.C.png
GRHL1.H13CORE.0.PSM.A.png      ZN510.H13CORE.0.P.C.png
IRF1.H13CORE.0.P.B.png	       ZN519.H13CORE.0.P.C.png
IRF4.H13CORE.0.P.B.png	       ZN580.H13CORE.0.SM.B.png
IRF4.H13CORE.1.S.B.png	       ZN587.H13CORE.0.P.C.png
IRF8.H13CORE.0.P.B.png	       ZN616.H13CORE.0.P.C.png
ITF2.H13CORE.1.PSM.A.png       ZN660.H13CORE.0.SM.B.png
JDP2.H13CORE.0.SM.B.png        ZN662.H13CORE.0.P.C.png
JUNB.H13CORE.0.PM.A.png        ZN664.H13CORE.0.P.C.png
JUND.H13CORE.0.PM.A.png        ZN681.H13CORE.0.P.C.png
JUN.H13CORE.0.P.B.png	       ZN682.H13CORE.1.P.B.png
JUN.H13CORE.1.S.C.png	       ZN730.H13CORE.0.P.C.png
KLF12.H13CORE.0.P.C.png        ZN766.H13CORE.0.P.B.png
KLF13.H13CORE.1.P.C.png        ZN768.H13CORE.0.P.B.png
KLF16.H13CORE.1.P.B.png        ZN770.H13CORE.1.P.B.png
KLF6.H13CORE.0.P.B.png	       ZN784.H13CORE.0.SM.B.png
KLF9.H13CORE.1.P.B.png	       ZN823.H13CORE.0.P.C.png
KMT2A.H13CORE.0.P.B.png        ZN875.H13CORE.0.P.C.png
KMT2B.H13CORE.0.P.B.png        ZNF12.H13CORE.1.M.C.png
motifs.html		       ZNF160.H13CORE.0.SG.A.png
MYT1.H13CORE.0.SG.A.png        ZNF43.H13CORE.0.PSGI.A.png
MYT1L.H13CORE.0.P.C.png        ZNF470.H13CORE.0.PSG.A.png
NR1I2.H13CORE.1.S.C.png        ZNF618.H13CORE.0.G.B.png
NR1I3.H13CORE.1.PSM.A.png      ZNF66.H13CORE.0.PG.A.png
NR5A2.H13CORE.0.PSM.A.png      ZNF746.H13CORE.0.PSG.A.png
NRF1.H13CORE.0.PS.A.png        ZNF800.H13CORE.0.PSG.A.png
ONEC2.H13CORE.3.P.B.png        ZNF813.H13CORE.0.SG.A.png
PATZ1.H13CORE.0.P.B.png        ZNF841.H13CORE.0.PSGI.A.png
PRGR.H13CORE.0.P.B.png	       ZNF85.H13CORE.1.P.B.png
RARG.H13CORE.0.P.B.png
</pre></div>
</div>
</div>
</div>
<p>The following file generates an HTML report with motif visualizations and detailed statistics for each discovered pattern.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">HTML</span><span class="p">,</span> <span class="n">Image</span>

<span class="n">HTML</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;example/modisco_neurons_report/motifs.html&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>pattern</th>
      <th>num_seqlets</th>
      <th>modisco_cwm_fwd</th>
      <th>modisco_cwm_rev</th>
      <th>match0</th>
      <th>qval0</th>
      <th>match0_logo</th>
      <th>match1</th>
      <th>qval1</th>
      <th>match1_logo</th>
      <th>match2</th>
      <th>qval2</th>
      <th>match2_logo</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>pos_patterns.pattern_0</td>
      <td>79</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_0.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_0.cwm.rev.png" width="240" ></td>
      <td>ZN362.H13CORE.0.P.C</td>
      <td>2.147410e-02</td>
      <td><img src="ZN362.H13CORE.0.P.C.png" width="240" ></td>
      <td>Z585A.H13CORE.0.P.C</td>
      <td>1.015950e-01</td>
      <td><img src="Z585A.H13CORE.0.P.C.png" width="240" ></td>
      <td>FOXC1.H13CORE.1.S.C</td>
      <td>1.015950e-01</td>
      <td><img src="FOXC1.H13CORE.1.S.C.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_1</td>
      <td>75</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_1.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_1.cwm.rev.png" width="240" ></td>
      <td>JUN.H13CORE.1.S.C</td>
      <td>1.000000e+00</td>
      <td><img src="JUN.H13CORE.1.S.C.png" width="240" ></td>
      <td>TF7L1.H13CORE.2.S.C</td>
      <td>1.000000e+00</td>
      <td><img src="TF7L1.H13CORE.2.S.C.png" width="240" ></td>
      <td>SPDEF.H13CORE.1.SM.B</td>
      <td>1.000000e+00</td>
      <td><img src="SPDEF.H13CORE.1.SM.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_2</td>
      <td>61</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_2.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_2.cwm.rev.png" width="240" ></td>
      <td>ZEB2.H13CORE.0.P.B</td>
      <td>2.037100e-02</td>
      <td><img src="ZEB2.H13CORE.0.P.B.png" width="240" ></td>
      <td>ITF2.H13CORE.1.PSM.A</td>
      <td>2.037100e-02</td>
      <td><img src="ITF2.H13CORE.1.PSM.A.png" width="240" ></td>
      <td>SNAI2.H13CORE.0.PSM.A</td>
      <td>2.037100e-02</td>
      <td><img src="SNAI2.H13CORE.0.PSM.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_3</td>
      <td>58</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_3.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_3.cwm.rev.png" width="240" ></td>
      <td>ZNF85.H13CORE.1.P.B</td>
      <td>7.257980e-01</td>
      <td><img src="ZNF85.H13CORE.1.P.B.png" width="240" ></td>
      <td>ZN766.H13CORE.0.P.B</td>
      <td>7.257980e-01</td>
      <td><img src="ZN766.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN250.H13CORE.0.P.C</td>
      <td>7.257980e-01</td>
      <td><img src="ZN250.H13CORE.0.P.C.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_4</td>
      <td>56</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_4.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_4.cwm.rev.png" width="240" ></td>
      <td>ZN519.H13CORE.0.P.C</td>
      <td>3.222360e-01</td>
      <td><img src="ZN519.H13CORE.0.P.C.png" width="240" ></td>
      <td>ZNF43.H13CORE.0.PSGI.A</td>
      <td>3.222360e-01</td>
      <td><img src="ZNF43.H13CORE.0.PSGI.A.png" width="240" ></td>
      <td>FAM200B.H13CORE.0.PSGIB.A</td>
      <td>3.222360e-01</td>
      <td><img src="FAM200B.H13CORE.0.PSGIB.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_5</td>
      <td>55</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_5.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_5.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>2.974570e-01</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN519.H13CORE.0.P.C</td>
      <td>3.238500e-01</td>
      <td><img src="ZN519.H13CORE.0.P.C.png" width="240" ></td>
      <td>ZNF66.H13CORE.0.PG.A</td>
      <td>6.090510e-01</td>
      <td><img src="ZNF66.H13CORE.0.PG.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_6</td>
      <td>54</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_6.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_6.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>6.289190e-06</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN519.H13CORE.0.P.C</td>
      <td>4.988970e-05</td>
      <td><img src="ZN519.H13CORE.0.P.C.png" width="240" ></td>
      <td>SP2.H13CORE.1.P.B</td>
      <td>1.925710e-02</td>
      <td><img src="SP2.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_7</td>
      <td>54</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_7.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_7.cwm.rev.png" width="240" ></td>
      <td>ZEB2.H13CORE.0.P.B</td>
      <td>9.978340e-01</td>
      <td><img src="ZEB2.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZEB1.H13CORE.0.P.B</td>
      <td>9.978340e-01</td>
      <td><img src="ZEB1.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN784.H13CORE.0.SM.B</td>
      <td>9.978340e-01</td>
      <td><img src="ZN784.H13CORE.0.SM.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_8</td>
      <td>52</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_8.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_8.cwm.rev.png" width="240" ></td>
      <td>SATB1.H13CORE.0.P.B</td>
      <td>1.405690e-01</td>
      <td><img src="SATB1.H13CORE.0.P.B.png" width="240" ></td>
      <td>FOXC2.H13CORE.0.SM.B</td>
      <td>2.268190e-01</td>
      <td><img src="FOXC2.H13CORE.0.SM.B.png" width="240" ></td>
      <td>CPEB1.H13CORE.1.S.B</td>
      <td>3.961930e-01</td>
      <td><img src="CPEB1.H13CORE.1.S.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_9</td>
      <td>51</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_9.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_9.cwm.rev.png" width="240" ></td>
      <td>ATF3.H13CORE.2.P.B</td>
      <td>9.799360e-03</td>
      <td><img src="ATF3.H13CORE.2.P.B.png" width="240" ></td>
      <td>CREB5.H13CORE.1.SM.B</td>
      <td>9.799360e-03</td>
      <td><img src="CREB5.H13CORE.1.SM.B.png" width="240" ></td>
      <td>CREB1.H13CORE.0.PSM.A</td>
      <td>9.799360e-03</td>
      <td><img src="CREB1.H13CORE.0.PSM.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_10</td>
      <td>51</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_10.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_10.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>1.162170e-05</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN519.H13CORE.0.P.C</td>
      <td>7.242570e-03</td>
      <td><img src="ZN519.H13CORE.0.P.C.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>9.900230e-03</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_11</td>
      <td>46</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_11.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_11.cwm.rev.png" width="240" ></td>
      <td>ZNF813.H13CORE.0.SG.A</td>
      <td>1.251000e-01</td>
      <td><img src="ZNF813.H13CORE.0.SG.A.png" width="240" ></td>
      <td>ZN681.H13CORE.0.P.C</td>
      <td>1.290180e-01</td>
      <td><img src="ZN681.H13CORE.0.P.C.png" width="240" ></td>
      <td>RXRA.H13CORE.3.P.B</td>
      <td>1.290180e-01</td>
      <td><img src="RXRA.H13CORE.3.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_12</td>
      <td>44</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_12.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_12.cwm.rev.png" width="240" ></td>
      <td>ERR3.H13CORE.0.PSM.A</td>
      <td>4.260500e-04</td>
      <td><img src="ERR3.H13CORE.0.PSM.A.png" width="240" ></td>
      <td>NR5A2.H13CORE.0.PSM.A</td>
      <td>7.454720e-04</td>
      <td><img src="NR5A2.H13CORE.0.PSM.A.png" width="240" ></td>
      <td>ERR1.H13CORE.0.PSM.A</td>
      <td>8.555690e-04</td>
      <td><img src="ERR1.H13CORE.0.PSM.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_13</td>
      <td>43</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_13.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_13.cwm.rev.png" width="240" ></td>
      <td>NRF1.H13CORE.0.PS.A</td>
      <td>1.698700e-02</td>
      <td><img src="NRF1.H13CORE.0.PS.A.png" width="240" ></td>
      <td>ZBT14.H13CORE.0.P.C</td>
      <td>1.773560e-02</td>
      <td><img src="ZBT14.H13CORE.0.P.C.png" width="240" ></td>
      <td>ZN454.H13CORE.1.SM.B</td>
      <td>1.026430e-01</td>
      <td><img src="ZN454.H13CORE.1.SM.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_14</td>
      <td>37</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_14.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_14.cwm.rev.png" width="240" ></td>
      <td>KMT2B.H13CORE.0.P.B</td>
      <td>4.570050e-02</td>
      <td><img src="KMT2B.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN100.H13CORE.1.P.B</td>
      <td>4.570050e-02</td>
      <td><img src="ZN100.H13CORE.1.P.B.png" width="240" ></td>
      <td>ZN519.H13CORE.0.P.C</td>
      <td>1.034090e-01</td>
      <td><img src="ZN519.H13CORE.0.P.C.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_15</td>
      <td>35</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_15.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_15.cwm.rev.png" width="240" ></td>
      <td>JUNB.H13CORE.0.PM.A</td>
      <td>8.438550e-02</td>
      <td><img src="JUNB.H13CORE.0.PM.A.png" width="240" ></td>
      <td>FOSB.H13CORE.0.P.B</td>
      <td>9.620290e-02</td>
      <td><img src="FOSB.H13CORE.0.P.B.png" width="240" ></td>
      <td>FOSL1.H13CORE.0.P.B</td>
      <td>1.398950e-01</td>
      <td><img src="FOSL1.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_16</td>
      <td>29</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_16.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_16.cwm.rev.png" width="240" ></td>
      <td>KLF6.H13CORE.0.P.B</td>
      <td>2.727660e-01</td>
      <td><img src="KLF6.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZNF800.H13CORE.0.PSG.A</td>
      <td>2.727660e-01</td>
      <td><img src="ZNF800.H13CORE.0.PSG.A.png" width="240" ></td>
      <td>KLF13.H13CORE.1.P.C</td>
      <td>3.620790e-01</td>
      <td><img src="KLF13.H13CORE.1.P.C.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_17</td>
      <td>25</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_17.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_17.cwm.rev.png" width="240" ></td>
      <td>CTCFL.H13CORE.0.P.B</td>
      <td>7.259900e-03</td>
      <td><img src="CTCFL.H13CORE.0.P.B.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>2.273880e-02</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>SP2.H13CORE.1.P.B</td>
      <td>2.332960e-02</td>
      <td><img src="SP2.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_18</td>
      <td>25</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_18.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_18.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>3.588490e-02</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>1.058890e-01</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
      <td>KLF16.H13CORE.1.P.B</td>
      <td>1.103100e-01</td>
      <td><img src="KLF16.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_19</td>
      <td>23</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_19.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_19.cwm.rev.png" width="240" ></td>
      <td>KLF12.H13CORE.0.P.C</td>
      <td>2.011500e-04</td>
      <td><img src="KLF12.H13CORE.0.P.C.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>2.011500e-04</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>2.011500e-04</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>pos_patterns.pattern_20</td>
      <td>20</td>
      <td><img src="trimmed_logos/pos_patterns.pattern_20.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/pos_patterns.pattern_20.cwm.rev.png" width="240" ></td>
      <td>ZN331.H13CORE.0.P.B</td>
      <td>5.969900e-02</td>
      <td><img src="ZN331.H13CORE.0.P.B.png" width="240" ></td>
      <td>ITF2.H13CORE.1.PSM.A</td>
      <td>7.167000e-02</td>
      <td><img src="ITF2.H13CORE.1.PSM.A.png" width="240" ></td>
      <td>ZEB2.H13CORE.0.P.B</td>
      <td>7.167000e-02</td>
      <td><img src="ZEB2.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_0</td>
      <td>287</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_0.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_0.cwm.rev.png" width="240" ></td>
      <td>ELK4.H13CORE.0.PSM.A</td>
      <td>1.988280e-03</td>
      <td><img src="ELK4.H13CORE.0.PSM.A.png" width="240" ></td>
      <td>ERG.H13CORE.0.P.B</td>
      <td>1.988280e-03</td>
      <td><img src="ERG.H13CORE.0.P.B.png" width="240" ></td>
      <td>ETV5.H13CORE.1.PM.A</td>
      <td>1.988280e-03</td>
      <td><img src="ETV5.H13CORE.1.PM.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_1</td>
      <td>122</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_1.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_1.cwm.rev.png" width="240" ></td>
      <td>ZNF618.H13CORE.0.G.B</td>
      <td>2.586510e-04</td>
      <td><img src="ZNF618.H13CORE.0.G.B.png" width="240" ></td>
      <td>ZN362.H13CORE.0.P.C</td>
      <td>2.586510e-04</td>
      <td><img src="ZN362.H13CORE.0.P.C.png" width="240" ></td>
      <td>SATB1.H13CORE.0.P.B</td>
      <td>2.586510e-04</td>
      <td><img src="SATB1.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_2</td>
      <td>120</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_2.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_2.cwm.rev.png" width="240" ></td>
      <td>MYT1.H13CORE.0.SG.A</td>
      <td>1.000000e+00</td>
      <td><img src="MYT1.H13CORE.0.SG.A.png" width="240" ></td>
      <td>MYT1L.H13CORE.0.P.C</td>
      <td>1.000000e+00</td>
      <td><img src="MYT1L.H13CORE.0.P.C.png" width="240" ></td>
      <td>NR1I3.H13CORE.1.PSM.A</td>
      <td>1.000000e+00</td>
      <td><img src="NR1I3.H13CORE.1.PSM.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_3</td>
      <td>100</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_3.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_3.cwm.rev.png" width="240" ></td>
      <td>JUN.H13CORE.0.P.B</td>
      <td>1.000000e+00</td>
      <td><img src="JUN.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN660.H13CORE.0.SM.B</td>
      <td>1.000000e+00</td>
      <td><img src="ZN660.H13CORE.0.SM.B.png" width="240" ></td>
      <td>FOSL2.H13CORE.0.P.B</td>
      <td>1.000000e+00</td>
      <td><img src="FOSL2.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_4</td>
      <td>79</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_4.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_4.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>1.266440e-02</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZN682.H13CORE.1.P.B</td>
      <td>7.867000e-02</td>
      <td><img src="ZN682.H13CORE.1.P.B.png" width="240" ></td>
      <td>SP3.H13CORE.0.P.B</td>
      <td>1.500860e-01</td>
      <td><img src="SP3.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_5</td>
      <td>75</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_5.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_5.cwm.rev.png" width="240" ></td>
      <td>ZEB1.H13CORE.0.P.B</td>
      <td>1.000000e+00</td>
      <td><img src="ZEB1.H13CORE.0.P.B.png" width="240" ></td>
      <td>ZKSC5.H13CORE.0.P.C</td>
      <td>1.000000e+00</td>
      <td><img src="ZKSC5.H13CORE.0.P.C.png" width="240" ></td>
      <td>ZN616.H13CORE.0.P.C</td>
      <td>1.000000e+00</td>
      <td><img src="ZN616.H13CORE.0.P.C.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_6</td>
      <td>66</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_6.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_6.cwm.rev.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>8.956970e-09</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
      <td>SP2.H13CORE.1.P.B</td>
      <td>1.211690e-08</td>
      <td><img src="SP2.H13CORE.1.P.B.png" width="240" ></td>
      <td>KLF16.H13CORE.1.P.B</td>
      <td>3.388310e-07</td>
      <td><img src="KLF16.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_7</td>
      <td>60</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_7.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_7.cwm.rev.png" width="240" ></td>
      <td>KLF16.H13CORE.1.P.B</td>
      <td>6.887900e-10</td>
      <td><img src="KLF16.H13CORE.1.P.B.png" width="240" ></td>
      <td>SP2.H13CORE.1.P.B</td>
      <td>1.595690e-09</td>
      <td><img src="SP2.H13CORE.1.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>1.994810e-09</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_8</td>
      <td>58</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_8.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_8.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>5.301210e-03</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>2.810640e-02</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
      <td>PATZ1.H13CORE.0.P.B</td>
      <td>3.475440e-02</td>
      <td><img src="PATZ1.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_9</td>
      <td>54</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_9.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_9.cwm.rev.png" width="240" ></td>
      <td>MYT1L.H13CORE.0.P.C</td>
      <td>1.000000e+00</td>
      <td><img src="MYT1L.H13CORE.0.P.C.png" width="240" ></td>
      <td>Z585A.H13CORE.0.P.C</td>
      <td>1.000000e+00</td>
      <td><img src="Z585A.H13CORE.0.P.C.png" width="240" ></td>
      <td>ZN768.H13CORE.0.P.B</td>
      <td>1.000000e+00</td>
      <td><img src="ZN768.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_10</td>
      <td>53</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_10.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_10.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>9.252870e-03</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>1.345110e-02</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
      <td>ZBT14.H13CORE.0.P.C</td>
      <td>3.046580e-02</td>
      <td><img src="ZBT14.H13CORE.0.P.C.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_11</td>
      <td>49</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_11.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_11.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>4.197120e-04</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>KLF16.H13CORE.1.P.B</td>
      <td>1.990510e-03</td>
      <td><img src="KLF16.H13CORE.1.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>3.379880e-03</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_12</td>
      <td>49</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_12.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_12.cwm.rev.png" width="240" ></td>
      <td>FOXI1.H13CORE.0.SM.B</td>
      <td>9.116190e-01</td>
      <td><img src="FOXI1.H13CORE.0.SM.B.png" width="240" ></td>
      <td>MYT1.H13CORE.0.SG.A</td>
      <td>9.116190e-01</td>
      <td><img src="MYT1.H13CORE.0.SG.A.png" width="240" ></td>
      <td>GRHL1.H13CORE.0.PSM.A</td>
      <td>9.116190e-01</td>
      <td><img src="GRHL1.H13CORE.0.PSM.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_13</td>
      <td>45</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_13.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_13.cwm.rev.png" width="240" ></td>
      <td>STAT2.H13CORE.0.P.B</td>
      <td>6.372390e-04</td>
      <td><img src="STAT2.H13CORE.0.P.B.png" width="240" ></td>
      <td>STAT1.H13CORE.1.P.B</td>
      <td>2.162320e-03</td>
      <td><img src="STAT1.H13CORE.1.P.B.png" width="240" ></td>
      <td>IRF1.H13CORE.0.P.B</td>
      <td>3.034750e-02</td>
      <td><img src="IRF1.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_14</td>
      <td>44</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_14.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_14.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>3.299920e-07</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>KLF16.H13CORE.1.P.B</td>
      <td>8.833780e-04</td>
      <td><img src="KLF16.H13CORE.1.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>8.833780e-04</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_15</td>
      <td>44</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_15.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_15.cwm.rev.png" width="240" ></td>
      <td>FOSL2.H13CORE.0.P.B</td>
      <td>1.462900e-03</td>
      <td><img src="FOSL2.H13CORE.0.P.B.png" width="240" ></td>
      <td>JUN.H13CORE.0.P.B</td>
      <td>1.462900e-03</td>
      <td><img src="JUN.H13CORE.0.P.B.png" width="240" ></td>
      <td>JUND.H13CORE.0.PM.A</td>
      <td>1.462900e-03</td>
      <td><img src="JUND.H13CORE.0.PM.A.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_16</td>
      <td>38</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_16.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_16.cwm.rev.png" width="240" ></td>
      <td>IRF8.H13CORE.0.P.B</td>
      <td>2.335030e-02</td>
      <td><img src="IRF8.H13CORE.0.P.B.png" width="240" ></td>
      <td>IRF4.H13CORE.0.P.B</td>
      <td>2.335030e-02</td>
      <td><img src="IRF4.H13CORE.0.P.B.png" width="240" ></td>
      <td>IRF1.H13CORE.0.P.B</td>
      <td>1.059590e-01</td>
      <td><img src="IRF1.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_17</td>
      <td>37</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_17.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_17.cwm.rev.png" width="240" ></td>
      <td>IRF4.H13CORE.1.S.B</td>
      <td>1.474470e-01</td>
      <td><img src="IRF4.H13CORE.1.S.B.png" width="240" ></td>
      <td>SPIB.H13CORE.0.P.B</td>
      <td>1.474470e-01</td>
      <td><img src="SPIB.H13CORE.0.P.B.png" width="240" ></td>
      <td>SP4.H13CORE.2.P.B</td>
      <td>1.474470e-01</td>
      <td><img src="SP4.H13CORE.2.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_18</td>
      <td>35</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_18.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_18.cwm.rev.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>3.329330e-05</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>3.723430e-03</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
      <td>ZN519.H13CORE.0.P.C</td>
      <td>7.556750e-03</td>
      <td><img src="ZN519.H13CORE.0.P.C.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_19</td>
      <td>30</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_19.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_19.cwm.rev.png" width="240" ></td>
      <td>MYT1L.H13CORE.0.P.C</td>
      <td>1.000000e+00</td>
      <td><img src="MYT1L.H13CORE.0.P.C.png" width="240" ></td>
      <td>TCF21.H13CORE.0.PS.A</td>
      <td>1.000000e+00</td>
      <td><img src="TCF21.H13CORE.0.PS.A.png" width="240" ></td>
      <td>PRGR.H13CORE.0.P.B</td>
      <td>1.000000e+00</td>
      <td><img src="PRGR.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_20</td>
      <td>29</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_20.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_20.cwm.rev.png" width="240" ></td>
      <td>KLF9.H13CORE.1.P.B</td>
      <td>1.010330e-01</td>
      <td><img src="KLF9.H13CORE.1.P.B.png" width="240" ></td>
      <td>ZN587.H13CORE.0.P.C</td>
      <td>1.010330e-01</td>
      <td><img src="ZN587.H13CORE.0.P.C.png" width="240" ></td>
      <td>KMT2A.H13CORE.0.P.B</td>
      <td>1.010330e-01</td>
      <td><img src="KMT2A.H13CORE.0.P.B.png" width="240" ></td>
    </tr>
    <tr>
      <td>neg_patterns.pattern_21</td>
      <td>29</td>
      <td><img src="trimmed_logos/neg_patterns.pattern_21.cwm.fwd.png" width="240" ></td>
      <td><img src="trimmed_logos/neg_patterns.pattern_21.cwm.rev.png" width="240" ></td>
      <td>CREB5.H13CORE.1.SM.B</td>
      <td>5.706230e-03</td>
      <td><img src="CREB5.H13CORE.1.SM.B.png" width="240" ></td>
      <td>CREB1.H13CORE.0.PSM.A</td>
      <td>6.479550e-03</td>
      <td><img src="CREB1.H13CORE.0.PSM.A.png" width="240" ></td>
      <td>ATF2.H13CORE.0.PSM.A</td>
      <td>8.492830e-03</td>
      <td><img src="ATF2.H13CORE.0.PSM.A.png" width="240" ></td>
    </tr>
  </tbody>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Image</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;example/modisco_neurons_report/MYT1.H13CORE.0.SG.A.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1b0bae2e0c95e3141d15d43a2a684aca2f39e8c506cf826225d27e7891039e6d.png" src="../_images/1b0bae2e0c95e3141d15d43a2a684aca2f39e8c506cf826225d27e7891039e6d.png" />
</div>
</div>
<section id="cli-subcommands">
<h3>CLI Subcommands<a class="headerlink" href="#cli-subcommands" title="Link to this heading">¶</a></h3>
<p>The Modisco API provides several subcommands to perform motif discovery and analysis in a stepwise manner.
Here is a detailed explanation of each step:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">modisco-attributions</span></code>: This subcommand computes attribution scores for each input sequence and gene. You can parallelize this step by running it on multiple GPUs. This is the only step that requires GPUs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">modisco-patterns</span></code>: After computing attributions, this subcommand discovers motifs (patterns) by clustering
the attribution scores using TF-Modisco. See the documentation of Modisco for the details of the output format.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">modisco-report</span></code>: This subcommand run tomtom to find motif matches for patterns motif images from the discovered patterns.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">modisco-seqlet-bed</span></code>: This subcommand extracts the genomic locations of seqlets (short, high-scoring
regions) from the modisco results and writes them to a BED file. This allows for downstream analysis
or visualization in genome browsers.</p></li>
</ol>
<p>By running these subcommands in sequence, you can go from raw model predictions to interpretable motif
discovery, visualization, and downstream analysis.</p>
<p>Let’s generate attributions for first replicate:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>modisco-attributions<span class="w"> </span><span class="err">\</span>
    <span class="o">--</span><span class="n">top</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">markers</span> <span class="mi">50</span> \
    <span class="o">--</span><span class="n">tasks</span> <span class="s2">&quot;cell_type.str.contains(&#39;neuron&#39;) and organ == &#39;CNS&#39; and disease == &#39;healthy&#39;&quot;</span> \
    <span class="o">--</span><span class="n">batch</span><span class="o">-</span><span class="n">size</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">model</span> <span class="n">v1_rep0</span> \
    <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">workers</span> <span class="mi">8</span> \
    <span class="o">-</span><span class="n">o</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons_0</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Using device: cuda
decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.9 (1684.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1859.3MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1204.6MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
100%|████████████████████████████████████████████| 1/1 [00:00&lt;00:00, 360.24it/s]
Computing attributions...:   8%|█▎               | 4/50 [00:04&lt;00:47,  1.02s/it]decima - WARNING - Gene XRCC5 has low correlation with the model. Pearson: 0.4621304058925145. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  10%|█▋               | 5/50 [00:05&lt;00:43,  1.03it/s]decima - WARNING - Gene FNIP1 has low correlation with the model. Pearson: 0.044592478397006044. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  40%|██████▍         | 20/50 [00:18&lt;00:26,  1.14it/s]decima - WARNING - Gene LRRFIP1 has low correlation with the model. Pearson: 0.4621130932978845. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  70%|███████████▏    | 35/50 [00:31&lt;00:13,  1.14it/s]decima - WARNING - Gene CHD4 has low correlation with the model. Pearson: 0.27165706153002717. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...: 100%|████████████████| 50/50 [00:44&lt;00:00,  1.11it/s]

</pre></div>
</div>
</div>
</div>
<p>end second replicate:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>modisco-attributions<span class="w"> </span><span class="err">\</span>
    <span class="o">--</span><span class="n">top</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">markers</span> <span class="mi">50</span> \
    <span class="o">--</span><span class="n">tasks</span> <span class="s2">&quot;cell_type.str.contains(&#39;neuron&#39;) and organ == &#39;CNS&#39; and disease == &#39;healthy&#39;&quot;</span> \
    <span class="o">--</span><span class="n">batch</span><span class="o">-</span><span class="n">size</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">model</span> <span class="n">v1_rep1</span> \
    <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">workers</span> <span class="mi">8</span> \
    <span class="o">-</span><span class="n">o</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons_1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Using device: cuda
decima - INFO - Loading model and metadata to compute attributions...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1831.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1823.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep1:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (417.2MB/s)
/home/celikm5/Projects/decima/src/decima/interpret/attributer.py:43: UserWarning: `off_tasks` is not provided. Using all other tasks as off_tasks.
100%|████████████████████████████████████████████| 1/1 [00:00&lt;00:00, 323.39it/s]
Computing attributions...:   8%|█▎               | 4/50 [00:04&lt;00:46,  1.02s/it]decima - WARNING - Gene XRCC5 has low correlation with the model. Pearson: 0.4621304058925145. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  10%|█▋               | 5/50 [00:05&lt;00:43,  1.03it/s]decima - WARNING - Gene FNIP1 has low correlation with the model. Pearson: 0.044592478397006044. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  40%|██████▍         | 20/50 [00:18&lt;00:26,  1.14it/s]decima - WARNING - Gene LRRFIP1 has low correlation with the model. Pearson: 0.4621130932978845. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...:  70%|███████████▏    | 35/50 [00:31&lt;00:13,  1.14it/s]decima - WARNING - Gene CHD4 has low correlation with the model. Pearson: 0.27165706153002717. Be careful with the predictions of the model for this gene. Check `DecimaResult.load().gene_metadata[&#39;pearson&#39;]` to see the correlation of the gene with the model.
Computing attributions...: 100%|████████████████| 50/50 [00:45&lt;00:00,  1.11it/s]

</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>example/modisco_subcommands/modisco_neurons*
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>example/modisco_subcommands/modisco_neurons_0.attributions.bigwig
example/modisco_subcommands/modisco_neurons_0.attributions.h5
example/modisco_subcommands/modisco_neurons_0.warnings.qc.log
example/modisco_subcommands/modisco_neurons_1.attributions.bigwig
example/modisco_subcommands/modisco_neurons_1.attributions.h5
example/modisco_subcommands/modisco_neurons_1.warnings.qc.log
example/modisco_subcommands/modisco_neurons.attributions.bigwig
example/modisco_subcommands/modisco_neurons.attributions.h5
example/modisco_subcommands/modisco_neurons.warnings.qc.log
</pre></div>
</div>
</div>
</div>
<p>Next, we will use the generated attributions as input for the pattern discovery step with MoDISco.
The <code class="docutils literal notranslate"><span class="pre">--attributions</span></code> argument expects a comma-separated list of attribution files (one per replicate or model),
which we produced in the previous step. These files contain the importance scores for each base in the input sequences,
and are required for MoDISco to identify recurring patterns (motifs).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>modisco-patterns<span class="w"> </span><span class="err">\</span>
    <span class="o">--</span><span class="n">attributions</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons_0</span><span class="o">.</span><span class="n">attributions</span><span class="o">.</span><span class="n">h5</span><span class="p">,</span><span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons_1</span><span class="o">.</span><span class="n">attributions</span><span class="o">.</span><span class="n">h5</span> \
    <span class="o">--</span><span class="n">top</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">markers</span> <span class="mi">50</span> \
    <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">seqlets</span> <span class="mi">5000</span> \
    <span class="o">--</span><span class="n">tss</span><span class="o">-</span><span class="n">distance</span> <span class="mi">5000</span>  \
    <span class="o">--</span><span class="n">tasks</span> <span class="s2">&quot;cell_type.str.contains(&#39;neuron&#39;) and organ == &#39;CNS&#39; and disease == &#39;healthy&#39;&quot;</span> \
    <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">workers</span> <span class="mi">8</span> \
    <span class="o">-</span><span class="n">o</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Loading metadata
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1833.5MB/s)
100%|████████████████████████████████████████████| 1/1 [00:00&lt;00:00, 359.72it/s]
Loading attributions and sequences...: 100%|████| 50/50 [00:05&lt;00:00,  9.94it/s]
2025-10-06 11:09:35,941 - modisco-lite - INFO - Running TFMoDISco version 2.4.0
2025-10-06 11:09:35,943 - modisco-lite - INFO - Extracting seqlets for 50 tasks:
2025-10-06 11:09:35,944 - modisco-lite - INFO - - Smoothing and splitting tracks
2025-10-06 11:09:35,980 - modisco-lite - INFO - - Computing null values with Laplacian null model
2025-10-06 11:09:36,039 - modisco-lite - INFO - - Computing isotonic thresholds
2025-10-06 11:09:36,060 - modisco-lite - INFO - - Refining thresholds
2025-10-06 11:09:36,361 - modisco-lite - INFO - - Extracting seqlets
2025-10-06 11:09:36,409 - modisco-lite - INFO - - Extracting 1381 positive seqlets
2025-10-06 11:09:36,416 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 1381 seqlets
100%|██████████████████████████████████████| 1381/1381 [00:05&lt;00:00, 232.84it/s]
2025-10-06 11:09:50,133 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 1381 seqlets and 1381 neighbors
100%|██████████████████████████████████████| 1381/1381 [00:02&lt;00:00, 561.49it/s]
100%|██████████████████████████████████████| 1381/1381 [00:02&lt;00:00, 558.80it/s]
2025-10-06 11:09:55,173 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 11:09:55,911 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 11:09:58,670 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    9.0s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    9.0s remaining:  1.1min
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    9.1s remaining:   39.6s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    9.5s remaining:   28.4s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    9.6s remaining:   21.1s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    9.7s remaining:   16.2s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    9.9s remaining:   12.7s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:   10.1s remaining:   10.1s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:   11.7s remaining:    9.1s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:   12.0s remaining:    7.2s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:   12.2s remaining:    5.5s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:   12.3s remaining:    4.1s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:   12.5s remaining:    2.9s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:   12.6s remaining:    1.8s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:   12.8s finished
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.19091861254544126
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.19091861254544126
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.19091861254544126
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.19091861254544126
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.19091861254544126
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.1909180465007868
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.19091861254544126
2025-10-06 11:10:11,573 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.1909180465007868
2025-10-06 11:10:11,576 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 3/3 [00:02&lt;00:00,  1.25it/s]
2025-10-06 11:10:13,977 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 1243 seqlets
100%|█████████████████████████████████████| 1243/1243 [00:00&lt;00:00, 1412.11it/s]
2025-10-06 11:10:15,217 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 1243 seqlets and 1243 neighbors
100%|██████████████████████████████████████| 1243/1243 [00:06&lt;00:00, 195.27it/s]
100%|██████████████████████████████████████| 1243/1243 [00:06&lt;00:00, 192.70it/s]
2025-10-06 11:10:28,149 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 11:10:30,807 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    2.3s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    2.4s remaining:   16.7s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    2.4s remaining:   10.4s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    2.6s remaining:    7.7s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    2.7s remaining:    6.0s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    3.1s remaining:    5.2s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    3.2s remaining:    4.1s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    3.9s remaining:    3.9s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    4.9s remaining:    3.8s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    5.5s remaining:    3.3s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    5.8s remaining:    2.6s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    5.9s remaining:    2.0s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    6.0s remaining:    1.4s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    6.2s remaining:    0.9s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    6.6s finished
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.16244580961606636
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.16244580961606636
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.1624362681795782
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.16244401853104182
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.16244476881502096
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.16244490347064755
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.16244476881502096
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.16244580961606636
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.16244774126381895
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.16244774126381895
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.16244401853104182
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.16244580961606636
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.16244476881502096
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.1624362681795782
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.1624362681795782
2025-10-06 11:10:37,493 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.16244580961606636
2025-10-06 11:10:37,496 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 3/3 [00:01&lt;00:00,  1.53it/s]
2025-10-06 11:10:39,464 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|████| 3/3 [00:16&lt;00:00,  5.63s/it]
2025-10-06 11:11:04,245 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|███████████████████| 21/21 [00:00&lt;00:00, 13219.33it/s]
Computing subpatterns:: 100%|███████████████████| 18/18 [00:01&lt;00:00, 16.44it/s]
2025-10-06 11:11:05,342 - modisco-lite - INFO - - Extracting 2245 negative seqlets
2025-10-06 11:11:05,356 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 2245 seqlets
100%|██████████████████████████████████████| 2245/2245 [00:03&lt;00:00, 723.88it/s]
2025-10-06 11:11:09,810 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 2245 seqlets and 2245 neighbors
100%|██████████████████████████████████████| 2245/2245 [00:04&lt;00:00, 550.75it/s]
100%|██████████████████████████████████████| 2245/2245 [00:04&lt;00:00, 557.49it/s]
2025-10-06 11:11:18,089 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 11:11:19,309 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 11:11:23,973 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:   11.8s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:   12.2s remaining:  1.4min
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:   12.4s remaining:   53.9s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:   13.9s remaining:   41.7s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:   14.3s remaining:   31.5s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:   14.4s remaining:   24.0s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:   14.5s remaining:   18.7s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:   15.8s remaining:   15.8s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:   18.9s remaining:   14.7s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:   19.4s remaining:   11.7s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:   19.5s remaining:    8.9s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:   20.8s remaining:    6.9s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:   20.9s remaining:    4.8s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:   21.1s remaining:    3.0s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:   23.2s finished
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.24321691671687068
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.24319041549731021
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.24321691671687068
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.24319289414836717
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.24319766197141385
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.24272769482516707
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.24319516585924073
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.24319454173648977
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.2432134283873644
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.2432143002042254
2025-10-06 11:11:47,320 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.24319766197141385
2025-10-06 11:11:47,321 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.24321691671687068
2025-10-06 11:11:47,321 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.24321691671687068
2025-10-06 11:11:47,321 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.24321691671687068
2025-10-06 11:11:47,321 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.24319516585924073
2025-10-06 11:11:47,321 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.24319454173648977
2025-10-06 11:11:47,322 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 4/4 [00:04&lt;00:00,  1.23s/it]
2025-10-06 11:11:52,242 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 1908 seqlets
100%|██████████████████████████████████████| 1908/1908 [00:02&lt;00:00, 939.08it/s]
2025-10-06 11:11:54,770 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 1908 seqlets and 1908 neighbors
100%|██████████████████████████████████████| 1908/1908 [00:09&lt;00:00, 196.68it/s]
100%|██████████████████████████████████████| 1908/1908 [00:09&lt;00:00, 195.64it/s]
2025-10-06 11:12:14,398 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 11:12:18,764 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    4.6s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    4.8s remaining:   33.8s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    5.0s remaining:   21.5s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    5.1s remaining:   15.4s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    6.6s remaining:   14.5s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    6.9s remaining:   11.4s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    7.5s remaining:    9.7s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    8.8s remaining:    8.8s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:   11.0s remaining:    8.5s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:   11.0s remaining:    6.6s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:   11.8s remaining:    5.4s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:   12.9s remaining:    4.3s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:   13.0s remaining:    3.0s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:   13.4s remaining:    1.9s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:   14.8s finished
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.1964854074309556
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.19708009479161454
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.1970347002260257
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.1970719073220813
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.19706807930037878
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.19710320331929443
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.19649090179343903
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.19708787969518912
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.19649196976106514
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.19649196976106514
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.19589979138080835
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.19649090179343903
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.19709496100658516
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.19709080388309483
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.19707540188328013
2025-10-06 11:12:33,631 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.19706748023467258
2025-10-06 11:12:33,633 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
Generating patterns from clusters:: 100%|█████████| 4/4 [00:06&lt;00:00,  1.60s/it]
2025-10-06 11:12:40,024 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|████| 4/4 [00:20&lt;00:00,  5.10s/it]
2025-10-06 11:13:20,175 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|███████████████████| 26/26 [00:00&lt;00:00, 16883.71it/s]
Computing subpatterns:: 100%|███████████████████| 19/19 [00:02&lt;00:00,  6.96it/s]

</pre></div>
</div>
</div>
</div>
<p>This step generates modisco.h5 file with discovered motifs and processed patterns.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>ls<span class="w"> </span>example/modisco_subcommands/modisco_neurons*
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>modisco-seqlet-bed<span class="w"> </span><span class="err">\</span>
    <span class="o">--</span><span class="n">modisco</span><span class="o">-</span><span class="n">h5</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons</span><span class="o">.</span><span class="n">modisco</span><span class="o">.</span><span class="n">h5</span> \
    <span class="o">-</span><span class="n">o</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1792.1MB/s)
Processing pos_patterns patterns...: 100%|█████| 18/18 [00:00&lt;00:00, 586.01it/s]
Processing neg_patterns patterns...: 100%|█████| 19/19 [00:00&lt;00:00, 591.46it/s]

</pre></div>
</div>
</div>
</div>
<p>Extracting seqlets used in motif detection:</p>
<p>The <code class="docutils literal notranslate"><span class="pre">.seqlets.bed</span></code> file contains the genomic coordinates of all seqlets (short, high-scoring subsequences)
that were used by MoDISco to identify motifs. Each line in the BED file follows
the standard BED format:</p>
<p>chrom    start    end    name    score    strand</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">chrom</span></code>: Chromosome name (e.g., “chr1”)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">start</span></code>: 0-based start position of the seqlet</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">end</span></code>: End position (not included)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">name</span></code>: Identifier for the seqlet (often includes pattern/cluster info)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">score</span></code>: Score assigned to the seqlet (e.g., importance or cluster assignment)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">strand</span></code>: “+” or “-” indicating the DNA strand</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">core</span> <span class="pre">motif</span> <span class="pre">start</span></code>: The starting position of the core motif within the seqlet, relative to the seqlet’s start coordinate.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">core</span> <span class="pre">motif</span> <span class="pre">end</span></code>: The ending position of the core motif within the seqlet, relative to the seqlet’s start coordinate.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">color</span></code>: An optional field indicating the color used to visualize the seqlet or motif in genome browsers.</p></li>
</ul>
<p>This file can be loaded in genome browsers or used for downstream analyses to visualize or further process the discovered seqlets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>head<span class="w"> </span>example/modisco_subcommands/modisco_neurons.seqlets.bed
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chr1	51514321	51514371	neg_patterns.pattern_3.seqlet_113.EPS15	-0.10036751271172761	-	51514339	51514355	65,105,225
chr1	51514389	51514439	pos_patterns.pattern_1.seqlet_48.EPS15	-0.021704078276343353	-	51514415	51514428	65,105,225
chr1	51514856	51514906	pos_patterns.pattern_4.seqlet_50.EPS15	0.20430506747379695	-	51514858	51514906	65,105,225
chr1	51515007	51515057	pos_patterns.pattern_4.seqlet_36.EPS15	0.11181213204145024	-	51515018	51515066	65,105,225
chr1	51515026	51515076	pos_patterns.pattern_2.seqlet_2.EPS15	-0.39258043596782954	+	51515021	51515067	65,105,225
chr1	51515047	51515097	neg_patterns.pattern_3.seqlet_7.EPS15	-0.5587078251555795	+	51515064	51515080	65,105,225
chr1	51515082	51515132	pos_patterns.pattern_2.seqlet_44.EPS15	0.19132944994089485	-	51515082	51515128	65,105,225
chr1	51515174	51515224	pos_patterns.pattern_4.seqlet_70.EPS15	0.12646874649226447	+	51515175	51515223	65,105,225
chr1	51517345	51517395	pos_patterns.pattern_12.seqlet_38.EPS15	0.08596134528943367	+	51517351	51517387	65,105,225
chr1	51517444	51517494	pos_patterns.pattern_15.seqlet_25.EPS15	0.10319471855837037	+	51517448	51517494	65,105,225
</pre></div>
</div>
</div>
</div>
<p>Tomtom is a computational tool from the MEME Suite that compares discovered motifs (such as those output by MoDISco)
to a database of known motifs, identifying statistically significant matches. This helps annotate de novo motifs
by suggesting potential transcription factors or motif families they correspond to.</p>
<p>In the MoDISco workflow, Tomtom is typically run on the discovered motif position weight matrices (PWMs) to
find similar motifs in reference databases (e.g., JASPAR, HOCOMOCO). The output includes a ranked list of
matches for each motif, along with similarity scores and p-values, which can be used to interpret the biological
relevance of the discovered patterns.</p>
<p>Tomtom can be run via the command line or through integrated pipelines, and its results are often included in
MoDISco reports for downstream analysis and visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>modisco-reports<span class="w"> </span><span class="err">\</span>
    <span class="o">--</span><span class="n">modisco</span><span class="o">-</span><span class="n">h5</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons</span><span class="o">.</span><span class="n">modisco</span><span class="o">.</span><span class="n">h5</span> \
    <span class="o">-</span><span class="n">o</span> <span class="n">example</span><span class="o">/</span><span class="n">modisco_subcommands</span><span class="o">/</span><span class="n">modisco_neurons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Creating modisco logos for pos_patterns: 100%|██| 18/18 [00:25&lt;00:00,  1.40s/it]
Creating modisco logos for neg_patterns: 100%|██| 19/19 [00:37&lt;00:00,  1.98s/it]
Generating patterns dataframe: 100%|█████████████| 2/2 [00:00&lt;00:00, 231.81it/s]
Reading patterns for pos_patterns: 100%|██████| 18/18 [00:00&lt;00:00, 3287.79it/s]
Reading patterns for neg_patterns: 100%|██████| 19/19 [00:00&lt;00:00, 3329.93it/s]
[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done   1 tasks      | elapsed:    7.0s
[Parallel(n_jobs=4)]: Done   2 tasks      | elapsed:    7.2s
[Parallel(n_jobs=4)]: Done   3 tasks      | elapsed:   12.4s
[Parallel(n_jobs=4)]: Done   4 tasks      | elapsed:   13.3s
[Parallel(n_jobs=4)]: Done   5 tasks      | elapsed:   16.0s
[Parallel(n_jobs=4)]: Done   6 tasks      | elapsed:   16.8s
[Parallel(n_jobs=4)]: Done   7 tasks      | elapsed:  1.2min
[Parallel(n_jobs=4)]: Done   8 tasks      | elapsed:  1.2min
[Parallel(n_jobs=4)]: Done   9 tasks      | elapsed:  1.3min
[Parallel(n_jobs=4)]: Done  10 tasks      | elapsed:  1.3min
[Parallel(n_jobs=4)]: Done  11 tasks      | elapsed:  1.4min
[Parallel(n_jobs=4)]: Done  12 tasks      | elapsed:  1.7min
[Parallel(n_jobs=4)]: Done  13 tasks      | elapsed:  1.7min
[Parallel(n_jobs=4)]: Done  14 tasks      | elapsed:  1.7min
[Parallel(n_jobs=4)]: Done  15 tasks      | elapsed:  1.8min
[Parallel(n_jobs=4)]: Done  16 tasks      | elapsed:  1.8min
[Parallel(n_jobs=4)]: Done  17 tasks      | elapsed:  1.8min
[Parallel(n_jobs=4)]: Done  18 tasks      | elapsed:  1.9min
[Parallel(n_jobs=4)]: Done  19 tasks      | elapsed:  1.9min
[Parallel(n_jobs=4)]: Done  20 tasks      | elapsed:  1.9min
[Parallel(n_jobs=4)]: Done  21 tasks      | elapsed:  2.4min
[Parallel(n_jobs=4)]: Done  22 tasks      | elapsed:  2.6min
[Parallel(n_jobs=4)]: Done  23 tasks      | elapsed:  2.9min
[Parallel(n_jobs=4)]: Done  24 tasks      | elapsed:  2.9min
[Parallel(n_jobs=4)]: Done  25 tasks      | elapsed:  3.0min
[Parallel(n_jobs=4)]: Done  26 tasks      | elapsed:  3.0min
[Parallel(n_jobs=4)]: Done  27 tasks      | elapsed:  3.2min
[Parallel(n_jobs=4)]: Done  28 tasks      | elapsed:  3.6min
[Parallel(n_jobs=4)]: Done  29 tasks      | elapsed:  3.6min
[Parallel(n_jobs=4)]: Done  30 tasks      | elapsed:  3.7min
[Parallel(n_jobs=4)]: Done  31 out of  37 | elapsed:  3.7min remaining:   43.2s
[Parallel(n_jobs=4)]: Done  32 out of  37 | elapsed:  3.9min remaining:   36.8s
[Parallel(n_jobs=4)]: Done  33 out of  37 | elapsed:  4.0min remaining:   29.2s
[Parallel(n_jobs=4)]: Done  34 out of  37 | elapsed:  4.0min remaining:   21.3s
[Parallel(n_jobs=4)]: Done  35 out of  37 | elapsed:  4.1min remaining:   14.1s
[Parallel(n_jobs=4)]: Done  37 out of  37 | elapsed:  4.3min finished

</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="python-user-api">
<h2>Python User API<a class="headerlink" href="#python-user-api" title="Link to this heading">¶</a></h2>
<p>MoDISco can also be run programmatically from Python using the Decima API.
The Python API provides more flexibility and allows you to integrate MoDISco directly into your analysis pipelines.</p>
<p>Notably, both the CLI and Python API support the <code class="docutils literal notranslate"><span class="pre">--off-task</span></code> (or <code class="docutils literal notranslate"><span class="pre">off_tasks</span></code> in Python) argument, which enables you to specify a contrasting set of tasks or conditions.
This allows you to perform contrastive motif discovery, identifying patterns that are specific to your target condition (e.g., a particular cell type or tissue)
by comparing against a background or “off” condition. This is useful for highlighting motifs that are enriched or specific to your biological question of interest.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.hub</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_decima_model</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.interpret.modisco</span><span class="w"> </span><span class="kn">import</span> <span class="n">modisco</span>


<span class="n">model</span> <span class="o">=</span> <span class="n">load_decima_model</span><span class="p">(</span><span class="s2">&quot;rep0&quot;</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># or 0, 1, 2, 3, &quot;ensemble&quot;</span>

<span class="n">modisco</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/modisco_python/modisco_fibroblast&quot;</span><span class="p">,</span>
    <span class="n">tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ == &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">off_tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ != &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">top_n_markers</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">max_seqlets_per_metacluster</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
    <span class="n">tss_distance</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.8 (399.4MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1814.6MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1851.5MB/s)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8aa63e2b882d468c9a0ef8d7b851b248", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Computing attributions...: 100%|██████████| 15/15 [00:16&lt;00:00,  1.13s/it]
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1871.4MB/s)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "f8130c94af3a4508b4a3323467862947", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading attributions and sequences...: 100%|██████████| 15/15 [00:00&lt;00:00, 21.31it/s]
2025-10-06 18:11:34,796 - modisco-lite - INFO - Running TFMoDISco version 2.4.0
2025-10-06 18:11:34,797 - modisco-lite - INFO - Extracting seqlets for 15 tasks:
2025-10-06 18:11:34,798 - modisco-lite - INFO - - Smoothing and splitting tracks
2025-10-06 18:11:34,808 - modisco-lite - INFO - - Computing null values with Laplacian null model
2025-10-06 18:11:34,854 - modisco-lite - INFO - - Computing isotonic thresholds
2025-10-06 18:11:34,863 - modisco-lite - INFO - - Refining thresholds
2025-10-06 18:11:34,943 - modisco-lite - INFO - - Extracting seqlets
2025-10-06 18:11:34,953 - modisco-lite - INFO - - Extracting 384 positive seqlets
2025-10-06 18:11:34,956 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 384 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "37cb7d6446064c82b9405af4d00df6fa", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:11:47,047 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 384 seqlets and 384 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "dad57a94f54a4334a6380abe48abfbd7", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "015a8689e88f46e98fe818ab4015d735", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:11:49,025 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 18:11:49,205 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 18:11:49,648 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    6.7s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    6.9s remaining:   48.0s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    6.9s remaining:   29.9s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    7.0s remaining:   21.0s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    7.1s remaining:   15.6s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    7.1s remaining:   11.8s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    7.1s remaining:    9.1s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    7.1s remaining:    7.1s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    7.2s remaining:    5.6s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    7.3s remaining:    4.4s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    7.4s remaining:    3.3s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    7.5s remaining:    2.5s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    7.5s remaining:    1.7s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    7.5s remaining:    1.1s
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:11:57,387 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.02360483048930951
2025-10-06 18:11:57,388 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.02361723253834695
2025-10-06 18:11:57,388 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.02361364332247759
2025-10-06 18:11:57,388 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.02360315002202235
2025-10-06 18:11:57,388 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.023621841682633215
2025-10-06 18:11:57,389 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.023621841682633215
2025-10-06 18:11:57,389 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.023613449615038574
2025-10-06 18:11:57,389 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.02361364332247759
2025-10-06 18:11:57,389 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.023615291366812573
2025-10-06 18:11:57,390 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.02360315002202235
2025-10-06 18:11:57,390 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.02360315002202235
2025-10-06 18:11:57,390 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.023614627921386563
2025-10-06 18:11:57,390 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.02360315002202235
2025-10-06 18:11:57,390 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.02360970287066969
2025-10-06 18:11:57,391 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.0236142640559574
2025-10-06 18:11:57,391 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.023606972537531266
2025-10-06 18:11:57,391 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    7.7s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generating patterns from clusters:: 100%|██████████| 6/6 [00:01&lt;00:00,  3.25it/s]
2025-10-06 18:11:59,240 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 332 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "714129c674814573bd61d460a09e567a", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:11:59,420 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 332 seqlets and 332 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "58fec8c99b984b2eadf1cf717116b54b", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "3b6c2f6ae07a4f97aa842edfc9c13bf0", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:01,732 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 18:12:02,149 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    0.4s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    0.4s remaining:    2.8s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    0.4s remaining:    1.8s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    0.4s remaining:    1.3s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    0.5s remaining:    1.1s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    0.5s remaining:    0.9s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    0.6s remaining:    0.7s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    0.7s remaining:    0.7s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    0.7s remaining:    0.6s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    0.8s remaining:    0.5s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    0.8s remaining:    0.4s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    0.9s remaining:    0.3s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    0.9s remaining:    0.2s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    1.0s remaining:    0.1s
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:03,644 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.026548278897844335
2025-10-06 18:12:03,644 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.026548278897844335
2025-10-06 18:12:03,645 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.026520905736640198
2025-10-06 18:12:03,645 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.026539255688272402
2025-10-06 18:12:03,645 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.026535249101320407
2025-10-06 18:12:03,645 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.026528099158945256
2025-10-06 18:12:03,646 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.026511587839415075
2025-10-06 18:12:03,646 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.026401092795887534
2025-10-06 18:12:03,646 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.02656779223573837
2025-10-06 18:12:03,646 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.02652991091394142
2025-10-06 18:12:03,647 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.02652052727856517
2025-10-06 18:12:03,647 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.026391695973984147
2025-10-06 18:12:03,647 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.026535249101320407
2025-10-06 18:12:03,647 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.026525402801011476
2025-10-06 18:12:03,647 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.026535249101320407
2025-10-06 18:12:03,648 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.026539255688272402
2025-10-06 18:12:03,648 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    1.5s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generating patterns from clusters:: 100%|██████████| 6/6 [00:01&lt;00:00,  3.04it/s]
2025-10-06 18:12:05,624 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|██████████| 6/6 [00:10&lt;00:00,  1.81s/it]
2025-10-06 18:12:17,634 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|██████████| 14/14 [00:00&lt;00:00, 16302.13it/s]
Computing subpatterns:: 100%|██████████| 9/9 [00:00&lt;00:00, 51.30it/s]
2025-10-06 18:12:17,813 - modisco-lite - INFO - - Extracting 340 negative seqlets
2025-10-06 18:12:17,816 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 340 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "9ee3085d0e3c4e06989f099104cb5346", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:18,024 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 340 seqlets and 340 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "b558b7056f4d47f7afcf77faaeef439d", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "ab7a6576e6c34bfc84a58ec0acc4c4e3", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:19,624 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 18:12:19,776 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 18:12:20,101 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    7.3s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    7.3s remaining:   51.4s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    7.4s remaining:   32.0s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    7.4s remaining:   22.2s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    7.4s remaining:   16.3s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    7.4s remaining:   12.4s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    7.5s remaining:    9.6s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    7.5s remaining:    7.5s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    7.7s remaining:    6.0s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    7.7s remaining:    4.6s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    7.7s remaining:    3.5s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    7.8s remaining:    2.6s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    7.8s remaining:    1.8s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    7.8s remaining:    1.1s
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:28,067 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.023850872058533996
2025-10-06 18:12:28,067 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.023768839260606318
2025-10-06 18:12:28,067 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.02385681078512026
2025-10-06 18:12:28,068 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.02387438972443567
2025-10-06 18:12:28,068 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.023876162687244136
2025-10-06 18:12:28,068 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.02386819542812389
2025-10-06 18:12:28,068 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.023585021130662808
2025-10-06 18:12:28,069 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.023854051195930443
2025-10-06 18:12:28,069 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.023659598169893423
2025-10-06 18:12:28,069 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.02378579897047768
2025-10-06 18:12:28,069 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.023863969323479898
2025-10-06 18:12:28,069 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.023795712308034282
2025-10-06 18:12:28,070 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.02387438972443567
2025-10-06 18:12:28,070 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.023852623184908793
2025-10-06 18:12:28,070 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.02387438972443567
2025-10-06 18:12:28,070 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.02377755429507505
2025-10-06 18:12:28,071 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    8.0s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generating patterns from clusters:: 100%|██████████| 5/5 [00:00&lt;00:00, 11.39it/s]
2025-10-06 18:12:28,512 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 283 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "80f124ca5742441887e5af95a6bbbb94", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:28,665 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 283 seqlets and 283 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5f7ce980c4294e479925cb865e83e05c", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "1185c2c90bf74f2783cc975d00595094", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:30,404 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 18:12:30,720 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    0.3s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    0.3s remaining:    1.9s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    0.3s remaining:    1.2s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    0.3s remaining:    0.9s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    0.3s remaining:    0.7s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    0.3s remaining:    0.6s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    0.4s remaining:    0.5s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    0.4s remaining:    0.4s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    0.5s remaining:    0.4s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    0.5s remaining:    0.3s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    0.6s remaining:    0.3s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    0.6s remaining:    0.2s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    0.6s remaining:    0.1s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    0.6s remaining:    0.1s
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:12:31,664 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.026586659803832666
2025-10-06 18:12:31,664 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.026586659803832666
2025-10-06 18:12:31,665 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.02660104643456792
2025-10-06 18:12:31,665 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.026634521581033865
2025-10-06 18:12:31,665 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.026616418993762592
2025-10-06 18:12:31,665 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.026392340322682274
2025-10-06 18:12:31,666 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.026586659803832666
2025-10-06 18:12:31,666 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.026603553780386605
2025-10-06 18:12:31,667 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.02660251974475603
2025-10-06 18:12:31,667 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.026586659803832666
2025-10-06 18:12:31,667 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.02658644850697434
2025-10-06 18:12:31,667 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.02666550122295357
2025-10-06 18:12:31,667 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.026517335933254525
2025-10-06 18:12:31,668 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.026629959060868624
2025-10-06 18:12:31,668 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.026586659803832666
2025-10-06 18:12:31,668 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.026634521581033865
2025-10-06 18:12:31,668 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    0.9s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generating patterns from clusters:: 100%|██████████| 5/5 [00:00&lt;00:00,  7.64it/s]
2025-10-06 18:12:32,325 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|██████████| 5/5 [00:07&lt;00:00,  1.56s/it]
2025-10-06 18:12:41,110 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|██████████| 13/13 [00:00&lt;00:00, 25952.38it/s]
Computing subpatterns:: 100%|██████████| 5/5 [00:00&lt;00:00, 42.82it/s]
Creating modisco logos for pos_patterns: 100%|██████████| 9/9 [00:15&lt;00:00,  1.67s/it]
Creating modisco logos for neg_patterns: 100%|██████████| 5/5 [00:10&lt;00:00,  2.13s/it]
Generating patterns dataframe: 100%|██████████| 2/2 [00:00&lt;00:00, 514.35it/s]
Reading patterns for pos_patterns: 100%|██████████| 9/9 [00:00&lt;00:00, 2967.90it/s]
Reading patterns for neg_patterns: 100%|██████████| 5/5 [00:00&lt;00:00, 3024.01it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    8.2s
[Parallel(n_jobs=8)]: Done   2 out of  14 | elapsed:   13.3s remaining:  1.3min
[Parallel(n_jobs=8)]: Done   3 out of  14 | elapsed:   14.1s remaining:   51.5s
[Parallel(n_jobs=8)]: Done   4 out of  14 | elapsed:   14.4s remaining:   35.9s
[Parallel(n_jobs=8)]: Done   5 out of  14 | elapsed:   15.6s remaining:   28.0s
[Parallel(n_jobs=8)]: Done   6 out of  14 | elapsed:   42.0s remaining:   56.0s
[Parallel(n_jobs=8)]: Done   7 out of  14 | elapsed:   52.6s remaining:   52.6s
[Parallel(n_jobs=8)]: Done   8 out of  14 | elapsed:  1.1min remaining:   50.0s
[Parallel(n_jobs=8)]: Done   9 out of  14 | elapsed:  1.4min remaining:   45.6s
[Parallel(n_jobs=8)]: Done  10 out of  14 | elapsed:  1.4min remaining:   33.1s
[Parallel(n_jobs=8)]: Done  11 out of  14 | elapsed:  1.4min remaining:   22.6s
[Parallel(n_jobs=8)]: Done  12 out of  14 | elapsed:  1.4min remaining:   13.9s
[Parallel(n_jobs=8)]: Done  14 out of  14 | elapsed:  1.5min finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1849.3MB/s)
Processing pos_patterns patterns...: 100%|██████████| 9/9 [00:00&lt;00:00, 594.76it/s]
Processing neg_patterns patterns...: 100%|██████████| 5/5 [00:00&lt;00:00, 621.42it/s]
</pre></div>
</div>
</div>
</div>
<p>The function above consists of the following functions:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.interpret.modisco</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">predict_save_modisco_attributions</span><span class="p">,</span>
    <span class="n">modisco_patterns</span><span class="p">,</span>
    <span class="n">modisco_reports</span><span class="p">,</span>
    <span class="n">modisco_seqlet_bed</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">predict_save_modisco_attributions</span></code> computes and saves attributions for the specified tasks and model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_save_modisco_attributions</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast_0&quot;</span><span class="p">,</span>
    <span class="n">tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ == &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">off_tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ != &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">top_n_markers</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1894.8MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.8 (1742.1MB/s)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "5e93a492fffc456f94bd315b24e56949", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Computing attributions...: 100%|██████████| 15/15 [00:13&lt;00:00,  1.10it/s]
</pre></div>
</div>
</div>
</div>
<p>Lets run it for the second replicate:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predict_save_modisco_attributions</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast_1&quot;</span><span class="p">,</span>
    <span class="n">tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ == &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">off_tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ != &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">top_n_markers</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1835.3MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.8 (1744.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep1:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1415.1MB/s)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "247b6a546d8248a89fde1430940a62e6", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Computing attributions...: 100%|██████████| 15/15 [00:13&lt;00:00,  1.10it/s]
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">AttributionResult</span></code> allows you to load and work with attribution results saved in HDF5 format.
It provides convenient access to gene names, sequences, and attributions for downstream analysis.
<code class="docutils literal notranslate"><span class="pre">agg_func</span></code> specifies how to aggregate attributions across multiple replicates or samples (e.g., ‘mean’, ‘sum’ or None) If None, attributions are not aggregated but retuned as new dimension.
<code class="docutils literal notranslate"><span class="pre">num_workers</span></code> sets the number of parallel worker processes to use for loading and processing data, which can speed up computation.
<code class="docutils literal notranslate"><span class="pre">correct_grad</span></code> determines whether to apply gradient correction to the attributions, which can improve interpretability in some cases.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.interpret.attributions</span><span class="w"> </span><span class="kn">import</span> <span class="n">AttributionResult</span>

<span class="k">with</span> <span class="n">AttributionResult</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast_0.attributions.h5&quot;</span><span class="p">,</span>
        <span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast_1.attributions.h5&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">correct_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">agg_func</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="k">as</span> <span class="n">ar</span><span class="p">:</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">ar</span><span class="o">.</span><span class="n">genes</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Genes: &quot;</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>

    <span class="n">seqs</span><span class="p">,</span> <span class="n">attrs</span> <span class="o">=</span> <span class="n">ar</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">genes</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Seqs: &quot;</span><span class="p">,</span> <span class="n">seqs</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Attrs: &quot;</span><span class="p">,</span> <span class="n">attrs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Genes:  [&#39;ANKRD1&#39;, &#39;CASQ2&#39;, &#39;TBX20&#39;, &#39;MYOZ2&#39;, &#39;HSPB3&#39;, &#39;PLN&#39;, &#39;NPPA&#39;, &#39;SMCO1&#39;, &#39;POPDC2&#39;, &#39;TNNI3&#39;, &#39;MYL7&#39;, &#39;ASB11&#39;, &#39;RBM24&#39;, &#39;NPPB&#39;, &#39;MIR133A1HG&#39;]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading attributions and sequences...: 100%|██████████| 5/5 [00:00&lt;00:00,  7.59it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Seqs:  [[[1. 1. 1. ... 1. 0. 0.]
  [0. 0. 0. ... 0. 1. 1.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[1. 0. 0. ... 1. 0. 0.]
  [0. 1. 1. ... 0. 0. 1.]
  [0. 0. 0. ... 0. 1. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[1. 0. 0. ... 0. 1. 0.]
  [0. 0. 1. ... 0. 0. 0.]
  [0. 1. 0. ... 1. 0. 1.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 1.]
  [0. 1. 0. ... 1. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [1. 0. 1. ... 0. 1. 0.]]

 [[0. 0. 0. ... 1. 0. 0.]
  [0. 0. 1. ... 0. 0. 0.]
  [0. 1. 0. ... 0. 0. 1.]
  [1. 0. 0. ... 0. 1. 0.]]]
Attrs:  [[[ 3.76145852e-06  3.03478555e-06  7.44035424e-06 ... -6.63653460e-04
    4.05797458e-04  3.16523459e-04]
  [ 7.55026849e-06  1.63838381e-05 -2.46401214e-06 ...  1.20408889e-03
   -5.13944813e-04 -1.87091660e-04]
  [-3.95166535e-06  2.12123268e-07 -1.75476748e-05 ... -3.38858253e-04
   -3.90247933e-05  1.54834273e-04]
  [ 1.02410469e-05  7.77581295e-06  4.79289201e-06 ...  2.04334607e-04
    4.03160076e-04  2.12350627e-04]]

 [[-6.97570385e-07 -8.62505237e-06 -1.01505470e-05 ...  3.63982608e-04
   -1.68484687e-04 -1.93401925e-05]
  [ 2.71305908e-07  6.53875031e-06  2.02391934e-05 ... -1.00724553e-04
   -4.64807979e-04  3.70103153e-05]
  [-5.30066410e-06  8.39225697e-06  1.32523191e-05 ... -5.41917134e-04
    1.13812412e-04  1.41916771e-04]
  [-8.00210637e-06 -1.03028333e-05 -1.16753581e-05 ...  1.12934016e-04
    1.06018910e-04 -3.66433733e-05]]

 [[ 6.05896548e-06 -2.37663748e-06  3.92000197e-06 ...  1.28393280e-04
   -4.03863789e-05  3.67814645e-05]
  [-8.24177278e-07 -5.30091552e-06  1.56831456e-05 ...  1.82856207e-04
   -8.50484323e-05  1.96156452e-06]
  [-1.29727255e-06 -5.29163117e-07 -1.32918512e-05 ... -2.37111172e-04
    5.91923515e-05  5.48859017e-05]
  [-3.15958550e-06 -8.84019664e-06  2.15492656e-06 ...  4.88488713e-05
    1.35330674e-04 -8.76247659e-05]]

 [[-1.36014714e-05 -5.10507289e-06  8.32582373e-07 ...  3.51886756e-05
    1.81178373e-05 -4.24103380e-05]
  [ 2.71168137e-05  2.33376031e-05  8.80534026e-06 ... -5.07025675e-05
    5.13899267e-05 -1.40666293e-05]
  [-2.49464001e-05 -6.53428560e-06 -3.16726541e-05 ...  2.62546517e-05
   -5.01153640e-05 -3.02389981e-06]
  [ 1.54220847e-05 -9.85898259e-06  6.44845568e-06 ...  6.47796769e-05
   -1.97882077e-05  1.18930311e-05]]

 [[-5.95685299e-06 -5.18014940e-06  6.04994420e-06 ... -2.10653108e-05
    2.29634907e-05 -9.31743455e-06]
  [ 5.62206424e-06  2.47596017e-06  6.78692459e-06 ...  1.00820489e-05
    1.86815123e-06  4.94928190e-06]
  [-1.69323505e-05  1.77675711e-05  3.42523000e-05 ...  1.01751598e-05
    3.79442545e-05  6.58590238e-06]
  [ 7.81596327e-07 -1.10149963e-05 -8.36851143e-06 ...  2.38808170e-05
   -4.04723414e-05  2.38439338e-05]]]
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">modisco_patterns</span></code> runs modisco and cluster seqlets into patterns:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">modisco_patterns</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast&quot;</span><span class="p">,</span>
    <span class="n">attributions</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast_0.attributions.h5&quot;</span><span class="p">,</span>
        <span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast_1.attributions.h5&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ == &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">off_tasks</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;fibroblast&#39; and organ != &#39;heart&#39;&quot;</span><span class="p">,</span>
    <span class="n">top_n_markers</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">max_seqlets_per_metacluster</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
    <span class="n">tss_distance</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.8 (1769.2MB/s)
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8cba5c8ca6dc4d48bf502c2421c986bd", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading attributions and sequences...: 100%|██████████| 15/15 [00:01&lt;00:00, 10.76it/s]
2025-10-06 18:16:03,703 - modisco-lite - INFO - Running TFMoDISco version 2.4.0
2025-10-06 18:16:03,703 - modisco-lite - INFO - Running TFMoDISco version 2.4.0
2025-10-06 18:16:03,704 - modisco-lite - INFO - Extracting seqlets for 15 tasks:
2025-10-06 18:16:03,704 - modisco-lite - INFO - Extracting seqlets for 15 tasks:
2025-10-06 18:16:03,705 - modisco-lite - INFO - - Smoothing and splitting tracks
2025-10-06 18:16:03,705 - modisco-lite - INFO - - Smoothing and splitting tracks
2025-10-06 18:16:03,716 - modisco-lite - INFO - - Computing null values with Laplacian null model
2025-10-06 18:16:03,716 - modisco-lite - INFO - - Computing null values with Laplacian null model
2025-10-06 18:16:03,762 - modisco-lite - INFO - - Computing isotonic thresholds
2025-10-06 18:16:03,762 - modisco-lite - INFO - - Computing isotonic thresholds
2025-10-06 18:16:03,771 - modisco-lite - INFO - - Refining thresholds
2025-10-06 18:16:03,771 - modisco-lite - INFO - - Refining thresholds
2025-10-06 18:16:03,851 - modisco-lite - INFO - - Extracting seqlets
2025-10-06 18:16:03,851 - modisco-lite - INFO - - Extracting seqlets
2025-10-06 18:16:03,863 - modisco-lite - INFO - - Extracting 454 positive seqlets
2025-10-06 18:16:03,863 - modisco-lite - INFO - - Extracting 454 positive seqlets
2025-10-06 18:16:03,866 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 454 seqlets
2025-10-06 18:16:03,866 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 454 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "dcd3b8ebdf454718b6bdbb3f794b5fe7", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:04,194 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 454 seqlets and 454 neighbors
2025-10-06 18:16:04,194 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 454 seqlets and 454 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c4a371939c0c45be90f1bc849637dc5c", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "3fedc310d36e41ae8ad95bd14d41cf26", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:06,868 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 18:16:06,868 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 18:16:07,102 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 18:16:07,102 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 18:16:07,782 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
2025-10-06 18:16:07,782 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:   12.2s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:   12.2s remaining:  1.4min
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:   12.2s remaining:   52.8s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:   12.3s remaining:   36.8s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:   12.3s remaining:   27.2s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:   12.4s remaining:   20.7s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:   12.4s remaining:   16.0s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:   12.5s remaining:   12.5s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:   12.8s remaining:   10.0s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:   13.0s remaining:    7.8s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:   13.1s remaining:    6.0s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:   13.2s remaining:    4.4s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:   13.2s remaining:    3.1s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:   13.4s remaining:    1.9s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:   13.5s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:21,397 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.02092204286706559
2025-10-06 18:16:21,397 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.02092204286706559
2025-10-06 18:16:21,398 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.020929666377409343
2025-10-06 18:16:21,398 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.020929666377409343
2025-10-06 18:16:21,398 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.020363825585150153
2025-10-06 18:16:21,398 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.020363825585150153
2025-10-06 18:16:21,398 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.020905953578916604
2025-10-06 18:16:21,398 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.020905953578916604
2025-10-06 18:16:21,399 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.020790482063174655
2025-10-06 18:16:21,399 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.020790482063174655
2025-10-06 18:16:21,399 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.02092843901868041
2025-10-06 18:16:21,399 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.02092843901868041
2025-10-06 18:16:21,400 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.020802270889506792
2025-10-06 18:16:21,400 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.020802270889506792
2025-10-06 18:16:21,400 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.020795473460761105
2025-10-06 18:16:21,400 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.020795473460761105
2025-10-06 18:16:21,401 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.02079413413859788
2025-10-06 18:16:21,401 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.02079413413859788
2025-10-06 18:16:21,401 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.020929666377409343
2025-10-06 18:16:21,401 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.020929666377409343
2025-10-06 18:16:21,402 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.020811249046675816
2025-10-06 18:16:21,402 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.020811249046675816
2025-10-06 18:16:21,402 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.020873329768449297
2025-10-06 18:16:21,402 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.020873329768449297
2025-10-06 18:16:21,403 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.020928540590008378
2025-10-06 18:16:21,403 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.020928540590008378
2025-10-06 18:16:21,403 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.020797440246157855
2025-10-06 18:16:21,403 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.020797440246157855
2025-10-06 18:16:21,403 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.02092204286706559
2025-10-06 18:16:21,403 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.02092204286706559
2025-10-06 18:16:21,404 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.020923814801806766
2025-10-06 18:16:21,404 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.020923814801806766
2025-10-06 18:16:21,404 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
2025-10-06 18:16:21,404 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
Generating patterns from clusters:: 100%|██████████| 5/5 [00:02&lt;00:00,  2.34it/s]
2025-10-06 18:16:23,545 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 407 seqlets
2025-10-06 18:16:23,545 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 407 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "0f1d86d10d1e48bb9ce27b29f10eeeb9", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:23,788 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 407 seqlets and 407 neighbors
2025-10-06 18:16:23,788 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 407 seqlets and 407 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "0ebc915da7884d8f99e3d6fb067e4896", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "8955f0ef77ee47b0ba5af658db3c65e8", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:27,265 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 18:16:27,265 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 18:16:27,911 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
2025-10-06 18:16:27,911 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    0.5s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    0.6s remaining:    3.9s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    0.6s remaining:    2.8s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    0.7s remaining:    2.1s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    0.7s remaining:    1.5s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    0.7s remaining:    1.2s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    0.8s remaining:    1.0s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    1.0s remaining:    1.0s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    1.2s remaining:    0.9s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    1.2s remaining:    0.7s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    1.4s remaining:    0.6s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    1.6s remaining:    0.5s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    1.7s remaining:    0.4s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    1.8s remaining:    0.3s
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:29,979 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.025002069983353303
2025-10-06 18:16:29,979 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.025002069983353303
2025-10-06 18:16:29,980 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.02508670994886661
2025-10-06 18:16:29,980 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.02508670994886661
2025-10-06 18:16:29,981 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.024970708770517408
2025-10-06 18:16:29,981 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.024970708770517408
2025-10-06 18:16:29,981 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.02513240247110472
2025-10-06 18:16:29,981 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.02513240247110472
2025-10-06 18:16:29,982 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.02514310327023221
2025-10-06 18:16:29,982 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.02514310327023221
2025-10-06 18:16:29,982 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.02499615731307966
2025-10-06 18:16:29,982 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.02499615731307966
2025-10-06 18:16:29,983 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.025149450184890565
2025-10-06 18:16:29,983 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.025149450184890565
2025-10-06 18:16:29,983 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.02508670994886661
2025-10-06 18:16:29,983 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.02508670994886661
2025-10-06 18:16:29,984 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.024964847410963064
2025-10-06 18:16:29,984 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.024964847410963064
2025-10-06 18:16:29,985 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.025146262320554554
2025-10-06 18:16:29,985 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.025146262320554554
2025-10-06 18:16:29,985 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.024962589681733253
2025-10-06 18:16:29,985 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.024962589681733253
2025-10-06 18:16:29,986 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.02508628049478435
2025-10-06 18:16:29,986 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.02508628049478435
2025-10-06 18:16:29,986 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.02513924665524891
2025-10-06 18:16:29,986 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.02513924665524891
2025-10-06 18:16:29,987 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.025084922401764815
2025-10-06 18:16:29,987 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.025084922401764815
2025-10-06 18:16:29,987 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.025131177812629434
2025-10-06 18:16:29,987 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.025131177812629434
2025-10-06 18:16:29,988 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.02499565203161923
2025-10-06 18:16:29,988 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.02499565203161923
2025-10-06 18:16:29,988 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
2025-10-06 18:16:29,988 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    2.0s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generating patterns from clusters:: 100%|██████████| 6/6 [00:02&lt;00:00,  2.59it/s]
2025-10-06 18:16:32,307 - modisco-lite - INFO - Detecting spurious merging of patterns
2025-10-06 18:16:32,307 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|██████████| 5/5 [00:07&lt;00:00,  1.59s/it]
2025-10-06 18:16:41,865 - modisco-lite - INFO - Filtering and merging patterns
2025-10-06 18:16:41,865 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|██████████| 11/11 [00:00&lt;00:00, 13007.43it/s]
Computing subpatterns:: 100%|██████████| 9/9 [00:00&lt;00:00, 38.31it/s]
2025-10-06 18:16:42,104 - modisco-lite - INFO - - Extracting 329 negative seqlets
2025-10-06 18:16:42,104 - modisco-lite - INFO - - Extracting 329 negative seqlets
2025-10-06 18:16:42,107 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 329 seqlets
2025-10-06 18:16:42,107 - modisco-lite - INFO - - Round 0: Generating coarse resolution affinity matrix for 329 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8985cc5d207a4055bf042d80eaa613d0", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:42,352 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 329 seqlets and 329 neighbors
2025-10-06 18:16:42,352 - modisco-lite - INFO - - Round 0: Generating fine resolution affinity matrix for 329 seqlets and 329 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "b161f4e6fabb4e52a2c7224c53f6ee91", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "b0ceb430551f41718a3745b90b1dfbd1", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:43,824 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 18:16:43,824 - modisco-lite - INFO - - Round 0: Filtering seqlets by correlation
2025-10-06 18:16:43,971 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 18:16:43,971 - modisco-lite - INFO - - Round 0: Density adaptation
2025-10-06 18:16:44,298 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
2025-10-06 18:16:44,298 - modisco-lite - INFO - - Round 0: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    7.2s
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    7.4s remaining:   51.5s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    7.4s remaining:   32.2s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    7.5s remaining:   22.6s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    7.6s remaining:   16.7s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    7.6s remaining:   12.6s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    7.6s remaining:    9.8s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    7.6s remaining:    7.6s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    7.7s remaining:    6.0s
[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    7.9s remaining:    4.7s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    7.9s remaining:    3.6s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    8.0s remaining:    2.7s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    8.0s remaining:    1.8s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    8.0s remaining:    1.1s
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:52,494 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.026763578635124848
2025-10-06 18:16:52,494 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.026763578635124848
2025-10-06 18:16:52,495 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.0265094500699511
2025-10-06 18:16:52,495 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.0265094500699511
2025-10-06 18:16:52,496 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.02654428903412237
2025-10-06 18:16:52,496 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.02654428903412237
2025-10-06 18:16:52,496 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.026721538949614872
2025-10-06 18:16:52,496 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.026721538949614872
2025-10-06 18:16:52,497 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.026735509288973492
2025-10-06 18:16:52,497 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.026735509288973492
2025-10-06 18:16:52,497 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.02667227717530735
2025-10-06 18:16:52,497 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.02667227717530735
2025-10-06 18:16:52,498 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.02657103680991128
2025-10-06 18:16:52,498 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.02657103680991128
2025-10-06 18:16:52,498 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.026660907084892486
2025-10-06 18:16:52,498 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.026660907084892486
2025-10-06 18:16:52,499 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.026594784402216663
2025-10-06 18:16:52,499 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.026594784402216663
2025-10-06 18:16:52,499 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.0265498166549464
2025-10-06 18:16:52,499 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.0265498166549464
2025-10-06 18:16:52,500 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.026736134589607066
2025-10-06 18:16:52,500 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.026736134589607066
2025-10-06 18:16:52,500 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.026543460372326553
2025-10-06 18:16:52,500 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.026543460372326553
2025-10-06 18:16:52,500 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.026605394300180895
2025-10-06 18:16:52,500 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.026605394300180895
2025-10-06 18:16:52,501 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.02666441289851534
2025-10-06 18:16:52,501 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.02666441289851534
2025-10-06 18:16:52,501 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.026607382962757012
2025-10-06 18:16:52,501 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.026607382962757012
2025-10-06 18:16:52,502 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.026671836351827943
2025-10-06 18:16:52,502 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.026671836351827943
2025-10-06 18:16:52,502 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
2025-10-06 18:16:52,502 - modisco-lite - INFO - - Round 0: Generating patterns from clusters
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    8.2s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generating patterns from clusters:: 100%|██████████| 6/6 [00:01&lt;00:00,  4.27it/s]
2025-10-06 18:16:53,909 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 232 seqlets
2025-10-06 18:16:53,909 - modisco-lite - INFO - - Round 1: Generating coarse resolution affinity matrix for 232 seqlets
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "46af569105d6453c8326381f66292057", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:54,025 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 232 seqlets and 232 neighbors
2025-10-06 18:16:54,025 - modisco-lite - INFO - - Round 1: Generating fine resolution affinity matrix for 232 seqlets and 232 neighbors
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d41cbdd7c77b4d5a9e6245f483a4d541", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "6943545224ca448487be97135b84be36", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:55,143 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 18:16:55,143 - modisco-lite - INFO - - Round 1: Density adaptation
2025-10-06 18:16:55,363 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
2025-10-06 18:16:55,363 - modisco-lite - INFO - - Round 1: Clustering with Leiden algorithm
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    0.2s
[Parallel(n_jobs=8)]: Batch computation too fast (0.19564294815063477s.) Setting batch_size=2.
[Parallel(n_jobs=8)]: Done   2 out of  16 | elapsed:    0.2s remaining:    1.4s
[Parallel(n_jobs=8)]: Done   3 out of  16 | elapsed:    0.2s remaining:    0.9s
[Parallel(n_jobs=8)]: Done   4 out of  16 | elapsed:    0.2s remaining:    0.6s
[Parallel(n_jobs=8)]: Done   5 out of  16 | elapsed:    0.2s remaining:    0.5s
[Parallel(n_jobs=8)]: Done   6 out of  16 | elapsed:    0.2s remaining:    0.4s
[Parallel(n_jobs=8)]: Done   7 out of  16 | elapsed:    0.2s remaining:    0.3s
[Parallel(n_jobs=8)]: Done   8 out of  16 | elapsed:    0.3s remaining:    0.3s
[Parallel(n_jobs=8)]: Done   9 out of  16 | elapsed:    0.4s remaining:    0.3s
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2025-10-06 18:16:55,864 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.036596707967135385
2025-10-06 18:16:55,864 - modisco-lite - INFO - Leiden clustering quality for seed 0: 0.036596707967135385
2025-10-06 18:16:55,865 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.036607288086638164
2025-10-06 18:16:55,865 - modisco-lite - INFO - Leiden clustering quality for seed 1: 0.036607288086638164
2025-10-06 18:16:55,866 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.03657606307660279
2025-10-06 18:16:55,866 - modisco-lite - INFO - Leiden clustering quality for seed 2: 0.03657606307660279
2025-10-06 18:16:55,866 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.03660305121676245
2025-10-06 18:16:55,866 - modisco-lite - INFO - Leiden clustering quality for seed 3: 0.03660305121676245
2025-10-06 18:16:55,867 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.036607288086638164
2025-10-06 18:16:55,867 - modisco-lite - INFO - Leiden clustering quality for seed 4: 0.036607288086638164
2025-10-06 18:16:55,867 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.03657606307660279
2025-10-06 18:16:55,867 - modisco-lite - INFO - Leiden clustering quality for seed 5: 0.03657606307660279
2025-10-06 18:16:55,868 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.036607288086638164
2025-10-06 18:16:55,868 - modisco-lite - INFO - Leiden clustering quality for seed 6: 0.036607288086638164
2025-10-06 18:16:55,868 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.03656872531629366
2025-10-06 18:16:55,868 - modisco-lite - INFO - Leiden clustering quality for seed 7: 0.03656872531629366
2025-10-06 18:16:55,870 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.03659722894571927
2025-10-06 18:16:55,870 - modisco-lite - INFO - Leiden clustering quality for seed 8: 0.03659722894571927
2025-10-06 18:16:55,870 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.03660305121676245
2025-10-06 18:16:55,870 - modisco-lite - INFO - Leiden clustering quality for seed 9: 0.03660305121676245
2025-10-06 18:16:55,871 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.036566275851921744
2025-10-06 18:16:55,871 - modisco-lite - INFO - Leiden clustering quality for seed 10: 0.036566275851921744
2025-10-06 18:16:55,872 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.03657606307660279
2025-10-06 18:16:55,872 - modisco-lite - INFO - Leiden clustering quality for seed 11: 0.03657606307660279
2025-10-06 18:16:55,872 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.036607288086638164
2025-10-06 18:16:55,872 - modisco-lite - INFO - Leiden clustering quality for seed 12: 0.036607288086638164
2025-10-06 18:16:55,873 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.036566275851921744
2025-10-06 18:16:55,873 - modisco-lite - INFO - Leiden clustering quality for seed 13: 0.036566275851921744
2025-10-06 18:16:55,873 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.036607288086638164
2025-10-06 18:16:55,873 - modisco-lite - INFO - Leiden clustering quality for seed 14: 0.036607288086638164
2025-10-06 18:16:55,874 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.036579117967440684
2025-10-06 18:16:55,874 - modisco-lite - INFO - Leiden clustering quality for seed 15: 0.036579117967440684
2025-10-06 18:16:55,874 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
2025-10-06 18:16:55,874 - modisco-lite - INFO - - Round 1: Generating patterns from clusters
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Done  10 out of  16 | elapsed:    0.4s remaining:    0.2s
[Parallel(n_jobs=8)]: Done  11 out of  16 | elapsed:    0.4s remaining:    0.2s
[Parallel(n_jobs=8)]: Done  12 out of  16 | elapsed:    0.4s remaining:    0.1s
[Parallel(n_jobs=8)]: Done  13 out of  16 | elapsed:    0.5s remaining:    0.1s
[Parallel(n_jobs=8)]: Done  14 out of  16 | elapsed:    0.5s remaining:    0.1s
[Parallel(n_jobs=8)]: Done  16 out of  16 | elapsed:    0.5s finished
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generating patterns from clusters:: 100%|██████████| 6/6 [00:00&lt;00:00,  6.97it/s]
2025-10-06 18:16:56,737 - modisco-lite - INFO - Detecting spurious merging of patterns
2025-10-06 18:16:56,737 - modisco-lite - INFO - Detecting spurious merging of patterns
Detecting spurious merging of patterns:: 100%|██████████| 5/5 [00:07&lt;00:00,  1.51s/it]
2025-10-06 18:17:05,761 - modisco-lite - INFO - Filtering and merging patterns
2025-10-06 18:17:05,761 - modisco-lite - INFO - Filtering and merging patterns
Filtering patterns:: 100%|██████████| 10/10 [00:00&lt;00:00, 17534.72it/s]
Computing subpatterns:: 100%|██████████| 6/6 [00:00&lt;00:00, 51.68it/s]
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">modisco_reports</span></code> runs tomtom and annotated patterns with now motifs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">modisco_reports</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast&quot;</span><span class="p">,</span>
    <span class="n">modisco_h5</span><span class="o">=</span><span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast.modisco.h5&quot;</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Creating modisco logos for pos_patterns: 100%|██████████| 9/9 [00:16&lt;00:00,  1.86s/it]
Creating modisco logos for neg_patterns: 100%|██████████| 6/6 [00:15&lt;00:00,  2.62s/it]
Generating patterns dataframe: 100%|██████████| 2/2 [00:00&lt;00:00, 457.69it/s]
Reading patterns for pos_patterns: 100%|██████████| 9/9 [00:00&lt;00:00, 2909.79it/s]
Reading patterns for neg_patterns: 100%|██████████| 6/6 [00:00&lt;00:00, 2932.74it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Parallel(n_jobs=8)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done   1 tasks      | elapsed:    8.0s
[Parallel(n_jobs=8)]: Done   2 out of  15 | elapsed:   15.0s remaining:  1.6min
[Parallel(n_jobs=8)]: Done   3 out of  15 | elapsed:   15.7s remaining:  1.0min
[Parallel(n_jobs=8)]: Done   4 out of  15 | elapsed:  1.0min remaining:  2.8min
[Parallel(n_jobs=8)]: Done   5 out of  15 | elapsed:  1.2min remaining:  2.4min
[Parallel(n_jobs=8)]: Done   6 out of  15 | elapsed:  1.2min remaining:  1.9min
[Parallel(n_jobs=8)]: Done   7 out of  15 | elapsed:  1.3min remaining:  1.5min
[Parallel(n_jobs=8)]: Done   8 out of  15 | elapsed:  1.4min remaining:  1.2min
[Parallel(n_jobs=8)]: Done   9 out of  15 | elapsed:  1.4min remaining:   55.7s
[Parallel(n_jobs=8)]: Done  10 out of  15 | elapsed:  1.4min remaining:   42.0s
[Parallel(n_jobs=8)]: Done  11 out of  15 | elapsed:  1.5min remaining:   33.4s
[Parallel(n_jobs=8)]: Done  12 out of  15 | elapsed:  1.9min remaining:   28.5s
[Parallel(n_jobs=8)]: Done  13 out of  15 | elapsed:  2.0min remaining:   18.1s
[Parallel(n_jobs=8)]: Done  15 out of  15 | elapsed:  2.4min finished
</pre></div>
</div>
</div>
</div>
<p>Seqlet bed extract seqlets from modisco output:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">modisco_seqlet_bed</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="o">=</span><span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast&quot;</span><span class="p">,</span>
    <span class="n">modisco_h5</span><span class="o">=</span><span class="s2">&quot;example/modisco_python_subcommands/modisco_fibroblast.modisco.h5&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.7 (1840.6MB/s)
Processing pos_patterns patterns...: 100%|██████████| 9/9 [00:00&lt;00:00, 559.61it/s]
Processing neg_patterns patterns...: 100%|██████████| 6/6 [00:00&lt;00:00, 599.99it/s]
</pre></div>
</div>
</div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="5-gene-expression-prediction.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Predicting Gene Expression with Decima</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="3-finetune.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Fine-tuning Borzoi to create a Decima model</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Gokcen Eraslan
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Detecting and Clustering Seqlets with Modisco</a><ul>
<li><a class="reference internal" href="#cli-api">CLI API</a><ul>
<li><a class="reference internal" href="#cli-subcommands">CLI Subcommands</a></li>
</ul>
</li>
<li><a class="reference internal" href="#python-user-api">Python User API</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=68a6d85b"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>