<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html"><link rel="search" title="Search" href="../search.html"><link rel="next" title="Contributing" href="../contributing.html"><link rel="prev" title="Detecting and Clustering Seqlets with Modisco" href="4-modisco.html">

    <!-- Generated with Sphinx 8.2.3 and Furo 2025.09.25 -->
        <title>Predicting Gene Expression with Decima - decima 0.0.post1.dev1+ga9d6eec85 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=580074bf" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">decima 0.0.post1.dev1+ga9d6eec85 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <span class="sidebar-brand-text">decima 0.0.post1.dev1+ga9d6eec85 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/decima.html">decima package</a><input aria-label="Toggle navigation of decima package" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.cli.html">decima.cli package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.core.html">decima.core package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.data.html">decima.data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.hub.html">decima.hub package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.interpret.html">decima.interpret package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.model.html">decima.model package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.plot.html">decima.plot package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.tools.html">decima.tools package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.train.html">decima.train package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.utils.html">decima.utils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/decima.vep.html">decima.vep package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="1-attribution-motif-discovery.html">Tutorial: Attribution Motif Discovery</a></li>
<li class="toctree-l1"><a class="reference internal" href="2-variant-effect-prediction.html">Tutorial: Variant Effect Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="3-finetune.html">Tutorial: Finetuning</a></li>
<li class="toctree-l1"><a class="reference internal" href="4-modisco.html">Tutorial: Modisco</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Tutorial: Gene Expression Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/5-gene-expression-prediction.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section class="tex2jax_ignore mathjax_ignore" id="predicting-gene-expression-with-decima">
<h1>Predicting Gene Expression with Decima<a class="headerlink" href="#predicting-gene-expression-with-decima" title="Link to this heading">¶</a></h1>
<p>Decima allows prediction of gene expression at the cell type level, and this tutorial demonstrates how to leverage the prediction API for both genes in the training data and custom genes.</p>
<section id="precomputed-predictions">
<h2>Precomputed Predictions<a class="headerlink" href="#precomputed-predictions" title="Link to this heading">¶</a></h2>
<p>Scores for all genes in the training data are precomputed and saved to metadata h5ad object for each model replicate and are available under the <code class="docutils literal notranslate"><span class="pre">DecimaResult</span></code> class. <code class="docutils literal notranslate"><span class="pre">predicted_expression_matrix</span></code> class returns predicted average gene expression across the replicates.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima</span><span class="w"> </span><span class="kn">import</span> <span class="n">DecimaResult</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">DecimaResult</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="n">result</span><span class="o">.</span><span class="n">predicted_expression_matrix</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/celikm5/miniforge3/envs/decima2/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The &#39;repr&#39; attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. &#39;repr&#39; is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.
  warnings.warn(
/home/celikm5/miniforge3/envs/decima2/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The &#39;frozen&#39; attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. &#39;frozen&#39; is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.
  warnings.warn(
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact &#39;metadata:latest&#39;, 3122.32MB. 1 files...
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 00:00:05.9 (533.2MB/s)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>STRADA</th>
      <th>ETV4</th>
      <th>USP25</th>
      <th>ZSWIM5</th>
      <th>C21orf58</th>
      <th>MIR497HG</th>
      <th>CFAP74</th>
      <th>GSE1</th>
      <th>LPP</th>
      <th>CLK1</th>
      <th>...</th>
      <th>STRIP2</th>
      <th>TNFRSF1A</th>
      <th>RBM14-RBM4</th>
      <th>C1orf21</th>
      <th>LINC00639</th>
      <th>NPDC1</th>
      <th>ZNF425</th>
      <th>COL5A1</th>
      <th>BRD3</th>
      <th>EVI5L</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>2.973438</td>
      <td>1.845565</td>
      <td>4.592531</td>
      <td>5.099802</td>
      <td>1.774879</td>
      <td>0.356812</td>
      <td>2.590836</td>
      <td>4.629774</td>
      <td>4.897171</td>
      <td>3.326940</td>
      <td>...</td>
      <td>2.836060</td>
      <td>0.297015</td>
      <td>1.883849</td>
      <td>4.293593</td>
      <td>1.463565</td>
      <td>3.183534</td>
      <td>2.340202</td>
      <td>2.374942</td>
      <td>2.911916</td>
      <td>3.230072</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>2.954213</td>
      <td>1.896726</td>
      <td>4.688557</td>
      <td>5.510440</td>
      <td>1.666929</td>
      <td>0.352725</td>
      <td>2.292625</td>
      <td>4.459535</td>
      <td>4.915286</td>
      <td>3.192858</td>
      <td>...</td>
      <td>3.125704</td>
      <td>0.242543</td>
      <td>1.908177</td>
      <td>4.439424</td>
      <td>1.236739</td>
      <td>3.494824</td>
      <td>2.425672</td>
      <td>2.054568</td>
      <td>2.713408</td>
      <td>3.491463</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>2.938851</td>
      <td>2.197247</td>
      <td>4.861410</td>
      <td>5.617520</td>
      <td>1.773381</td>
      <td>0.380867</td>
      <td>2.394917</td>
      <td>4.415038</td>
      <td>4.836399</td>
      <td>3.390717</td>
      <td>...</td>
      <td>3.082098</td>
      <td>0.263285</td>
      <td>2.006456</td>
      <td>4.383455</td>
      <td>1.208590</td>
      <td>4.013819</td>
      <td>2.408381</td>
      <td>2.297343</td>
      <td>2.892222</td>
      <td>3.695785</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>3.045972</td>
      <td>2.138573</td>
      <td>4.863791</td>
      <td>5.273604</td>
      <td>1.760097</td>
      <td>0.463555</td>
      <td>2.391702</td>
      <td>3.940975</td>
      <td>4.857763</td>
      <td>3.410926</td>
      <td>...</td>
      <td>2.882890</td>
      <td>0.290327</td>
      <td>1.922963</td>
      <td>4.550189</td>
      <td>1.430520</td>
      <td>3.693118</td>
      <td>2.297103</td>
      <td>2.121887</td>
      <td>2.626117</td>
      <td>3.223912</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>3.025518</td>
      <td>2.019096</td>
      <td>4.602948</td>
      <td>5.257001</td>
      <td>1.755338</td>
      <td>0.382190</td>
      <td>2.432810</td>
      <td>4.392480</td>
      <td>4.959488</td>
      <td>3.250500</td>
      <td>...</td>
      <td>3.082296</td>
      <td>0.258540</td>
      <td>2.038277</td>
      <td>4.464807</td>
      <td>1.249043</td>
      <td>3.665800</td>
      <td>2.400820</td>
      <td>2.255862</td>
      <td>2.925619</td>
      <td>3.471005</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>2.333562</td>
      <td>0.633322</td>
      <td>4.675825</td>
      <td>2.793023</td>
      <td>0.752030</td>
      <td>0.692083</td>
      <td>0.503531</td>
      <td>4.327948</td>
      <td>6.903193</td>
      <td>3.695593</td>
      <td>...</td>
      <td>0.549795</td>
      <td>2.270181</td>
      <td>1.563218</td>
      <td>4.395422</td>
      <td>0.550088</td>
      <td>1.330252</td>
      <td>1.044471</td>
      <td>3.759369</td>
      <td>2.491346</td>
      <td>1.872717</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>0.835037</td>
      <td>0.358773</td>
      <td>1.964896</td>
      <td>0.307449</td>
      <td>0.337240</td>
      <td>0.834196</td>
      <td>0.093885</td>
      <td>1.853794</td>
      <td>3.700790</td>
      <td>4.467302</td>
      <td>...</td>
      <td>0.176885</td>
      <td>1.370898</td>
      <td>1.022708</td>
      <td>3.400267</td>
      <td>0.052162</td>
      <td>1.908870</td>
      <td>0.253417</td>
      <td>1.448111</td>
      <td>1.622033</td>
      <td>1.064292</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>3.008039</td>
      <td>1.209324</td>
      <td>4.798392</td>
      <td>3.931870</td>
      <td>1.401328</td>
      <td>1.638555</td>
      <td>0.969720</td>
      <td>4.779201</td>
      <td>6.631931</td>
      <td>4.127797</td>
      <td>...</td>
      <td>1.174298</td>
      <td>1.870530</td>
      <td>2.506874</td>
      <td>5.151776</td>
      <td>0.967644</td>
      <td>1.809947</td>
      <td>2.205356</td>
      <td>4.244005</td>
      <td>2.974467</td>
      <td>2.659873</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>1.241936</td>
      <td>0.455059</td>
      <td>2.919995</td>
      <td>0.571672</td>
      <td>0.486448</td>
      <td>1.175586</td>
      <td>0.145397</td>
      <td>2.412148</td>
      <td>4.759118</td>
      <td>4.913945</td>
      <td>...</td>
      <td>0.371035</td>
      <td>1.361073</td>
      <td>1.668085</td>
      <td>4.005738</td>
      <td>0.078611</td>
      <td>1.571750</td>
      <td>0.508187</td>
      <td>2.067150</td>
      <td>2.323764</td>
      <td>1.429850</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>1.715507</td>
      <td>0.700955</td>
      <td>3.044732</td>
      <td>0.858696</td>
      <td>0.903406</td>
      <td>1.763168</td>
      <td>0.215304</td>
      <td>2.604478</td>
      <td>4.549708</td>
      <td>4.839124</td>
      <td>...</td>
      <td>0.594310</td>
      <td>1.801298</td>
      <td>2.075996</td>
      <td>3.933860</td>
      <td>0.165590</td>
      <td>1.970268</td>
      <td>0.993521</td>
      <td>2.232347</td>
      <td>2.473388</td>
      <td>1.902884</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 18457 columns</p>
</div></div></div>
</div>
<p>To access the predicted expression matrix for a specific model, you can use the <code class="docutils literal notranslate"><span class="pre">model_name</span></code> parameter. In this example, we obtain the predicted gene expression for first model replicate.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">predicted_expression_matrix</span><span class="p">(</span><span class="n">model_name</span><span class="o">=</span><span class="s2">&quot;v1_rep0&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>STRADA</th>
      <th>ETV4</th>
      <th>USP25</th>
      <th>ZSWIM5</th>
      <th>C21orf58</th>
      <th>MIR497HG</th>
      <th>CFAP74</th>
      <th>GSE1</th>
      <th>LPP</th>
      <th>CLK1</th>
      <th>...</th>
      <th>STRIP2</th>
      <th>TNFRSF1A</th>
      <th>RBM14-RBM4</th>
      <th>C1orf21</th>
      <th>LINC00639</th>
      <th>NPDC1</th>
      <th>ZNF425</th>
      <th>COL5A1</th>
      <th>BRD3</th>
      <th>EVI5L</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>2.932758</td>
      <td>2.020476</td>
      <td>4.795636</td>
      <td>5.142194</td>
      <td>1.721651</td>
      <td>0.403188</td>
      <td>3.207011</td>
      <td>4.647735</td>
      <td>4.823714</td>
      <td>3.241457</td>
      <td>...</td>
      <td>1.804592</td>
      <td>0.392356</td>
      <td>1.923160</td>
      <td>4.256888</td>
      <td>1.837033</td>
      <td>3.382362</td>
      <td>2.358378</td>
      <td>2.471895</td>
      <td>2.808940</td>
      <td>3.232095</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>2.816624</td>
      <td>1.938934</td>
      <td>4.917684</td>
      <td>5.402302</td>
      <td>1.319163</td>
      <td>0.393221</td>
      <td>2.809423</td>
      <td>4.842717</td>
      <td>4.939364</td>
      <td>2.980938</td>
      <td>...</td>
      <td>1.753353</td>
      <td>0.307677</td>
      <td>2.096857</td>
      <td>4.221365</td>
      <td>1.679307</td>
      <td>3.917950</td>
      <td>2.459049</td>
      <td>2.356942</td>
      <td>2.696189</td>
      <td>3.447138</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>2.742950</td>
      <td>2.394729</td>
      <td>4.805934</td>
      <td>5.102471</td>
      <td>1.731927</td>
      <td>0.452663</td>
      <td>2.778222</td>
      <td>4.538105</td>
      <td>4.443571</td>
      <td>3.041557</td>
      <td>...</td>
      <td>1.804225</td>
      <td>0.364422</td>
      <td>2.145009</td>
      <td>4.004273</td>
      <td>1.471316</td>
      <td>4.249442</td>
      <td>2.204638</td>
      <td>2.687080</td>
      <td>2.793549</td>
      <td>3.760900</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>2.804515</td>
      <td>2.358393</td>
      <td>4.676318</td>
      <td>5.462350</td>
      <td>1.543962</td>
      <td>0.483929</td>
      <td>2.695394</td>
      <td>4.411092</td>
      <td>4.355472</td>
      <td>3.220853</td>
      <td>...</td>
      <td>1.663766</td>
      <td>0.429275</td>
      <td>2.065255</td>
      <td>4.045637</td>
      <td>1.837513</td>
      <td>3.879115</td>
      <td>2.242636</td>
      <td>2.191440</td>
      <td>2.614539</td>
      <td>3.287658</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>2.815917</td>
      <td>2.098659</td>
      <td>4.594968</td>
      <td>4.648893</td>
      <td>1.517963</td>
      <td>0.426446</td>
      <td>3.173781</td>
      <td>4.829787</td>
      <td>4.389949</td>
      <td>3.011941</td>
      <td>...</td>
      <td>2.119307</td>
      <td>0.320393</td>
      <td>2.064416</td>
      <td>3.868720</td>
      <td>1.499770</td>
      <td>4.144956</td>
      <td>2.074824</td>
      <td>2.375941</td>
      <td>2.745383</td>
      <td>3.339186</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>2.510139</td>
      <td>0.811121</td>
      <td>5.367220</td>
      <td>2.944120</td>
      <td>0.668071</td>
      <td>0.898573</td>
      <td>0.737285</td>
      <td>4.750776</td>
      <td>7.324188</td>
      <td>3.585325</td>
      <td>...</td>
      <td>0.258857</td>
      <td>1.437364</td>
      <td>1.913985</td>
      <td>4.471622</td>
      <td>0.823436</td>
      <td>1.486110</td>
      <td>1.288418</td>
      <td>3.211422</td>
      <td>2.712792</td>
      <td>1.665431</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>1.245692</td>
      <td>0.397959</td>
      <td>2.112123</td>
      <td>0.537773</td>
      <td>0.231708</td>
      <td>0.898190</td>
      <td>0.114970</td>
      <td>2.234345</td>
      <td>4.391264</td>
      <td>4.418188</td>
      <td>...</td>
      <td>0.132106</td>
      <td>0.980370</td>
      <td>1.672495</td>
      <td>3.558780</td>
      <td>0.072055</td>
      <td>2.029817</td>
      <td>0.418454</td>
      <td>1.192156</td>
      <td>1.853873</td>
      <td>1.099853</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>2.808824</td>
      <td>1.503026</td>
      <td>4.660236</td>
      <td>3.781434</td>
      <td>1.292014</td>
      <td>1.926411</td>
      <td>1.174456</td>
      <td>4.833678</td>
      <td>6.598332</td>
      <td>4.183246</td>
      <td>...</td>
      <td>0.833774</td>
      <td>1.562745</td>
      <td>2.589504</td>
      <td>5.134478</td>
      <td>1.171923</td>
      <td>1.833690</td>
      <td>2.468382</td>
      <td>4.154689</td>
      <td>2.939265</td>
      <td>2.304627</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>1.579961</td>
      <td>0.518564</td>
      <td>2.777915</td>
      <td>0.865460</td>
      <td>0.373625</td>
      <td>1.167617</td>
      <td>0.173764</td>
      <td>2.826529</td>
      <td>4.751182</td>
      <td>4.572947</td>
      <td>...</td>
      <td>0.219681</td>
      <td>0.967018</td>
      <td>2.106584</td>
      <td>4.237208</td>
      <td>0.087220</td>
      <td>1.733176</td>
      <td>0.613084</td>
      <td>1.758543</td>
      <td>2.333462</td>
      <td>1.247311</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>2.013838</td>
      <td>0.808357</td>
      <td>3.148299</td>
      <td>1.180536</td>
      <td>0.818062</td>
      <td>1.499798</td>
      <td>0.267851</td>
      <td>3.099998</td>
      <td>4.809313</td>
      <td>4.769134</td>
      <td>...</td>
      <td>0.497330</td>
      <td>1.301714</td>
      <td>2.513882</td>
      <td>4.128883</td>
      <td>0.159118</td>
      <td>2.036867</td>
      <td>1.212433</td>
      <td>1.859695</td>
      <td>2.457608</td>
      <td>1.816770</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 18457 columns</p>
</div></div></div>
</div>
<p>and for the second model replicate.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">predicted_expression_matrix</span><span class="p">(</span><span class="n">model_name</span><span class="o">=</span><span class="s2">&quot;v1_rep1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>STRADA</th>
      <th>ETV4</th>
      <th>USP25</th>
      <th>ZSWIM5</th>
      <th>C21orf58</th>
      <th>MIR497HG</th>
      <th>CFAP74</th>
      <th>GSE1</th>
      <th>LPP</th>
      <th>CLK1</th>
      <th>...</th>
      <th>STRIP2</th>
      <th>TNFRSF1A</th>
      <th>RBM14-RBM4</th>
      <th>C1orf21</th>
      <th>LINC00639</th>
      <th>NPDC1</th>
      <th>ZNF425</th>
      <th>COL5A1</th>
      <th>BRD3</th>
      <th>EVI5L</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>2.892922</td>
      <td>1.268846</td>
      <td>4.257362</td>
      <td>5.062403</td>
      <td>1.708644</td>
      <td>0.509606</td>
      <td>2.129140</td>
      <td>5.102046</td>
      <td>3.987716</td>
      <td>3.229576</td>
      <td>...</td>
      <td>3.279328</td>
      <td>0.496723</td>
      <td>2.626515</td>
      <td>3.294918</td>
      <td>1.754605</td>
      <td>3.018026</td>
      <td>2.396543</td>
      <td>1.611294</td>
      <td>2.942932</td>
      <td>3.334204</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>2.812950</td>
      <td>1.338056</td>
      <td>4.490562</td>
      <td>5.533390</td>
      <td>1.759490</td>
      <td>0.507474</td>
      <td>1.839018</td>
      <td>4.959751</td>
      <td>4.505932</td>
      <td>3.428190</td>
      <td>...</td>
      <td>3.428484</td>
      <td>0.463658</td>
      <td>2.532722</td>
      <td>4.212840</td>
      <td>1.994184</td>
      <td>3.312843</td>
      <td>2.755963</td>
      <td>1.515888</td>
      <td>2.539843</td>
      <td>3.585655</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>2.971495</td>
      <td>1.536687</td>
      <td>4.427551</td>
      <td>5.517713</td>
      <td>1.655871</td>
      <td>0.504365</td>
      <td>1.900368</td>
      <td>4.990572</td>
      <td>4.717015</td>
      <td>3.675014</td>
      <td>...</td>
      <td>3.258069</td>
      <td>0.498031</td>
      <td>2.830813</td>
      <td>3.899091</td>
      <td>1.731513</td>
      <td>3.000879</td>
      <td>2.582952</td>
      <td>1.570875</td>
      <td>2.792023</td>
      <td>3.834818</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>3.346772</td>
      <td>1.558132</td>
      <td>4.612154</td>
      <td>4.838746</td>
      <td>1.690119</td>
      <td>0.695964</td>
      <td>2.075206</td>
      <td>3.900921</td>
      <td>4.603170</td>
      <td>3.397104</td>
      <td>...</td>
      <td>3.024333</td>
      <td>0.464141</td>
      <td>2.634533</td>
      <td>4.260577</td>
      <td>1.873491</td>
      <td>3.353543</td>
      <td>2.285160</td>
      <td>1.407212</td>
      <td>2.382554</td>
      <td>3.494318</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>3.088684</td>
      <td>1.522742</td>
      <td>4.351163</td>
      <td>4.806933</td>
      <td>1.668947</td>
      <td>0.553315</td>
      <td>1.892262</td>
      <td>4.495190</td>
      <td>4.887375</td>
      <td>3.282181</td>
      <td>...</td>
      <td>3.286668</td>
      <td>0.475796</td>
      <td>2.699739</td>
      <td>4.285741</td>
      <td>1.986780</td>
      <td>3.199118</td>
      <td>2.432456</td>
      <td>1.463589</td>
      <td>2.867734</td>
      <td>3.607585</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>2.760382</td>
      <td>0.603564</td>
      <td>4.133419</td>
      <td>2.195248</td>
      <td>1.040137</td>
      <td>0.813743</td>
      <td>0.439841</td>
      <td>4.018094</td>
      <td>6.189613</td>
      <td>3.845854</td>
      <td>...</td>
      <td>0.460953</td>
      <td>2.616969</td>
      <td>2.051131</td>
      <td>3.720593</td>
      <td>0.373556</td>
      <td>1.720908</td>
      <td>0.799354</td>
      <td>2.626200</td>
      <td>2.094780</td>
      <td>2.103047</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>1.192427</td>
      <td>0.355300</td>
      <td>1.690948</td>
      <td>0.294813</td>
      <td>0.529883</td>
      <td>1.071292</td>
      <td>0.109688</td>
      <td>1.716427</td>
      <td>2.887877</td>
      <td>4.564609</td>
      <td>...</td>
      <td>0.117188</td>
      <td>1.749936</td>
      <td>1.504746</td>
      <td>2.753326</td>
      <td>0.032030</td>
      <td>2.539737</td>
      <td>0.283777</td>
      <td>1.469010</td>
      <td>1.537342</td>
      <td>1.323113</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>3.369931</td>
      <td>0.922407</td>
      <td>4.246971</td>
      <td>3.432078</td>
      <td>1.621070</td>
      <td>1.825625</td>
      <td>0.817042</td>
      <td>4.820827</td>
      <td>6.642282</td>
      <td>4.045234</td>
      <td>...</td>
      <td>0.992599</td>
      <td>1.805059</td>
      <td>3.023350</td>
      <td>4.650090</td>
      <td>0.711800</td>
      <td>2.415006</td>
      <td>1.855187</td>
      <td>3.172574</td>
      <td>2.799354</td>
      <td>2.962450</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>1.533311</td>
      <td>0.472664</td>
      <td>2.386347</td>
      <td>0.445858</td>
      <td>0.655848</td>
      <td>1.248309</td>
      <td>0.157859</td>
      <td>1.858983</td>
      <td>3.740174</td>
      <td>5.334047</td>
      <td>...</td>
      <td>0.221859</td>
      <td>1.986270</td>
      <td>2.305407</td>
      <td>3.003855</td>
      <td>0.045920</td>
      <td>1.986692</td>
      <td>0.455313</td>
      <td>1.756422</td>
      <td>2.054446</td>
      <td>1.544962</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>2.077998</td>
      <td>0.637752</td>
      <td>2.782059</td>
      <td>0.618168</td>
      <td>1.017707</td>
      <td>1.815663</td>
      <td>0.241223</td>
      <td>2.105849</td>
      <td>4.217391</td>
      <td>4.784609</td>
      <td>...</td>
      <td>0.379386</td>
      <td>2.361434</td>
      <td>2.744125</td>
      <td>3.384300</td>
      <td>0.075347</td>
      <td>2.696805</td>
      <td>0.822292</td>
      <td>2.332086</td>
      <td>2.385205</td>
      <td>2.119385</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 18457 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">anndata</span><span class="o">.</span><span class="n">layers</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Layers with keys: preds, v1_rep0, v1_rep1, v1_rep2, v1_rep3
</pre></div>
</div>
</div>
</div>
</section>
<section id="cli-api">
<h2>CLI API<a class="headerlink" href="#cli-api" title="Link to this heading">¶</a></h2>
<p>If you want to perform gene expression prediction again, rather than using the precomputed scores, you can use the Decima command-line interface (CLI) to generate new predictions for any set of genes you specify. For example, you can run the <code class="docutils literal notranslate"><span class="pre">decima</span> <span class="pre">predict-genes</span></code> command with the <code class="docutils literal notranslate"><span class="pre">--genes</span></code> argument to provide a comma-separated list of gene names (such as “STRADA,ETV4,USP25”) if no gene provided it will perform expression predictions for all genes, select the prediction model with the <code class="docutils literal notranslate"><span class="pre">--model</span></code> option (for instance, “ensemble” or a specific replicate like “0”), and use <code class="docutils literal notranslate"><span class="pre">--save-replicates</span></code> to save predictions for each replicate. The <code class="docutils literal notranslate"><span class="pre">-o</span></code> flag lets you specify the output file path for the predictions in <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code> format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>decima<span class="w"> </span>predict-genes<span class="w"> </span>--genes<span class="w"> </span><span class="s2">&quot;STRADA,ETV4,USP25&quot;</span><span class="w"> </span>--model<span class="w"> </span>ensemble<span class="w"> </span>--save-replicates<span class="w"> </span>-o<span class="w"> </span>example/predict_genes/predictions.h5ad<span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>decima - INFO - Using device: cuda and genome: hg38 for prediction.
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Currently logged in as: <span class=" -Color -Color-Yellow">mhcelik</span> (<span class=" -Color -Color-Yellow">mhcw</span>) to <span class=" -Color -Color-Green">https://api.wandb.ai</span>. Use <span class=" -Color -Color-Bold">`wandb login --relogin`</span> to force relogin
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1433.8MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep1:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1374.8MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep2:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.7 (1016.5MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep3:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1371.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.5 (2022.8MB/s)
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/lightning_fabric/plugins/environments/slurm.py:204: PossibleUserWarning: The `srun` command is available on your system but is not used. HINT: If your intention is to run Lightning on SLURM, prepend your python command with `srun` like so: srun python3.11 /home/celikm5/miniforge3/envs/decima/bin/decima  ...
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Predicting DataLoader 0: 100%|████████████████████| 1/1 [00:06&lt;00:00,  0.15it/s]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/torchmetrics/utilities/prints.py:43: UserWarning: The ``compute`` method of metric WarningCounter was called before the ``update`` method which may lead to errors, as metric states have not yet been updated.

</pre></div>
</div>
</div>
</div>
<p>After running this command, you can load the resulting predictions in Python using the <code class="docutils literal notranslate"><span class="pre">DecimaResult</span></code> class and access the predicted expression matrix as:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">DecimaResult</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;example/predict_genes/predictions.h5ad&quot;</span><span class="p">)</span>
<span class="n">result</span><span class="o">.</span><span class="n">predicted_expression_matrix</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>STRADA</th>
      <th>ETV4</th>
      <th>USP25</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>2.973423</td>
      <td>1.845874</td>
      <td>4.592660</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>2.954493</td>
      <td>1.897248</td>
      <td>4.688962</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>2.938927</td>
      <td>2.197576</td>
      <td>4.861925</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>3.046052</td>
      <td>2.138910</td>
      <td>4.864000</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>3.025393</td>
      <td>2.019534</td>
      <td>4.602815</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>2.334128</td>
      <td>0.633340</td>
      <td>4.675878</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>0.835045</td>
      <td>0.358672</td>
      <td>1.964816</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>3.008091</td>
      <td>1.209339</td>
      <td>4.798233</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>1.241775</td>
      <td>0.455031</td>
      <td>2.919700</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>1.715322</td>
      <td>0.700902</td>
      <td>3.044498</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 3 columns</p>
</div></div></div>
</div>
<p>or for a specific replicate:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span><span class="o">.</span><span class="n">predicted_expression_matrix</span><span class="p">(</span><span class="n">model_name</span><span class="o">=</span><span class="s2">&quot;preds_v1_rep0&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>STRADA</th>
      <th>ETV4</th>
      <th>USP25</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>agg_0</th>
      <td>2.932055</td>
      <td>2.022040</td>
      <td>4.795146</td>
    </tr>
    <tr>
      <th>agg_1</th>
      <td>2.816373</td>
      <td>1.940856</td>
      <td>4.917033</td>
    </tr>
    <tr>
      <th>agg_2</th>
      <td>2.742780</td>
      <td>2.396595</td>
      <td>4.806089</td>
    </tr>
    <tr>
      <th>agg_3</th>
      <td>2.804382</td>
      <td>2.360139</td>
      <td>4.676453</td>
    </tr>
    <tr>
      <th>agg_4</th>
      <td>2.815311</td>
      <td>2.100366</td>
      <td>4.594069</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>agg_9533</th>
      <td>2.510625</td>
      <td>0.811365</td>
      <td>5.366706</td>
    </tr>
    <tr>
      <th>agg_9535</th>
      <td>1.245751</td>
      <td>0.397869</td>
      <td>2.111997</td>
    </tr>
    <tr>
      <th>agg_9536</th>
      <td>2.808623</td>
      <td>1.503265</td>
      <td>4.659208</td>
    </tr>
    <tr>
      <th>agg_9537</th>
      <td>1.580037</td>
      <td>0.518710</td>
      <td>2.777559</td>
    </tr>
    <tr>
      <th>agg_9538</th>
      <td>2.013807</td>
      <td>0.808542</td>
      <td>3.147933</td>
    </tr>
  </tbody>
</table>
<p>8856 rows × 3 columns</p>
</div></div></div>
</div>
</section>
<section id="python-api">
<h2>Python API<a class="headerlink" href="#python-api" title="Link to this heading">¶</a></h2>
<p>The same functionality is available through the Python API, allowing you to perform gene expression prediction programmatically. You can specify the genes, model, and other options directly in your Python code using the provided classes and functions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.tools.inference</span><span class="w"> </span><span class="kn">import</span> <span class="n">predict_gene_expression</span>

<span class="n">ad</span> <span class="o">=</span> <span class="n">predict_gene_expression</span><span class="p">(</span>
    <span class="n">genes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;STRADA&quot;</span><span class="p">,</span> <span class="s2">&quot;ETV4&quot;</span><span class="p">,</span> <span class="s2">&quot;USP25&quot;</span><span class="p">],</span>
    <span class="n">model</span><span class="o">=</span><span class="s2">&quot;ensemble&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1404.0MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep1:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1277.2MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep2:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.6 (1157.4MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep3:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1321.9MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (1932.7MB/s)
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/lightning_fabric/plugins/environments/slurm.py:204: PossibleUserWarning: The `srun` command is available on your system but is not used. HINT: If your intention is to run Lightning on SLURM, prepend your python command with `srun` like so: srun python /home/celikm5/miniforge3/envs/decima/lib/python3.11/ ...
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "3ab38f07a0724d099bcbaee7e9211463", "version_major": 2, "version_minor": 0}</script><div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/torchmetrics/utilities/prints.py:43: UserWarning: The ``compute`` method of metric WarningCounter was called before the ``update`` method which may lead to errors, as metric states have not yet been updated.
</pre></div>
</div>
</div>
</div>
</section>
<section id="developer-api">
<h2>Developer API<a class="headerlink" href="#developer-api" title="Link to this heading">¶</a></h2>
<p>Under the hood, the Decima prediction API uses the <code class="docutils literal notranslate"><span class="pre">GeneDataset</span></code> and <code class="docutils literal notranslate"><span class="pre">SeqDataset</span></code> pytorch Dataset classes to prepare the data for prediction. These classes provide a flexible way to handle different types of input data, including custom genes and DNA sequences. Internally, these datasets represent sequences using one-hot encoding and apply a gene mask to indicate which positions correspond to gene regions.</p>
<p>For example, you can create a <code class="docutils literal notranslate"><span class="pre">GeneDataset</span></code> object to predict expression for the genes in your metadata. The <code class="docutils literal notranslate"><span class="pre">predict_on_dataset</span></code> method returns a dictionary containing the predicted expression values and if there is any warnings.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pprint</span><span class="w"> </span><span class="kn">import</span> <span class="n">pprint</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.data.dataset</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneDataset</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.hub</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_decima_model</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">load_decima_model</span><span class="p">(</span><span class="s2">&quot;rep0&quot;</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">GeneDataset</span><span class="p">(</span><span class="n">genes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;STRADA&quot;</span><span class="p">,</span> <span class="s2">&quot;ETV4&quot;</span><span class="p">,</span> <span class="s2">&quot;USP25&quot;</span><span class="p">])</span>

<span class="n">preds</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_on_dataset</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">preds</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact rep0:latest, 720.03MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:0.5 (1333.5MB/s)
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>: Downloading large artifact metadata:latest, 3122.32MB. 1 files... 
<span class=" -Color -Color-Bold -Color-Bold-Blue">wandb</span>:   1 of 1 files downloaded.  
Done. 0:0:1.6 (2010.7MB/s)
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/lightning_fabric/plugins/environments/slurm.py:204: PossibleUserWarning: The `srun` command is available on your system but is not used. HINT: If your intention is to run Lightning on SLURM, prepend your python command with `srun` like so: srun python /home/celikm5/miniforge3/envs/decima/lib/python3.11/ ...
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:425: PossibleUserWarning: The &#39;predict_dataloader&#39; does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=7` in the `DataLoader` to improve performance.
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "3ff66d87baf44d02998ba910f6ed5217", "version_major": 2, "version_minor": 0}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;expression&#39;: array([[2.9320552 , 2.8163733 , 2.7427804 , ..., 2.808623  , 1.5800365 ,
        2.0138068 ],
       [2.0220397 , 1.9408565 , 2.3965948 , ..., 1.5032653 , 0.51870984,
        0.80854166],
       [4.7951455 , 4.9170327 , 4.806089  , ..., 4.659208  , 2.7775593 ,
        3.1479335 ]], dtype=float32),
 &#39;warnings&#39;: Counter({&#39;unknown&#39;: tensor(0),
                      &#39;allele_mismatch_with_reference_genome&#39;: tensor(0)})}
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/celikm5/miniforge3/envs/decima/lib/python3.11/site-packages/torchmetrics/utilities/prints.py:43: UserWarning: The ``compute`` method of metric WarningCounter was called before the ``update`` method which may lead to errors, as metric states have not yet been updated.
</pre></div>
</div>
</div>
</div>
</section>
<section id="custom-expression-for-custom-genes">
<h2>Custom Expression for custom genes<a class="headerlink" href="#custom-expression-for-custom-genes" title="Link to this heading">¶</a></h2>
<p>If you have custom genes, you can create a <code class="docutils literal notranslate"><span class="pre">SeqDataset</span></code> object to predict expression for those genes.</p>
<p>To do this, prepare a FASTA file where:</p>
<ul class="simple">
<li><p>Each sequence is exactly the Decima context size (524,288 bases).</p></li>
<li><p>The FASTA header for each sequence must include the gene name and the gene mask coordinates, using the format:
<code class="docutils literal notranslate"><span class="pre">&gt;gene_name|gene_mask_start=X|gene_mask_end=Y</span></code>
where <code class="docutils literal notranslate"><span class="pre">X</span></code> and <code class="docutils literal notranslate"><span class="pre">Y</span></code> specify the start and end positions (0-based, inclusive) of the gene region within the sequence. The gene mask indicates which region of the sequence corresponds to the gene for which expression will be predicted.</p></li>
</ul>
<p>For example, seqs.fasta contains these information:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>cat<span class="w"> </span>../tests/data/seqs.fasta<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span>-200
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&gt;CD68|gene_mask_start=163840|gene_mask_end=166460
CTCTGCAGAGAGCGAGGACGGTGTGTCTGCCAGCGCCTTTGACTTCACTGTCTCCAACTTTGTGGACAACCTGTATGGCTACCCGGAAGGCAAGGATGTGCTTCGGGAGACCATCAAGTTTATGTACACAGACTGGGCCGACCGGGACAATGGCGAAATGCGCCGCAAAACCCTGCTGGCGCTCTTTACTGACCACCAAT
&gt;SPI1|gene_mask_start=163840|gene_mask_end=187556
TGCCACTTTTAGATATGTTCATGGGTGCAGATACGGCTTTATTTATTTGAGACAGAGTTTCACTCTTGTTGCCCAGGGTGGAGTGCAGTGGTGCGATCTCAGCTCACTGCAGCCTTCGCCTCCCGGGTTGAAGCGATTCTTCTGCCTCAACCTCGAGTAGCTGGGATTATAGGCACCTGCCAGCATGCCTGGCTAATTTT
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">decima.data.dataset</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqDataset</span>


<span class="n">ds</span> <span class="o">=</span> <span class="n">SeqDataset</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;../tests/data/seqs.fasta&quot;</span><span class="p">)</span>

<span class="n">preds</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_on_dataset</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">preds</span><span class="p">[</span><span class="s2">&quot;expression&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "cca4f2bc479748b2af20bbc7361a15e4", "version_major": 2, "version_minor": 0}</script><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[0.22032928, 0.21946219, 0.230885  , ..., 1.0364848 , 1.2134479 ,
        1.4428278 ],
       [0.2652336 , 0.13350041, 0.14896178, ..., 0.44437772, 0.26560807,
        0.30437455]], dtype=float32)
</pre></div>
</div>
</div>
</div>
<p>See the documentation of SeqDataset for more details. SeqDataset can be created from a pandas DataFrame with the following columns: <code class="docutils literal notranslate"><span class="pre">seq</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_mask_start</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_mask_end</span></code>, and <code class="docutils literal notranslate"><span class="pre">gene_name</span></code> with <code class="docutils literal notranslate"><span class="pre">SeqDataset.from_dataframe</span></code> or from a one-hot encoded tensor with <code class="docutils literal notranslate"><span class="pre">SeqDataset.from_one_hot</span></code>.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../contributing.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Contributing</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="4-modisco.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Detecting and Clustering Seqlets with Modisco</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Gokcen Eraslan
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Predicting Gene Expression with Decima</a><ul>
<li><a class="reference internal" href="#precomputed-predictions">Precomputed Predictions</a></li>
<li><a class="reference internal" href="#cli-api">CLI API</a></li>
<li><a class="reference internal" href="#python-api">Python API</a></li>
<li><a class="reference internal" href="#developer-api">Developer API</a></li>
<li><a class="reference internal" href="#custom-expression-for-custom-genes">Custom Expression for custom genes</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=68a6d85b"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>