<!doctype html>
<html class="no-js" lang="en" data-content_root="../../../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../../genindex.html"><link rel="search" title="Search" href="../../../search.html">

    <!-- Generated with Sphinx 8.2.3 and Furo 2025.12.19 -->
        <title>decima.interpret.attributions - decima 0.5.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=7bdb33bb" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">decima 0.5.2 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  <span class="sidebar-brand-text">decima 0.5.2 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">Overview</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api/modules.html">Module Reference</a><input aria-label="Toggle navigation of Module Reference" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../api/decima.html">decima package</a><input aria-label="Toggle navigation of decima package" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.cli.html">decima.cli package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.core.html">decima.core package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.data.html">decima.data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.hub.html">decima.hub package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.interpret.html">decima.interpret package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.model.html">decima.model package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.plot.html">decima.plot package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.tools.html">decima.tools package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.train.html">decima.train package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.utils.html">decima.utils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/decima.vep.html">decima.vep package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/1-attribution-motif-discovery.html">Tutorial: Attribution Motif Discovery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/2-variant-effect-prediction.html">Tutorial: Variant Effect Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/3-finetune.html">Tutorial: Finetuning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/4-modisco.html">Tutorial: Modisco</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/5-gene-expression-prediction.html">Tutorial: Gene Expression Prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <h1>Source code for decima.interpret.attributions</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Attributions module predict attributes and performs recursive seqlet calling.</span>

<span class="sd">Examples:</span>
<span class="sd">    &gt;&gt;&gt; predict_save_attributions(</span>
<span class="sd">    ...     output_prefix=&quot;output_prefix&quot;,</span>
<span class="sd">    ...     tasks=[</span>
<span class="sd">    ...         &quot;agg1&quot;,</span>
<span class="sd">    ...         &quot;agg2&quot;,</span>
<span class="sd">    ...     ],</span>
<span class="sd">    ...     off_tasks=[</span>
<span class="sd">    ...         &quot;agg3&quot;,</span>
<span class="sd">    ...         &quot;agg4&quot;,</span>
<span class="sd">    ...     ],</span>
<span class="sd">    ... )</span>
<span class="sd">    &gt;&gt;&gt; recursive_seqlet_calling(</span>
<span class="sd">    ...     output_prefix=&quot;output_prefix&quot;,</span>
<span class="sd">    ...     attributions=&quot;attributions.h5&quot;,</span>
<span class="sd">    ...     tasks=[</span>
<span class="sd">    ...         &quot;agg1&quot;,</span>
<span class="sd">    ...         &quot;agg2&quot;,</span>
<span class="sd">    ...     ],</span>
<span class="sd">    ...     off_tasks=[</span>
<span class="sd">    ...         &quot;agg3&quot;,</span>
<span class="sd">    ...         &quot;agg4&quot;,</span>
<span class="sd">    ...     ],</span>
<span class="sd">    ... )</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">more_itertools</span><span class="w"> </span><span class="kn">import</span> <span class="n">chunked</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">torch.utils.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">DataLoader</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pyfaidx</span><span class="w"> </span><span class="kn">import</span> <span class="n">Faidx</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">decima.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">DEFAULT_ENSEMBLE</span><span class="p">,</span> <span class="n">MODEL_METADATA</span><span class="p">,</span> <span class="n">ENSEMBLE_MODELS</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.core.attribution</span><span class="w"> </span><span class="kn">import</span> <span class="n">AttributionResult</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.core.result</span><span class="w"> </span><span class="kn">import</span> <span class="n">DecimaResult</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.data.dataset</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneDataset</span><span class="p">,</span> <span class="n">SeqDataset</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.hub</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_decima_model</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.interpret.attributer</span><span class="w"> </span><span class="kn">import</span> <span class="n">DecimaAttributer</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_compute_device</span><span class="p">,</span> <span class="n">_get_on_off_tasks</span><span class="p">,</span> <span class="n">_get_genes</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.utils.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">AttributionWriter</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">decima.utils.qc</span><span class="w"> </span><span class="kn">import</span> <span class="n">QCLogger</span>


<div class="viewcode-block" id="predict_save_attributions">
<a class="viewcode-back" href="../../../api/decima.interpret.html#decima.predict_save_attributions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">predict_save_attributions</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">off_tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">DEFAULT_ENSEMBLE</span><span class="p">,</span>
    <span class="n">metadata_anndata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;inputxgradient&quot;</span><span class="p">,</span>
    <span class="n">transform</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;specificity&quot;</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seqs</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">top_n_markers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">bigwig</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">correct_grad_bigwig</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
    <span class="n">device</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">genome</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate and save attribution analysis results for a gene.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_prefix: Prefix for the output files where attribution results will be saved.</span>
<span class="sd">        tasks: Tasks to attribute for prediction either list of task names or query string to filter cell types to analyze attributions for (e.g. &#39;cell_type == &#39;classical monocyte&#39;&#39;). If not provided, all tasks will be analyzed.</span>
<span class="sd">        off_tasks: Off tasks to attribute for prediction either list of task names or query string to filter cell types to contrast against (e.g. &#39;cell_type == &#39;classical monocyte&#39;&#39;). If not provided, no contrast will be performed.</span>
<span class="sd">        model: Model to attribute for prediction default is 0. Can be replicate number (0-3) or path to custom model.</span>
<span class="sd">        metadata_anndata: Metadata anndata path or DecimaResult object. If not provided, the default metadata will be downloaded from wandb.</span>
<span class="sd">        method: Method to use for attribution analysis available options: &quot;saliency&quot;, &quot;inputxgradient&quot;, &quot;integratedgradients&quot;. Default is &quot;inputxgradient&quot;.</span>
<span class="sd">        transform: Transform to use for attribution analysis available options: &quot;specificity&quot;, &quot;aggregate&quot;. Default is &quot;specificity&quot;.</span>
<span class="sd">        batch_size: Batch size for attribution analysis default is 1. Increasing batch size may speed up computation but requires more memory.</span>
<span class="sd">        genes: Genes to attribute for prediction if not provided, all genes will be used. Can be list of gene symbols or IDs.</span>
<span class="sd">        seqs: Sequences to attribute for prediction. Can be path to fasta file, DataFrame, or numpy/torch tensor. Mutually exclusive with genes parameter.</span>
<span class="sd">        top_n_markers: Top n markers for prediction if not provided, genes will be used. Useful for analyzing only the most important marker genes.</span>
<span class="sd">        bigwig: Whether to save attribution scores as a bigwig file default is True. Bigwig files can be loaded in genome browsers for visualization.</span>
<span class="sd">        correct_grad_bigwig: Whether to correct the gradient bigwig file default is True. Applies gradient correction for better visualization.</span>
<span class="sd">        num_workers: Number of workers for attribution analysis default is 4. Increasing number of workers will speed up the process.</span>
<span class="sd">        device: Device to use for attribution analysis (e.g. &#39;cuda&#39;, &#39;cpu&#39;). If not provided, the best available device will be used automatically.</span>
<span class="sd">        genome: Genome to use for attribution analysis default is &quot;hg38&quot;. Can be genome name or path to custom genome fasta file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Path to the attribution file.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; predict_save_attributions(</span>
<span class="sd">        ...     output_prefix=&quot;output_prefix&quot;,</span>
<span class="sd">        ...     tasks=[</span>
<span class="sd">        ...         &quot;task1&quot;,</span>
<span class="sd">        ...         &quot;task2&quot;,</span>
<span class="sd">        ...     ],</span>
<span class="sd">        ...     off_tasks=[</span>
<span class="sd">        ...         &quot;task3&quot;,</span>
<span class="sd">        ...         &quot;task4&quot;,</span>
<span class="sd">        ...     ],</span>
<span class="sd">        ...     model=0,</span>
<span class="sd">        ...     metadata_anndata=&quot;metadata_anndata.h5ad&quot;,</span>
<span class="sd">        ...     method=&quot;inputxgradient&quot;,</span>
<span class="sd">        ...     transform=&quot;specificity&quot;,</span>
<span class="sd">        ...     batch_size=1,</span>
<span class="sd">        ...     genes=[</span>
<span class="sd">        ...         &quot;gene1&quot;,</span>
<span class="sd">        ...         &quot;gene2&quot;,</span>
<span class="sd">        ...     ],</span>
<span class="sd">        ...     seqs=&quot;seqs.fasta&quot;,</span>
<span class="sd">        ...     top_n_markers=10,</span>
<span class="sd">        ...     bigwig=True,</span>
<span class="sd">        ...     correct_grad_bigwig=True,</span>
<span class="sd">        ...     num_workers=4,</span>
<span class="sd">        ...     device=&quot;cuda&quot;,</span>
<span class="sd">        ...     genome=&quot;hg38&quot;,</span>
<span class="sd">        ... )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">model</span> <span class="ow">in</span> <span class="n">ENSEMBLE_MODELS</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">ENSEMBLE_MODELS</span><span class="p">:</span>
            <span class="n">models</span> <span class="o">=</span> <span class="n">MODEL_METADATA</span><span class="p">[</span><span class="n">model</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">models</span> <span class="o">=</span> <span class="n">model</span>
        <span class="k">return</span> <span class="p">[</span>
            <span class="n">predict_save_attributions</span><span class="p">(</span>
                <span class="n">output_prefix</span><span class="o">=</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_</span><span class="si">{model}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">model</span><span class="o">=</span><span class="n">idx</span><span class="p">),</span>
                <span class="n">tasks</span><span class="o">=</span><span class="n">tasks</span><span class="p">,</span>
                <span class="n">off_tasks</span><span class="o">=</span><span class="n">off_tasks</span><span class="p">,</span>
                <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
                <span class="n">metadata_anndata</span><span class="o">=</span><span class="n">metadata_anndata</span><span class="p">,</span>
                <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
                <span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">,</span>
                <span class="n">batch_size</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
                <span class="n">genes</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span>
                <span class="n">seqs</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span>
                <span class="n">top_n_markers</span><span class="o">=</span><span class="n">top_n_markers</span><span class="p">,</span>
                <span class="n">bigwig</span><span class="o">=</span><span class="n">bigwig</span><span class="p">,</span>
                <span class="n">correct_grad_bigwig</span><span class="o">=</span><span class="n">correct_grad_bigwig</span><span class="p">,</span>
                <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">,</span>
                <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
                <span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">model</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">models</span><span class="p">)</span>
        <span class="p">]</span>

    <span class="n">output_prefix</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span>
    <span class="n">output_prefix</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;decima&quot;</span><span class="p">)</span>

    <span class="n">device</span> <span class="o">=</span> <span class="n">get_compute_device</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using device: </span><span class="si">{</span><span class="n">device</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading model </span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2"> and metadata to compute attributions...&quot;</span><span class="p">)</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">load_decima_model</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">DecimaResult</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">metadata_anndata</span> <span class="ow">or</span> <span class="n">model</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>

    <span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span> <span class="o">=</span> <span class="n">_get_on_off_tasks</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span><span class="p">)</span>
    <span class="n">attributer</span> <span class="o">=</span> <span class="n">DecimaAttributer</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">transform</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">QCLogger</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.warnings.qc.log&quot;</span><span class="p">,</span> <span class="n">metadata_anndata</span><span class="o">=</span><span class="n">result</span><span class="p">)</span> <span class="k">as</span> <span class="n">qc</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">ground_truth</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">qc</span><span class="o">.</span><span class="n">log_correlation</span><span class="p">(</span><span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span><span class="n">genes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Only one of `genes` or `seqs` arguments must be provided not both.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">seqs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">top_n_markers</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;`top_n_markers` is not supported when `seqs` is provided.&quot;</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">dataset</span> <span class="o">=</span> <span class="n">SeqDataset</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
                <span class="n">dataset</span> <span class="o">=</span> <span class="n">SeqDataset</span><span class="o">.</span><span class="n">from_dataframe</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
                <span class="k">assert</span> <span class="n">seqs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">5</span><span class="p">,</span> <span class="p">(</span>
                    <span class="s2">&quot;`seqs` must be 5-dimensional with shape (batch_size, 5, seq_len) &quot;</span>
                    <span class="s2">&quot;where the 2th dimension is a one_hot encoded seq and binary mask gene mask.&quot;</span>
                <span class="p">)</span>
                <span class="n">dataset</span> <span class="o">=</span> <span class="n">SeqDataset</span><span class="o">.</span><span class="n">from_one_hot</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid type for seqs: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span><span class="si">}</span><span class="s2">. Must be a path to fasta file or pd.DataFrame.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dataset</span> <span class="o">=</span> <span class="n">GeneDataset</span><span class="p">(</span>
                <span class="n">genes</span><span class="o">=</span><span class="n">_get_genes</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">top_n_markers</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span><span class="p">),</span> <span class="n">metadata_anndata</span><span class="o">=</span><span class="n">result</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="n">genome</span>
            <span class="p">)</span>

        <span class="n">genes_batch</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">chunked</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">))</span>
        <span class="n">gene_mask_starts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">chunked</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">gene_mask_starts</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">))</span>
        <span class="n">gene_mask_ends</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">chunked</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">gene_mask_ends</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">))</span>

        <span class="n">dl</span> <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span>
            <span class="n">dataset</span><span class="p">,</span>
            <span class="n">batch_size</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
            <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">pin_memory</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.attributions.h5&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">AttributionWriter</span><span class="p">(</span>
            <span class="n">path</span><span class="o">=</span><span class="n">output_path</span><span class="p">,</span>
            <span class="n">genes</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span>
            <span class="n">model_name</span><span class="o">=</span><span class="n">attributer</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="n">metadata_anndata</span><span class="o">=</span><span class="n">result</span><span class="p">,</span>
            <span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">,</span>
            <span class="n">bigwig</span><span class="o">=</span><span class="n">bigwig</span><span class="p">,</span>
            <span class="n">correct_grad_bigwig</span><span class="o">=</span><span class="n">correct_grad_bigwig</span><span class="p">,</span>
            <span class="n">custom_genes</span><span class="o">=</span><span class="n">seqs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">)</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">inputs</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tqdm</span><span class="p">(</span><span class="n">dl</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Computing attributions...&quot;</span><span class="p">)):</span>
                <span class="n">attrs</span> <span class="o">=</span> <span class="n">attributer</span><span class="o">.</span><span class="n">attribute</span><span class="p">(</span><span class="n">inputs</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">))</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
                <span class="n">_seqs</span> <span class="o">=</span> <span class="n">inputs</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>

                <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">attr</span><span class="p">,</span> <span class="n">g_start</span><span class="p">,</span> <span class="n">g_end</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                    <span class="n">genes_batch</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">_seqs</span><span class="p">,</span> <span class="n">attrs</span><span class="p">,</span> <span class="n">gene_mask_starts</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">gene_mask_ends</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="p">):</span>
                    <span class="n">writer</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span> <span class="n">seqs</span><span class="o">=</span><span class="n">seq</span><span class="p">,</span> <span class="n">attrs</span><span class="o">=</span><span class="n">attr</span><span class="p">,</span> <span class="n">gene_mask_start</span><span class="o">=</span><span class="n">g_start</span><span class="p">,</span> <span class="n">gene_mask_end</span><span class="o">=</span><span class="n">g_end</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">seqs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">qc</span><span class="o">.</span><span class="n">log_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">seqs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Saving sequences...&quot;</span><span class="p">)</span>
            <span class="n">fasta_path</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.seqs.fasta&quot;</span><span class="p">))</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">genes</span><span class="p">,</span> <span class="n">dataset</span><span class="o">.</span><span class="n">seqs</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Saving sequences...&quot;</span><span class="p">):</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">Faidx</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="n">build_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">output_path</span></div>



<div class="viewcode-block" id="recursive_seqlet_calling">
<a class="viewcode-back" href="../../../api/decima.interpret.html#decima.recursive_seqlet_calling">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">recursive_seqlet_calling</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">attributions</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">off_tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">tss_distance</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">metadata_anndata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">top_n_markers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
    <span class="n">agg_func</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;mean&quot;</span><span class="p">,</span>
    <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-4</span><span class="p">,</span>
    <span class="n">min_seqlet_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
    <span class="n">max_seqlet_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">25</span><span class="p">,</span>
    <span class="n">additional_flanks</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">pattern_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span>
    <span class="n">custom_genome</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">meme_motif_db</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;hocomoco_v13&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Recursive seqlet calling for attribution analysis.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_prefix: Prefix for the output files where seqlet calling results will be saved.</span>
<span class="sd">        attributions: Attributions to use for recursive seqlet calling generated by `decima attributions-predict` or `decima attributions` commands. Can be single file path or list of attribution files.</span>
<span class="sd">        tasks: Tasks to attribute for recursive seqlet calling either list of task names or query string to filter cell types to analyze attributions for (e.g. &#39;cell_type == &#39;classical monocyte&#39;&#39;). If not provided, all tasks will be analyzed.</span>
<span class="sd">        off_tasks: Off tasks to attribute for recursive seqlet calling either list of task names or query string to filter cell types to contrast against (e.g. &#39;cell_type == &#39;classical monocyte&#39;&#39;). If not provided, no contrast will be performed.</span>
<span class="sd">        tss_distance: TSS distance for recursive seqlet calling default is full context size of decima of 524288. Controls the genomic window size around TSS for analysis.</span>
<span class="sd">        metadata_anndata: Metadata anndata path or DecimaResult object. If not provided, the default metadata will be used from the attribution files.</span>
<span class="sd">        genes: Genes to attribute for recursive seqlet calling if not provided, all genes will be used. Can be list of gene symbols or IDs to focus analysis on specific genes.</span>
<span class="sd">        top_n_markers: Top n markers for recursive seqlet calling if not provided, genes will be used. Useful for analyzing only the most important marker genes for the specified tasks.</span>
<span class="sd">        num_workers: Number of workers for recursive seqlet calling default is 4. Increasing number of workers will speed up the process but requires more memory.</span>
<span class="sd">        agg_func: Aggregation function for recursive seqlet calling default is &#39;mean&#39;. Available options: &#39;mean&#39;, &#39;max&#39;. Determines how attribution scores are aggregated across replicates.</span>
<span class="sd">        threshold: P-value threshold for recursive seqlet calling default is 5e-4. Lower values result in more stringent peak calling and fewer detected seqlets.</span>
<span class="sd">        min_seqlet_len: Minimum seqlet length for recursive seqlet calling default is 4. Shorter sequences will be filtered out from the analysis.</span>
<span class="sd">        max_seqlet_len: Maximum seqlet length for recursive seqlet calling default is 25. Longer sequences will be truncated or filtered based on the algorithm.</span>
<span class="sd">        additional_flanks: Additional flanks for recursive seqlet calling default is 0. Extends seqlet regions by this number of base pairs on each side.</span>
<span class="sd">        pattern_type: Pattern type for recursive seqlet calling default is &#39;both&#39;. Available options: &#39;both&#39;, &#39;pos&#39;, &#39;neg&#39;. Controls whether to consider positive peaks, negative peaks, or both.</span>
<span class="sd">        custom_genome: Custom genome flag for recursive seqlet calling default is False. If True, bigwig files will be generated with each gene as a different chromosome for custom sequences.</span>
<span class="sd">        meme_motif_db: MEME motif database for motif discovery default is &#39;hocomoco_v13&#39;. Specifies which motif database to use for downstream motif enrichment analysis.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; recursive_seqlet_calling(</span>
<span class="sd">        ...     output_prefix=&quot;output_prefix&quot;,</span>
<span class="sd">        ...     attributions=&quot;attributions.h5&quot;,</span>
<span class="sd">        ...     tasks=[</span>
<span class="sd">        ...         &quot;task1&quot;,</span>
<span class="sd">        ...         &quot;task2&quot;,</span>
<span class="sd">        ...     ],</span>
<span class="sd">        ...     off_tasks=[</span>
<span class="sd">        ...         &quot;task3&quot;,</span>
<span class="sd">        ...         &quot;task4&quot;,</span>
<span class="sd">        ...     ],</span>
<span class="sd">        ... )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">output_prefix</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span>
    <span class="n">output_prefix</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># TODO: update dependencies so we do not get future errors.</span>
    <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span>

    <span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;decima&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Loading model and metadata to compute attributions...&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">attributions</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">)):</span>
        <span class="n">attributions_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">Path</span><span class="p">(</span><span class="n">attributions</span><span class="p">)</span><span class="o">.</span><span class="n">as_posix</span><span class="p">()]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">attributions_files</span> <span class="o">=</span> <span class="n">attributions</span>

    <span class="k">with</span> <span class="n">AttributionResult</span><span class="p">(</span>
        <span class="n">attributions_files</span><span class="p">,</span> <span class="n">tss_distance</span><span class="p">,</span> <span class="n">correct_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">,</span> <span class="n">agg_func</span><span class="o">=</span><span class="n">agg_func</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">ar</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">DecimaResult</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">metadata_anndata</span> <span class="ow">or</span> <span class="n">ar</span><span class="o">.</span><span class="n">model_name</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">top_n_markers</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span> <span class="o">=</span> <span class="n">_get_on_off_tasks</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span><span class="p">)</span>
            <span class="n">all_genes</span> <span class="o">=</span> <span class="n">_get_genes</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">top_n_markers</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="n">off_tasks</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">genes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">all_genes</span> <span class="o">=</span> <span class="n">genes</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="k">else</span> <span class="p">[</span><span class="n">genes</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">all_genes</span> <span class="o">=</span> <span class="n">ar</span><span class="o">.</span><span class="n">genes</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No genes provided, using all </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">all_genes</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes in the attribution files.&quot;</span><span class="p">)</span>

        <span class="n">df_peaks</span><span class="p">,</span> <span class="n">df_motifs</span> <span class="o">=</span> <span class="n">ar</span><span class="o">.</span><span class="n">recursive_seqlet_calling</span><span class="p">(</span>
            <span class="n">all_genes</span><span class="p">,</span>
            <span class="n">metadata_anndata</span><span class="p">,</span>
            <span class="n">custom_genome</span><span class="p">,</span>
            <span class="n">threshold</span><span class="p">,</span>
            <span class="n">min_seqlet_len</span><span class="p">,</span>
            <span class="n">max_seqlet_len</span><span class="p">,</span>
            <span class="n">additional_flanks</span><span class="p">,</span>
            <span class="n">pattern_type</span><span class="p">,</span>
            <span class="n">meme_motif_db</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="n">df_peaks</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="n">output_prefix</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.seqlets.bed&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>
    <span class="n">df_motifs</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_prefix</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.motifs.tsv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>



<div class="viewcode-block" id="predict_attributions_seqlet_calling">
<a class="viewcode-back" href="../../../api/decima.interpret.html#decima.predict_attributions_seqlet_calling">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">predict_attributions_seqlet_calling</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seqs</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">off_tasks</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="n">DEFAULT_ENSEMBLE</span><span class="p">,</span>
    <span class="n">metadata_anndata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;inputxgradient&quot;</span><span class="p">,</span>
    <span class="n">transform</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;specificity&quot;</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">tss_distance</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">top_n_markers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">device</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-4</span><span class="p">,</span>
    <span class="n">min_seqlet_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
    <span class="n">max_seqlet_len</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">25</span><span class="p">,</span>
    <span class="n">additional_flanks</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">pattern_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;both&quot;</span><span class="p">,</span>
    <span class="n">meme_motif_db</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;hocomoco_v13&quot;</span><span class="p">,</span>
    <span class="n">genome</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate and save attribution analysis results for a gene.</span>
<span class="sd">    This function performs attribution analysis for a given gene and cell types, saving</span>
<span class="sd">    the following output files to the specified directory:</span>

<span class="sd">    Output files:</span>
<span class="sd">         {output_prefix}.attributions.h5      # Raw attribution score matrix per gene.</span>

<span class="sd">         {output_prefix}.attributions.bigwig  # Genome browser track of attribution as bigwig file.</span>

<span class="sd">         {output_prefix}.seqlets.bed          # List of attribution peaks in BED format.</span>

<span class="sd">         {output_prefix}.motifs.tsv           # Detected motifs in peak regions.</span>

<span class="sd">         {output_prefix}.warnings.qc.log      # QC warnings about prediction reliability.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_dir: Directory to save output files</span>
<span class="sd">        gene: Gene symbol or ID to analyze</span>
<span class="sd">        tasks: List of cell types to analyze attributions for either list of task names or query string to filter cell types to analyze attributions for (e.g. &#39;cell_type == &#39;classical monocyte&#39;&#39;). If not provided, all tasks will be analyzed.</span>
<span class="sd">        off_tasks: Optional list of cell types to contrast against either list of task names or query string to filter cell types to contrast against (e.g. &#39;cell_type == &#39;classical monocyte&#39;&#39;). If not provided, all tasks will be used as off tasks.</span>
<span class="sd">        model: Optional model to use for attribution analysis either replicate number or path to the model.</span>
<span class="sd">        method: Method to use for attribution analysis available options: &quot;saliency&quot;, &quot;inputxgradient&quot;, &quot;integratedgradients&quot;.</span>
<span class="sd">        device: Device to use for attribution analysis (e.g. &#39;cuda&#39;, &#39;cpu&#39;). If not provided, the best available device will be used automatically.</span>
<span class="sd">        dpi: DPI for attribution plots.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileExistsError: If output directory already exists.</span>

<span class="sd">    Examples:</span>
<span class="sd">    &gt;&gt;&gt; predict_save_attributions(</span>
<span class="sd">    ...     output_dir=&quot;output_dir&quot;,</span>
<span class="sd">    ...     genes=[</span>
<span class="sd">    ...         &quot;SPI1&quot;,</span>
<span class="sd">    ...         &quot;CD68&quot;,</span>
<span class="sd">    ...     ],</span>
<span class="sd">    ...     tasks=&quot;cell_type == &#39;classical monocyte&#39;&quot;,</span>
<span class="sd">    ... )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">output_prefix</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span>
    <span class="n">output_prefix</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">attributions_paths</span> <span class="o">=</span> <span class="n">predict_save_attributions</span><span class="p">(</span>
        <span class="n">output_prefix</span><span class="o">=</span><span class="n">output_prefix</span><span class="p">,</span>
        <span class="n">genes</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span>
        <span class="n">seqs</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span>
        <span class="n">tasks</span><span class="o">=</span><span class="n">tasks</span><span class="p">,</span>
        <span class="n">off_tasks</span><span class="o">=</span><span class="n">off_tasks</span><span class="p">,</span>
        <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
        <span class="n">metadata_anndata</span><span class="o">=</span><span class="n">metadata_anndata</span><span class="p">,</span>
        <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
        <span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">,</span>
        <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">,</span>
        <span class="n">batch_size</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
        <span class="n">top_n_markers</span><span class="o">=</span><span class="n">top_n_markers</span><span class="p">,</span>
        <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
        <span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">custom_genome</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">seqs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">custom_genome</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">assert</span> <span class="n">genes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;`genes` must be None when `seqs` is provided.&quot;</span>

    <span class="n">recursive_seqlet_calling</span><span class="p">(</span>
        <span class="n">output_prefix</span><span class="o">=</span><span class="n">output_prefix</span><span class="p">,</span>
        <span class="n">attributions</span><span class="o">=</span><span class="n">attributions_paths</span><span class="p">,</span>
        <span class="n">metadata_anndata</span><span class="o">=</span><span class="n">metadata_anndata</span><span class="p">,</span>
        <span class="n">genes</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span>
        <span class="n">tasks</span><span class="o">=</span><span class="n">tasks</span><span class="p">,</span>
        <span class="n">off_tasks</span><span class="o">=</span><span class="n">off_tasks</span><span class="p">,</span>
        <span class="n">tss_distance</span><span class="o">=</span><span class="n">tss_distance</span><span class="p">,</span>
        <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">,</span>
        <span class="n">custom_genome</span><span class="o">=</span><span class="n">custom_genome</span><span class="p">,</span>
        <span class="n">threshold</span><span class="o">=</span><span class="n">threshold</span><span class="p">,</span>
        <span class="n">min_seqlet_len</span><span class="o">=</span><span class="n">min_seqlet_len</span><span class="p">,</span>
        <span class="n">max_seqlet_len</span><span class="o">=</span><span class="n">max_seqlet_len</span><span class="p">,</span>
        <span class="n">additional_flanks</span><span class="o">=</span><span class="n">additional_flanks</span><span class="p">,</span>
        <span class="n">pattern_type</span><span class="o">=</span><span class="n">pattern_type</span><span class="p">,</span>
        <span class="n">meme_motif_db</span><span class="o">=</span><span class="n">meme_motif_db</span><span class="p">,</span>
    <span class="p">)</span></div>



<div class="viewcode-block" id="plot_attributions">
<a class="viewcode-back" href="../../../api/decima.interpret.html#decima.plot_attributions">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">plot_attributions</span><span class="p">(</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">metadata_anndata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">tss_distance</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seqlogo_window</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span>
    <span class="n">agg_func</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;mean&quot;</span><span class="p">,</span>
    <span class="n">custom_genome</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">dpi</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot attributions.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_prefix: Prefix for the output files.</span>
<span class="sd">        genes: Genes to attribute if not provided, all genes will be used.</span>
<span class="sd">        metadata_anndata: Metadata anndata.</span>
<span class="sd">        tss_distance: TSS distance for attribution for plotting.</span>
<span class="sd">        seqlogo_window: Seqlogo window.</span>
<span class="sd">        agg_func: Agg func for aggregation of attributions across replicates. Available options: &#39;mean&#39;, &#39;max&#39;.</span>
<span class="sd">        custom_genome: Custom genome if custom genome bigwig files will be generated as each gene is difference chromosome.</span>
<span class="sd">        dpi: DPI for attribution plots.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plot_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_plots&quot;</span><span class="p">)</span>
    <span class="n">plot_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">AttributionResult</span><span class="p">(</span>
        <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;*.attributions.h5&quot;</span><span class="p">),</span> <span class="n">tss_distance</span><span class="p">,</span> <span class="n">correct_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">agg_func</span><span class="o">=</span><span class="n">agg_func</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">ar</span><span class="p">:</span>
        <span class="c1"># TODO: if we save seqlets as h5 we do not need to recalculate them</span>
        <span class="c1"># TODO: create html reports</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Plotting attributions...&quot;</span><span class="p">):</span>
            <span class="n">attribution</span> <span class="o">=</span> <span class="n">ar</span><span class="o">.</span><span class="n">load_attribution</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">metadata_anndata</span><span class="p">,</span> <span class="n">custom_genome</span><span class="p">)</span>
            <span class="n">attribution</span><span class="o">.</span><span class="n">plot_peaks</span><span class="p">()</span><span class="o">.</span><span class="n">save</span><span class="p">((</span><span class="n">plot_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">.peaks.png&quot;</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span>

            <span class="n">seqlogo_plot_dir</span> <span class="o">=</span> <span class="n">plot_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_seqlogos&quot;</span>
            <span class="n">seqlogo_plot_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">peak</span> <span class="ow">in</span> <span class="n">attribution</span><span class="o">.</span><span class="n">peaks</span><span class="o">.</span><span class="n">itertuples</span><span class="p">():</span>
                <span class="n">logo</span> <span class="o">=</span> <span class="n">attribution</span><span class="o">.</span><span class="n">plot_seqlogo</span><span class="p">(</span><span class="n">relative_loc</span><span class="o">=</span><span class="n">peak</span><span class="o">.</span><span class="n">from_tss</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="n">seqlogo_window</span><span class="p">)</span>
                <span class="n">logo</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">seqlogo_plot_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">attribution</span><span class="o">.</span><span class="n">gene</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">peak</span><span class="o">.</span><span class="n">from_tss</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span></div>

</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Gokcen Eraslan
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../../_static/documentation_options.js?v=8e576371"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/furo.js?v=46bd48cc"></script>
    </body>
</html>